<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Orchestrating Hi-C analysis with Bioconductor - 2&nbsp; Hi-C data structures in R</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../pages/parsing.html" rel="next">
<link href="../pages/principles.html" rel="prev">
<link href="../assets/cover.jpg" rel="icon" type="image/jpeg">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
</head>
<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/principles.html">Fundamentals concepts</a></li><li class="breadcrumb-item"><a href="../pages/data-representation.html"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Hi-C data structures in R</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../">Orchestrating Hi-C analysis with Bioconductor</a> 
        <div class="sidebar-tools-main">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label=""><i class="bi bi-git"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
<li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://github.com/js2264/OHCA/">
            Source Code
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://js2264.github.io/OHCA/devel/">
            Browse version `devel`
            </a>
          </li>
      </ul>
</div>
    <a href="../Orchestrating-Hi-C-analysis-with-Bioconductor.pdf" rel="" title="Download PDF" class="quarto-navigation-tool px-1" aria-label="Download PDF"><i class="bi bi-file-pdf"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/preamble.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preamble</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Fundamentals concepts</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/principles.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Hi-C pre-processing steps</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/data-representation.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Hi-C data structures in R</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/parsing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Manipulating Hi-C data in R</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/visualization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Hi-C data visualization</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">In-depth Hi-C analysis</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/matrix-centric.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Matrix-centric analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/interactions-centric.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Interactions-centric analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/topological-features.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Finding topological features in Hi-C</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Advanced Hi-C topics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/disseminating.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Data gateways: accessing public Hi-C data portals</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/interoperability.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Interoperability: using HiCExperiment with other R packages</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/workflow-yeast.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Workflow 1: Distance-dependent interactions across yeast mutants</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/workflow-chicken.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Workflow 2: Chromosome compartment cohesion upon mitosis entry</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/workflow-centros.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Workflow 3: Inter-centromere interactions in yeast</span></span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li>
<a href="#granges-class" id="toc-granges-class" class="nav-link active" data-scroll-target="#granges-class"><span class="header-section-number">2.1</span> <code>GRanges</code> class</a>
  <ul class="collapse">
<li><a href="#granges-fundamentals" id="toc-granges-fundamentals" class="nav-link" data-scroll-target="#granges-fundamentals"><span class="header-section-number">2.1.1</span> <code>GRanges</code> fundamentals</a></li>
  <li><a href="#granges-metadata" id="toc-granges-metadata" class="nav-link" data-scroll-target="#granges-metadata"><span class="header-section-number">2.1.2</span> <code>GRanges</code> metadata</a></li>
  <li><a href="#genomic-arithmetics-on-individual-granges-objects" id="toc-genomic-arithmetics-on-individual-granges-objects" class="nav-link" data-scroll-target="#genomic-arithmetics-on-individual-granges-objects"><span class="header-section-number">2.1.3</span> Genomic arithmetics on individual <code>GRanges</code> objects</a></li>
  <li><a href="#comparing-multiple-granges-objects" id="toc-comparing-multiple-granges-objects" class="nav-link" data-scroll-target="#comparing-multiple-granges-objects"><span class="header-section-number">2.1.4</span> Comparing multiple <code>GRanges</code> objects</a></li>
  </ul>
</li>
  <li>
<a href="#ginteractions-class" id="toc-ginteractions-class" class="nav-link" data-scroll-target="#ginteractions-class"><span class="header-section-number">2.2</span> <code>GInteractions</code> class</a>
  <ul class="collapse">
<li><a href="#building-a-ginteractions-object-from-scratch" id="toc-building-a-ginteractions-object-from-scratch" class="nav-link" data-scroll-target="#building-a-ginteractions-object-from-scratch"><span class="header-section-number">2.2.1</span> Building a <code>GInteractions</code> object from scratch</a></li>
  <li><a href="#ginteractions-specific-slots" id="toc-ginteractions-specific-slots" class="nav-link" data-scroll-target="#ginteractions-specific-slots"><span class="header-section-number">2.2.2</span> <code>GInteractions</code> specific slots</a></li>
  <li><a href="#ginteractions-methods" id="toc-ginteractions-methods" class="nav-link" data-scroll-target="#ginteractions-methods"><span class="header-section-number">2.2.3</span> <code>GInteractions</code> methods</a></li>
  </ul>
</li>
  <li>
<a href="#contactfile-class" id="toc-contactfile-class" class="nav-link" data-scroll-target="#contactfile-class"><span class="header-section-number">2.3</span> <code>ContactFile</code> class</a>
  <ul class="collapse">
<li><a href="#accessing-example-hi-c-files" id="toc-accessing-example-hi-c-files" class="nav-link" data-scroll-target="#accessing-example-hi-c-files"><span class="header-section-number">2.3.1</span> Accessing example Hi-C files</a></li>
  <li><a href="#contactfile-fundamentals" id="toc-contactfile-fundamentals" class="nav-link" data-scroll-target="#contactfile-fundamentals"><span class="header-section-number">2.3.2</span> <code>ContactFile</code> fundamentals</a></li>
  <li><a href="#contactfile-slots" id="toc-contactfile-slots" class="nav-link" data-scroll-target="#contactfile-slots"><span class="header-section-number">2.3.3</span> <code>ContactFile</code> slots</a></li>
  <li><a href="#contactfile-methods" id="toc-contactfile-methods" class="nav-link" data-scroll-target="#contactfile-methods"><span class="header-section-number">2.3.4</span> <code>ContactFile</code> methods</a></li>
  </ul>
</li>
  <li>
<a href="#hicexperiment-class" id="toc-hicexperiment-class" class="nav-link" data-scroll-target="#hicexperiment-class"><span class="header-section-number">2.4</span> <code>HiCExperiment</code> class</a>
  <ul class="collapse">
<li><a href="#creating-a-hicexperiment-object" id="toc-creating-a-hicexperiment-object" class="nav-link" data-scroll-target="#creating-a-hicexperiment-object"><span class="header-section-number">2.4.1</span> Creating a <code>HiCExperiment</code> object</a></li>
  <li><a href="#interacting-with-hicexperiment-data" id="toc-interacting-with-hicexperiment-data" class="nav-link" data-scroll-target="#interacting-with-hicexperiment-data"><span class="header-section-number">2.4.2</span> Interacting with <code>HiCExperiment</code> data</a></li>
  </ul>
</li>
  <li><a href="#visual-summary-of-the-hicexperiment-data-structure" id="toc-visual-summary-of-the-hicexperiment-data-structure" class="nav-link" data-scroll-target="#visual-summary-of-the-hicexperiment-data-structure"><span class="header-section-number">2.5</span> Visual summary of the <code>HiCExperiment</code> data structure</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title">
<span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Hi-C data structures in R</span>
</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Aims
</div>
</div>
<div class="callout-body-container callout-body">
<p>This chapter introduces the four main classes offered by <code>Bioconductor</code> leveraged to perform Hi-C analysis, describes their structure and how to interact with them:</p>
<ul>
<li>
<code>GRanges</code> (jump to <a href="#granges-class">the section</a>)<br>
</li>
<li>
<code>GInteractions</code> (jump to <a href="#ginteractions-class">the section</a>)<br>
</li>
<li>
<code>ContactFile</code> (jump to <a href="#contactfile-class">the section</a>)<br>
</li>
<li>
<code>HiCExperiment</code> (jump to <a href="#hicexperiment-class">the section</a>)</li>
</ul>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
TL;DR
</div>
</div>
<div class="callout-body-container callout-body">
<p>Directly jump to the <a href="#visual-summary-of-the-hicexperiment-data-structure">last section of this chapter</a> to get a visual representation of these data structures.</p>
</div>
</div>
<section id="granges-class" class="level2" data-number="2.1"><h2 data-number="2.1" class="anchored" data-anchor-id="granges-class">
<span class="header-section-number">2.1</span> <code>GRanges</code> class</h2>
<p><code>GRanges</code> is a shorthand for <code>GenomicRanges</code>, a core class in <code>Bioconductor</code>. This class is primarily used to describe genomic ranges of any nature, e.g.&nbsp; sets of promoters, SNPs, chromatin loop anchors, ….<br>
The data structure has been published in the seminal 2015 publication by the <code>Bioconductor</code> team (<span class="citation" data-cites="Huber_2015">Huber et al. (<a href="#ref-Huber_2015" role="doc-biblioref">2015</a>)</span>).</p>
<section id="granges-fundamentals" class="level3" data-number="2.1.1"><h3 data-number="2.1.1" class="anchored" data-anchor-id="granges-fundamentals">
<span class="header-section-number">2.1.1</span> <code>GRanges</code> fundamentals</h3>
<p>The easiest way to generate a <code>GRanges</code> object is to coerce it from a vector of genomic coordinates in the UCSC format (e.g.&nbsp;<code>"chr2:2004-4853"</code>):</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges">GenomicRanges</a></span><span class="op">)</span></span>
<span><span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html">GRanges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"chr2:2004-7853:+"</span>, </span>
<span>    <span class="st">"chr4:4482-9873:-"</span>, </span>
<span>    <span class="st">"chr5:1943-4203:+"</span>, </span>
<span>    <span class="st">"chr5:4103-5004:+"</span>  </span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">gr</span></span>
<span><span class="co">##  GRanges object with 4 ranges and 0 metadata columns:</span></span>
<span><span class="co">##        seqnames    ranges strand</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##    [1]     chr2 2004-7853      +</span></span>
<span><span class="co">##    [2]     chr4 4482-9873      -</span></span>
<span><span class="co">##    [3]     chr5 1943-4203      +</span></span>
<span><span class="co">##    [4]     chr5 4103-5004      +</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>A single <code>GRanges</code> object can contain one or several <strong>“ranges”</strong>, or genomic intervals. To navigate between these ranges, <code>GRanges</code> can be subset using the standard <code>R</code> single bracket notation <code>[</code>:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gr</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="co">##  GRanges object with 1 range and 0 metadata columns:</span></span>
<span><span class="co">##        seqnames    ranges strand</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##    [1]     chr2 2004-7853      +</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span>
<span></span>
<span><span class="va">gr</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span>
<span><span class="co">##  GRanges object with 3 ranges and 0 metadata columns:</span></span>
<span><span class="co">##        seqnames    ranges strand</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##    [1]     chr2 2004-7853      +</span></span>
<span><span class="co">##    [2]     chr4 4482-9873      -</span></span>
<span><span class="co">##    [3]     chr5 1943-4203      +</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><code>GenomicRanges</code> objects aim to provide a natural description of genomic intervals (ranges) and are incredibly versatile. They extend the <code>data.frame</code> object and have four required pieces of information:</p>
<ul>
<li>
<code>seqnames</code> (i.e.&nbsp;chromosome names) (accessible with <code><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html">seqnames()</a></code>)</li>
<li>
<code>start</code> (accessible with <code><a href="https://rdrr.io/r/stats/start.html">start()</a></code>)</li>
<li>
<code>end</code> (accessible with <code><a href="https://rdrr.io/r/stats/start.html">end()</a></code>)</li>
<li>
<code>strand</code> (accessible with <code><a href="https://rdrr.io/pkg/BiocGenerics/man/strand.html">strand()</a></code>)</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html">seqnames</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span></span>
<span><span class="co">##  factor-Rle of length 4 with 3 runs</span></span>
<span><span class="co">##    Lengths:    1    1    2</span></span>
<span><span class="co">##    Values : chr2 chr4 chr5</span></span>
<span><span class="co">##  Levels(3): chr2 chr4 chr5</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/start.html">start</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span></span>
<span><span class="co">##  [1] 2004 4482 1943 4103</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/start.html">end</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span></span>
<span><span class="co">##  [1] 7853 9873 4203 5004</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/strand.html">strand</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span></span>
<span><span class="co">##  factor-Rle of length 4 with 3 runs</span></span>
<span><span class="co">##    Lengths: 1 1 2</span></span>
<span><span class="co">##    Values : + - +</span></span>
<span><span class="co">##  Levels(3): + - *</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here is a graphical representation of a <code>GRanges</code> object, taken from <a href="https://www.bioconductor.org/help/course-materials/2015/UseBioconductorFeb2015/"><code>Bioconductor</code> course material</a>:</p>
<p><img src="images/20230306102639.png" class="img-fluid"></p>
<p>We will now delve into the detailed structure and operability of <code>GRanges</code> objects.</p>
</section><section id="granges-metadata" class="level3" data-number="2.1.2"><h3 data-number="2.1.2" class="anchored" data-anchor-id="granges-metadata">
<span class="header-section-number">2.1.2</span> <code>GRanges</code> metadata</h3>
<p>An important aspect of <code>GRanges</code> objects is that each entry (range) can have extra optional metadata. This metadata is stored in a rectangular <code>DataFrame</code>. Each column can contain a different type of information, e.g.&nbsp;a <code>numerical</code> vector, a <code>factor</code>, a list, …</p>
<p>One can directly access this <code>DataFrame</code> using the <code><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">mcols()</a></code> function, and individual columns of metadata using the <code>$</code> notation:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">mcols</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span></span>
<span><span class="co">##  DataFrame with 4 rows and 0 columns</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">mcols</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span><span class="op">$</span><span class="va">GC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.45</span>, <span class="fl">0.43</span>, <span class="fl">0.44</span>, <span class="fl">0.42</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">mcols</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span><span class="op">$</span><span class="va">annotation</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="cn">NA</span>, <span class="st">'promoter'</span>, <span class="st">'enhancer'</span>, <span class="st">'centromere'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">mcols</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span><span class="op">$</span><span class="va">extended.info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="cn">NA</span><span class="op">)</span><span class="op">)</span>, </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>date <span class="op">=</span> <span class="fl">2023</span>, source <span class="op">=</span> <span class="st">'manual'</span><span class="op">)</span><span class="op">)</span>, </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>date <span class="op">=</span> <span class="fl">2021</span>, source <span class="op">=</span> <span class="st">'manual'</span><span class="op">)</span><span class="op">)</span>, </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>date <span class="op">=</span> <span class="fl">2019</span>, source <span class="op">=</span> <span class="st">'homology'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">mcols</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span></span>
<span><span class="co">##  DataFrame with 4 rows and 3 columns</span></span>
<span><span class="co">##           GC annotation extended.info</span></span>
<span><span class="co">##    &lt;numeric&gt;   &lt;factor&gt;        &lt;list&gt;</span></span>
<span><span class="co">##  1      0.45 NA                    NA</span></span>
<span><span class="co">##  2      0.43 promoter     2023,manual</span></span>
<span><span class="co">##  3      0.44 enhancer     2021,manual</span></span>
<span><span class="co">##  4      0.42 centromere 2019,homology</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>When metadata columns are defined for a <code>GRanges</code> object, they are pasted next to the minimal 4 required <code>GRanges</code> fields, separated by a <code>|</code> character.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gr</span></span>
<span><span class="co">##  GRanges object with 4 ranges and 3 metadata columns:</span></span>
<span><span class="co">##        seqnames    ranges strand |        GC annotation extended.info</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;   &lt;factor&gt;        &lt;list&gt;</span></span>
<span><span class="co">##    [1]     chr2 2004-7853      + |      0.45 NA                  &lt;NA&gt;</span></span>
<span><span class="co">##    [2]     chr4 4482-9873      - |      0.43 promoter     2023,manual</span></span>
<span><span class="co">##    [3]     chr5 1943-4203      + |      0.44 enhancer     2021,manual</span></span>
<span><span class="co">##    [4]     chr5 4103-5004      + |      0.42 centromere 2019,homology</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="genomic-arithmetics-on-individual-granges-objects" class="level3" data-number="2.1.3"><h3 data-number="2.1.3" class="anchored" data-anchor-id="genomic-arithmetics-on-individual-granges-objects">
<span class="header-section-number">2.1.3</span> Genomic arithmetics on individual <code>GRanges</code> objects</h3>
<p>A <code>GRanges</code> object primarily describes a set of genomic ranges (<em>it is in the name!</em>). Useful genomic-oriented methods have been implemented to investigate individual <code>GRanges</code> object from a genomic perspective.</p>
<section id="intra-range-methods" class="level4" data-number="2.1.3.1"><h4 data-number="2.1.3.1" class="anchored" data-anchor-id="intra-range-methods">
<span class="header-section-number">2.1.3.1</span> Intra-range methods</h4>
<p>Standard genomic arithmetics are possible with <code>GRanges</code>, e.g.&nbsp; shifting ranges, resizing, trimming, … These methods are referred to as “intra-range” methods as they work <em>“one-region-at-a-time”</em>.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Each range of the input <code>GRanges</code> object is modified independently from the other ranges in the following code chunks.</li>
<li>Intra-range operations are <strong>endomorphisms</strong>: they all take <code>GRanges</code> inputs and always return <code>GRanges</code> objects.</li>
</ul>
</div>
</div>
<ul>
<li>Shifting each genomic range in a <code>GRanges</code> object by a certain number of bases:</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gr</span></span>
<span><span class="co">##  GRanges object with 4 ranges and 3 metadata columns:</span></span>
<span><span class="co">##        seqnames    ranges strand |        GC annotation extended.info</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;   &lt;factor&gt;        &lt;list&gt;</span></span>
<span><span class="co">##    [1]     chr2 2004-7853      + |      0.45 NA                  &lt;NA&gt;</span></span>
<span><span class="co">##    [2]     chr4 4482-9873      - |      0.43 promoter     2023,manual</span></span>
<span><span class="co">##    [3]     chr5 1943-4203      + |      0.44 enhancer     2021,manual</span></span>
<span><span class="co">##    [4]     chr5 4103-5004      + |      0.42 centromere 2019,homology</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span>
<span></span>
<span><span class="co"># ----- Shift all genomic ranges towards the "right" (downstream in `+` strand), by 1000bp:</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/intra-range-methods.html">shift</a></span><span class="op">(</span><span class="va">gr</span>, <span class="fl">1000</span><span class="op">)</span></span>
<span><span class="co">##  GRanges object with 4 ranges and 3 metadata columns:</span></span>
<span><span class="co">##        seqnames     ranges strand |        GC annotation extended.info</span></span>
<span><span class="co">##           &lt;Rle&gt;  &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;   &lt;factor&gt;        &lt;list&gt;</span></span>
<span><span class="co">##    [1]     chr2  3004-8853      + |      0.45 NA                  &lt;NA&gt;</span></span>
<span><span class="co">##    [2]     chr4 5482-10873      - |      0.43 promoter     2023,manual</span></span>
<span><span class="co">##    [3]     chr5  2943-5203      + |      0.44 enhancer     2021,manual</span></span>
<span><span class="co">##    [4]     chr5  5103-6004      + |      0.42 centromere 2019,homology</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span>
<span></span>
<span><span class="co"># ----- Shift all genomic ranges towards the "left" (upstream in `+` strand), by 1000bp:</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/intra-range-methods.html">shift</a></span><span class="op">(</span><span class="va">gr</span>, <span class="op">-</span><span class="fl">1000</span><span class="op">)</span></span>
<span><span class="co">##  GRanges object with 4 ranges and 3 metadata columns:</span></span>
<span><span class="co">##        seqnames    ranges strand |        GC annotation extended.info</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;   &lt;factor&gt;        &lt;list&gt;</span></span>
<span><span class="co">##    [1]     chr2 1004-6853      + |      0.45 NA                  &lt;NA&gt;</span></span>
<span><span class="co">##    [2]     chr4 3482-8873      - |      0.43 promoter     2023,manual</span></span>
<span><span class="co">##    [3]     chr5  943-3203      + |      0.44 enhancer     2021,manual</span></span>
<span><span class="co">##    [4]     chr5 3103-4004      + |      0.42 centromere 2019,homology</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Narrowing each genomic range in a <code>GRanges</code> object by a certain number of bases:</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gr</span></span>
<span><span class="co">##  GRanges object with 4 ranges and 3 metadata columns:</span></span>
<span><span class="co">##        seqnames    ranges strand |        GC annotation extended.info</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;   &lt;factor&gt;        &lt;list&gt;</span></span>
<span><span class="co">##    [1]     chr2 2004-7853      + |      0.45 NA                  &lt;NA&gt;</span></span>
<span><span class="co">##    [2]     chr4 4482-9873      - |      0.43 promoter     2023,manual</span></span>
<span><span class="co">##    [3]     chr5 1943-4203      + |      0.44 enhancer     2021,manual</span></span>
<span><span class="co">##    [4]     chr5 4103-5004      + |      0.42 centromere 2019,homology</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span>
<span></span>
<span><span class="co"># ----- Extract 21st-40th subrange for each range in `gr`:</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/intra-range-methods.html">narrow</a></span><span class="op">(</span><span class="va">gr</span>, start <span class="op">=</span> <span class="fl">21</span>, end <span class="op">=</span> <span class="fl">40</span><span class="op">)</span></span>
<span><span class="co">##  GRanges object with 4 ranges and 3 metadata columns:</span></span>
<span><span class="co">##        seqnames    ranges strand |        GC annotation extended.info</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;   &lt;factor&gt;        &lt;list&gt;</span></span>
<span><span class="co">##    [1]     chr2 2024-2043      + |      0.45 NA                  &lt;NA&gt;</span></span>
<span><span class="co">##    [2]     chr4 4502-4521      - |      0.43 promoter     2023,manual</span></span>
<span><span class="co">##    [3]     chr5 1963-1982      + |      0.44 enhancer     2021,manual</span></span>
<span><span class="co">##    [4]     chr5 4123-4142      + |      0.42 centromere 2019,homology</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/start.html">width</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/intra-range-methods.html">narrow</a></span><span class="op">(</span><span class="va">gr</span>, start <span class="op">=</span> <span class="fl">21</span>, end <span class="op">=</span> <span class="fl">40</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  [1] 20 20 20 20</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Resizing each genomic range in a <code>GRanges</code> object to a certain number of bases:</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gr</span></span>
<span><span class="co">##  GRanges object with 4 ranges and 3 metadata columns:</span></span>
<span><span class="co">##        seqnames    ranges strand |        GC annotation extended.info</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;   &lt;factor&gt;        &lt;list&gt;</span></span>
<span><span class="co">##    [1]     chr2 2004-7853      + |      0.45 NA                  &lt;NA&gt;</span></span>
<span><span class="co">##    [2]     chr4 4482-9873      - |      0.43 promoter     2023,manual</span></span>
<span><span class="co">##    [3]     chr5 1943-4203      + |      0.44 enhancer     2021,manual</span></span>
<span><span class="co">##    [4]     chr5 4103-5004      + |      0.42 centromere 2019,homology</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span>
<span></span>
<span><span class="co"># ----- Resize `gr` entries to 100, fixed at the start of each range:</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/intra-range-methods.html">resize</a></span><span class="op">(</span><span class="va">gr</span>, <span class="fl">100</span>, fix <span class="op">=</span> <span class="st">"start"</span><span class="op">)</span></span>
<span><span class="co">##  GRanges object with 4 ranges and 3 metadata columns:</span></span>
<span><span class="co">##        seqnames    ranges strand |        GC annotation extended.info</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;   &lt;factor&gt;        &lt;list&gt;</span></span>
<span><span class="co">##    [1]     chr2 2004-2103      + |      0.45 NA                  &lt;NA&gt;</span></span>
<span><span class="co">##    [2]     chr4 9774-9873      - |      0.43 promoter     2023,manual</span></span>
<span><span class="co">##    [3]     chr5 1943-2042      + |      0.44 enhancer     2021,manual</span></span>
<span><span class="co">##    [4]     chr5 4103-4202      + |      0.42 centromere 2019,homology</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span>
<span></span>
<span><span class="co"># ----- Resize `gr` entries to 100, fixed at the start of each range, disregarding strand information:</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/intra-range-methods.html">resize</a></span><span class="op">(</span><span class="va">gr</span>, <span class="fl">100</span>, fix <span class="op">=</span> <span class="st">"start"</span>, ignore.strand <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">##  GRanges object with 4 ranges and 3 metadata columns:</span></span>
<span><span class="co">##        seqnames    ranges strand |        GC annotation extended.info</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;   &lt;factor&gt;        &lt;list&gt;</span></span>
<span><span class="co">##    [1]     chr2 2004-2103      + |      0.45 NA                  &lt;NA&gt;</span></span>
<span><span class="co">##    [2]     chr4 4482-4581      - |      0.43 promoter     2023,manual</span></span>
<span><span class="co">##    [3]     chr5 1943-2042      + |      0.44 enhancer     2021,manual</span></span>
<span><span class="co">##    [4]     chr5 4103-4202      + |      0.42 centromere 2019,homology</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span>
<span></span>
<span><span class="co"># ----- Resize `gr` entries to 1 bp, fixed at the center of each range:</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/intra-range-methods.html">resize</a></span><span class="op">(</span><span class="va">gr</span>, <span class="fl">1</span>, fix <span class="op">=</span> <span class="st">"center"</span><span class="op">)</span></span>
<span><span class="co">##  GRanges object with 4 ranges and 3 metadata columns:</span></span>
<span><span class="co">##        seqnames    ranges strand |        GC annotation extended.info</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;   &lt;factor&gt;        &lt;list&gt;</span></span>
<span><span class="co">##    [1]     chr2      4928      + |      0.45 NA                  &lt;NA&gt;</span></span>
<span><span class="co">##    [2]     chr4      7177      - |      0.43 promoter     2023,manual</span></span>
<span><span class="co">##    [3]     chr5      3073      + |      0.44 enhancer     2021,manual</span></span>
<span><span class="co">##    [4]     chr5      4553      + |      0.42 centromere 2019,homology</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Extracting flanking coordinates for each entry in <code>gr</code>:</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gr</span></span>
<span><span class="co">##  GRanges object with 4 ranges and 3 metadata columns:</span></span>
<span><span class="co">##        seqnames    ranges strand |        GC annotation extended.info</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;   &lt;factor&gt;        &lt;list&gt;</span></span>
<span><span class="co">##    [1]     chr2 2004-7853      + |      0.45 NA                  &lt;NA&gt;</span></span>
<span><span class="co">##    [2]     chr4 4482-9873      - |      0.43 promoter     2023,manual</span></span>
<span><span class="co">##    [3]     chr5 1943-4203      + |      0.44 enhancer     2021,manual</span></span>
<span><span class="co">##    [4]     chr5 4103-5004      + |      0.42 centromere 2019,homology</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span>
<span></span>
<span><span class="co"># ----- Extract 100bp UPSTREAM of each genomic range, according to range strandness:</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/intra-range-methods.html">flank</a></span><span class="op">(</span><span class="va">gr</span>, <span class="fl">100</span>, start <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">##  GRanges object with 4 ranges and 3 metadata columns:</span></span>
<span><span class="co">##        seqnames    ranges strand |        GC annotation extended.info</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;   &lt;factor&gt;        &lt;list&gt;</span></span>
<span><span class="co">##    [1]     chr2 1904-2003      + |      0.45 NA                  &lt;NA&gt;</span></span>
<span><span class="co">##    [2]     chr4 9874-9973      - |      0.43 promoter     2023,manual</span></span>
<span><span class="co">##    [3]     chr5 1843-1942      + |      0.44 enhancer     2021,manual</span></span>
<span><span class="co">##    [4]     chr5 4003-4102      + |      0.42 centromere 2019,homology</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span>
<span></span>
<span><span class="co"># ----- Extract 1bp DOWNSTREAM of each genomic range, according to range strandness:</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/intra-range-methods.html">flank</a></span><span class="op">(</span><span class="va">gr</span>, <span class="fl">1</span>, start <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">##  GRanges object with 4 ranges and 3 metadata columns:</span></span>
<span><span class="co">##        seqnames    ranges strand |        GC annotation extended.info</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;   &lt;factor&gt;        &lt;list&gt;</span></span>
<span><span class="co">##    [1]     chr2      7854      + |      0.45 NA                  &lt;NA&gt;</span></span>
<span><span class="co">##    [2]     chr4      4481      - |      0.43 promoter     2023,manual</span></span>
<span><span class="co">##    [3]     chr5      4204      + |      0.44 enhancer     2021,manual</span></span>
<span><span class="co">##    [4]     chr5      5005      + |      0.42 centromere 2019,homology</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note how here again, strand information is crucial and correctly leveraged to extract “upstream” or “downstream” flanking regions in agreement with genomic range orientation.</p>
<ul>
<li>Several arithmetics operators can also directly work with <code>GRanges</code>:</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gr</span></span>
<span><span class="co">##  GRanges object with 4 ranges and 3 metadata columns:</span></span>
<span><span class="co">##        seqnames    ranges strand |        GC annotation extended.info</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;   &lt;factor&gt;        &lt;list&gt;</span></span>
<span><span class="co">##    [1]     chr2 2004-7853      + |      0.45 NA                  &lt;NA&gt;</span></span>
<span><span class="co">##    [2]     chr4 4482-9873      - |      0.43 promoter     2023,manual</span></span>
<span><span class="co">##    [3]     chr5 1943-4203      + |      0.44 enhancer     2021,manual</span></span>
<span><span class="co">##    [4]     chr5 4103-5004      + |      0.42 centromere 2019,homology</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span>
<span></span>
<span><span class="va">gr</span> <span class="op">+</span> <span class="fl">100</span> <span class="co"># ----- Extend each side of the `GRanges` by a given number of bases</span></span>
<span><span class="co">##  GRanges object with 4 ranges and 3 metadata columns:</span></span>
<span><span class="co">##        seqnames    ranges strand |        GC annotation extended.info</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;   &lt;factor&gt;        &lt;list&gt;</span></span>
<span><span class="co">##    [1]     chr2 1904-7953      + |      0.45 NA                  &lt;NA&gt;</span></span>
<span><span class="co">##    [2]     chr4 4382-9973      - |      0.43 promoter     2023,manual</span></span>
<span><span class="co">##    [3]     chr5 1843-4303      + |      0.44 enhancer     2021,manual</span></span>
<span><span class="co">##    [4]     chr5 4003-5104      + |      0.42 centromere 2019,homology</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span>
<span></span>
<span><span class="va">gr</span> <span class="op">-</span> <span class="fl">200</span> <span class="co"># ----- Shrink each side of the `GRanges` by a given number of bases </span></span>
<span><span class="co">##  GRanges object with 4 ranges and 3 metadata columns:</span></span>
<span><span class="co">##        seqnames    ranges strand |        GC annotation extended.info</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;   &lt;factor&gt;        &lt;list&gt;</span></span>
<span><span class="co">##    [1]     chr2 2204-7653      + |      0.45 NA                  &lt;NA&gt;</span></span>
<span><span class="co">##    [2]     chr4 4682-9673      - |      0.43 promoter     2023,manual</span></span>
<span><span class="co">##    [3]     chr5 2143-4003      + |      0.44 enhancer     2021,manual</span></span>
<span><span class="co">##    [4]     chr5 4303-4804      + |      0.42 centromere 2019,homology</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span>
<span></span>
<span><span class="va">gr</span> <span class="op">*</span> <span class="fl">1000</span> <span class="co"># ----- Zoom in by a given factor (effectively decreasing the `GRanges` width by the same factor)</span></span>
<span><span class="co">##  GRanges object with 4 ranges and 3 metadata columns:</span></span>
<span><span class="co">##        seqnames    ranges strand |        GC annotation extended.info</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;   &lt;factor&gt;        &lt;list&gt;</span></span>
<span><span class="co">##    [1]     chr2 4926-4930      + |      0.45 NA                  &lt;NA&gt;</span></span>
<span><span class="co">##    [2]     chr4 7175-7179      - |      0.43 promoter     2023,manual</span></span>
<span><span class="co">##    [3]     chr5 3072-3073      + |      0.44 enhancer     2021,manual</span></span>
<span><span class="co">##    [4]     chr5 4554-4553      + |      0.42 centromere 2019,homology</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Going further
</div>
</div>
<div class="callout-body-container callout-body">
<p>To fully grasp how to operate <code>GRanges</code> objects, we highly recommend reading the detailed documentation for this class by typing <code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html">?GenomicRanges</a></code> and <code><a href="https://rdrr.io/pkg/GenomicRanges/man/intra-range-methods.html">?GenomicRanges::`intra-range-methods`</a></code>.</p>
</div>
</div>
</section><section id="inter-range" class="level4" data-number="2.1.3.2"><h4 data-number="2.1.3.2" class="anchored" data-anchor-id="inter-range">
<span class="header-section-number">2.1.3.2</span> Inter-range methods</h4>
<p>Compared to “intra-range” methods described above, <strong>inter-range</strong> methods involve comparisons <em>between</em> ranges <strong>in a single</strong> GRanges object.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Compared to previous section, the result of each function described below depends on the <strong>entire set of ranges</strong> in the input <code>GRanges</code> object.</p>
</div>
</div>
<ul>
<li>Computing the “inverse” genomic ranges, i.e.&nbsp;ranges in-between the input ranges:</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/inter-range-methods.html">gaps</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span></span>
<span><span class="co">##  GRanges object with 3 ranges and 0 metadata columns:</span></span>
<span><span class="co">##        seqnames    ranges strand</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##    [1]     chr2    1-2003      +</span></span>
<span><span class="co">##    [2]     chr4    1-4481      -</span></span>
<span><span class="co">##    [3]     chr5    1-1942      +</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 3 sequences from an unspecified genome; no seqlengths</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>For each entry in a <code>GRanges</code>, finding the index of the preceding/following/nearest genomic range:</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/nearest-methods.html">precede</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span></span>
<span><span class="co">##  [1] NA NA NA NA</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/nearest-methods.html">follow</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span></span>
<span><span class="co">##  [1] NA NA NA NA</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/nearest-methods.html">nearest</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span></span>
<span><span class="co">##  [1] NA NA  4  3</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Computing a coverage over a genome, optionally indicated a “score” column from metadata:</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/coverage-methods.html">coverage</a></span><span class="op">(</span><span class="va">gr</span>, weight <span class="op">=</span> <span class="st">'GC'</span><span class="op">)</span></span>
<span><span class="co">##  RleList of length 3</span></span>
<span><span class="co">##  $chr2</span></span>
<span><span class="co">##  numeric-Rle of length 7853 with 2 runs</span></span>
<span><span class="co">##    Lengths: 2003 5850</span></span>
<span><span class="co">##    Values : 0.00 0.45</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $chr4</span></span>
<span><span class="co">##  numeric-Rle of length 9873 with 2 runs</span></span>
<span><span class="co">##    Lengths: 4481 5392</span></span>
<span><span class="co">##    Values : 0.00 0.43</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $chr5</span></span>
<span><span class="co">##  numeric-Rle of length 5004 with 4 runs</span></span>
<span><span class="co">##    Lengths: 1942 2160  101  801</span></span>
<span><span class="co">##    Values : 0.00 0.44 0.86 0.42</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Going further
</div>
</div>
<div class="callout-body-container callout-body">
<p>To fully grasp how to operate <code>GRanges</code> objects, we highly recommend reading the detailed documentation for this class by typing <code><a href="https://rdrr.io/pkg/GenomicRanges/man/inter-range-methods.html">?GenomicRanges::`inter-range-methods`</a></code>.</p>
</div>
</div>
</section></section><section id="comparing-multiple-granges-objects" class="level3" data-number="2.1.4"><h3 data-number="2.1.4" class="anchored" data-anchor-id="comparing-multiple-granges-objects">
<span class="header-section-number">2.1.4</span> Comparing multiple <code>GRanges</code> objects</h3>
<p>Genomic analysis typically requires intersection of two sets of genomic ranges, e.g.&nbsp;to find which ranges from one set overlap with those from another set.</p>
<p>In the next examples, we will use two <code>GRanges</code>:</p>
<ul>
<li>
<code>peaks</code> represents dummy 8 ChIP-seq peaks</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">peaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html">GRanges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>    <span class="st">'chr1:320-418'</span>,</span>
<span>    <span class="st">'chr1:512-567'</span>,</span>
<span>    <span class="st">'chr1:843-892'</span>,</span>
<span>    <span class="st">'chr1:1221-1317'</span>, </span>
<span>    <span class="st">'chr1:1329-1372'</span>, </span>
<span>    <span class="st">'chr1:1852-1909'</span>, </span>
<span>    <span class="st">'chr1:2489-2532'</span>, </span>
<span>    <span class="st">'chr1:2746-2790'</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">peaks</span></span>
<span><span class="co">##  GRanges object with 8 ranges and 0 metadata columns:</span></span>
<span><span class="co">##        seqnames    ranges strand</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##    [1]     chr1   320-418      *</span></span>
<span><span class="co">##    [2]     chr1   512-567      *</span></span>
<span><span class="co">##    [3]     chr1   843-892      *</span></span>
<span><span class="co">##    [4]     chr1 1221-1317      *</span></span>
<span><span class="co">##    [5]     chr1 1329-1372      *</span></span>
<span><span class="co">##    [6]     chr1 1852-1909      *</span></span>
<span><span class="co">##    [7]     chr1 2489-2532      *</span></span>
<span><span class="co">##    [8]     chr1 2746-2790      *</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>
<code>TSSs</code> represents dummy 3 gene promoters (± 10bp around the TSS)</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html">GRanges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>    <span class="st">'chr1:358-1292:+'</span>,</span>
<span>    <span class="st">'chr1:1324-2343:+'</span>, </span>
<span>    <span class="st">'chr1:2732-2751:+'</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">TSSs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/intra-range-methods.html">resize</a></span><span class="op">(</span><span class="va">genes</span>, width <span class="op">=</span> <span class="fl">1</span>, fix <span class="op">=</span> <span class="st">'start'</span><span class="op">)</span> <span class="op">+</span> <span class="fl">10</span></span>
<span><span class="va">TSSs</span></span>
<span><span class="co">##  GRanges object with 3 ranges and 0 metadata columns:</span></span>
<span><span class="co">##        seqnames    ranges strand</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##    [1]     chr1   348-368      +</span></span>
<span><span class="co">##    [2]     chr1 1314-1334      +</span></span>
<span><span class="co">##    [3]     chr1 2722-2742      +</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s see how they overlap by plotting them:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="va">peaks</span><span class="op">$</span><span class="va">type</span> <span class="op">&lt;-</span> <span class="st">'peaks'</span></span>
<span><span class="va">TSSs</span><span class="op">$</span><span class="va">type</span> <span class="op">&lt;-</span> <span class="st">'TSSs'</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">ggbio</span><span class="fu">::</span><span class="fu">geom_rect</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">peaks</span>, <span class="va">TSSs</span><span class="op">)</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="va">type</span><span class="op">)</span>, facets <span class="op">=</span> <span class="va">type</span><span class="op">~</span><span class="va">.</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">ggbio</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggbio/man/theme.html">theme_alignment</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html">coord_fixed</a></span><span class="op">(</span>ratio <span class="op">=</span> <span class="fl">300</span><span class="op">)</span></span>
<span><span class="co">##  Registered S3 method overwritten by 'GGally':</span></span>
<span><span class="co">##    method from   </span></span>
<span><span class="co">##    +.gg   ggplot2</span></span>
<span><span class="co">##  Warning: The `facets` argument of `facet_grid()` is deprecated as of ggplot2 2.2.0.</span></span>
<span><span class="co">##  ℹ Please use the `rows` argument instead.</span></span>
<span><span class="co">##  ℹ The deprecated feature was likely used in the ggbio package.</span></span>
<span><span class="co">##    Please report the issue at &lt;https://github.com/lawremi/ggbio/issues&gt;.</span></span>
<span><span class="co">##  Scale for y is already present.</span></span>
<span><span class="co">##  Adding another scale for y, which will replace the existing scale.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="data-representation_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<section id="finding-overlaps-between-two-granges-sets" class="level4" data-number="2.1.4.1"><h4 data-number="2.1.4.1" class="anchored" data-anchor-id="finding-overlaps-between-two-granges-sets">
<span class="header-section-number">2.1.4.1</span> Finding overlaps between two <code>GRanges</code> sets</h4>
<ul>
<li>Finding overlaps between a query and a subject</li>
</ul>
<p>In our case, we want to identify which ChIP-seq peaks overlap with a TSS: the query is the set of peaks and the subject is the set of TSSs.</p>
<p><code>findOverlaps</code> returns a <code>Hits</code> object listing which <code>query</code> ranges overlap with which <code>subject</code> ranges.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ov</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html">findOverlaps</a></span><span class="op">(</span>query <span class="op">=</span> <span class="va">peaks</span>, subject <span class="op">=</span> <span class="va">TSSs</span><span class="op">)</span></span>
<span><span class="va">ov</span></span>
<span><span class="co">##  Hits object with 3 hits and 0 metadata columns:</span></span>
<span><span class="co">##        queryHits subjectHits</span></span>
<span><span class="co">##        &lt;integer&gt;   &lt;integer&gt;</span></span>
<span><span class="co">##    [1]         1           1</span></span>
<span><span class="co">##    [2]         4           2</span></span>
<span><span class="co">##    [3]         5           2</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    queryLength: 8 / subjectLength: 3</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code>Hits</code> output clearly describes what overlaps with what:</p>
<ul>
<li>The query (peak) <code>#1</code> overlaps with subject (TSS) <code>#1</code>
</li>
<li>The query (peak) <code>#5</code> overlaps with subject (TSS) <code>#2</code>
</li>
</ul>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Because no other query index or subject index is listed in the <code>ov</code> output, none of the remaining ranges from <code>query</code> overlap with ranges from <code>subject</code>.</p>
</div>
</div>
<ul>
<li>Subsetting by overlaps between a query and a subject</li>
</ul>
<p>To directly <strong>subset</strong> ranges from <code>query</code> overlapping with ranges from a <code>subject</code> (e.g.&nbsp;to only keep <em>peaks</em> overlapping a <em>TSS</em>), we can use the <code>subsetByOverlaps</code> function. The output of <code>subsetByOverlaps</code> is a subset of the original <code>GRanges</code> object provided as a <code>query</code>, with retained ranges being unmodified.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html">subsetByOverlaps</a></span><span class="op">(</span><span class="va">peaks</span>, <span class="va">TSSs</span><span class="op">)</span></span>
<span><span class="co">##  GRanges object with 3 ranges and 1 metadata column:</span></span>
<span><span class="co">##        seqnames    ranges strand |        type</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</span></span>
<span><span class="co">##    [1]     chr1   320-418      * |       peaks</span></span>
<span><span class="co">##    [2]     chr1 1221-1317      * |       peaks</span></span>
<span><span class="co">##    [3]     chr1 1329-1372      * |       peaks</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Counting overlaps between a query and a subject</li>
</ul>
<p>Finally, the <code>countOverlaps</code> is used to count, for each range in a <code>query</code>, how many ranges in the <code>subject</code> it overlaps with.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html">countOverlaps</a></span><span class="op">(</span>query <span class="op">=</span> <span class="va">peaks</span>, subject <span class="op">=</span> <span class="va">TSSs</span><span class="op">)</span></span>
<span><span class="co">##  [1] 1 0 0 1 1 0 0 0</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Note that which <code>GRanges</code> goes in <code>query</code> or <code>subject</code> is crucial! Counting <strong>for each peak, the number of TSSs it overlaps with</strong> is very different from <strong>for each TSS, how many peaks it overlaps with</strong>.</p>
<p>In our case example, it would also be informative to count how many peaks overlap with each TSS, so we’d need to swap <code>query</code> and <code>subject</code>:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html">countOverlaps</a></span><span class="op">(</span>query <span class="op">=</span> <span class="va">TSSs</span>, subject <span class="op">=</span> <span class="va">peaks</span><span class="op">)</span></span>
<span><span class="co">##  [1] 1 2 0</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can add these counts to the original <code>query</code> object:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">TSSs</span><span class="op">$</span><span class="va">n_peaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html">countOverlaps</a></span><span class="op">(</span>query <span class="op">=</span> <span class="va">TSSs</span>, subject <span class="op">=</span> <span class="va">peaks</span><span class="op">)</span></span>
<span><span class="va">TSSs</span></span>
<span><span class="co">##  GRanges object with 3 ranges and 2 metadata columns:</span></span>
<span><span class="co">##        seqnames    ranges strand |        type   n_peaks</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;integer&gt;</span></span>
<span><span class="co">##    [1]     chr1   348-368      + |        TSSs         1</span></span>
<span><span class="co">##    [2]     chr1 1314-1334      + |        TSSs         2</span></span>
<span><span class="co">##    [3]     chr1 2722-2742      + |        TSSs         0</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<ul>
<li>
<code>%over%</code>, <code>%within%</code>, <code>%outside%</code> : handy operators</li>
</ul>
<p>Handy operators exist that return logical vectors (same length as the <code>query</code>). They essentially are short-hands for specific <code><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html">findOverlaps()</a></code> cases.</p>
<p><code>&lt;query&gt; %over% &lt;subject&gt;</code>:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">peaks</span> <span class="op"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html">%over%</a></span> <span class="va">TSSs</span></span>
<span><span class="co">##  [1]  TRUE FALSE FALSE  TRUE  TRUE FALSE FALSE FALSE</span></span>
<span></span>
<span><span class="va">peaks</span><span class="op">[</span><span class="va">peaks</span> <span class="op"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html">%over%</a></span> <span class="va">TSSs</span><span class="op">]</span> <span class="co"># ----- Equivalent to `subsetByOverlaps(peaks, TSSs)`</span></span>
<span><span class="co">##  GRanges object with 3 ranges and 1 metadata column:</span></span>
<span><span class="co">##        seqnames    ranges strand |        type</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</span></span>
<span><span class="co">##    [1]     chr1   320-418      * |       peaks</span></span>
<span><span class="co">##    [2]     chr1 1221-1317      * |       peaks</span></span>
<span><span class="co">##    [3]     chr1 1329-1372      * |       peaks</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><code>&lt;query&gt; %within% &lt;subject&gt;</code>:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">peaks</span> <span class="op"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html">%within%</a></span> <span class="va">TSSs</span></span>
<span><span class="co">##  [1] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span></span>
<span></span>
<span><span class="va">TSSs</span> <span class="op"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html">%within%</a></span> <span class="va">peaks</span></span>
<span><span class="co">##  [1]  TRUE FALSE FALSE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><code>&lt;query&gt; %outside% &lt;subject&gt;</code>:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">peaks</span> <span class="op"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html">%outside%</a></span> <span class="va">TSSs</span></span>
<span><span class="co">##  [1] FALSE  TRUE  TRUE FALSE FALSE  TRUE  TRUE  TRUE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Going further
</div>
</div>
<div class="callout-body-container callout-body">
<p>To fully grasp how to find overlaps between <code>GRanges</code> objects, we highly recommend reading the detailed documentation by typing <code><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html">?IRanges::`findOverlaps-methods`</a></code>.</p>
</div>
</div>
</section><section id="find-nearest-range-from-a-subject-for-each-range-in-a-query" class="level4" data-number="2.1.4.2"><h4 data-number="2.1.4.2" class="anchored" data-anchor-id="find-nearest-range-from-a-subject-for-each-range-in-a-query">
<span class="header-section-number">2.1.4.2</span> Find nearest range from a subject for each range in a query</h4>
<p><code>*Overlaps</code> methods are not always enough to match a <code>query</code> to a <code>subject</code>. For instance, some peaks in the <code>query</code> might be very near to some TSSs in the <code>subject</code>, but not quite overlapping.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">peaks</span><span class="op">[</span><span class="fl">8</span><span class="op">]</span></span>
<span><span class="co">##  GRanges object with 1 range and 1 metadata column:</span></span>
<span><span class="co">##        seqnames    ranges strand |        type</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</span></span>
<span><span class="co">##    [1]     chr1 2746-2790      * |       peaks</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span>
<span></span>
<span><span class="va">TSSs</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span></span>
<span><span class="co">##  GRanges object with 1 range and 2 metadata columns:</span></span>
<span><span class="co">##        seqnames    ranges strand |        type   n_peaks</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;integer&gt;</span></span>
<span><span class="co">##    [1]     chr1 2722-2742      + |        TSSs         0</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li><code><a href="https://rdrr.io/pkg/IRanges/man/nearest-methods.html">nearest()</a></code></li>
</ul>
<p>Rather than finding the <em>overlapping</em> range in a <code>subject</code> for each range in a <code>query</code>, we can find the <code>nearest</code> range.</p>
<p>For each range in the <code>query</code>, this returns the <em>index</em> of the range in the <code>subject</code> to which the <code>query</code> is the nearest.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/nearest-methods.html">nearest</a></span><span class="op">(</span><span class="va">peaks</span>, <span class="va">TSSs</span><span class="op">)</span></span>
<span><span class="co">##  [1] 1 1 2 2 2 2 3 3</span></span>
<span></span>
<span><span class="va">TSSs</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/nearest-methods.html">nearest</a></span><span class="op">(</span><span class="va">peaks</span>, <span class="va">TSSs</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">##  GRanges object with 8 ranges and 2 metadata columns:</span></span>
<span><span class="co">##        seqnames    ranges strand |        type   n_peaks</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;integer&gt;</span></span>
<span><span class="co">##    [1]     chr1   348-368      + |        TSSs         1</span></span>
<span><span class="co">##    [2]     chr1   348-368      + |        TSSs         1</span></span>
<span><span class="co">##    [3]     chr1 1314-1334      + |        TSSs         2</span></span>
<span><span class="co">##    [4]     chr1 1314-1334      + |        TSSs         2</span></span>
<span><span class="co">##    [5]     chr1 1314-1334      + |        TSSs         2</span></span>
<span><span class="co">##    [6]     chr1 1314-1334      + |        TSSs         2</span></span>
<span><span class="co">##    [7]     chr1 2722-2742      + |        TSSs         0</span></span>
<span><span class="co">##    [8]     chr1 2722-2742      + |        TSSs         0</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li><code><a href="https://rdrr.io/pkg/IRanges/man/nearest-methods.html">distance()</a></code></li>
</ul>
<p>Alternatively, one can simply ask to calculate the <code>distanceToNearest</code> between ranges in a <code>query</code> and ranges in a <code>subject</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/nearest-methods.html">distanceToNearest</a></span><span class="op">(</span><span class="va">peaks</span>, <span class="va">TSSs</span><span class="op">)</span></span>
<span><span class="co">##  Hits object with 8 hits and 1 metadata column:</span></span>
<span><span class="co">##        queryHits subjectHits |  distance</span></span>
<span><span class="co">##        &lt;integer&gt;   &lt;integer&gt; | &lt;integer&gt;</span></span>
<span><span class="co">##    [1]         1           1 |         0</span></span>
<span><span class="co">##    [2]         2           1 |       143</span></span>
<span><span class="co">##    [3]         3           2 |       421</span></span>
<span><span class="co">##    [4]         4           2 |         0</span></span>
<span><span class="co">##    [5]         5           2 |         0</span></span>
<span><span class="co">##    [6]         6           2 |       517</span></span>
<span><span class="co">##    [7]         7           3 |       189</span></span>
<span><span class="co">##    [8]         8           3 |         3</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    queryLength: 8 / subjectLength: 3</span></span>
<span></span>
<span><span class="va">peaks</span><span class="op">$</span><span class="va">distance_to_nearest_TSS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">mcols</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/nearest-methods.html">distanceToNearest</a></span><span class="op">(</span><span class="va">peaks</span>, <span class="va">TSSs</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">distance</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note how close from a TSS the 8th peak was. It could be worth considering this as an overlap!</p>
</section></section></section><section id="ginteractions-class" class="level2" data-number="2.2"><h2 data-number="2.2" class="anchored" data-anchor-id="ginteractions-class">
<span class="header-section-number">2.2</span> <code>GInteractions</code> class</h2>
<p><code>GRanges</code> describe genomic ranges and hence are of general use to study 1D genome organization. To study chromatin interactions, we need a way to link pairs of <code>GRanges</code>. This is exactly what the <code>GInteractions</code> class does. This data structure is defined in the <code>InteractionSet</code> package and has been published in the 2016 paper by <code>Lun et al.</code> (<span class="citation" data-cites="Lun_2016">Lun et al. (<a href="#ref-Lun_2016" role="doc-biblioref">2016</a>)</span>).</p>
<p><img src="images/20230309114047.png" class="img-fluid"></p>
<section id="building-a-ginteractions-object-from-scratch" class="level3" data-number="2.2.1"><h3 data-number="2.2.1" class="anchored" data-anchor-id="building-a-ginteractions-object-from-scratch">
<span class="header-section-number">2.2.1</span> Building a <code>GInteractions</code> object from scratch</h3>
<p>Let’s first define two parallel <code>GRanges</code> objects (i.e.&nbsp;two <code>GRanges</code> of same length). Each <code>GRanges</code> will contain 5 ranges.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gr_first</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html">GRanges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>    <span class="st">'chr1:1-100'</span>, </span>
<span>    <span class="st">'chr1:1001-2000'</span>, </span>
<span>    <span class="st">'chr1:5001-6000'</span>, </span>
<span>    <span class="st">'chr1:8001-9000'</span>, </span>
<span>    <span class="st">'chr1:7001-8000'</span>  </span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">gr_second</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html">GRanges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>    <span class="st">'chr1:1-100'</span>, </span>
<span>    <span class="st">'chr1:3001-4000'</span>, </span>
<span>    <span class="st">'chr1:8001-9000'</span>, </span>
<span>    <span class="st">'chr1:7001-8000'</span>, </span>
<span>    <span class="st">'chr2:13000-14000'</span>  </span>
<span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Because these two <code>GRanges</code> objects are of same length (5), one can “bind” them together by using the <code>GInteractions</code>function. This effectively associate each entry from one <code>GRanges</code> to the entry aligned in the other <code>GRanges</code> object.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">InteractionSet</span><span class="op">)</span></span>
<span><span class="va">gi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/GInteractions-class.html">GInteractions</a></span><span class="op">(</span><span class="va">gr_first</span>, <span class="va">gr_second</span><span class="op">)</span></span>
<span><span class="va">gi</span></span>
<span><span class="co">##  GInteractions object with 5 interactions and 0 metadata columns:</span></span>
<span><span class="co">##        seqnames1   ranges1     seqnames2     ranges2</span></span>
<span><span class="co">##            &lt;Rle&gt; &lt;IRanges&gt;         &lt;Rle&gt;   &lt;IRanges&gt;</span></span>
<span><span class="co">##    [1]      chr1     1-100 ---      chr1       1-100</span></span>
<span><span class="co">##    [2]      chr1 1001-2000 ---      chr1   3001-4000</span></span>
<span><span class="co">##    [3]      chr1 5001-6000 ---      chr1   8001-9000</span></span>
<span><span class="co">##    [4]      chr1 8001-9000 ---      chr1   7001-8000</span></span>
<span><span class="co">##    [5]      chr1 7001-8000 ---      chr2 13000-14000</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    regions: 7 ranges and 0 metadata columns</span></span>
<span><span class="co">##    seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The way <code>GInteractions</code> objects are printed in an R console mimics that of <code>GRanges</code>, but pairs two “ends” (a.k.a. <em>anchors</em>) of an interaction together, each end being represented as a separate <code>GRanges</code> range.</p>
<ul>
<li>Note that it is possible to have interactions joining two identical anchors.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gi</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="co">##  GInteractions object with 1 interaction and 0 metadata columns:</span></span>
<span><span class="co">##        seqnames1   ranges1     seqnames2   ranges2</span></span>
<span><span class="co">##            &lt;Rle&gt; &lt;IRanges&gt;         &lt;Rle&gt; &lt;IRanges&gt;</span></span>
<span><span class="co">##    [1]      chr1     1-100 ---      chr1     1-100</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    regions: 7 ranges and 0 metadata columns</span></span>
<span><span class="co">##    seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>It is also technically possible (though not advised) to have interactions for which the “first” end is located after the “second” end along the chromosome.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gi</span><span class="op">[</span><span class="fl">4</span><span class="op">]</span></span>
<span><span class="co">##  GInteractions object with 1 interaction and 0 metadata columns:</span></span>
<span><span class="co">##        seqnames1   ranges1     seqnames2   ranges2</span></span>
<span><span class="co">##            &lt;Rle&gt; &lt;IRanges&gt;         &lt;Rle&gt; &lt;IRanges&gt;</span></span>
<span><span class="co">##    [1]      chr1 8001-9000 ---      chr1 7001-8000</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    regions: 7 ranges and 0 metadata columns</span></span>
<span><span class="co">##    seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Finally, it is possible to define inter-chromosomal interactions (a.k.a. trans interactions).</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gi</span><span class="op">[</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">##  GInteractions object with 1 interaction and 0 metadata columns:</span></span>
<span><span class="co">##        seqnames1   ranges1     seqnames2     ranges2</span></span>
<span><span class="co">##            &lt;Rle&gt; &lt;IRanges&gt;         &lt;Rle&gt;   &lt;IRanges&gt;</span></span>
<span><span class="co">##    [1]      chr1 7001-8000 ---      chr2 13000-14000</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    regions: 7 ranges and 0 metadata columns</span></span>
<span><span class="co">##    seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="ginteractions-specific-slots" class="level3" data-number="2.2.2"><h3 data-number="2.2.2" class="anchored" data-anchor-id="ginteractions-specific-slots">
<span class="header-section-number">2.2.2</span> <code>GInteractions</code> specific slots</h3>
<p>Compared to <code>GRanges</code>, extra slots are available for <code>GInteractions</code> objects, e.g. <code>anchors</code> and <code>regions</code>.</p>
<section id="anchors" class="level4" data-number="2.2.2.1"><h4 data-number="2.2.2.1" class="anchored" data-anchor-id="anchors">
<span class="header-section-number">2.2.2.1</span> Anchors</h4>
<p>“Anchors” of a single genomic interaction refer to the two ends of this interaction. These anchors can be extracted from a <code>GInteractions</code> object using the <code><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">anchors()</a></code> function. This outputs a list of two <code>GRanges</code>, the first corresponding to the “left” end of interactions (when printed to the console) and the second corresponding to the “right” end of interactions (when printed to the console).</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># ----- This extracts the two sets of anchors ("first" and "second") from a GInteractions object</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">anchors</a></span><span class="op">(</span><span class="va">gi</span><span class="op">)</span></span>
<span><span class="co">##  $first</span></span>
<span><span class="co">##  GRanges object with 5 ranges and 0 metadata columns:</span></span>
<span><span class="co">##        seqnames    ranges strand</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##    [1]     chr1     1-100      *</span></span>
<span><span class="co">##    [2]     chr1 1001-2000      *</span></span>
<span><span class="co">##    [3]     chr1 5001-6000      *</span></span>
<span><span class="co">##    [4]     chr1 8001-9000      *</span></span>
<span><span class="co">##    [5]     chr1 7001-8000      *</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $second</span></span>
<span><span class="co">##  GRanges object with 5 ranges and 0 metadata columns:</span></span>
<span><span class="co">##        seqnames      ranges strand</span></span>
<span><span class="co">##           &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##    [1]     chr1       1-100      *</span></span>
<span><span class="co">##    [2]     chr1   3001-4000      *</span></span>
<span><span class="co">##    [3]     chr1   8001-9000      *</span></span>
<span><span class="co">##    [4]     chr1   7001-8000      *</span></span>
<span><span class="co">##    [5]     chr2 13000-14000      *</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span>
<span></span>
<span><span class="co"># ----- We can query for the "first" or "second" set of anchors directly</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">anchors</a></span><span class="op">(</span><span class="va">gi</span>, <span class="st">"first"</span><span class="op">)</span></span>
<span><span class="co">##  GRanges object with 5 ranges and 0 metadata columns:</span></span>
<span><span class="co">##        seqnames    ranges strand</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##    [1]     chr1     1-100      *</span></span>
<span><span class="co">##    [2]     chr1 1001-2000      *</span></span>
<span><span class="co">##    [3]     chr1 5001-6000      *</span></span>
<span><span class="co">##    [4]     chr1 8001-9000      *</span></span>
<span><span class="co">##    [5]     chr1 7001-8000      *</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">anchors</a></span><span class="op">(</span><span class="va">gi</span>, <span class="st">"second"</span><span class="op">)</span></span>
<span><span class="co">##  GRanges object with 5 ranges and 0 metadata columns:</span></span>
<span><span class="co">##        seqnames      ranges strand</span></span>
<span><span class="co">##           &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##    [1]     chr1       1-100      *</span></span>
<span><span class="co">##    [2]     chr1   3001-4000      *</span></span>
<span><span class="co">##    [3]     chr1   8001-9000      *</span></span>
<span><span class="co">##    [4]     chr1   7001-8000      *</span></span>
<span><span class="co">##    [5]     chr2 13000-14000      *</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="regions" class="level4" data-number="2.2.2.2"><h4 data-number="2.2.2.2" class="anchored" data-anchor-id="regions">
<span class="header-section-number">2.2.2.2</span> Regions</h4>
<p>“Regions” of a <em>set</em> of interactions refer to the universe of unique anchors represented in a set of interactions. Therefore, the length of the <code>regions</code> can only be equal to or strictly lower than twice the length of <code>anchors</code>.</p>
<p>The <code>regions</code> function returns the regions associated with a <code>GInteractions</code> object, stored as a <code>GRanges</code> object.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">regions</a></span><span class="op">(</span><span class="va">gi</span><span class="op">)</span></span>
<span><span class="co">##  GRanges object with 7 ranges and 0 metadata columns:</span></span>
<span><span class="co">##        seqnames      ranges strand</span></span>
<span><span class="co">##           &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##    [1]     chr1       1-100      *</span></span>
<span><span class="co">##    [2]     chr1   1001-2000      *</span></span>
<span><span class="co">##    [3]     chr1   3001-4000      *</span></span>
<span><span class="co">##    [4]     chr1   5001-6000      *</span></span>
<span><span class="co">##    [5]     chr1   7001-8000      *</span></span>
<span><span class="co">##    [6]     chr1   8001-9000      *</span></span>
<span><span class="co">##    [7]     chr2 13000-14000      *</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">regions</a></span><span class="op">(</span><span class="va">gi</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  [1] 7</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">anchors</a></span><span class="op">(</span><span class="va">gi</span>, <span class="st">"first"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  [1] 5</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section><section id="ginteractions-methods" class="level3" data-number="2.2.3"><h3 data-number="2.2.3" class="anchored" data-anchor-id="ginteractions-methods">
<span class="header-section-number">2.2.3</span> <code>GInteractions</code> methods</h3>
<p><code>GInteractions</code> behave as an extension of <code>GRanges</code>. For this reason, many methods that work with <code>GRanges</code> will work seamlessly with <code>GInteractions</code>.</p>
<section id="metadata" class="level4" data-number="2.2.3.1"><h4 data-number="2.2.3.1" class="anchored" data-anchor-id="metadata">
<span class="header-section-number">2.2.3.1</span> Metadata</h4>
<p>One can add metadata columns directly to a <code>GInteractions</code> object.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">mcols</a></span><span class="op">(</span><span class="va">gi</span><span class="op">)</span></span>
<span><span class="co">##  DataFrame with 5 rows and 0 columns</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">mcols</a></span><span class="op">(</span><span class="va">gi</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span></span>
<span>    idx <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">gi</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"cis"</span>, <span class="st">"cis"</span>, <span class="st">"cis"</span>, <span class="st">"trans"</span>, <span class="st">"cis"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">gi</span></span>
<span><span class="co">##  GInteractions object with 5 interactions and 2 metadata columns:</span></span>
<span><span class="co">##        seqnames1   ranges1     seqnames2     ranges2 |       idx        type</span></span>
<span><span class="co">##            &lt;Rle&gt; &lt;IRanges&gt;         &lt;Rle&gt;   &lt;IRanges&gt; | &lt;integer&gt; &lt;character&gt;</span></span>
<span><span class="co">##    [1]      chr1     1-100 ---      chr1       1-100 |         1         cis</span></span>
<span><span class="co">##    [2]      chr1 1001-2000 ---      chr1   3001-4000 |         2         cis</span></span>
<span><span class="co">##    [3]      chr1 5001-6000 ---      chr1   8001-9000 |         3         cis</span></span>
<span><span class="co">##    [4]      chr1 8001-9000 ---      chr1   7001-8000 |         4       trans</span></span>
<span><span class="co">##    [5]      chr1 7001-8000 ---      chr2 13000-14000 |         5         cis</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    regions: 7 ranges and 0 metadata columns</span></span>
<span><span class="co">##    seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span>
<span></span>
<span><span class="va">gi</span><span class="op">$</span><span class="va">type</span></span>
<span><span class="co">##  [1] "cis"   "cis"   "cis"   "trans" "cis"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Importantly, metadata columns can also be directly added to <strong>regions</strong> of a <code>GInteractions</code> object, since these <code>regions</code> are a <code>GRanges</code> object themselves!</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">regions</a></span><span class="op">(</span><span class="va">gi</span><span class="op">)</span></span>
<span><span class="co">##  GRanges object with 7 ranges and 0 metadata columns:</span></span>
<span><span class="co">##        seqnames      ranges strand</span></span>
<span><span class="co">##           &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##    [1]     chr1       1-100      *</span></span>
<span><span class="co">##    [2]     chr1   1001-2000      *</span></span>
<span><span class="co">##    [3]     chr1   3001-4000      *</span></span>
<span><span class="co">##    [4]     chr1   5001-6000      *</span></span>
<span><span class="co">##    [5]     chr1   7001-8000      *</span></span>
<span><span class="co">##    [6]     chr1   8001-9000      *</span></span>
<span><span class="co">##    [7]     chr2 13000-14000      *</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">regions</a></span><span class="op">(</span><span class="va">gi</span><span class="op">)</span><span class="op">$</span><span class="va">binID</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_along</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">regions</a></span><span class="op">(</span><span class="va">gi</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">regions</a></span><span class="op">(</span><span class="va">gi</span><span class="op">)</span><span class="op">$</span><span class="va">type</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"P"</span>, <span class="st">"P"</span>, <span class="st">"P"</span>, <span class="st">"E"</span>, <span class="st">"E"</span>, <span class="st">"P"</span>, <span class="st">"P"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">regions</a></span><span class="op">(</span><span class="va">gi</span><span class="op">)</span></span>
<span><span class="co">##  GRanges object with 7 ranges and 2 metadata columns:</span></span>
<span><span class="co">##        seqnames      ranges strand |     binID        type</span></span>
<span><span class="co">##           &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;</span></span>
<span><span class="co">##    [1]     chr1       1-100      * |         1           P</span></span>
<span><span class="co">##    [2]     chr1   1001-2000      * |         2           P</span></span>
<span><span class="co">##    [3]     chr1   3001-4000      * |         3           P</span></span>
<span><span class="co">##    [4]     chr1   5001-6000      * |         4           E</span></span>
<span><span class="co">##    [5]     chr1   7001-8000      * |         5           E</span></span>
<span><span class="co">##    [6]     chr1   8001-9000      * |         6           P</span></span>
<span><span class="co">##    [7]     chr2 13000-14000      * |         7           P</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="sorting-ginteractions" class="level4" data-number="2.2.3.2"><h4 data-number="2.2.3.2" class="anchored" data-anchor-id="sorting-ginteractions">
<span class="header-section-number">2.2.3.2</span> Sorting <code>GInteractions</code>
</h4>
<p>The <code>sort</code> function works seamlessly with <code>GInteractions</code> objects. It sorts the interactions using a similar approach to that performed by <code>pairtools sort ...</code> for disk-stored <code>.pairs</code> files, sorting on the “first” anchor first, then for interactions with the same “first” anchors, sorting on the “second” anchor.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gi</span></span>
<span><span class="co">##  GInteractions object with 5 interactions and 2 metadata columns:</span></span>
<span><span class="co">##        seqnames1   ranges1     seqnames2     ranges2 |       idx        type</span></span>
<span><span class="co">##            &lt;Rle&gt; &lt;IRanges&gt;         &lt;Rle&gt;   &lt;IRanges&gt; | &lt;integer&gt; &lt;character&gt;</span></span>
<span><span class="co">##    [1]      chr1     1-100 ---      chr1       1-100 |         1         cis</span></span>
<span><span class="co">##    [2]      chr1 1001-2000 ---      chr1   3001-4000 |         2         cis</span></span>
<span><span class="co">##    [3]      chr1 5001-6000 ---      chr1   8001-9000 |         3         cis</span></span>
<span><span class="co">##    [4]      chr1 8001-9000 ---      chr1   7001-8000 |         4       trans</span></span>
<span><span class="co">##    [5]      chr1 7001-8000 ---      chr2 13000-14000 |         5         cis</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    regions: 7 ranges and 2 metadata columns</span></span>
<span><span class="co">##    seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html">sort</a></span><span class="op">(</span><span class="va">gi</span><span class="op">)</span></span>
<span><span class="co">##  GInteractions object with 5 interactions and 2 metadata columns:</span></span>
<span><span class="co">##        seqnames1   ranges1     seqnames2     ranges2 |       idx        type</span></span>
<span><span class="co">##            &lt;Rle&gt; &lt;IRanges&gt;         &lt;Rle&gt;   &lt;IRanges&gt; | &lt;integer&gt; &lt;character&gt;</span></span>
<span><span class="co">##    [1]      chr1     1-100 ---      chr1       1-100 |         1         cis</span></span>
<span><span class="co">##    [2]      chr1 1001-2000 ---      chr1   3001-4000 |         2         cis</span></span>
<span><span class="co">##    [3]      chr1 5001-6000 ---      chr1   8001-9000 |         3         cis</span></span>
<span><span class="co">##    [4]      chr1 7001-8000 ---      chr2 13000-14000 |         5         cis</span></span>
<span><span class="co">##    [5]      chr1 8001-9000 ---      chr1   7001-8000 |         4       trans</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    regions: 7 ranges and 2 metadata columns</span></span>
<span><span class="co">##    seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="swapping-ginteractions-anchors" class="level4" data-number="2.2.3.3"><h4 data-number="2.2.3.3" class="anchored" data-anchor-id="swapping-ginteractions-anchors">
<span class="header-section-number">2.2.3.3</span> Swapping <code>GInteractions</code> anchors</h4>
<p>For an individual interaction contained in a <code>GInteractions</code> object, the “first” and “second” anchors themselves can be sorted as well. This is called “pairs swapping”, and it is performed similarly to <code>pairtools flip ...</code> for disk-stored <code>.pairs</code> files. This ensures that interactions, when represented as a contact matrix, generate an upper-triangular matrix.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gi</span></span>
<span><span class="co">##  GInteractions object with 5 interactions and 2 metadata columns:</span></span>
<span><span class="co">##        seqnames1   ranges1     seqnames2     ranges2 |       idx        type</span></span>
<span><span class="co">##            &lt;Rle&gt; &lt;IRanges&gt;         &lt;Rle&gt;   &lt;IRanges&gt; | &lt;integer&gt; &lt;character&gt;</span></span>
<span><span class="co">##    [1]      chr1     1-100 ---      chr1       1-100 |         1         cis</span></span>
<span><span class="co">##    [2]      chr1 1001-2000 ---      chr1   3001-4000 |         2         cis</span></span>
<span><span class="co">##    [3]      chr1 5001-6000 ---      chr1   8001-9000 |         3         cis</span></span>
<span><span class="co">##    [4]      chr1 8001-9000 ---      chr1   7001-8000 |         4       trans</span></span>
<span><span class="co">##    [5]      chr1 7001-8000 ---      chr2 13000-14000 |         5         cis</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    regions: 7 ranges and 2 metadata columns</span></span>
<span><span class="co">##    seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-compare.html">swapAnchors</a></span><span class="op">(</span><span class="va">gi</span><span class="op">)</span></span>
<span><span class="co">##  GInteractions object with 5 interactions and 2 metadata columns:</span></span>
<span><span class="co">##        seqnames1   ranges1     seqnames2     ranges2 |       idx        type</span></span>
<span><span class="co">##            &lt;Rle&gt; &lt;IRanges&gt;         &lt;Rle&gt;   &lt;IRanges&gt; | &lt;integer&gt; &lt;character&gt;</span></span>
<span><span class="co">##    [1]      chr1     1-100 ---      chr1       1-100 |         1         cis</span></span>
<span><span class="co">##    [2]      chr1 1001-2000 ---      chr1   3001-4000 |         2         cis</span></span>
<span><span class="co">##    [3]      chr1 5001-6000 ---      chr1   8001-9000 |         3         cis</span></span>
<span><span class="co">##    [4]      chr1 7001-8000 ---      chr1   8001-9000 |         4       trans</span></span>
<span><span class="co">##    [5]      chr1 7001-8000 ---      chr2 13000-14000 |         5         cis</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    regions: 7 ranges and 2 metadata columns</span></span>
<span><span class="co">##    seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>“Sorting” and “swapping” a <code>GInteractions</code> object are two <strong>entirely different actions</strong>:</p>
<ul>
<li>“sorting” reorganizes all <em>rows</em> (interactions);</li>
<li>“swapping” <em>anchors</em> reorganizes <em>“first” and “second” anchors</em> for each interaction independently.</li>
</ul>
</div>
</div>
</section><section id="ginteractions-distance-method" class="level4" data-number="2.2.3.4"><h4 data-number="2.2.3.4" class="anchored" data-anchor-id="ginteractions-distance-method">
<span class="header-section-number">2.2.3.4</span> <code>GInteractions</code> distance method</h4>
<p>“Distance”, when applied to genomic interactions, typically refers to the genomic distance between the two anchors of a single interaction. For <code>GInteractions</code>, this is computed using the <code>pairdist</code> function.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gi</span></span>
<span><span class="co">##  GInteractions object with 5 interactions and 2 metadata columns:</span></span>
<span><span class="co">##        seqnames1   ranges1     seqnames2     ranges2 |       idx        type</span></span>
<span><span class="co">##            &lt;Rle&gt; &lt;IRanges&gt;         &lt;Rle&gt;   &lt;IRanges&gt; | &lt;integer&gt; &lt;character&gt;</span></span>
<span><span class="co">##    [1]      chr1     1-100 ---      chr1       1-100 |         1         cis</span></span>
<span><span class="co">##    [2]      chr1 1001-2000 ---      chr1   3001-4000 |         2         cis</span></span>
<span><span class="co">##    [3]      chr1 5001-6000 ---      chr1   8001-9000 |         3         cis</span></span>
<span><span class="co">##    [4]      chr1 8001-9000 ---      chr1   7001-8000 |         4       trans</span></span>
<span><span class="co">##    [5]      chr1 7001-8000 ---      chr2 13000-14000 |         5         cis</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    regions: 7 ranges and 2 metadata columns</span></span>
<span><span class="co">##    seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/distances.html">pairdist</a></span><span class="op">(</span><span class="va">gi</span><span class="op">)</span></span>
<span><span class="co">##  [1]    0 2000 3000 1000   NA</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note that for “trans” inter-chromosomal interactions, i.e.&nbsp;interactions with anchors on different chromosomes, the notion of genomic distance is meaningless and for this reason, <code>pairdist</code> returns a <code>NA</code> value.</p>
<p>The <code>type</code> argument of the <code><a href="https://rdrr.io/pkg/InteractionSet/man/distances.html">pairdist()</a></code> function can be tweaked to specify which type of “distance” should be computed:</p>
<ul>
<li>
<code>mid</code>: The distance between the midpoints of the two regions (rounded down to the nearest integer) is returned (Default).</li>
<li>
<code>gap</code>: The length of the gap between the closest points of the two regions is computed - negative lengths are returned for overlapping regions, indicating the length of the overlap.</li>
<li>
<code>span</code>: The distance between the furthermost points of the two regions is computed.</li>
<li>
<code>diag</code>: The difference between the anchor indices is returned. This corresponds to a diagonal on the interaction space when bins are used in the ‘regions’ slot of ‘x’.</li>
</ul></section><section id="ginteractions-overlap-methods" class="level4" data-number="2.2.3.5"><h4 data-number="2.2.3.5" class="anchored" data-anchor-id="ginteractions-overlap-methods">
<span class="header-section-number">2.2.3.5</span> <code>GInteractions</code> overlap methods</h4>
<p>“Overlaps” for genomic interactions could be computed in different contexts:</p>
<ul>
<li>Case 1: Overlap between any of the two anchors of an interaction with a genomic range</li>
<li>Case 2: Overlap between anchors of an interaction with anchors of another interaction</li>
<li>Case 3: Spanning of the interaction “across” a genomic range</li>
</ul>
<blockquote class="blockquote">
<p>Case 1: Overlap between any of the two anchors of an interaction with a genomic range</p>
</blockquote>
<p>This is the default behavior of <code>findOverlaps</code> when providing a <code>GInteractions</code> object as <code>query</code> and a <code>GRanges</code> as a <code>subject</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html">GRanges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"chr1:7501-7600"</span>, <span class="st">"chr1:8501-8600"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html">findOverlaps</a></span><span class="op">(</span>query <span class="op">=</span> <span class="va">gi</span>, subject <span class="op">=</span> <span class="va">gr</span><span class="op">)</span></span>
<span><span class="co">##  Hits object with 4 hits and 0 metadata columns:</span></span>
<span><span class="co">##        queryHits subjectHits</span></span>
<span><span class="co">##        &lt;integer&gt;   &lt;integer&gt;</span></span>
<span><span class="co">##    [1]         3           2</span></span>
<span><span class="co">##    [2]         4           1</span></span>
<span><span class="co">##    [3]         4           2</span></span>
<span><span class="co">##    [4]         5           1</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    queryLength: 5 / subjectLength: 2</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html">countOverlaps</a></span><span class="op">(</span><span class="va">gi</span>, <span class="va">gr</span><span class="op">)</span></span>
<span><span class="co">##  [1] 0 0 1 2 1</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html">subsetByOverlaps</a></span><span class="op">(</span><span class="va">gi</span>, <span class="va">gr</span><span class="op">)</span></span>
<span><span class="co">##  GInteractions object with 3 interactions and 2 metadata columns:</span></span>
<span><span class="co">##        seqnames1   ranges1     seqnames2     ranges2 |       idx        type</span></span>
<span><span class="co">##            &lt;Rle&gt; &lt;IRanges&gt;         &lt;Rle&gt;   &lt;IRanges&gt; | &lt;integer&gt; &lt;character&gt;</span></span>
<span><span class="co">##    [1]      chr1 5001-6000 ---      chr1   8001-9000 |         3         cis</span></span>
<span><span class="co">##    [2]      chr1 8001-9000 ---      chr1   7001-8000 |         4       trans</span></span>
<span><span class="co">##    [3]      chr1 7001-8000 ---      chr2 13000-14000 |         5         cis</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    regions: 7 ranges and 2 metadata columns</span></span>
<span><span class="co">##    seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here again, the order matters!</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html">countOverlaps</a></span><span class="op">(</span><span class="va">gr</span>, <span class="va">gi</span><span class="op">)</span></span>
<span><span class="co">##  [1] 2 2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>And again, the <code>%over%</code> operator can be used here:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gi</span> <span class="op"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html">%over%</a></span> <span class="va">gr</span></span>
<span><span class="co">##  [1] FALSE FALSE  TRUE  TRUE  TRUE</span></span>
<span></span>
<span><span class="va">gi</span><span class="op">[</span><span class="va">gi</span> <span class="op"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html">%over%</a></span> <span class="va">gr</span><span class="op">]</span> <span class="co"># ----- Equivalent to `subsetByOverlaps(gi, gr)`</span></span>
<span><span class="co">##  GInteractions object with 3 interactions and 2 metadata columns:</span></span>
<span><span class="co">##        seqnames1   ranges1     seqnames2     ranges2 |       idx        type</span></span>
<span><span class="co">##            &lt;Rle&gt; &lt;IRanges&gt;         &lt;Rle&gt;   &lt;IRanges&gt; | &lt;integer&gt; &lt;character&gt;</span></span>
<span><span class="co">##    [1]      chr1 5001-6000 ---      chr1   8001-9000 |         3         cis</span></span>
<span><span class="co">##    [2]      chr1 8001-9000 ---      chr1   7001-8000 |         4       trans</span></span>
<span><span class="co">##    [3]      chr1 7001-8000 ---      chr2 13000-14000 |         5         cis</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    regions: 7 ranges and 2 metadata columns</span></span>
<span><span class="co">##    seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<blockquote class="blockquote">
<p>Case 2: Overlap between anchors of an interaction with anchors of another interaction</p>
</blockquote>
<p>This slightly different scenario involves overlapping two sets of interactions, to see whether any interaction in <code>Set-1</code> has its two anchors overlapping anchors from an interaction in <code>Set-2</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gi2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/GInteractions-class.html">GInteractions</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html">GRanges</a></span><span class="op">(</span><span class="st">"chr1:1081-1090"</span><span class="op">)</span>, </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html">GRanges</a></span><span class="op">(</span><span class="st">"chr1:3401-3501"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">gi</span> <span class="op"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html">%over%</a></span> <span class="va">gi2</span></span>
<span><span class="co">##  [1] FALSE  TRUE FALSE FALSE FALSE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note that both anchors of an interaction from a <code>query</code> have to overlap to a pair of anchors of a single interaction from a <code>subject</code> with this method!</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gi3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/GInteractions-class.html">GInteractions</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html">GRanges</a></span><span class="op">(</span><span class="st">"chr1:1-1000"</span><span class="op">)</span>, </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html">GRanges</a></span><span class="op">(</span><span class="st">"chr1:3401-3501"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">gi</span> <span class="op"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html">%over%</a></span> <span class="va">gi3</span></span>
<span><span class="co">##  [1] FALSE FALSE FALSE FALSE FALSE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<blockquote class="blockquote">
<p>Case 3 : Spanning of the interaction “accross” a genomic range</p>
</blockquote>
<p>This requires a bit of wrangling, to mimic an overlap between two <code>GRanges</code> objects:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-compare.html">swapAnchors</a></span><span class="op">(</span><span class="va">gi</span><span class="op">)</span> <span class="co"># ----- Make sure anchors are correctly sorted</span></span>
<span><span class="va">gi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html">sort</a></span><span class="op">(</span><span class="va">gi</span><span class="op">)</span> <span class="co"># ----- Make sure interactions are correctly sorted</span></span>
<span><span class="va">gi</span> <span class="op">&lt;-</span> <span class="va">gi</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/distances.html">pairdist</a></span><span class="op">(</span><span class="va">gi</span><span class="op">)</span><span class="op">)</span><span class="op">]</span> <span class="co"># ----- Remove inter-chromosomal interactions</span></span>
<span><span class="va">spanning_gi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html">GRanges</a></span><span class="op">(</span></span>
<span>    seqnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html">seqnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">anchors</a></span><span class="op">(</span><span class="va">gi</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, </span>
<span>    ranges <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html">IRanges</a></span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/stats/start.html">start</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">anchors</a></span><span class="op">(</span><span class="va">gi</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, </span>
<span>        <span class="fu"><a href="https://rdrr.io/r/stats/start.html">end</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">anchors</a></span><span class="op">(</span><span class="va">gi</span><span class="op">)</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">spanning_gi</span> </span>
<span><span class="co">##  GRanges object with 4 ranges and 0 metadata columns:</span></span>
<span><span class="co">##        seqnames    ranges strand</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##    [1]     chr1     1-100      *</span></span>
<span><span class="co">##    [2]     chr1 1001-4000      *</span></span>
<span><span class="co">##    [3]     chr1 5001-9000      *</span></span>
<span><span class="co">##    [4]     chr1 7001-9000      *</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></span>
<span></span>
<span><span class="va">spanning_gi</span> <span class="op"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html">%over%</a></span> <span class="va">gr</span></span>
<span><span class="co">##  [1] FALSE FALSE  TRUE  TRUE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Going further
</div>
</div>
<div class="callout-body-container callout-body">
<p>A detailed manual of overlap methods available for <code>GInteractions</code> object can be read by typing <code><a href="https://rdrr.io/pkg/InteractionSet/man/overlaps.html">?`Interaction-overlaps`</a></code> in R.</p>
</div>
</div>
</section></section></section><section id="contactfile-class" class="level2" data-number="2.3"><h2 data-number="2.3" class="anchored" data-anchor-id="contactfile-class">
<span class="header-section-number">2.3</span> <code>ContactFile</code> class</h2>
<p>Hi-C contacts can be stored in four different formats (see <a href="../pages/principles.html#binned-contact-matrix-files">previous chapter</a>):</p>
<ul>
<li>As a <code>.(m)cool</code> matrix (multi-scores, multi-resolution, indexed)</li>
<li>As a <code>.hic</code> matrix (multi-scores, multi-resolution, indexed)</li>
<li>As a HiC-pro derived matrix (single-score, single-resolution, non-indexed)</li>
<li>Unbinned, Hi-C contacts can be stored in <code>.pairs</code> files</li>
</ul>
<section id="accessing-example-hi-c-files" class="level3" data-number="2.3.1"><h3 data-number="2.3.1" class="anchored" data-anchor-id="accessing-example-hi-c-files">
<span class="header-section-number">2.3.1</span> Accessing example Hi-C files</h3>
<p>Example contact files can be downloaded using <code>HiContactsData</code> function.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/js2264/HiContactsData">HiContactsData</a></span><span class="op">)</span></span>
<span><span class="va">coolf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/HiContactsData/man/HiContactsData.html">HiContactsData</a></span><span class="op">(</span><span class="st">'yeast_wt'</span>, <span class="st">'mcool'</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This fetches files from the cloud, download them locally and returns the path of the local file.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">coolf</span></span>
<span><span class="co">##                                          EH7702 </span></span>
<span><span class="co">##  "/root/.cache/R/ExperimentHub/16dd6c5d5f_7752"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Similarly, example files are available for other file formats:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hicf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/HiContactsData/man/HiContactsData.html">HiContactsData</a></span><span class="op">(</span><span class="st">'yeast_wt'</span>, <span class="st">'hic'</span><span class="op">)</span></span>
<span><span class="va">hicpromatrixf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/HiContactsData/man/HiContactsData.html">HiContactsData</a></span><span class="op">(</span><span class="st">'yeast_wt'</span>, <span class="st">'hicpro_matrix'</span><span class="op">)</span></span>
<span><span class="va">hicproregionsf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/HiContactsData/man/HiContactsData.html">HiContactsData</a></span><span class="op">(</span><span class="st">'yeast_wt'</span>, <span class="st">'hicpro_bed'</span><span class="op">)</span></span>
<span><span class="va">pairsf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/HiContactsData/man/HiContactsData.html">HiContactsData</a></span><span class="op">(</span><span class="st">'yeast_wt'</span>, <span class="st">'pairs.gz'</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can even check the content of some of these files to make sure they are actually what they are:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># ---- HiC-Pro generates a tab-separated `regions.bed` file</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readLines.html">readLines</a></span><span class="op">(</span><span class="va">hicproregionsf</span>, <span class="fl">25</span><span class="op">)</span></span>
<span><span class="co">##   [1] "I\t0\t1000"      "I\t1000\t2000"   "I\t2000\t3000"   "I\t3000\t4000"  </span></span>
<span><span class="co">##   [5] "I\t4000\t5000"   "I\t5000\t6000"   "I\t6000\t7000"   "I\t7000\t8000"  </span></span>
<span><span class="co">##   [9] "I\t8000\t9000"   "I\t9000\t10000"  "I\t10000\t11000" "I\t11000\t12000"</span></span>
<span><span class="co">##  [13] "I\t12000\t13000" "I\t13000\t14000" "I\t14000\t15000" "I\t15000\t16000"</span></span>
<span><span class="co">##  [17] "I\t16000\t17000" "I\t17000\t18000" "I\t18000\t19000" "I\t19000\t20000"</span></span>
<span><span class="co">##  [21] "I\t20000\t21000" "I\t21000\t22000" "I\t22000\t23000" "I\t23000\t24000"</span></span>
<span><span class="co">##  [25] "I\t24000\t25000"</span></span>
<span></span>
<span><span class="co"># ---- Pairs are also tab-separated </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readLines.html">readLines</a></span><span class="op">(</span><span class="va">pairsf</span>, <span class="fl">25</span><span class="op">)</span></span>
<span><span class="co">##   [1] "## pairs format v1.0"                                                             </span></span>
<span><span class="co">##   [2] "#sorted: chr1-pos1-chr2-pos2"                                                     </span></span>
<span><span class="co">##   [3] "#columns: readID chr1 pos1 chr2 pos2 strand1 strand2 frag1 frag2"                 </span></span>
<span><span class="co">##   [4] "#chromsize: I 230218"                                                             </span></span>
<span><span class="co">##   [5] "#chromsize: II 813184"                                                            </span></span>
<span><span class="co">##   [6] "#chromsize: III 316620"                                                           </span></span>
<span><span class="co">##   [7] "#chromsize: IV 1531933"                                                           </span></span>
<span><span class="co">##   [8] "#chromsize: V 576874"                                                             </span></span>
<span><span class="co">##   [9] "#chromsize: VI 270161"                                                            </span></span>
<span><span class="co">##  [10] "#chromsize: VII 1090940"                                                          </span></span>
<span><span class="co">##  [11] "#chromsize: VIII 562643"                                                          </span></span>
<span><span class="co">##  [12] "#chromsize: IX 439888"                                                            </span></span>
<span><span class="co">##  [13] "#chromsize: X 745751"                                                             </span></span>
<span><span class="co">##  [14] "#chromsize: XI 666816"                                                            </span></span>
<span><span class="co">##  [15] "#chromsize: XII 1078177"                                                          </span></span>
<span><span class="co">##  [16] "#chromsize: XIII 924431"                                                          </span></span>
<span><span class="co">##  [17] "#chromsize: XIV 784333"                                                           </span></span>
<span><span class="co">##  [18] "#chromsize: XV 1091291"                                                           </span></span>
<span><span class="co">##  [19] "#chromsize: XVI 948066"                                                           </span></span>
<span><span class="co">##  [20] "#chromsize: Mito 85779"                                                           </span></span>
<span><span class="co">##  [21] "NS500150:527:HHGYNBGXF:3:21611:19085:3986\tII\t105\tII\t48548\t+\t-\t1358\t1681"  </span></span>
<span><span class="co">##  [22] "NS500150:527:HHGYNBGXF:4:13604:19734:2406\tII\t113\tII\t45003\t-\t+\t1358\t1658"  </span></span>
<span><span class="co">##  [23] "NS500150:527:HHGYNBGXF:2:11108:25178:11036\tII\t119\tII\t687251\t-\t+\t1358\t5550"</span></span>
<span><span class="co">##  [24] "NS500150:527:HHGYNBGXF:1:22301:8468:1586\tII\t160\tII\t26124\t+\t-\t1358\t1510"   </span></span>
<span><span class="co">##  [25] "NS500150:527:HHGYNBGXF:4:23606:24037:2076\tII\t169\tII\t39052\t+\t+\t1358\t1613"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="contactfile-fundamentals" class="level3" data-number="2.3.2"><h3 data-number="2.3.2" class="anchored" data-anchor-id="contactfile-fundamentals">
<span class="header-section-number">2.3.2</span> <code>ContactFile</code> fundamentals</h3>
<p>A <code>ContactFile</code> object <strong>establishes a connection with a disk-stored Hi-C file</strong> (e.g.&nbsp;a <code>.cool</code> file, or a <code>.pairs</code> file, …). <code>ContactFile</code> classes are defined in the <code>HiCExperiment</code> package.</p>
<p><code>ContactFile</code>s come in four different flavors:</p>
<ul>
<li>
<code>CoolFile</code>: connection to a <code>.(m)cool</code> file</li>
<li>
<code>HicFile</code>: connection to a <code>.hic</code> file</li>
<li>
<code>HicproFile</code>: connection to output files generated by HiC-Pro</li>
<li>
<code>PairsFile</code>: connection to a <code>.pairs</code> file</li>
</ul>
<p>To create each flavor of <code>ContactFile</code>, one can use the corresponding function:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/js2264/HiCExperiment">HiCExperiment</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># ----- This creates a connection to a `.(m)cool` file (path stored in `coolf`)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/CoolFile-class.html">CoolFile</a></span><span class="op">(</span><span class="va">coolf</span><span class="op">)</span></span>
<span><span class="co">##  CoolFile object</span></span>
<span><span class="co">##  .mcool file: /root/.cache/R/ExperimentHub/16dd6c5d5f_7752 </span></span>
<span><span class="co">##  resolution: 1000 </span></span>
<span><span class="co">##  pairs file: </span></span>
<span><span class="co">##  metadata(0):</span></span>
<span></span>
<span><span class="co"># ----- This creates a connection to a `.hic` file (path stored in `hicf`)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/HicFile-class.html">HicFile</a></span><span class="op">(</span><span class="va">hicf</span><span class="op">)</span></span>
<span><span class="co">##  HicFile object</span></span>
<span><span class="co">##  .hic file: /root/.cache/R/ExperimentHub/16d160e6aeb_7836 </span></span>
<span><span class="co">##  resolution: 1000 </span></span>
<span><span class="co">##  pairs file: </span></span>
<span><span class="co">##  metadata(0):</span></span>
<span></span>
<span><span class="co"># ----- This creates a connection to output files from HiC-Pro</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/HicproFile-class.html">HicproFile</a></span><span class="op">(</span><span class="va">hicpromatrixf</span>, <span class="va">hicproregionsf</span><span class="op">)</span></span>
<span><span class="co">##  HicproFile object</span></span>
<span><span class="co">##  HiC-Pro files:</span></span>
<span><span class="co">##    $ matrix:   /root/.cache/R/ExperimentHub/16dfa774ff_7837 </span></span>
<span><span class="co">##    $ regions:  /root/.cache/R/ExperimentHub/16d75113d35_7838 </span></span>
<span><span class="co">##  resolution: 1000 </span></span>
<span><span class="co">##  pairs file: </span></span>
<span><span class="co">##  metadata(0):</span></span>
<span></span>
<span><span class="co"># ----- This creates a connection to a pairs file</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/PairsFile-class.html">PairsFile</a></span><span class="op">(</span><span class="va">pairsf</span><span class="op">)</span></span>
<span><span class="co">##  PairsFile object</span></span>
<span><span class="co">##  resource: /root/.cache/R/ExperimentHub/16d2ae89407_7753</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="contactfile-slots" class="level3" data-number="2.3.3"><h3 data-number="2.3.3" class="anchored" data-anchor-id="contactfile-slots">
<span class="header-section-number">2.3.3</span> <code>ContactFile</code> slots</h3>
<p>Several “slots” (i.e.&nbsp;pieces of information) are attached to a <code>ContactFile</code> object:</p>
<ul>
<li>The path to the disk-stored contact matrix;</li>
<li>The active resolution (by default, the finest resolution available in a multi-resolution contact matrix);</li>
<li>Optionally, the path to a matching <code>pairs</code> file (see below);</li>
<li>Some metadata.</li>
</ul>
<p>Slots of a <code>CoolFile</code> object can be accessed as follow:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/CoolFile-class.html">CoolFile</a></span><span class="op">(</span><span class="va">coolf</span><span class="op">)</span></span>
<span><span class="va">cf</span></span>
<span><span class="co">##  CoolFile object</span></span>
<span><span class="co">##  .mcool file: /root/.cache/R/ExperimentHub/16dd6c5d5f_7752 </span></span>
<span><span class="co">##  resolution: 1000 </span></span>
<span><span class="co">##  pairs file: </span></span>
<span><span class="co">##  metadata(0):</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">resolution</a></span><span class="op">(</span><span class="va">cf</span><span class="op">)</span></span>
<span><span class="co">##  [1] 1000</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">pairsFile</a></span><span class="op">(</span><span class="va">cf</span><span class="op">)</span></span>
<span><span class="co">##  NULL</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html">metadata</a></span><span class="op">(</span><span class="va">cf</span><span class="op">)</span></span>
<span><span class="co">##  list()</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important!
</div>
</div>
<div class="callout-body-container callout-body">
<p><code>ContactFile</code> objects are only <strong>connections</strong> to a disk-stored HiC file. Although metadata is available, they do not contain actual data!</p>
</div>
</div>
</section><section id="contactfile-methods" class="level3" data-number="2.3.4"><h3 data-number="2.3.4" class="anchored" data-anchor-id="contactfile-methods">
<span class="header-section-number">2.3.4</span> <code>ContactFile</code> methods</h3>
<p>Two useful methods are available for <code>ContactFile</code>s:</p>
<ul>
<li>
<code>availableResolutions</code> checks which resolutions are available in a <code>ContactFile</code>.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">availableResolutions</a></span><span class="op">(</span><span class="va">cf</span><span class="op">)</span></span>
<span><span class="co">##  resolutions(5): 1000 2000 4000 8000 16000</span></span>
<span><span class="co">##  </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>
<code>availableChromosomes</code> checks which chromosomes are available in a <code>ContactFile</code>, along with their length.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">availableChromosomes</a></span><span class="op">(</span><span class="va">cf</span><span class="op">)</span></span>
<span><span class="co">##  Seqinfo object with 16 sequences from an unspecified genome:</span></span>
<span><span class="co">##    seqnames seqlengths isCircular genome</span></span>
<span><span class="co">##    I            230218       &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">##    II           813184       &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">##    III          316620       &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">##    IV          1531933       &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">##    V            576874       &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">##    ...             ...        ...    ...</span></span>
<span><span class="co">##    XII         1078177       &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">##    XIII         924431       &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">##    XIV          784333       &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">##    XV          1091291       &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">##    XVI          948066       &lt;NA&gt;   &lt;NA&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section><section id="hicexperiment-class" class="level2" data-number="2.4"><h2 data-number="2.4" class="anchored" data-anchor-id="hicexperiment-class">
<span class="header-section-number">2.4</span> <code>HiCExperiment</code> class</h2>
<p>Based on the previous sections, we have different Bioconductor classes relevant for Hi-C:</p>
<ul>
<li>
<code>GInteractions</code> which can be used to represent genomic interactions in R</li>
<li>
<code>ContactFile</code>s which can be used to establish a connection with disk-stored Hi-C files</li>
</ul>
<p><code>HiCExperiment</code> objects are created when <em>parsing</em> a <code>ContactFile</code> in R. The <code>HiCExperiment</code> class reads a <code>ContactFile</code> in memory and store genomic interactions as <code>GInteractions</code>. The <code>HiCExperiment</code> class is, quite obviously, defined in the <code>HiCExperiment</code> package.</p>
<section id="creating-a-hicexperiment-object" class="level3" data-number="2.4.1"><h3 data-number="2.4.1" class="anchored" data-anchor-id="creating-a-hicexperiment-object">
<span class="header-section-number">2.4.1</span> Creating a <code>HiCExperiment</code> object</h3>
<section id="importing-a-contactfile" class="level4" data-number="2.4.1.1"><h4 data-number="2.4.1.1" class="anchored" data-anchor-id="importing-a-contactfile">
<span class="header-section-number">2.4.1.1</span> Importing a <code>ContactFile</code>
</h4>
<p>In practice, to create a <code>HiCExperiment</code> object from a <code>ContactFile</code>, one can use the <code>import</code> method.</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Caution
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Creating a <code>HiCExperiment</code> object means <strong><em>importing data from a Hi-C matrix</em></strong> (e.g.&nbsp; from a <code>ContactFile</code>) in memory in R.<br>
</li>
<li>Creating a <code>HiCExperiment</code> object from large disk-stored contact matrices can potentially take a long time.</li>
</ul>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/CoolFile-class.html">CoolFile</a></span><span class="op">(</span><span class="va">coolf</span><span class="op">)</span></span>
<span><span class="va">hic</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocIO/man/IO.html">import</a></span><span class="op">(</span><span class="va">cf</span><span class="op">)</span></span>
<span><span class="va">hic</span></span>
<span><span class="co">##  `HiCExperiment` object with 8,757,906 contacts over 12,079 regions </span></span>
<span><span class="co">##  -------</span></span>
<span><span class="co">##  fileName: "/root/.cache/R/ExperimentHub/16dd6c5d5f_7752" </span></span>
<span><span class="co">##  focus: "whole genome" </span></span>
<span><span class="co">##  resolutions(5): 1000 2000 4000 8000 16000</span></span>
<span><span class="co">##  active resolution: 1000 </span></span>
<span><span class="co">##  interactions: 2945692 </span></span>
<span><span class="co">##  scores(2): count balanced </span></span>
<span><span class="co">##  topologicalFeatures: compartments(0) borders(0) loops(0) viewpoints(0) </span></span>
<span><span class="co">##  pairsFile: N/A </span></span>
<span><span class="co">##  metadata(0):</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Printing a <code>HiCExperiment</code> to the console will not reveal the actual data stored in the object (it would most likely crash your R session!). Instead, it gives a <strong>summary of the data</strong> stored in the object:</p>
<ul>
<li>The <code>fileName</code>, i.e.&nbsp;the path to the disk-stored data file</li>
<li>The <code>focus</code>, i.e.&nbsp;the genomic location for which data has been imported (in the example above, <code>"whole genome"</code> implies that all the data has been imported in R)</li>
<li>
<code>resolutions</code> available in the disk-stored data file (this will be identical to <code>availableResolutions(cf)</code>)</li>
<li>
<code>active resolution</code> indicates at which resolution the data is currently imported</li>
<li>
<code>interactions</code> refers to the actual <code>GInteractions</code> imported in R and “hidden” (for now!) in the <code>HiCExperiment</code> object</li>
<li>
<code>scores</code> refer to different interaction frequency estimates. These can be raw <code>count</code>s, <code>balanced</code> (if the contact matrix has been previously normalized), or whatever score the end-user want to attribute to each interaction (e.g.&nbsp;ratio of counts between two Hi-C maps, …)</li>
<li>
<code>topologicalFeatures</code> is a <code>list</code> of <code>GRanges</code> or <code>GInteractions</code> objects to describe important topological features.</li>
<li>
<code>pairsFile</code> is a pointer to an optional disk-stored <code>.pairs</code> file from which the contact matrix has been created. This is often useful to estimate some Hi-C metrics.</li>
<li>
<code>metadata</code> is a <code>list</code> to further describe the experiment.</li>
</ul>
<p>These pieces of information are called <code>slots</code>. They can be directly accessed using <code>getter</code> functions, bearing the same name than the slot.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/fileName.html">fileName</a></span><span class="op">(</span><span class="va">hic</span><span class="op">)</span></span>
<span><span class="co">##  [1] "/root/.cache/R/ExperimentHub/16dd6c5d5f_7752"</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">focus</a></span><span class="op">(</span><span class="va">hic</span><span class="op">)</span></span>
<span><span class="co">##  NULL</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">resolutions</a></span><span class="op">(</span><span class="va">hic</span><span class="op">)</span></span>
<span><span class="co">##  [1]  1000  2000  4000  8000 16000</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">resolution</a></span><span class="op">(</span><span class="va">hic</span><span class="op">)</span></span>
<span><span class="co">##  [1] 1000</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">interactions</a></span><span class="op">(</span><span class="va">hic</span><span class="op">)</span></span>
<span><span class="co">##  GInteractions object with 2945692 interactions and 4 metadata columns:</span></span>
<span><span class="co">##              seqnames1       ranges1     seqnames2       ranges2 |   bin_id1</span></span>
<span><span class="co">##                  &lt;Rle&gt;     &lt;IRanges&gt;         &lt;Rle&gt;     &lt;IRanges&gt; | &lt;numeric&gt;</span></span>
<span><span class="co">##          [1]         I        1-1000 ---         I        1-1000 |         0</span></span>
<span><span class="co">##          [2]         I        1-1000 ---         I     1001-2000 |         0</span></span>
<span><span class="co">##          [3]         I        1-1000 ---         I     2001-3000 |         0</span></span>
<span><span class="co">##          [4]         I        1-1000 ---         I     3001-4000 |         0</span></span>
<span><span class="co">##          [5]         I        1-1000 ---         I     4001-5000 |         0</span></span>
<span><span class="co">##          ...       ...           ... ...       ...           ... .       ...</span></span>
<span><span class="co">##    [2945688]       XVI 940001-941000 ---       XVI 942001-943000 |     12070</span></span>
<span><span class="co">##    [2945689]       XVI 940001-941000 ---       XVI 943001-944000 |     12070</span></span>
<span><span class="co">##    [2945690]       XVI 941001-942000 ---       XVI 941001-942000 |     12071</span></span>
<span><span class="co">##    [2945691]       XVI 941001-942000 ---       XVI 942001-943000 |     12071</span></span>
<span><span class="co">##    [2945692]       XVI 941001-942000 ---       XVI 943001-944000 |     12071</span></span>
<span><span class="co">##                bin_id2     count  balanced</span></span>
<span><span class="co">##              &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##          [1]         0        15 0.0663491</span></span>
<span><span class="co">##          [2]         1        21 0.1273505</span></span>
<span><span class="co">##          [3]         2        21 0.0738691</span></span>
<span><span class="co">##          [4]         3        38 0.0827051</span></span>
<span><span class="co">##          [5]         4        17 0.0591984</span></span>
<span><span class="co">##          ...       ...       ...       ...</span></span>
<span><span class="co">##    [2945688]     12072        11 0.0575550</span></span>
<span><span class="co">##    [2945689]     12073         1       NaN</span></span>
<span><span class="co">##    [2945690]     12071        74 0.0504615</span></span>
<span><span class="co">##    [2945691]     12072        39 0.1624599</span></span>
<span><span class="co">##    [2945692]     12073         1       NaN</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    regions: 12079 ranges and 4 metadata columns</span></span>
<span><span class="co">##    seqinfo: 16 sequences from an unspecified genome</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">scores</a></span><span class="op">(</span><span class="va">hic</span><span class="op">)</span></span>
<span><span class="co">##  List of length 2</span></span>
<span><span class="co">##  names(2): count balanced</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">topologicalFeatures</a></span><span class="op">(</span><span class="va">hic</span><span class="op">)</span></span>
<span><span class="co">##  List of length 4</span></span>
<span><span class="co">##  names(4): compartments borders loops viewpoints</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">pairsFile</a></span><span class="op">(</span><span class="va">hic</span><span class="op">)</span></span>
<span><span class="co">##  NULL</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html">metadata</a></span><span class="op">(</span><span class="va">hic</span><span class="op">)</span></span>
<span><span class="co">##  list()</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><code>import</code> also works for other types of <code>ContactFile</code> (<code>HicFile</code>, <code>HicproFile</code>, <code>PairsFile</code>), e.g.&nbsp;</p>
<ul>
<li>For <code>HicFile</code> and <code>HicproFile</code>, <code>import</code> seamlessly returns a <code>HiCExperiment</code> as well:</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/HicFile-class.html">HicFile</a></span><span class="op">(</span><span class="va">hicf</span><span class="op">)</span></span>
<span><span class="va">hic</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocIO/man/IO.html">import</a></span><span class="op">(</span><span class="va">hf</span><span class="op">)</span></span>
<span><span class="va">hic</span></span>
<span><span class="co">##  `HiCExperiment` object with 13,681,280 contacts over 12,165 regions </span></span>
<span><span class="co">##  -------</span></span>
<span><span class="co">##  fileName: "/root/.cache/R/ExperimentHub/16d160e6aeb_7836" </span></span>
<span><span class="co">##  focus: "whole genome" </span></span>
<span><span class="co">##  resolutions(5): 1000 2000 4000 8000 16000</span></span>
<span><span class="co">##  active resolution: 1000 </span></span>
<span><span class="co">##  interactions: 2965693 </span></span>
<span><span class="co">##  scores(2): count balanced </span></span>
<span><span class="co">##  topologicalFeatures: compartments(0) borders(0) loops(0) viewpoints(0) </span></span>
<span><span class="co">##  pairsFile: N/A </span></span>
<span><span class="co">##  metadata(0):</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>For <code>PairsFile</code>, the returned object is a representation of Hi-C “pairs” in R, i.e.&nbsp;<code>GInteractions</code>
</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/PairsFile-class.html">PairsFile</a></span><span class="op">(</span><span class="va">pairsf</span><span class="op">)</span></span>
<span><span class="va">pairs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocIO/man/IO.html">import</a></span><span class="op">(</span><span class="va">pf</span><span class="op">)</span></span>
<span><span class="va">pairs</span></span>
<span><span class="co">##  GInteractions object with 471364 interactions and 3 metadata columns:</span></span>
<span><span class="co">##             seqnames1   ranges1     seqnames2   ranges2 |     frag1     frag2</span></span>
<span><span class="co">##                 &lt;Rle&gt; &lt;IRanges&gt;         &lt;Rle&gt; &lt;IRanges&gt; | &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##         [1]        II       105 ---        II     48548 |      1358      1681</span></span>
<span><span class="co">##         [2]        II       113 ---        II     45003 |      1358      1658</span></span>
<span><span class="co">##         [3]        II       119 ---        II    687251 |      1358      5550</span></span>
<span><span class="co">##         [4]        II       160 ---        II     26124 |      1358      1510</span></span>
<span><span class="co">##         [5]        II       169 ---        II     39052 |      1358      1613</span></span>
<span><span class="co">##         ...       ...       ... ...       ...       ... .       ...       ...</span></span>
<span><span class="co">##    [471360]        II    808605 ---        II    809683 |      6316      6320</span></span>
<span><span class="co">##    [471361]        II    808609 ---        II    809917 |      6316      6324</span></span>
<span><span class="co">##    [471362]        II    808617 ---        II    809506 |      6316      6319</span></span>
<span><span class="co">##    [471363]        II    809447 ---        II    809685 |      6319      6321</span></span>
<span><span class="co">##    [471364]        II    809472 ---        II    809675 |      6319      6320</span></span>
<span><span class="co">##              distance</span></span>
<span><span class="co">##             &lt;integer&gt;</span></span>
<span><span class="co">##         [1]     48443</span></span>
<span><span class="co">##         [2]     44890</span></span>
<span><span class="co">##         [3]    687132</span></span>
<span><span class="co">##         [4]     25964</span></span>
<span><span class="co">##         [5]     38883</span></span>
<span><span class="co">##         ...       ...</span></span>
<span><span class="co">##    [471360]      1078</span></span>
<span><span class="co">##    [471361]      1308</span></span>
<span><span class="co">##    [471362]       889</span></span>
<span><span class="co">##    [471363]       238</span></span>
<span><span class="co">##    [471364]       203</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    regions: 549331 ranges and 0 metadata columns</span></span>
<span><span class="co">##    seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="customizing-the-import" class="level4" data-number="2.4.1.2"><h4 data-number="2.4.1.2" class="anchored" data-anchor-id="customizing-the-import">
<span class="header-section-number">2.4.1.2</span> Customizing the <code>import</code>
</h4>
<p>To reduce the <code>import</code> to only parse the data that is relevant to the study, two arguments can be passed to <code>import</code>, along with a <code>ContactFile</code>.</p>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Key <code>import</code> arguments:
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>
<code>focus</code>: This can be used to <strong>only parse data for a specific genomic location</strong>.</li>
<li>
<code>resolution</code>: This can be used to choose which resolution to parse the contact matrix at (this is ignored if the <code>ContactFile</code> is not multi-resolution, e.g.&nbsp;<code>.cool</code> or HiC-Pro generated matrices)</li>
</ul>
</div>
</div>
<ul>
<li>Import interactions within a single chromosome:</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hic</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocIO/man/IO.html">import</a></span><span class="op">(</span><span class="va">cf</span>, focus <span class="op">=</span> <span class="st">'II'</span>, resolution <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">regions</a></span><span class="op">(</span><span class="va">hic</span><span class="op">)</span> <span class="co"># ---- `regions()` work on `HiCExperiment` the same way than on `GInteractions`</span></span>
<span><span class="co">##  GRanges object with 407 ranges and 4 metadata columns:</span></span>
<span><span class="co">##                     seqnames        ranges strand |    bin_id    weight   chr</span></span>
<span><span class="co">##                        &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;numeric&gt; &lt;Rle&gt;</span></span>
<span><span class="co">##           II_1_2000       II        1-2000      * |       116       NaN    II</span></span>
<span><span class="co">##        II_2001_4000       II     2001-4000      * |       117       NaN    II</span></span>
<span><span class="co">##        II_4001_6000       II     4001-6000      * |       118       NaN    II</span></span>
<span><span class="co">##        II_6001_8000       II     6001-8000      * |       119       NaN    II</span></span>
<span><span class="co">##       II_8001_10000       II    8001-10000      * |       120 0.0461112    II</span></span>
<span><span class="co">##                 ...      ...           ...    ... .       ...       ...   ...</span></span>
<span><span class="co">##    II_804001_806000       II 804001-806000      * |       518 0.0493107    II</span></span>
<span><span class="co">##    II_806001_808000       II 806001-808000      * |       519 0.0611355    II</span></span>
<span><span class="co">##    II_808001_810000       II 808001-810000      * |       520       NaN    II</span></span>
<span><span class="co">##    II_810001_812000       II 810001-812000      * |       521       NaN    II</span></span>
<span><span class="co">##    II_812001_813184       II 812001-813184      * |       522       NaN    II</span></span>
<span><span class="co">##                        center</span></span>
<span><span class="co">##                     &lt;integer&gt;</span></span>
<span><span class="co">##           II_1_2000      1000</span></span>
<span><span class="co">##        II_2001_4000      3000</span></span>
<span><span class="co">##        II_4001_6000      5000</span></span>
<span><span class="co">##        II_6001_8000      7000</span></span>
<span><span class="co">##       II_8001_10000      9000</span></span>
<span><span class="co">##                 ...       ...</span></span>
<span><span class="co">##    II_804001_806000    805000</span></span>
<span><span class="co">##    II_806001_808000    807000</span></span>
<span><span class="co">##    II_808001_810000    809000</span></span>
<span><span class="co">##    II_810001_812000    811000</span></span>
<span><span class="co">##    II_812001_813184    812592</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 16 sequences from an unspecified genome</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html">seqnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">regions</a></span><span class="op">(</span><span class="va">hic</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##     I   II  III   IV    V   VI  VII VIII   IX    X   XI  XII XIII  XIV   XV </span></span>
<span><span class="co">##     0  407    0    0    0    0    0    0    0    0    0    0    0    0    0 </span></span>
<span><span class="co">##   XVI </span></span>
<span><span class="co">##     0</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">anchors</a></span><span class="op">(</span><span class="va">hic</span><span class="op">)</span> <span class="co"># ---- `anchors()` work on `HiCExperiment` the same way than on `GInteractions`</span></span>
<span><span class="co">##  $first</span></span>
<span><span class="co">##  GRanges object with 34063 ranges and 4 metadata columns:</span></span>
<span><span class="co">##            seqnames        ranges strand |    bin_id    weight   chr</span></span>
<span><span class="co">##               &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;numeric&gt; &lt;Rle&gt;</span></span>
<span><span class="co">##        [1]       II        1-2000      * |       116       NaN    II</span></span>
<span><span class="co">##        [2]       II        1-2000      * |       116       NaN    II</span></span>
<span><span class="co">##        [3]       II        1-2000      * |       116       NaN    II</span></span>
<span><span class="co">##        [4]       II        1-2000      * |       116       NaN    II</span></span>
<span><span class="co">##        [5]       II        1-2000      * |       116       NaN    II</span></span>
<span><span class="co">##        ...      ...           ...    ... .       ...       ...   ...</span></span>
<span><span class="co">##    [34059]       II 804001-806000      * |       518 0.0493107    II</span></span>
<span><span class="co">##    [34060]       II 806001-808000      * |       519 0.0611355    II</span></span>
<span><span class="co">##    [34061]       II 806001-808000      * |       519 0.0611355    II</span></span>
<span><span class="co">##    [34062]       II 806001-808000      * |       519 0.0611355    II</span></span>
<span><span class="co">##    [34063]       II 808001-810000      * |       520       NaN    II</span></span>
<span><span class="co">##               center</span></span>
<span><span class="co">##            &lt;integer&gt;</span></span>
<span><span class="co">##        [1]      1000</span></span>
<span><span class="co">##        [2]      1000</span></span>
<span><span class="co">##        [3]      1000</span></span>
<span><span class="co">##        [4]      1000</span></span>
<span><span class="co">##        [5]      1000</span></span>
<span><span class="co">##        ...       ...</span></span>
<span><span class="co">##    [34059]    805000</span></span>
<span><span class="co">##    [34060]    807000</span></span>
<span><span class="co">##    [34061]    807000</span></span>
<span><span class="co">##    [34062]    807000</span></span>
<span><span class="co">##    [34063]    809000</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 16 sequences from an unspecified genome</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $second</span></span>
<span><span class="co">##  GRanges object with 34063 ranges and 4 metadata columns:</span></span>
<span><span class="co">##            seqnames        ranges strand |    bin_id    weight   chr</span></span>
<span><span class="co">##               &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;numeric&gt; &lt;Rle&gt;</span></span>
<span><span class="co">##        [1]       II        1-2000      * |       116       NaN    II</span></span>
<span><span class="co">##        [2]       II     4001-6000      * |       118       NaN    II</span></span>
<span><span class="co">##        [3]       II     6001-8000      * |       119       NaN    II</span></span>
<span><span class="co">##        [4]       II    8001-10000      * |       120 0.0461112    II</span></span>
<span><span class="co">##        [5]       II   10001-12000      * |       121 0.0334807    II</span></span>
<span><span class="co">##        ...      ...           ...    ... .       ...       ...   ...</span></span>
<span><span class="co">##    [34059]       II 810001-812000      * |       521       NaN    II</span></span>
<span><span class="co">##    [34060]       II 806001-808000      * |       519 0.0611355    II</span></span>
<span><span class="co">##    [34061]       II 808001-810000      * |       520       NaN    II</span></span>
<span><span class="co">##    [34062]       II 810001-812000      * |       521       NaN    II</span></span>
<span><span class="co">##    [34063]       II 808001-810000      * |       520       NaN    II</span></span>
<span><span class="co">##               center</span></span>
<span><span class="co">##            &lt;integer&gt;</span></span>
<span><span class="co">##        [1]      1000</span></span>
<span><span class="co">##        [2]      5000</span></span>
<span><span class="co">##        [3]      7000</span></span>
<span><span class="co">##        [4]      9000</span></span>
<span><span class="co">##        [5]     11000</span></span>
<span><span class="co">##        ...       ...</span></span>
<span><span class="co">##    [34059]    811000</span></span>
<span><span class="co">##    [34060]    807000</span></span>
<span><span class="co">##    [34061]    809000</span></span>
<span><span class="co">##    [34062]    811000</span></span>
<span><span class="co">##    [34063]    809000</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 16 sequences from an unspecified genome</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Import interactions within a segment of a chromosome:</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hic</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocIO/man/IO.html">import</a></span><span class="op">(</span><span class="va">cf</span>, focus <span class="op">=</span> <span class="st">'II:40000-60000'</span>, resolution <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">regions</a></span><span class="op">(</span><span class="va">hic</span><span class="op">)</span> </span>
<span><span class="co">##  GRanges object with 21 ranges and 4 metadata columns:</span></span>
<span><span class="co">##                   seqnames      ranges strand |    bin_id    weight   chr</span></span>
<span><span class="co">##                      &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;numeric&gt; &lt;Rle&gt;</span></span>
<span><span class="co">##    II_39001_40000       II 39001-40000      * |       270 0.0220798    II</span></span>
<span><span class="co">##    II_40001_41000       II 40001-41000      * |       271 0.0246775    II</span></span>
<span><span class="co">##    II_41001_42000       II 41001-42000      * |       272 0.0269232    II</span></span>
<span><span class="co">##    II_42001_43000       II 42001-43000      * |       273 0.0341849    II</span></span>
<span><span class="co">##    II_43001_44000       II 43001-44000      * |       274 0.0265386    II</span></span>
<span><span class="co">##               ...      ...         ...    ... .       ...       ...   ...</span></span>
<span><span class="co">##    II_55001_56000       II 55001-56000      * |       286 0.0213532    II</span></span>
<span><span class="co">##    II_56001_57000       II 56001-57000      * |       287 0.0569839    II</span></span>
<span><span class="co">##    II_57001_58000       II 57001-58000      * |       288 0.0338612    II</span></span>
<span><span class="co">##    II_58001_59000       II 58001-59000      * |       289 0.0294531    II</span></span>
<span><span class="co">##    II_59001_60000       II 59001-60000      * |       290 0.0306662    II</span></span>
<span><span class="co">##                      center</span></span>
<span><span class="co">##                   &lt;integer&gt;</span></span>
<span><span class="co">##    II_39001_40000     39500</span></span>
<span><span class="co">##    II_40001_41000     40500</span></span>
<span><span class="co">##    II_41001_42000     41500</span></span>
<span><span class="co">##    II_42001_43000     42500</span></span>
<span><span class="co">##    II_43001_44000     43500</span></span>
<span><span class="co">##               ...       ...</span></span>
<span><span class="co">##    II_55001_56000     55500</span></span>
<span><span class="co">##    II_56001_57000     56500</span></span>
<span><span class="co">##    II_57001_58000     57500</span></span>
<span><span class="co">##    II_58001_59000     58500</span></span>
<span><span class="co">##    II_59001_60000     59500</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 16 sequences from an unspecified genome</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">anchors</a></span><span class="op">(</span><span class="va">hic</span><span class="op">)</span></span>
<span><span class="co">##  $first</span></span>
<span><span class="co">##  GRanges object with 210 ranges and 4 metadata columns:</span></span>
<span><span class="co">##          seqnames      ranges strand |    bin_id    weight   chr    center</span></span>
<span><span class="co">##             &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;numeric&gt; &lt;Rle&gt; &lt;integer&gt;</span></span>
<span><span class="co">##      [1]       II 40001-41000      * |       271 0.0246775    II     40500</span></span>
<span><span class="co">##      [2]       II 40001-41000      * |       271 0.0246775    II     40500</span></span>
<span><span class="co">##      [3]       II 40001-41000      * |       271 0.0246775    II     40500</span></span>
<span><span class="co">##      [4]       II 40001-41000      * |       271 0.0246775    II     40500</span></span>
<span><span class="co">##      [5]       II 40001-41000      * |       271 0.0246775    II     40500</span></span>
<span><span class="co">##      ...      ...         ...    ... .       ...       ...   ...       ...</span></span>
<span><span class="co">##    [206]       II 57001-58000      * |       288 0.0338612    II     57500</span></span>
<span><span class="co">##    [207]       II 57001-58000      * |       288 0.0338612    II     57500</span></span>
<span><span class="co">##    [208]       II 58001-59000      * |       289 0.0294531    II     58500</span></span>
<span><span class="co">##    [209]       II 58001-59000      * |       289 0.0294531    II     58500</span></span>
<span><span class="co">##    [210]       II 59001-60000      * |       290 0.0306662    II     59500</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 16 sequences from an unspecified genome</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $second</span></span>
<span><span class="co">##  GRanges object with 210 ranges and 4 metadata columns:</span></span>
<span><span class="co">##          seqnames      ranges strand |    bin_id    weight   chr    center</span></span>
<span><span class="co">##             &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;numeric&gt; &lt;Rle&gt; &lt;integer&gt;</span></span>
<span><span class="co">##      [1]       II 40001-41000      * |       271 0.0246775    II     40500</span></span>
<span><span class="co">##      [2]       II 41001-42000      * |       272 0.0269232    II     41500</span></span>
<span><span class="co">##      [3]       II 42001-43000      * |       273 0.0341849    II     42500</span></span>
<span><span class="co">##      [4]       II 43001-44000      * |       274 0.0265386    II     43500</span></span>
<span><span class="co">##      [5]       II 44001-45000      * |       275 0.0488968    II     44500</span></span>
<span><span class="co">##      ...      ...         ...    ... .       ...       ...   ...       ...</span></span>
<span><span class="co">##    [206]       II 58001-59000      * |       289 0.0294531    II     58500</span></span>
<span><span class="co">##    [207]       II 59001-60000      * |       290 0.0306662    II     59500</span></span>
<span><span class="co">##    [208]       II 58001-59000      * |       289 0.0294531    II     58500</span></span>
<span><span class="co">##    [209]       II 59001-60000      * |       290 0.0306662    II     59500</span></span>
<span><span class="co">##    [210]       II 59001-60000      * |       290 0.0306662    II     59500</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 16 sequences from an unspecified genome</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Import interactions between two chromosomes:</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hic2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocIO/man/IO.html">import</a></span><span class="op">(</span><span class="va">cf</span>, focus <span class="op">=</span> <span class="st">'II|XV'</span>, resolution <span class="op">=</span> <span class="fl">4000</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">regions</a></span><span class="op">(</span><span class="va">hic2</span><span class="op">)</span></span>
<span><span class="co">##  GRanges object with 477 ranges and 4 metadata columns:</span></span>
<span><span class="co">##                       seqnames          ranges strand |    bin_id    weight</span></span>
<span><span class="co">##                          &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##             II_1_4000       II          1-4000      * |        58       NaN</span></span>
<span><span class="co">##          II_4001_8000       II       4001-8000      * |        59       NaN</span></span>
<span><span class="co">##         II_8001_12000       II      8001-12000      * |        60 0.0274474</span></span>
<span><span class="co">##        II_12001_16000       II     12001-16000      * |        61 0.0342116</span></span>
<span><span class="co">##        II_16001_20000       II     16001-20000      * |        62 0.0195128</span></span>
<span><span class="co">##                   ...      ...             ...    ... .       ...       ...</span></span>
<span><span class="co">##    XV_1072001_1076000       XV 1072001-1076000      * |      2783  0.041763</span></span>
<span><span class="co">##    XV_1076001_1080000       XV 1076001-1080000      * |      2784       NaN</span></span>
<span><span class="co">##    XV_1080001_1084000       XV 1080001-1084000      * |      2785       NaN</span></span>
<span><span class="co">##    XV_1084001_1088000       XV 1084001-1088000      * |      2786       NaN</span></span>
<span><span class="co">##    XV_1088001_1091291       XV 1088001-1091291      * |      2787       NaN</span></span>
<span><span class="co">##                         chr    center</span></span>
<span><span class="co">##                       &lt;Rle&gt; &lt;integer&gt;</span></span>
<span><span class="co">##             II_1_4000    II      2000</span></span>
<span><span class="co">##          II_4001_8000    II      6000</span></span>
<span><span class="co">##         II_8001_12000    II     10000</span></span>
<span><span class="co">##        II_12001_16000    II     14000</span></span>
<span><span class="co">##        II_16001_20000    II     18000</span></span>
<span><span class="co">##                   ...   ...       ...</span></span>
<span><span class="co">##    XV_1072001_1076000    XV   1074000</span></span>
<span><span class="co">##    XV_1076001_1080000    XV   1078000</span></span>
<span><span class="co">##    XV_1080001_1084000    XV   1082000</span></span>
<span><span class="co">##    XV_1084001_1088000    XV   1086000</span></span>
<span><span class="co">##    XV_1088001_1091291    XV   1089646</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 16 sequences from an unspecified genome</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">anchors</a></span><span class="op">(</span><span class="va">hic2</span><span class="op">)</span></span>
<span><span class="co">##  $first</span></span>
<span><span class="co">##  GRanges object with 18032 ranges and 4 metadata columns:</span></span>
<span><span class="co">##            seqnames        ranges strand |    bin_id    weight   chr</span></span>
<span><span class="co">##               &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;numeric&gt; &lt;Rle&gt;</span></span>
<span><span class="co">##        [1]       II        1-4000      * |        58       NaN    II</span></span>
<span><span class="co">##        [2]       II        1-4000      * |        58       NaN    II</span></span>
<span><span class="co">##        [3]       II        1-4000      * |        58       NaN    II</span></span>
<span><span class="co">##        [4]       II        1-4000      * |        58       NaN    II</span></span>
<span><span class="co">##        [5]       II        1-4000      * |        58       NaN    II</span></span>
<span><span class="co">##        ...      ...           ...    ... .       ...       ...   ...</span></span>
<span><span class="co">##    [18028]       II 808001-812000      * |       260       NaN    II</span></span>
<span><span class="co">##    [18029]       II 808001-812000      * |       260       NaN    II</span></span>
<span><span class="co">##    [18030]       II 808001-812000      * |       260       NaN    II</span></span>
<span><span class="co">##    [18031]       II 808001-812000      * |       260       NaN    II</span></span>
<span><span class="co">##    [18032]       II 808001-812000      * |       260       NaN    II</span></span>
<span><span class="co">##               center</span></span>
<span><span class="co">##            &lt;integer&gt;</span></span>
<span><span class="co">##        [1]      2000</span></span>
<span><span class="co">##        [2]      2000</span></span>
<span><span class="co">##        [3]      2000</span></span>
<span><span class="co">##        [4]      2000</span></span>
<span><span class="co">##        [5]      2000</span></span>
<span><span class="co">##        ...       ...</span></span>
<span><span class="co">##    [18028]    810000</span></span>
<span><span class="co">##    [18029]    810000</span></span>
<span><span class="co">##    [18030]    810000</span></span>
<span><span class="co">##    [18031]    810000</span></span>
<span><span class="co">##    [18032]    810000</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 16 sequences from an unspecified genome</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $second</span></span>
<span><span class="co">##  GRanges object with 18032 ranges and 4 metadata columns:</span></span>
<span><span class="co">##            seqnames          ranges strand |    bin_id    weight   chr</span></span>
<span><span class="co">##               &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;numeric&gt; &lt;Rle&gt;</span></span>
<span><span class="co">##        [1]       XV     48001-52000      * |      2527 0.0185354    XV</span></span>
<span><span class="co">##        [2]       XV   348001-352000      * |      2602 0.0233750    XV</span></span>
<span><span class="co">##        [3]       XV   468001-472000      * |      2632 0.0153615    XV</span></span>
<span><span class="co">##        [4]       XV   472001-476000      * |      2633 0.0189624    XV</span></span>
<span><span class="co">##        [5]       XV   584001-588000      * |      2661 0.0167715    XV</span></span>
<span><span class="co">##        ...      ...             ...    ... .       ...       ...   ...</span></span>
<span><span class="co">##    [18028]       XV   980001-984000      * |      2760 0.0187827    XV</span></span>
<span><span class="co">##    [18029]       XV   984001-988000      * |      2761 0.0250094    XV</span></span>
<span><span class="co">##    [18030]       XV   992001-996000      * |      2763 0.0185599    XV</span></span>
<span><span class="co">##    [18031]       XV 1004001-1008000      * |      2766 0.0196942    XV</span></span>
<span><span class="co">##    [18032]       XV 1064001-1068000      * |      2781 0.0208220    XV</span></span>
<span><span class="co">##               center</span></span>
<span><span class="co">##            &lt;integer&gt;</span></span>
<span><span class="co">##        [1]     50000</span></span>
<span><span class="co">##        [2]    350000</span></span>
<span><span class="co">##        [3]    470000</span></span>
<span><span class="co">##        [4]    474000</span></span>
<span><span class="co">##        [5]    586000</span></span>
<span><span class="co">##        ...       ...</span></span>
<span><span class="co">##    [18028]    982000</span></span>
<span><span class="co">##    [18029]    986000</span></span>
<span><span class="co">##    [18030]    994000</span></span>
<span><span class="co">##    [18031]   1006000</span></span>
<span><span class="co">##    [18032]   1066000</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 16 sequences from an unspecified genome</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Import interactions between segments of two chromosomes:</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hic3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocIO/man/IO.html">import</a></span><span class="op">(</span><span class="va">cf</span>, focus <span class="op">=</span> <span class="st">'III:10000-40000|XV:10000-40000'</span>, resolution <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">regions</a></span><span class="op">(</span><span class="va">hic3</span><span class="op">)</span></span>
<span><span class="co">##  GRanges object with 32 ranges and 4 metadata columns:</span></span>
<span><span class="co">##                    seqnames      ranges strand |    bin_id    weight   chr</span></span>
<span><span class="co">##                       &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;numeric&gt; &lt;Rle&gt;</span></span>
<span><span class="co">##     III_8001_10000      III  8001-10000      * |       527       NaN   III</span></span>
<span><span class="co">##    III_10001_12000      III 10001-12000      * |       528       NaN   III</span></span>
<span><span class="co">##    III_12001_14000      III 12001-14000      * |       529       NaN   III</span></span>
<span><span class="co">##    III_14001_16000      III 14001-16000      * |       530 0.0356351   III</span></span>
<span><span class="co">##    III_16001_18000      III 16001-18000      * |       531 0.0230693   III</span></span>
<span><span class="co">##                ...      ...         ...    ... .       ...       ...   ...</span></span>
<span><span class="co">##     XV_30001_32000       XV 30001-32000      * |      5039 0.0482465    XV</span></span>
<span><span class="co">##     XV_32001_34000       XV 32001-34000      * |      5040 0.0241580    XV</span></span>
<span><span class="co">##     XV_34001_36000       XV 34001-36000      * |      5041 0.0273166    XV</span></span>
<span><span class="co">##     XV_36001_38000       XV 36001-38000      * |      5042 0.0542235    XV</span></span>
<span><span class="co">##     XV_38001_40000       XV 38001-40000      * |      5043 0.0206849    XV</span></span>
<span><span class="co">##                       center</span></span>
<span><span class="co">##                    &lt;integer&gt;</span></span>
<span><span class="co">##     III_8001_10000      9000</span></span>
<span><span class="co">##    III_10001_12000     11000</span></span>
<span><span class="co">##    III_12001_14000     13000</span></span>
<span><span class="co">##    III_14001_16000     15000</span></span>
<span><span class="co">##    III_16001_18000     17000</span></span>
<span><span class="co">##                ...       ...</span></span>
<span><span class="co">##     XV_30001_32000     31000</span></span>
<span><span class="co">##     XV_32001_34000     33000</span></span>
<span><span class="co">##     XV_34001_36000     35000</span></span>
<span><span class="co">##     XV_36001_38000     37000</span></span>
<span><span class="co">##     XV_38001_40000     39000</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 16 sequences from an unspecified genome</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">anchors</a></span><span class="op">(</span><span class="va">hic3</span><span class="op">)</span></span>
<span><span class="co">##  $first</span></span>
<span><span class="co">##  GRanges object with 11 ranges and 4 metadata columns:</span></span>
<span><span class="co">##         seqnames      ranges strand |    bin_id    weight   chr    center</span></span>
<span><span class="co">##            &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;numeric&gt; &lt;Rle&gt; &lt;integer&gt;</span></span>
<span><span class="co">##     [1]      III 14001-16000      * |       530 0.0356351   III     15000</span></span>
<span><span class="co">##     [2]      III 16001-18000      * |       531 0.0230693   III     17000</span></span>
<span><span class="co">##     [3]      III 16001-18000      * |       531 0.0230693   III     17000</span></span>
<span><span class="co">##     [4]      III 20001-22000      * |       533 0.0343250   III     21000</span></span>
<span><span class="co">##     [5]      III 22001-24000      * |       534 0.0258604   III     23000</span></span>
<span><span class="co">##     [6]      III 24001-26000      * |       535 0.0290757   III     25000</span></span>
<span><span class="co">##     [7]      III 28001-30000      * |       537 0.0290713   III     29000</span></span>
<span><span class="co">##     [8]      III 30001-32000      * |       538 0.0266373   III     31000</span></span>
<span><span class="co">##     [9]      III 32001-34000      * |       539 0.0201137   III     33000</span></span>
<span><span class="co">##    [10]      III 32001-34000      * |       539 0.0201137   III     33000</span></span>
<span><span class="co">##    [11]      III 36001-38000      * |       541 0.0220603   III     37000</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 16 sequences from an unspecified genome</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $second</span></span>
<span><span class="co">##  GRanges object with 11 ranges and 4 metadata columns:</span></span>
<span><span class="co">##         seqnames      ranges strand |    bin_id    weight   chr    center</span></span>
<span><span class="co">##            &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;numeric&gt; &lt;Rle&gt; &lt;integer&gt;</span></span>
<span><span class="co">##     [1]       XV 16001-18000      * |      5032 0.0187250    XV     17000</span></span>
<span><span class="co">##     [2]       XV 16001-18000      * |      5032 0.0187250    XV     17000</span></span>
<span><span class="co">##     [3]       XV 20001-22000      * |      5034 0.0247973    XV     21000</span></span>
<span><span class="co">##     [4]       XV 14001-16000      * |      5031 0.0379727    XV     15000</span></span>
<span><span class="co">##     [5]       XV 10001-12000      * |      5029 0.0296913    XV     11000</span></span>
<span><span class="co">##     [6]       XV 32001-34000      * |      5040 0.0241580    XV     33000</span></span>
<span><span class="co">##     [7]       XV 16001-18000      * |      5032 0.0187250    XV     17000</span></span>
<span><span class="co">##     [8]       XV 38001-40000      * |      5043 0.0206849    XV     39000</span></span>
<span><span class="co">##     [9]       XV 22001-24000      * |      5035 0.0613856    XV     23000</span></span>
<span><span class="co">##    [10]       XV 30001-32000      * |      5039 0.0482465    XV     31000</span></span>
<span><span class="co">##    [11]       XV 10001-12000      * |      5029 0.0296913    XV     11000</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 16 sequences from an unspecified genome</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section><section id="interacting-with-hicexperiment-data" class="level3" data-number="2.4.2"><h3 data-number="2.4.2" class="anchored" data-anchor-id="interacting-with-hicexperiment-data">
<span class="header-section-number">2.4.2</span> Interacting with <code>HiCExperiment</code> data</h3>
<ul>
<li>An <code>HiCExperiment</code> object allows parsing of a disk-stored contact matrix.</li>
<li>An <code>HiCExperiment</code> object operates by wrapping together (1) a <code>ContactFile</code> (i.e.&nbsp;a connection to a disk-stored data file) and (2) a <code>GInteractions</code> generated by parsing the data file.</li>
</ul>
<p>We will use the <code>yeast_hic</code> <code>HiCExperiment</code> object to demonstrate how to parse information from a <code>HiCExperiment</code> object.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">yeast_hic</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/data.html">contacts_yeast</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">yeast_hic</span></span>
<span><span class="co">##  `HiCExperiment` object with 8,757,906 contacts over 763 regions </span></span>
<span><span class="co">##  -------</span></span>
<span><span class="co">##  fileName: "/root/.cache/R/ExperimentHub/16dd6c5d5f_7752" </span></span>
<span><span class="co">##  focus: "whole genome" </span></span>
<span><span class="co">##  resolutions(5): 1000 2000 4000 8000 16000</span></span>
<span><span class="co">##  active resolution: 16000 </span></span>
<span><span class="co">##  interactions: 267709 </span></span>
<span><span class="co">##  scores(2): count balanced </span></span>
<span><span class="co">##  topologicalFeatures: compartments(0) borders(0) loops(0) viewpoints(0) centromeres(16) </span></span>
<span><span class="co">##  pairsFile: /root/.cache/R/ExperimentHub/16d2ae89407_7753 </span></span>
<span><span class="co">##  metadata(3): ID org date</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="interactions" class="level4" data-number="2.4.2.1"><h4 data-number="2.4.2.1" class="anchored" data-anchor-id="interactions">
<span class="header-section-number">2.4.2.1</span> Interactions</h4>
<p>The imported genomic interactions can be directly <strong>exposed</strong> using the <code>interactions</code> function and are returned as a <code>GInteractions</code> object.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">interactions</a></span><span class="op">(</span><span class="va">yeast_hic</span><span class="op">)</span></span>
<span><span class="co">##  GInteractions object with 267709 interactions and 4 metadata columns:</span></span>
<span><span class="co">##             seqnames1       ranges1     seqnames2       ranges2 |   bin_id1</span></span>
<span><span class="co">##                 &lt;Rle&gt;     &lt;IRanges&gt;         &lt;Rle&gt;     &lt;IRanges&gt; | &lt;numeric&gt;</span></span>
<span><span class="co">##         [1]         I       1-16000 ---         I       1-16000 |         0</span></span>
<span><span class="co">##         [2]         I       1-16000 ---         I   16001-32000 |         0</span></span>
<span><span class="co">##         [3]         I       1-16000 ---         I   32001-48000 |         0</span></span>
<span><span class="co">##         [4]         I       1-16000 ---         I   48001-64000 |         0</span></span>
<span><span class="co">##         [5]         I       1-16000 ---         I   64001-80000 |         0</span></span>
<span><span class="co">##         ...       ...           ... ...       ...           ... .       ...</span></span>
<span><span class="co">##    [267705]       XVI 896001-912000 ---       XVI 912001-928000 |       759</span></span>
<span><span class="co">##    [267706]       XVI 896001-912000 ---       XVI 928001-944000 |       759</span></span>
<span><span class="co">##    [267707]       XVI 912001-928000 ---       XVI 912001-928000 |       760</span></span>
<span><span class="co">##    [267708]       XVI 912001-928000 ---       XVI 928001-944000 |       760</span></span>
<span><span class="co">##    [267709]       XVI 928001-944000 ---       XVI 928001-944000 |       761</span></span>
<span><span class="co">##               bin_id2     count  balanced</span></span>
<span><span class="co">##             &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##         [1]         0      2836 1.0943959</span></span>
<span><span class="co">##         [2]         1      2212 0.9592069</span></span>
<span><span class="co">##         [3]         2      1183 0.4385242</span></span>
<span><span class="co">##         [4]         3       831 0.2231192</span></span>
<span><span class="co">##         [5]         4       310 0.0821255</span></span>
<span><span class="co">##         ...       ...       ...       ...</span></span>
<span><span class="co">##    [267705]       760      3565  1.236371</span></span>
<span><span class="co">##    [267706]       761      1359  0.385016</span></span>
<span><span class="co">##    [267707]       760      3534  2.103988</span></span>
<span><span class="co">##    [267708]       761      3055  1.485794</span></span>
<span><span class="co">##    [267709]       761      4308  1.711565</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    regions: 763 ranges and 4 metadata columns</span></span>
<span><span class="co">##    seqinfo: 16 sequences from an unspecified genome</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Because genomic interactions are actually stored as <code>GInteractions</code>, <code>regions</code> and <code>anchors</code> work on <code>HiCExperiment</code> objects just as they work with <code>GInteractions</code>!</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">regions</a></span><span class="op">(</span><span class="va">yeast_hic</span><span class="op">)</span></span>
<span><span class="co">##  GRanges object with 763 ranges and 4 metadata columns:</span></span>
<span><span class="co">##                      seqnames        ranges strand |    bin_id     weight</span></span>
<span><span class="co">##                         &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">##            I_1_16000        I       1-16000      * |         0  0.0196442</span></span>
<span><span class="co">##        I_16001_32000        I   16001-32000      * |         1  0.0220746</span></span>
<span><span class="co">##        I_32001_48000        I   32001-48000      * |         2  0.0188701</span></span>
<span><span class="co">##        I_48001_64000        I   48001-64000      * |         3  0.0136679</span></span>
<span><span class="co">##        I_64001_80000        I   64001-80000      * |         4  0.0134860</span></span>
<span><span class="co">##                  ...      ...           ...    ... .       ...        ...</span></span>
<span><span class="co">##    XVI_880001_896000      XVI 880001-896000      * |       758 0.00910873</span></span>
<span><span class="co">##    XVI_896001_912000      XVI 896001-912000      * |       759 0.01421350</span></span>
<span><span class="co">##    XVI_912001_928000      XVI 912001-928000      * |       760 0.02439992</span></span>
<span><span class="co">##    XVI_928001_944000      XVI 928001-944000      * |       761 0.01993237</span></span>
<span><span class="co">##    XVI_944001_948066      XVI 944001-948066      * |       762        NaN</span></span>
<span><span class="co">##                        chr    center</span></span>
<span><span class="co">##                      &lt;Rle&gt; &lt;integer&gt;</span></span>
<span><span class="co">##            I_1_16000     I      8000</span></span>
<span><span class="co">##        I_16001_32000     I     24000</span></span>
<span><span class="co">##        I_32001_48000     I     40000</span></span>
<span><span class="co">##        I_48001_64000     I     56000</span></span>
<span><span class="co">##        I_64001_80000     I     72000</span></span>
<span><span class="co">##                  ...   ...       ...</span></span>
<span><span class="co">##    XVI_880001_896000   XVI    888000</span></span>
<span><span class="co">##    XVI_896001_912000   XVI    904000</span></span>
<span><span class="co">##    XVI_912001_928000   XVI    920000</span></span>
<span><span class="co">##    XVI_928001_944000   XVI    936000</span></span>
<span><span class="co">##    XVI_944001_948066   XVI    946033</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 16 sequences from an unspecified genome</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">anchors</a></span><span class="op">(</span><span class="va">yeast_hic</span><span class="op">)</span></span>
<span><span class="co">##  $first</span></span>
<span><span class="co">##  GRanges object with 267709 ranges and 4 metadata columns:</span></span>
<span><span class="co">##             seqnames        ranges strand |    bin_id    weight   chr</span></span>
<span><span class="co">##                &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;numeric&gt; &lt;Rle&gt;</span></span>
<span><span class="co">##         [1]        I       1-16000      * |         0 0.0196442     I</span></span>
<span><span class="co">##         [2]        I       1-16000      * |         0 0.0196442     I</span></span>
<span><span class="co">##         [3]        I       1-16000      * |         0 0.0196442     I</span></span>
<span><span class="co">##         [4]        I       1-16000      * |         0 0.0196442     I</span></span>
<span><span class="co">##         [5]        I       1-16000      * |         0 0.0196442     I</span></span>
<span><span class="co">##         ...      ...           ...    ... .       ...       ...   ...</span></span>
<span><span class="co">##    [267705]      XVI 896001-912000      * |       759 0.0142135   XVI</span></span>
<span><span class="co">##    [267706]      XVI 896001-912000      * |       759 0.0142135   XVI</span></span>
<span><span class="co">##    [267707]      XVI 912001-928000      * |       760 0.0243999   XVI</span></span>
<span><span class="co">##    [267708]      XVI 912001-928000      * |       760 0.0243999   XVI</span></span>
<span><span class="co">##    [267709]      XVI 928001-944000      * |       761 0.0199324   XVI</span></span>
<span><span class="co">##                center</span></span>
<span><span class="co">##             &lt;integer&gt;</span></span>
<span><span class="co">##         [1]      8000</span></span>
<span><span class="co">##         [2]      8000</span></span>
<span><span class="co">##         [3]      8000</span></span>
<span><span class="co">##         [4]      8000</span></span>
<span><span class="co">##         [5]      8000</span></span>
<span><span class="co">##         ...       ...</span></span>
<span><span class="co">##    [267705]    904000</span></span>
<span><span class="co">##    [267706]    904000</span></span>
<span><span class="co">##    [267707]    920000</span></span>
<span><span class="co">##    [267708]    920000</span></span>
<span><span class="co">##    [267709]    936000</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 16 sequences from an unspecified genome</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $second</span></span>
<span><span class="co">##  GRanges object with 267709 ranges and 4 metadata columns:</span></span>
<span><span class="co">##             seqnames        ranges strand |    bin_id    weight   chr</span></span>
<span><span class="co">##                &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;numeric&gt; &lt;Rle&gt;</span></span>
<span><span class="co">##         [1]        I       1-16000      * |         0 0.0196442     I</span></span>
<span><span class="co">##         [2]        I   16001-32000      * |         1 0.0220746     I</span></span>
<span><span class="co">##         [3]        I   32001-48000      * |         2 0.0188701     I</span></span>
<span><span class="co">##         [4]        I   48001-64000      * |         3 0.0136679     I</span></span>
<span><span class="co">##         [5]        I   64001-80000      * |         4 0.0134860     I</span></span>
<span><span class="co">##         ...      ...           ...    ... .       ...       ...   ...</span></span>
<span><span class="co">##    [267705]      XVI 912001-928000      * |       760 0.0243999   XVI</span></span>
<span><span class="co">##    [267706]      XVI 928001-944000      * |       761 0.0199324   XVI</span></span>
<span><span class="co">##    [267707]      XVI 912001-928000      * |       760 0.0243999   XVI</span></span>
<span><span class="co">##    [267708]      XVI 928001-944000      * |       761 0.0199324   XVI</span></span>
<span><span class="co">##    [267709]      XVI 928001-944000      * |       761 0.0199324   XVI</span></span>
<span><span class="co">##                center</span></span>
<span><span class="co">##             &lt;integer&gt;</span></span>
<span><span class="co">##         [1]      8000</span></span>
<span><span class="co">##         [2]     24000</span></span>
<span><span class="co">##         [3]     40000</span></span>
<span><span class="co">##         [4]     56000</span></span>
<span><span class="co">##         [5]     72000</span></span>
<span><span class="co">##         ...       ...</span></span>
<span><span class="co">##    [267705]    920000</span></span>
<span><span class="co">##    [267706]    936000</span></span>
<span><span class="co">##    [267707]    920000</span></span>
<span><span class="co">##    [267708]    936000</span></span>
<span><span class="co">##    [267709]    936000</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 16 sequences from an unspecified genome</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="bins-and-seqinfo" class="level4" data-number="2.4.2.2"><h4 data-number="2.4.2.2" class="anchored" data-anchor-id="bins-and-seqinfo">
<span class="header-section-number">2.4.2.2</span> Bins and seqinfo</h4>
<p>Additional useful information can be recovered from a <code>HiCExperiment</code> object. This includes:</p>
<ul>
<li>The <code>seqinfo</code> of the <code>HiCExperiment</code>:</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html">seqinfo</a></span><span class="op">(</span><span class="va">yeast_hic</span><span class="op">)</span></span>
<span><span class="co">##  Seqinfo object with 16 sequences from an unspecified genome:</span></span>
<span><span class="co">##    seqnames seqlengths isCircular genome</span></span>
<span><span class="co">##    I            230218       &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">##    II           813184       &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">##    III          316620       &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">##    IV          1531933       &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">##    V            576874       &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">##    ...             ...        ...    ...</span></span>
<span><span class="co">##    XII         1078177       &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">##    XIII         924431       &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">##    XIV          784333       &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">##    XV          1091291       &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">##    XVI          948066       &lt;NA&gt;   &lt;NA&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This lists the different chromosomes available to parse along with their length.</p>
<ul>
<li>The <code>bins</code> of the <code>HiCExperiment</code>:</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">bins</a></span><span class="op">(</span><span class="va">yeast_hic</span><span class="op">)</span></span>
<span><span class="co">##  GRanges object with 763 ranges and 2 metadata columns:</span></span>
<span><span class="co">##                      seqnames        ranges strand |    bin_id     weight</span></span>
<span><span class="co">##                         &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">##            I_1_16000        I       1-16000      * |         0  0.0196442</span></span>
<span><span class="co">##        I_16001_32000        I   16001-32000      * |         1  0.0220746</span></span>
<span><span class="co">##        I_32001_48000        I   32001-48000      * |         2  0.0188701</span></span>
<span><span class="co">##        I_48001_64000        I   48001-64000      * |         3  0.0136679</span></span>
<span><span class="co">##        I_64001_80000        I   64001-80000      * |         4  0.0134860</span></span>
<span><span class="co">##                  ...      ...           ...    ... .       ...        ...</span></span>
<span><span class="co">##    XVI_880001_896000      XVI 880001-896000      * |       758 0.00910873</span></span>
<span><span class="co">##    XVI_896001_912000      XVI 896001-912000      * |       759 0.01421350</span></span>
<span><span class="co">##    XVI_912001_928000      XVI 912001-928000      * |       760 0.02439992</span></span>
<span><span class="co">##    XVI_928001_944000      XVI 928001-944000      * |       761 0.01993237</span></span>
<span><span class="co">##    XVI_944001_948066      XVI 944001-948066      * |       762        NaN</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 16 sequences from an unspecified genome</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Difference between <code>bins</code> and <code>regions</code>
</div>
</div>
<div class="callout-body-container callout-body">
<p><code>bins</code> are <strong>not</strong> equivalent to <code>regions</code> of an <code>HiCExperiment</code>.</p>
<ul>
<li>
<code>bins</code> refer to all the <strong>possible</strong> <code>regions</code> of a <code>HiCExperiment</code>. For instance, for a <code>HiCExperiment</code> with a total genome size of <code>1,000,000</code> and a resolution of <code>2000</code>, <code>bins</code> will always return a <code>GRanges</code> object with <code>500</code> ranges.</li>
<li>
<code>regions</code>, on the opposite, refer to the union of <code>anchors</code> of all the <code>interactions</code> imported in a <code>HiCExperiment</code> object.</li>
</ul>
<p>Thus, all the <code>regions</code> will necessarily be a subset of the <code>HiCExperiment</code> <code>bins</code>, or equal to <code>bins</code> if no focus has been specified when importing a <code>ContactFile</code>.</p>
</div>
</div>
</section><section id="scores" class="level4" data-number="2.4.2.3"><h4 data-number="2.4.2.3" class="anchored" data-anchor-id="scores">
<span class="header-section-number">2.4.2.3</span> Scores</h4>
<p>Of course, what the end-user would be looking for is the <strong>frequency</strong> for each genomic interaction. Such frequency scores are available using the <code>scores</code> function. <code>scores</code> returns a list with a number of different types of scores.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">scores</a></span><span class="op">(</span><span class="va">yeast_hic</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  List of length 2</span></span>
<span><span class="co">##  names(2): count balanced</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">scores</a></span><span class="op">(</span><span class="va">yeast_hic</span>, <span class="st">"count"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  [1] 2836 2212 1183  831  310  159</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">scores</a></span><span class="op">(</span><span class="va">yeast_hic</span>, <span class="st">"balanced"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  [1] 1.09439586 0.95920688 0.43852417 0.22311917 0.08212549 0.03345221</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Calling <code>interactions(hic)</code> returns a <code>GInteractions</code> with <code>scores</code> already stored in extra columns. This short-hand allows one to dynamically check <code>scores</code> directly from the <code>interactions</code> output.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">interactions</a></span><span class="op">(</span><span class="va">yeast_hic</span><span class="op">)</span></span>
<span><span class="co">##  GInteractions object with 267709 interactions and 4 metadata columns:</span></span>
<span><span class="co">##             seqnames1       ranges1     seqnames2       ranges2 |   bin_id1</span></span>
<span><span class="co">##                 &lt;Rle&gt;     &lt;IRanges&gt;         &lt;Rle&gt;     &lt;IRanges&gt; | &lt;numeric&gt;</span></span>
<span><span class="co">##         [1]         I       1-16000 ---         I       1-16000 |         0</span></span>
<span><span class="co">##         [2]         I       1-16000 ---         I   16001-32000 |         0</span></span>
<span><span class="co">##         [3]         I       1-16000 ---         I   32001-48000 |         0</span></span>
<span><span class="co">##         [4]         I       1-16000 ---         I   48001-64000 |         0</span></span>
<span><span class="co">##         [5]         I       1-16000 ---         I   64001-80000 |         0</span></span>
<span><span class="co">##         ...       ...           ... ...       ...           ... .       ...</span></span>
<span><span class="co">##    [267705]       XVI 896001-912000 ---       XVI 912001-928000 |       759</span></span>
<span><span class="co">##    [267706]       XVI 896001-912000 ---       XVI 928001-944000 |       759</span></span>
<span><span class="co">##    [267707]       XVI 912001-928000 ---       XVI 912001-928000 |       760</span></span>
<span><span class="co">##    [267708]       XVI 912001-928000 ---       XVI 928001-944000 |       760</span></span>
<span><span class="co">##    [267709]       XVI 928001-944000 ---       XVI 928001-944000 |       761</span></span>
<span><span class="co">##               bin_id2     count  balanced</span></span>
<span><span class="co">##             &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##         [1]         0      2836 1.0943959</span></span>
<span><span class="co">##         [2]         1      2212 0.9592069</span></span>
<span><span class="co">##         [3]         2      1183 0.4385242</span></span>
<span><span class="co">##         [4]         3       831 0.2231192</span></span>
<span><span class="co">##         [5]         4       310 0.0821255</span></span>
<span><span class="co">##         ...       ...       ...       ...</span></span>
<span><span class="co">##    [267705]       760      3565  1.236371</span></span>
<span><span class="co">##    [267706]       761      1359  0.385016</span></span>
<span><span class="co">##    [267707]       760      3534  2.103988</span></span>
<span><span class="co">##    [267708]       761      3055  1.485794</span></span>
<span><span class="co">##    [267709]       761      4308  1.711565</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    regions: 763 ranges and 4 metadata columns</span></span>
<span><span class="co">##    seqinfo: 16 sequences from an unspecified genome</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">interactions</a></span><span class="op">(</span><span class="va">yeast_hic</span><span class="op">)</span><span class="op">$</span><span class="va">count</span><span class="op">)</span></span>
<span><span class="co">##  [1] 2836 2212 1183  831  310  159</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="topologicalfeatures" class="level4" data-number="2.4.2.4"><h4 data-number="2.4.2.4" class="anchored" data-anchor-id="topologicalfeatures">
<span class="header-section-number">2.4.2.4</span> topologicalFeatures</h4>
<p>In Hi-C studies, “topological features” refer to genomic structures identified (usually from a Hi-C map, but not necessarily). For instance, one may want to study known structural loops anchored at CTCF sites, or interactions around or over centromeres, or simply specific genomic “viewpoints”.</p>
<p><code>HiCExperiment</code> objects can store <code>topologicalFeatures</code> to facilitate this analysis. By default, four empty <code>topologicalFeatures</code> are stored in a list:</p>
<ul>
<li><code>compartments</code></li>
<li><code>borders</code></li>
<li><code>loops</code></li>
<li><code>viewpoints</code></li>
</ul>
<p>Additional <code>topologicalFeatures</code> can be added to this list (read <a href="parsing.html">next chapter</a> for more detail).</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">topologicalFeatures</a></span><span class="op">(</span><span class="va">yeast_hic</span><span class="op">)</span></span>
<span><span class="co">##  List of length 5</span></span>
<span><span class="co">##  names(5): compartments borders loops viewpoints centromeres</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">topologicalFeatures</a></span><span class="op">(</span><span class="va">yeast_hic</span>, <span class="st">'centromeres'</span><span class="op">)</span></span>
<span><span class="co">##  GRanges object with 16 ranges and 0 metadata columns:</span></span>
<span><span class="co">##         seqnames        ranges strand</span></span>
<span><span class="co">##            &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##     [1]        I 151583-151641      +</span></span>
<span><span class="co">##     [2]       II 238361-238419      +</span></span>
<span><span class="co">##     [3]      III 114322-114380      +</span></span>
<span><span class="co">##     [4]       IV 449879-449937      +</span></span>
<span><span class="co">##     [5]        V 152522-152580      +</span></span>
<span><span class="co">##     ...      ...           ...    ...</span></span>
<span><span class="co">##    [12]      XII 151366-151424      +</span></span>
<span><span class="co">##    [13]     XIII 268222-268280      +</span></span>
<span><span class="co">##    [14]      XIV 628588-628646      +</span></span>
<span><span class="co">##    [15]       XV 326897-326955      +</span></span>
<span><span class="co">##    [16]      XVI 556255-556313      +</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 17 sequences (1 circular) from R64-1-1 genome</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="pairsfile" class="level4" data-number="2.4.2.5"><h4 data-number="2.4.2.5" class="anchored" data-anchor-id="pairsfile">
<span class="header-section-number">2.4.2.5</span> pairsFile</h4>
<p>As a contact matrix is typically obtained from binning a <code>.pairs</code> file, it is often the case that the matching <code>.pairs</code> file is available to then end-user. A <code>PairsFile</code> can thus be created and associated to the corresponding <code>HiCExperiment</code> object. This allows more accurate estimation of contact distribution, e.g.&nbsp;when calculating distance-dependent genomic interaction frequency.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">pairsFile</a></span><span class="op">(</span><span class="va">yeast_hic</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">pairsf</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">pairsFile</a></span><span class="op">(</span><span class="va">yeast_hic</span><span class="op">)</span></span>
<span><span class="co">##                                           EH7703 </span></span>
<span><span class="co">##  "/root/.cache/R/ExperimentHub/16d2ae89407_7753"</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readLines.html">readLines</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">pairsFile</a></span><span class="op">(</span><span class="va">yeast_hic</span><span class="op">)</span>, <span class="fl">25</span><span class="op">)</span></span>
<span><span class="co">##   [1] "## pairs format v1.0"                                                             </span></span>
<span><span class="co">##   [2] "#sorted: chr1-pos1-chr2-pos2"                                                     </span></span>
<span><span class="co">##   [3] "#columns: readID chr1 pos1 chr2 pos2 strand1 strand2 frag1 frag2"                 </span></span>
<span><span class="co">##   [4] "#chromsize: I 230218"                                                             </span></span>
<span><span class="co">##   [5] "#chromsize: II 813184"                                                            </span></span>
<span><span class="co">##   [6] "#chromsize: III 316620"                                                           </span></span>
<span><span class="co">##   [7] "#chromsize: IV 1531933"                                                           </span></span>
<span><span class="co">##   [8] "#chromsize: V 576874"                                                             </span></span>
<span><span class="co">##   [9] "#chromsize: VI 270161"                                                            </span></span>
<span><span class="co">##  [10] "#chromsize: VII 1090940"                                                          </span></span>
<span><span class="co">##  [11] "#chromsize: VIII 562643"                                                          </span></span>
<span><span class="co">##  [12] "#chromsize: IX 439888"                                                            </span></span>
<span><span class="co">##  [13] "#chromsize: X 745751"                                                             </span></span>
<span><span class="co">##  [14] "#chromsize: XI 666816"                                                            </span></span>
<span><span class="co">##  [15] "#chromsize: XII 1078177"                                                          </span></span>
<span><span class="co">##  [16] "#chromsize: XIII 924431"                                                          </span></span>
<span><span class="co">##  [17] "#chromsize: XIV 784333"                                                           </span></span>
<span><span class="co">##  [18] "#chromsize: XV 1091291"                                                           </span></span>
<span><span class="co">##  [19] "#chromsize: XVI 948066"                                                           </span></span>
<span><span class="co">##  [20] "#chromsize: Mito 85779"                                                           </span></span>
<span><span class="co">##  [21] "NS500150:527:HHGYNBGXF:3:21611:19085:3986\tII\t105\tII\t48548\t+\t-\t1358\t1681"  </span></span>
<span><span class="co">##  [22] "NS500150:527:HHGYNBGXF:4:13604:19734:2406\tII\t113\tII\t45003\t-\t+\t1358\t1658"  </span></span>
<span><span class="co">##  [23] "NS500150:527:HHGYNBGXF:2:11108:25178:11036\tII\t119\tII\t687251\t-\t+\t1358\t5550"</span></span>
<span><span class="co">##  [24] "NS500150:527:HHGYNBGXF:1:22301:8468:1586\tII\t160\tII\t26124\t+\t-\t1358\t1510"   </span></span>
<span><span class="co">##  [25] "NS500150:527:HHGYNBGXF:4:23606:24037:2076\tII\t169\tII\t39052\t+\t+\t1358\t1613"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="importing-a-pairsfile" class="level4" data-number="2.4.2.6"><h4 data-number="2.4.2.6" class="anchored" data-anchor-id="importing-a-pairsfile">
<span class="header-section-number">2.4.2.6</span> Importing a <code>PairsFile</code>
</h4>
<p>The <code>.pairs</code> file linked to a <code>HiCExperiment</code> object can itself be imported in a <code>GInteractions</code> object:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocIO/man/IO.html">import</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">pairsFile</a></span><span class="op">(</span><span class="va">yeast_hic</span><span class="op">)</span>, format <span class="op">=</span> <span class="st">'pairs'</span><span class="op">)</span></span>
<span><span class="co">##  GInteractions object with 471364 interactions and 3 metadata columns:</span></span>
<span><span class="co">##             seqnames1   ranges1     seqnames2   ranges2 |     frag1     frag2</span></span>
<span><span class="co">##                 &lt;Rle&gt; &lt;IRanges&gt;         &lt;Rle&gt; &lt;IRanges&gt; | &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##         [1]        II       105 ---        II     48548 |      1358      1681</span></span>
<span><span class="co">##         [2]        II       113 ---        II     45003 |      1358      1658</span></span>
<span><span class="co">##         [3]        II       119 ---        II    687251 |      1358      5550</span></span>
<span><span class="co">##         [4]        II       160 ---        II     26124 |      1358      1510</span></span>
<span><span class="co">##         [5]        II       169 ---        II     39052 |      1358      1613</span></span>
<span><span class="co">##         ...       ...       ... ...       ...       ... .       ...       ...</span></span>
<span><span class="co">##    [471360]        II    808605 ---        II    809683 |      6316      6320</span></span>
<span><span class="co">##    [471361]        II    808609 ---        II    809917 |      6316      6324</span></span>
<span><span class="co">##    [471362]        II    808617 ---        II    809506 |      6316      6319</span></span>
<span><span class="co">##    [471363]        II    809447 ---        II    809685 |      6319      6321</span></span>
<span><span class="co">##    [471364]        II    809472 ---        II    809675 |      6319      6320</span></span>
<span><span class="co">##              distance</span></span>
<span><span class="co">##             &lt;integer&gt;</span></span>
<span><span class="co">##         [1]     48443</span></span>
<span><span class="co">##         [2]     44890</span></span>
<span><span class="co">##         [3]    687132</span></span>
<span><span class="co">##         [4]     25964</span></span>
<span><span class="co">##         [5]     38883</span></span>
<span><span class="co">##         ...       ...</span></span>
<span><span class="co">##    [471360]      1078</span></span>
<span><span class="co">##    [471361]      1308</span></span>
<span><span class="co">##    [471362]       889</span></span>
<span><span class="co">##    [471363]       238</span></span>
<span><span class="co">##    [471364]       203</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    regions: 549331 ranges and 0 metadata columns</span></span>
<span><span class="co">##    seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note that these <code>GInteractions</code> are <strong>not</strong> binned, contrary to <code>interactions</code> extracted from a <code>HiCExperiment</code>. Anchors of the interactions listed in the <code>GInteractions</code> imported from a disk-stored <code>.pairs</code> file are all of width <code>1</code>.</p>
</section></section></section><section id="visual-summary-of-the-hicexperiment-data-structure" class="level2" data-number="2.5"><h2 data-number="2.5" class="anchored" data-anchor-id="visual-summary-of-the-hicexperiment-data-structure">
<span class="header-section-number">2.5</span> Visual summary of the <code>HiCExperiment</code> data structure</h2>
<p>The <code>HiCExperiment</code> data structure provided by the <code>HiCExperiment</code> package inherits methods from core <code>GInteractions</code> and <code>BiocFile</code> classes to provide a flexible representation of Hi-C data in <code>R</code>. It allows random access-based queries to seamlessly import parts or all the data contained in disk-stored Hi-C contact matrices in a variety of formats.</p>
<p><img src="images/20230309114202.png" class="img-fluid"></p>
</section><section id="references" class="level1 unnumbered"><h1 class="unnumbered">References</h1>


<div id="refs" class="references csl-bib-body hanging-indent" data-line-spacing="2" role="list">
<div id="ref-Huber_2015" class="csl-entry" role="listitem">
Huber, W., Carey, V. J., Gentleman, R., Anders, S., Carlson, M., Carvalho, B. S., Bravo, H. C., Davis, S., Gatto, L., Girke, T., Gottardo, R., Hahne, F., Hansen, K. D., Irizarry, R. A., Lawrence, M., Love, M. I., MacDonald, J., Obenchain, V., Oleś, A. K., … Morgan, M. (2015). Orchestrating high-throughput genomic analysis with bioconductor. <em>Nature Methods</em>, <em>12</em>(2), 115–121. <a href="https://doi.org/10.1038/nmeth.3252">https://doi.org/10.1038/nmeth.3252</a>
</div>
<div id="ref-Lun_2016" class="csl-entry" role="listitem">
Lun, A. T. L., Perry, M., &amp; Ing-Simmons, E. (2016). Infrastructure for genomic interactions: Bioconductor classes for hi-c, <span>ChIA</span>-<span>PET</span> and related experiments. <em>F1000Research</em>, <em>5</em>, 950. <a href="https://doi.org/10.12688/f1000research.8759.2">https://doi.org/10.12688/f1000research.8759.2</a>
</div>
</div>
</section><a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../pages/principles.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Hi-C pre-processing steps</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../pages/parsing.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Manipulating Hi-C data in R</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left"><i>OHCA: Orchestrating Hi-C analysis with Bioconductor</i> was written by J. Serizay</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">This book was built with <a href="https://github.com/js2264/BiocBook/">BiocBook</a> with <span class="emoji" data-emoji="heart">❤️</span></div>
  </div>
</footer>


</body></html>