<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Orchestrating Hi-C analysis with Bioconductor - 8&nbsp; Data gateways: accessing public Hi-C data portals</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../pages/interoperability.html" rel="next">
<link href="../pages/topological-features.html" rel="prev">
<link href="../assets/cover.jpg" rel="icon" type="image/jpeg">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
</head>
<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/disseminating.html">Advanced Hi-C topics</a></li><li class="breadcrumb-item"><a href="../pages/disseminating.html"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Data gateways: accessing public Hi-C data portals</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../">Orchestrating Hi-C analysis with Bioconductor</a> 
        <div class="sidebar-tools-main">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label=""><i class="bi bi-git"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
<li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://github.com/js2264/OHCA/">
            Source Code
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://js2264.github.io/OHCA/devel/">
            Browse version `devel`
            </a>
          </li>
      </ul>
</div>
    <a href="../Orchestrating-Hi-C-analysis-with-Bioconductor.pdf" rel="" title="Download PDF" class="quarto-navigation-tool px-1" aria-label="Download PDF"><i class="bi bi-file-pdf"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/preamble.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preamble</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Fundamentals concepts</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/principles.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Hi-C pre-processing steps</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/data-representation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Hi-C data structures in R</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/parsing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Manipulating Hi-C data in R</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/visualization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Hi-C data visualization</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">In-depth Hi-C analysis</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/matrix-centric.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Matrix-centric analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/interactions-centric.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Interactions-centric analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/topological-features.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Finding topological features in Hi-C</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Advanced Hi-C topics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/disseminating.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Data gateways: accessing public Hi-C data portals</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/interoperability.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Interoperability: using HiCExperiment with other R packages</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/workflow-yeast.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workflow 1: Distance-dependent interactions across yeast mutants</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/workflow-chicken.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workflow 2: Chromosome compartment cohesion upon mitosis entry</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/workflow-centros.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workflow 3: Inter-centromere interactions in yeast</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li>
<a href="#dn-data-portal" id="toc-dn-data-portal" class="nav-link active" data-scroll-target="#dn-data-portal"><span class="header-section-number">8.1</span> 4DN data portal</a>
  <ul class="collapse">
<li><a href="#querying-individual-files" id="toc-querying-individual-files" class="nav-link" data-scroll-target="#querying-individual-files"><span class="header-section-number">8.1.1</span> Querying individual files</a></li>
  <li><a href="#querying-a-complete-experiment-dataset" id="toc-querying-a-complete-experiment-dataset" class="nav-link" data-scroll-target="#querying-a-complete-experiment-dataset"><span class="header-section-number">8.1.2</span> Querying a complete experiment dataset</a></li>
  </ul>
</li>
  <li><a href="#dna-zoo" id="toc-dna-zoo" class="nav-link" data-scroll-target="#dna-zoo"><span class="header-section-number">8.2</span> DNA Zoo</a></li>
  <li><a href="#session-info" id="toc-session-info" class="nav-link" data-scroll-target="#session-info">Session info</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title">
<span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Data gateways: accessing public Hi-C data portals</span>
</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Pre-loading packages and objects 📦
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges">GenomicRanges</a></span><span class="op">)</span></span>
<span><span class="co">##  Loading required package: stats4</span></span>
<span><span class="co">##  Loading required package: BiocGenerics</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Attaching package: 'BiocGenerics'</span></span>
<span><span class="co">##  The following objects are masked from 'package:stats':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      IQR, mad, sd, var, xtabs</span></span>
<span><span class="co">##  The following objects are masked from 'package:base':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      Filter, Find, Map, Position, Reduce, anyDuplicated, aperm,</span></span>
<span><span class="co">##      append, as.data.frame, basename, cbind, colnames, dirname,</span></span>
<span><span class="co">##      do.call, duplicated, eval, evalq, get, grep, grepl, intersect,</span></span>
<span><span class="co">##      is.unsorted, lapply, mapply, match, mget, order, paste, pmax,</span></span>
<span><span class="co">##      pmax.int, pmin, pmin.int, rank, rbind, rownames, sapply,</span></span>
<span><span class="co">##      setdiff, table, tapply, union, unique, unsplit, which.max,</span></span>
<span><span class="co">##      which.min</span></span>
<span><span class="co">##  Loading required package: S4Vectors</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Attaching package: 'S4Vectors'</span></span>
<span><span class="co">##  The following object is masked from 'package:utils':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      findMatches</span></span>
<span><span class="co">##  The following objects are masked from 'package:base':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      I, expand.grid, unname</span></span>
<span><span class="co">##  Loading required package: IRanges</span></span>
<span><span class="co">##  Loading required package: GenomeInfoDb</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">InteractionSet</span><span class="op">)</span></span>
<span><span class="co">##  Loading required package: SummarizedExperiment</span></span>
<span><span class="co">##  Loading required package: MatrixGenerics</span></span>
<span><span class="co">##  Loading required package: matrixStats</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Attaching package: 'MatrixGenerics'</span></span>
<span><span class="co">##  The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span></span>
<span><span class="co">##      colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span></span>
<span><span class="co">##      colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span></span>
<span><span class="co">##      colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span></span>
<span><span class="co">##      colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span></span>
<span><span class="co">##      colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span></span>
<span><span class="co">##      colWeightedMeans, colWeightedMedians, colWeightedSds,</span></span>
<span><span class="co">##      colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span></span>
<span><span class="co">##      rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span></span>
<span><span class="co">##      rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span></span>
<span><span class="co">##      rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span></span>
<span><span class="co">##      rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span></span>
<span><span class="co">##      rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs,</span></span>
<span><span class="co">##      rowVars, rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span></span>
<span><span class="co">##      rowWeightedSds, rowWeightedVars</span></span>
<span><span class="co">##  Loading required package: Biobase</span></span>
<span><span class="co">##  Welcome to Bioconductor</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      Vignettes contain introductory material; view with</span></span>
<span><span class="co">##      'browseVignettes()'. To cite Bioconductor, see</span></span>
<span><span class="co">##      'citation("Biobase")', and for packages 'citation("pkgname")'.</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Attaching package: 'Biobase'</span></span>
<span><span class="co">##  The following object is masked from 'package:MatrixGenerics':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      rowMedians</span></span>
<span><span class="co">##  The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      anyMissing, rowMedians</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/js2264/HiCExperiment">HiCExperiment</a></span><span class="op">)</span></span>
<span><span class="co">##  Consider using the `HiContacts` package to perform advanced genomic operations </span></span>
<span><span class="co">##  on `HiCExperiment` objects.</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Read "Orchestrating Hi-C analysis with Bioconductor" online book to learn more:</span></span>
<span><span class="co">##  https://js2264.github.io/OHCA/</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Attaching package: 'HiCExperiment'</span></span>
<span><span class="co">##  The following object is masked from 'package:SummarizedExperiment':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      metadata&lt;-</span></span>
<span><span class="co">##  The following object is masked from 'package:S4Vectors':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      metadata&lt;-</span></span>
<span><span class="co">##  The following object is masked from 'package:ggplot2':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      resolution</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/js2264/fourDNData">fourDNData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/js2264/DNAZooData">DNAZooData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Aims
</div>
</div>
<div class="callout-body-container callout-body">
<p>This chapter focuses on introducing two important portals hosting public Hi-C datasets: the <a href="https://www.4dnucleome.org/">4DN Consortium</a> and the <a href="https://www.dnazoo.org/">DNA Zoo project</a>. Two <code>R</code> packages provide a programmatic access to these portals:</p>
<ul>
<li><code>fourDNData</code></li>
<li><code>DNAZooData</code></li>
</ul>
</div>
</div>
<p>The Hi-C experimental approach has gained significant traction across multiple fields related to genome biology, and several consortia developed large-scale programs based on this technique. The <code>fourDNData</code> and <code>DNAZooData</code> R packages were designed to accelerate the investigation of chromatin structure using these public resources.</p>
<section id="dn-data-portal" class="level2" data-number="8.1"><h2 data-number="8.1" class="anchored" data-anchor-id="dn-data-portal">
<span class="header-section-number">8.1</span> 4DN data portal</h2>
<p>The 4D Nucleome Data Coordination and Integration Center (<a href="http://dcic.4dnucleome.org/">DCIC</a>) has developed and actively maintains a <a href="https://data.4dnucleome.org/browse">data portal</a> providing public access to a wealth of resources to investigate 3D chromatin architecture. Notably, 3D chromatin conformation libraries relying on different technologies (“in situ” or “dilution” Hi-C, Capture Hi-C, Micro-C, DNase Hi-C, …), generated by 50+ collaborating labs, were homogeneously processed, yielding more than 350 sets of processed files.</p>
<p><code>fourDNData</code> (read <em>4DN-Data</em>) is a package giving programmatic access to these uniformly processed Hi-C contact files.</p>
<p>The <code><a href="https://rdrr.io/pkg/fourDNData/man/fourDNData.html">fourDNData()</a></code> function provides a gateway to 4DN-hosted Hi-C files, including contact matrices (in <code>.hic</code> or <code>.mcool</code>) and other Hi-C derived files such as annotated compartments, domains, insulation scores, or <code>.pairs</code> files.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/js2264/fourDNData">fourDNData</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/fourDNData/man/fourDNData.html">fourDNData</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##    experimentSetAccession     fileType     size organism experimentType</span></span>
<span><span class="co">##  1           4DNES18BMU79        pairs 10151.53    mouse   in situ Hi-C</span></span>
<span><span class="co">##  3           4DNES18BMU79          hic  5285.82    mouse   in situ Hi-C</span></span>
<span><span class="co">##  4           4DNES18BMU79        mcool  6110.75    mouse   in situ Hi-C</span></span>
<span><span class="co">##  5           4DNES18BMU79   boundaries     0.12    mouse   in situ Hi-C</span></span>
<span><span class="co">##  6           4DNES18BMU79   insulation     7.18    mouse   in situ Hi-C</span></span>
<span><span class="co">##  7           4DNES18BMU79 compartments     0.18    mouse   in situ Hi-C</span></span>
<span><span class="co">##    details                              dataset</span></span>
<span><span class="co">##  1   DpnII Hi-C on Mouse Olfactory System cells</span></span>
<span><span class="co">##  3   DpnII Hi-C on Mouse Olfactory System cells</span></span>
<span><span class="co">##  4   DpnII Hi-C on Mouse Olfactory System cells</span></span>
<span><span class="co">##  5   DpnII Hi-C on Mouse Olfactory System cells</span></span>
<span><span class="co">##  6   DpnII Hi-C on Mouse Olfactory System cells</span></span>
<span><span class="co">##  7   DpnII Hi-C on Mouse Olfactory System cells</span></span>
<span><span class="co">##                                                          condition</span></span>
<span><span class="co">##  1 Mature olfactory sensory neurons with conditional Ldb1 knockout</span></span>
<span><span class="co">##  3 Mature olfactory sensory neurons with conditional Ldb1 knockout</span></span>
<span><span class="co">##  4 Mature olfactory sensory neurons with conditional Ldb1 knockout</span></span>
<span><span class="co">##  5 Mature olfactory sensory neurons with conditional Ldb1 knockout</span></span>
<span><span class="co">##  6 Mature olfactory sensory neurons with conditional Ldb1 knockout</span></span>
<span><span class="co">##  7 Mature olfactory sensory neurons with conditional Ldb1 knockout</span></span>
<span><span class="co">##                  biosource biosourceType             publication</span></span>
<span><span class="co">##  1 olfactory receptor cell  primary cell Monahan K et al. (2019)</span></span>
<span><span class="co">##  3 olfactory receptor cell  primary cell Monahan K et al. (2019)</span></span>
<span><span class="co">##  4 olfactory receptor cell  primary cell Monahan K et al. (2019)</span></span>
<span><span class="co">##  5 olfactory receptor cell  primary cell Monahan K et al. (2019)</span></span>
<span><span class="co">##  6 olfactory receptor cell  primary cell Monahan K et al. (2019)</span></span>
<span><span class="co">##  7 olfactory receptor cell  primary cell Monahan K et al. (2019)</span></span>
<span><span class="co">##                                                                                                                                    URL</span></span>
<span><span class="co">##  1 https://4dn-open-data-public.s3.amazonaws.com/fourfront-webprod/wfoutput/49504f97-904e-48c1-8c20-1033680b66da/4DNFIC5AHBPV.pairs.gz</span></span>
<span><span class="co">##  3      https://4dn-open-data-public.s3.amazonaws.com/fourfront-webprod/wfoutput/6cd4378a-8f51-4e65-99eb-15f5c80abf8d/4DNFIT4I5C6Z.hic</span></span>
<span><span class="co">##  4    https://4dn-open-data-public.s3.amazonaws.com/fourfront-webprod/wfoutput/01fb704f-2fd7-48c6-91af-c5f4584529ed/4DNFIVPAXJO8.mcool</span></span>
<span><span class="co">##  5   https://4dn-open-data-public.s3.amazonaws.com/fourfront-webprod/wfoutput/5c07cdee-53e2-43e0-8853-cfe5f057b3f1/4DNFIR3XCIMA.bed.gz</span></span>
<span><span class="co">##  6       https://4dn-open-data-public.s3.amazonaws.com/fourfront-webprod/wfoutput/d1f4beb9-701f-4188-abe2-6271fe658770/4DNFIXKKNMS7.bw</span></span>
<span><span class="co">##  7       https://4dn-open-data-public.s3.amazonaws.com/fourfront-webprod/wfoutput/3d429647-51c8-4e3a-a18b-eec0b1480905/4DNFIN13N8C1.bw</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="querying-individual-files" class="level3" data-number="8.1.1"><h3 data-number="8.1.1" class="anchored" data-anchor-id="querying-individual-files">
<span class="header-section-number">8.1.1</span> Querying individual files</h3>
<p>The <code><a href="https://rdrr.io/pkg/fourDNData/man/fourDNData.html">fourDNData()</a></code> function can be used to directly fetch specific files from the 4DN data portal:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/fourDNData/man/fourDNData.html">fourDNData</a></span><span class="op">(</span>experimentSetAccession <span class="op">=</span> <span class="st">'4DNESJNPEKZD'</span>, type <span class="op">=</span> <span class="st">'mcool'</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This effectively downloads and caches the queried file locally.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cf</span></span>
<span><span class="co">##  [1] "/root/.cache/R/fourDNData/150e6c94d4df_4DNFIZL8OZE1.mcool"</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">availableChromosomes</a></span><span class="op">(</span><span class="va">cf</span><span class="op">)</span></span>
<span><span class="co">##  Seqinfo object with 24 sequences from an unspecified genome:</span></span>
<span><span class="co">##    seqnames seqlengths isCircular genome</span></span>
<span><span class="co">##    chr1      248956422       &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">##    chr2      242193529       &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">##    chr3      198295559       &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">##    chr4      190214555       &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">##    chr5      181538259       &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">##    ...             ...        ...    ...</span></span>
<span><span class="co">##    chr20      64444167       &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">##    chr21      46709983       &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">##    chr22      50818468       &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">##    chrX      156040895       &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">##    chrY       57227415       &lt;NA&gt;   &lt;NA&gt;</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">availableResolutions</a></span><span class="op">(</span><span class="va">cf</span><span class="op">)</span></span>
<span><span class="co">##  resolutions(13): 1000 2000 ... 5e+06 1e+07</span></span>
<span><span class="co">##  </span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocIO/man/IO.html">import</a></span><span class="op">(</span><span class="va">cf</span>, focus <span class="op">=</span> <span class="st">"chr4:10000001-20000000"</span>, resolution <span class="op">=</span> <span class="fl">5000</span><span class="op">)</span></span>
<span><span class="co">##  `HiCExperiment` object with 656 contacts over 2,000 regions </span></span>
<span><span class="co">##  -------</span></span>
<span><span class="co">##  fileName: "/root/.cache/R/fourDNData/150e6c94d4df_4DNFIZL8OZE1.mcool" </span></span>
<span><span class="co">##  focus: "chr4:10,000,001-20,000,000" </span></span>
<span><span class="co">##  resolutions(13): 1000 2000 ... 5000000 10000000</span></span>
<span><span class="co">##  active resolution: 5000 </span></span>
<span><span class="co">##  interactions: 614 </span></span>
<span><span class="co">##  scores(2): count balanced </span></span>
<span><span class="co">##  topologicalFeatures: compartments(0) borders(0) loops(0) viewpoints(0) </span></span>
<span><span class="co">##  pairsFile: N/A </span></span>
<span><span class="co">##  metadata(0):</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Different Hi-C related genomic files are provided by the 4DN consortium. The type of file to fetch can be specified with the <code>type</code> argument:</p>
<ul>
<li>
<code>type = 'pairs'</code> will fetch the pairs file which was generated by the 4DN consortium and binned into a contact matrix. Once fetched from the 4DN data portal, the local file can be imported in <code>R</code> using the <code>import</code> function, which will generate a <code>GInteractions</code> object.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## Not evaluated for now</span></span>
<span><span class="va">pairs_f</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/fourDNData/man/fourDNData.html">fourDNData</a></span><span class="op">(</span>experimentSetAccession <span class="op">=</span> <span class="st">'4DNESJNPEKZD'</span>, type <span class="op">=</span> <span class="st">'pairs'</span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">pairs_f</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocIO/man/IO.html">import</a></span><span class="op">(</span><span class="va">pairs_f</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Watch out
</div>
</div>
<div class="callout-body-container callout-body">
<p><code>.pairs</code> files can be particularly large and therefore will take both and long time to download and a large storage footprint.</p>
</div>
</div>
<ul>
<li>
<code>type = 'insulation'</code> will fetch a <code>.bigwig</code> track file precomputed by the 4DN consortium. This track corresponds to the genome-wide insulation score computed by <code>cooltools</code> as described in <span class="citation" data-cites="Crane_2015">Crane et al. (<a href="#ref-Crane_2015" role="doc-biblioref">2015</a>)</span>. To know more about this, read <a href="https://data.4dnucleome.org/resources/data-analysis/insulation_compartment_scores#insulation_scores_and_boundaries_page_all">the excerpt from 4DN data portal</a>. Once fetched from the 4DN data portal, the local file can be imported in <code>R</code> using the <code>import</code> function, which will generate a <code>RleList</code> object.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/fourDNData/man/fourDNData.html">fourDNData</a></span><span class="op">(</span>experimentSetAccession <span class="op">=</span> <span class="st">'4DNES25ABNZ1'</span>, type <span class="op">=</span> <span class="st">'insulation'</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocIO/man/IO.html">import</a></span><span class="op">(</span>as <span class="op">=</span> <span class="st">'Rle'</span><span class="op">)</span></span>
<span><span class="co">##  RleList of length 21</span></span>
<span><span class="co">##  $chr1</span></span>
<span><span class="co">##  numeric-Rle of length 195471971 with 38145 runs</span></span>
<span><span class="co">##    Lengths:      3065000         5000 ...         5000       171971</span></span>
<span><span class="co">##    Values :  0.00000e+00  1.01441e-01 ...     0.807009     0.000000</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $chr10</span></span>
<span><span class="co">##  numeric-Rle of length 130694993 with 25100 runs</span></span>
<span><span class="co">##    Lengths:     3175000        5000        5000 ...        5000      169993</span></span>
<span><span class="co">##    Values :  0.00000000  0.37584546  0.33597839 ...    0.628601    0.000000</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $chr11</span></span>
<span><span class="co">##  numeric-Rle of length 122082543 with 23536 runs</span></span>
<span><span class="co">##    Lengths:    3165000       5000       5000 ...       5000     162543</span></span>
<span><span class="co">##    Values :  0.0000000 -0.7906257 -0.7930040 ...   0.515919   0.000000</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $chr12</span></span>
<span><span class="co">##  numeric-Rle of length 120129022 with 22578 runs</span></span>
<span><span class="co">##    Lengths:   3075000      5000      5000 ...      5000      5000    164022</span></span>
<span><span class="co">##    Values :  0.000000  0.411216  0.400357 ... 0.1650951 0.2175749 0.0000000</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $chr13</span></span>
<span><span class="co">##  numeric-Rle of length 120421639 with 22807 runs</span></span>
<span><span class="co">##    Lengths:     3080000        5000        5000 ...        5000      171639</span></span>
<span><span class="co">##    Values :  0.00000000  0.17005745  0.10652249 ...  1.14856148  0.00000000</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  ...</span></span>
<span><span class="co">##  &lt;16 more elements&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>
<code>type = 'boundaries'</code> will fetch a <code>.bed</code> file precomputed by the 4DN consortium, listing the annotated borders between topological domains. These borders correspond to local minima identified from the genome-wide insulation track. It can also be imported in <code>R</code> using the <code>import</code> function, which will generate a <code>GRanges</code> object.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/fourDNData/man/fourDNData.html">fourDNData</a></span><span class="op">(</span>experimentSetAccession <span class="op">=</span> <span class="st">'4DNES25ABNZ1'</span>, type <span class="op">=</span> <span class="st">'boundaries'</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocIO/man/IO.html">import</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">##  GRanges object with 6103 ranges and 2 metadata columns:</span></span>
<span><span class="co">##           seqnames            ranges strand |        name     score</span></span>
<span><span class="co">##              &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##       [1]     chr1   4380001-4385000      * |      Strong  0.695274</span></span>
<span><span class="co">##       [2]     chr1   4760001-4765000      * |        Weak  0.444476</span></span>
<span><span class="co">##       [3]     chr1   4910001-4915000      * |        Weak  0.353184</span></span>
<span><span class="co">##       [4]     chr1   5180001-5185000      * |      Strong  0.565763</span></span>
<span><span class="co">##       [5]     chr1   6170001-6175000      * |      Strong  1.644911</span></span>
<span><span class="co">##       ...      ...               ...    ... .         ...       ...</span></span>
<span><span class="co">##    [6099]     chrY 89725001-89730000      * |        Weak  0.258094</span></span>
<span><span class="co">##    [6100]     chrY 89790001-89795000      * |        Weak  0.442186</span></span>
<span><span class="co">##    [6101]     chrY 89895001-89900000      * |        Weak  0.279879</span></span>
<span><span class="co">##    [6102]     chrY 90025001-90030000      * |      Strong  0.660699</span></span>
<span><span class="co">##    [6103]     chrY 90410001-90415000      * |      Strong  1.160018</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 21 sequences from an unspecified genome; no seqlengths</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>
<code>type = 'compartments'</code> will fetch a <code>.bigwig</code> track file precomputed by the 4DN consortium. This track corresponds to the selected genome-wide eigenvector computed by <code>cooltools</code> and representing A/B compartments. To know more about this, read <a href="https://data.4dnucleome.org/resources/data-analysis/insulation_compartment_scores#compartments_page_all">the excerpt from 4DN data portal</a>. Once fetched from the 4DN data portal, the local file can be imported in <code>R</code> using the <code>import</code> function, which will generate a <code>RleList</code> object. The score represents the eigenvector values, and by convention a genomic bin with a positive score is associated with the A compartment whereas a genomic bin with a negative score is associated with the B compartment.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/fourDNData/man/fourDNData.html">fourDNData</a></span><span class="op">(</span>experimentSetAccession <span class="op">=</span> <span class="st">'4DNES25ABNZ1'</span>, type <span class="op">=</span> <span class="st">'compartments'</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocIO/man/IO.html">import</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">##  GRanges object with 10911 ranges and 1 metadata column:</span></span>
<span><span class="co">##            seqnames            ranges strand |     score</span></span>
<span><span class="co">##               &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;</span></span>
<span><span class="co">##        [1]     chr1          1-250000      * |       NaN</span></span>
<span><span class="co">##        [2]     chr1     250001-500000      * |       NaN</span></span>
<span><span class="co">##        [3]     chr1     500001-750000      * |       NaN</span></span>
<span><span class="co">##        [4]     chr1    750001-1000000      * |       NaN</span></span>
<span><span class="co">##        [5]     chr1   1000001-1250000      * |       NaN</span></span>
<span><span class="co">##        ...      ...               ...    ... .       ...</span></span>
<span><span class="co">##    [10907]     chrY 90500001-90750000      * | 0.0237907</span></span>
<span><span class="co">##    [10908]     chrY 90750001-91000000      * |       NaN</span></span>
<span><span class="co">##    [10909]     chrY 91000001-91250000      * |       NaN</span></span>
<span><span class="co">##    [10910]     chrY 91250001-91500000      * |       NaN</span></span>
<span><span class="co">##    [10911]     chrY 91500001-91744698      * |       NaN</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 21 sequences from an unspecified genome</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="querying-a-complete-experiment-dataset" class="level3" data-number="8.1.2"><h3 data-number="8.1.2" class="anchored" data-anchor-id="querying-a-complete-experiment-dataset">
<span class="header-section-number">8.1.2</span> Querying a complete experiment dataset</h3>
<p>Rather than importing multiple files corresponding to a single experimentSet accession ID one by one, one can import all the available files associated with a experimentSet accession ID into a <code>HiCExperiment</code> object by using the <code><a href="https://rdrr.io/pkg/fourDNData/man/fourDNData.html">fourDNHiCExperiment()</a></code> function.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hic</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/fourDNData/man/fourDNData.html">fourDNHiCExperiment</a></span><span class="op">(</span><span class="st">'4DNESJNPEKZD'</span><span class="op">)</span></span>
<span><span class="co">##  Fetching local Hi-C contact map from Bioc cache</span></span>
<span><span class="co">##  Fetching compartments bigwig file from 4DN portal</span></span>
<span><span class="co">##  Insulation not found for the provided experimentSet accession.</span></span>
<span><span class="co">##  Borders not found for the provided experimentSet accession.</span></span>
<span><span class="co">##  Importing contacts in memory</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This is a more efficient way to import datasets, as it aggregates the different bits together into a single <code>HiCExperiment</code> object with populated <code>topologicalFeatures</code> and <code>metadata</code> slots.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hic</span></span>
<span><span class="co">##  `HiCExperiment` object with 453,301 contacts over 12,366 regions </span></span>
<span><span class="co">##  -------</span></span>
<span><span class="co">##  fileName: "/root/.cache/R/fourDNData/150e6c94d4df_4DNFIZL8OZE1.mcool" </span></span>
<span><span class="co">##  focus: "whole genome" </span></span>
<span><span class="co">##  resolutions(13): 1000 2000 ... 5000000 10000000</span></span>
<span><span class="co">##  active resolution: 250000 </span></span>
<span><span class="co">##  interactions: 289086 </span></span>
<span><span class="co">##  scores(2): count balanced </span></span>
<span><span class="co">##  topologicalFeatures: compartments(5437) borders(0) </span></span>
<span><span class="co">##  pairsFile: N/A </span></span>
<span><span class="co">##  metadata(2): 4DN_info eigens</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html">metadata</a></span><span class="op">(</span><span class="va">hic</span><span class="op">)</span></span>
<span><span class="co">##  $`4DN_info`</span></span>
<span><span class="co">##       experimentSetAccession     fileType   size organism experimentType</span></span>
<span><span class="co">##  1376           4DNESJNPEKZD        pairs   6.67    human   in situ Hi-C</span></span>
<span><span class="co">##  1378           4DNESJNPEKZD          hic 179.51    human   in situ Hi-C</span></span>
<span><span class="co">##  1379           4DNESJNPEKZD        mcool  30.17    human   in situ Hi-C</span></span>
<span><span class="co">##  1380           4DNESJNPEKZD compartments   0.21    human   in situ Hi-C</span></span>
<span><span class="co">##       details                                     dataset</span></span>
<span><span class="co">##  1376    MboI Hi-C on GM12878 cells - protocol variations</span></span>
<span><span class="co">##  1378    MboI Hi-C on GM12878 cells - protocol variations</span></span>
<span><span class="co">##  1379    MboI Hi-C on GM12878 cells - protocol variations</span></span>
<span><span class="co">##  1380    MboI Hi-C on GM12878 cells - protocol variations</span></span>
<span><span class="co">##                                                       condition biosource</span></span>
<span><span class="co">##  1376 in situ Hi-C on GM12878 with MboI and bio-dUTP (Tri-Link)   GM12878</span></span>
<span><span class="co">##  1378 in situ Hi-C on GM12878 with MboI and bio-dUTP (Tri-Link)   GM12878</span></span>
<span><span class="co">##  1379 in situ Hi-C on GM12878 with MboI and bio-dUTP (Tri-Link)   GM12878</span></span>
<span><span class="co">##  1380 in situ Hi-C on GM12878 with MboI and bio-dUTP (Tri-Link)   GM12878</span></span>
<span><span class="co">##                biosourceType          publication</span></span>
<span><span class="co">##  1376 immortalized cell line Rao SS et al. (2014)</span></span>
<span><span class="co">##  1378 immortalized cell line Rao SS et al. (2014)</span></span>
<span><span class="co">##  1379 immortalized cell line Rao SS et al. (2014)</span></span>
<span><span class="co">##  1380 immortalized cell line Rao SS et al. (2014)</span></span>
<span><span class="co">##                                                                                                                                       URL</span></span>
<span><span class="co">##  1376 https://4dn-open-data-public.s3.amazonaws.com/fourfront-webprod/wfoutput/0bdd4745-7203-49d0-adf6-291cef1a96b7/4DNFIOZ7D1OQ.pairs.gz</span></span>
<span><span class="co">##  1378      https://4dn-open-data-public.s3.amazonaws.com/fourfront-webprod/wfoutput/1201682a-a223-482d-913d-3c3972b8eb65/4DNFIIRIHBR2.hic</span></span>
<span><span class="co">##  1379    https://4dn-open-data-public.s3.amazonaws.com/fourfront-webprod/wfoutput/356fab42-5562-4cfd-a3f8-592aa060b992/4DNFIZL8OZE1.mcool</span></span>
<span><span class="co">##  1380       https://4dn-open-data-public.s3.amazonaws.com/fourfront-webprod/wfoutput/333aabfd-b747-447c-b93a-8138f9488fad/4DNFIO9V5G93.bw</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $eigens</span></span>
<span><span class="co">##  GRanges object with 11280 ranges and 2 metadata columns:</span></span>
<span><span class="co">##            seqnames              ranges strand |       score       eigen</span></span>
<span><span class="co">##               &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |   &lt;numeric&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">##        [1]     chr1      750001-1000000      * |   1.6911879   1.6911879</span></span>
<span><span class="co">##        [2]     chr1     1000001-1250000      * |   0.0809129   0.0809129</span></span>
<span><span class="co">##        [3]     chr1     1250001-1500000      * |   0.0690173   0.0690173</span></span>
<span><span class="co">##        [4]     chr1     1500001-1750000      * |  -0.1903324  -0.1903324</span></span>
<span><span class="co">##        [5]     chr1     1750001-2000000      * |   0.3283633   0.3283633</span></span>
<span><span class="co">##        ...      ...                 ...    ... .         ...         ...</span></span>
<span><span class="co">##    [11276]     chrX 154750001-155000000      * | -0.10909061 -0.10909061</span></span>
<span><span class="co">##    [11277]     chrX 155000001-155250000      * | -1.39655280 -1.39655280</span></span>
<span><span class="co">##    [11278]     chrX 155250001-155500000      * |  0.00264734  0.00264734</span></span>
<span><span class="co">##    [11279]     chrX 155500001-155750000      * | -0.15279847 -0.15279847</span></span>
<span><span class="co">##    [11280]     chrX 155750001-156000000      * | -1.41699576 -1.41699576</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 24 sequences from an unspecified genome</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section><section id="dna-zoo" class="level2" data-number="8.2"><h2 data-number="8.2" class="anchored" data-anchor-id="dna-zoo">
<span class="header-section-number">8.2</span> DNA Zoo</h2>
<p>The <a href="https://www.dnazoo.org/">DNA Zoo Consortium</a> is a collaborative group whose aim is to correct and refine genome assemblies across the tree of life using Hi-C approaches. As of 2023, they have performed Hi-C across more than 300 animal, plant and fungi species.</p>
<p><code>DNAZooData</code> is a package giving programmatic access to these uniformly processed Hi-C contact files, as well as the refined genome assemblies.</p>
<p>The <code><a href="https://rdrr.io/pkg/DNAZooData/man/DNAZooData.html">DNAZooData()</a></code> function provides a gateway to DNA Zoo-hosted Hi-C files, fetching and caching relevant contact matrices in <code>.hic</code> format It returns a <code>HicFile</code> object, which can then be imported in memory using <code><a href="https://rdrr.io/pkg/BiocIO/man/IO.html">import()</a></code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/js2264/DNAZooData">DNAZooData</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/DNAZooData/man/DNAZooData.html">DNAZooData</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##                    species                              readme</span></span>
<span><span class="co">##  1        Acinonyx_jubatus        Acinonyx_jubatus/README.json</span></span>
<span><span class="co">##  2      Acropora_millepora      Acropora_millepora/README.json</span></span>
<span><span class="co">##  3     Addax_nasomaculatus     Addax_nasomaculatus/README.json</span></span>
<span><span class="co">##  4           Aedes_aegypti           Aedes_aegypti/README.json</span></span>
<span><span class="co">##  5   Aedes_aegypti__AaegL4   Aedes_aegypti__AaegL4/README.json</span></span>
<span><span class="co">##  6 Aedes_aegypti__AaegL5.0 Aedes_aegypti__AaegL5.0/README.json</span></span>
<span><span class="co">##                                                            readme_link</span></span>
<span><span class="co">##  1        https://dnazoo.s3.wasabisys.com/Acinonyx_jubatus/README.json</span></span>
<span><span class="co">##  2      https://dnazoo.s3.wasabisys.com/Acropora_millepora/README.json</span></span>
<span><span class="co">##  3     https://dnazoo.s3.wasabisys.com/Addax_nasomaculatus/README.json</span></span>
<span><span class="co">##  4           https://dnazoo.s3.wasabisys.com/Aedes_aegypti/README.json</span></span>
<span><span class="co">##  5   https://dnazoo.s3.wasabisys.com/Aedes_aegypti__AaegL4/README.json</span></span>
<span><span class="co">##  6 https://dnazoo.s3.wasabisys.com/Aedes_aegypti__AaegL5.0/README.json</span></span>
<span><span class="co">##    original_assembly     new_assembly</span></span>
<span><span class="co">##  1           aciJub1      aciJub1_HiC</span></span>
<span><span class="co">##  2       amil_sf_1.1  amil_sf_1.1_HiC</span></span>
<span><span class="co">##  3      ASM1959352v1 ASM1959352v1_HiC</span></span>
<span><span class="co">##  4        AGWG.draft         AaegL5.0</span></span>
<span><span class="co">##  5            AaegL3           AaegL4</span></span>
<span><span class="co">##  6        AGWG.draft         AaegL5.0</span></span>
<span><span class="co">##                                                                new_assembly_link</span></span>
<span><span class="co">##  1         https://dnazoo.s3.wasabisys.com/Acinonyx_jubatus/aciJub1_HiC.fasta.gz</span></span>
<span><span class="co">##  2   https://dnazoo.s3.wasabisys.com/Acropora_millepora/amil_sf_1.1_HiC.fasta.gz</span></span>
<span><span class="co">##  3 https://dnazoo.s3.wasabisys.com/Addax_nasomaculatus/ASM1959352v1_HiC.fasta.gz</span></span>
<span><span class="co">##  4               https://dnazoo.s3.wasabisys.com/Aedes_aegypti/AaegL5.0.fasta.gz</span></span>
<span><span class="co">##  5         https://dnazoo.s3.wasabisys.com/Aedes_aegypti__AaegL4/AaegL4.fasta.gz</span></span>
<span><span class="co">##  6     https://dnazoo.s3.wasabisys.com/Aedes_aegypti__AaegL5.0/AaegL5.0.fasta.gz</span></span>
<span><span class="co">##    new_assembly_link_status</span></span>
<span><span class="co">##  1                      200</span></span>
<span><span class="co">##  2                      200</span></span>
<span><span class="co">##  3                      200</span></span>
<span><span class="co">##  4                      404</span></span>
<span><span class="co">##  5                      200</span></span>
<span><span class="co">##  6                      200</span></span>
<span><span class="co">##                                                                    hic_link</span></span>
<span><span class="co">##  1    https://dnazoo.s3.wasabisys.com/Acinonyx_jubatus/aciJub1.rawchrom.hic</span></span>
<span><span class="co">##  2   https://dnazoo.s3.wasabisys.com/Acropora_millepora/amil_sf_1.1_HiC.hic</span></span>
<span><span class="co">##  3 https://dnazoo.s3.wasabisys.com/Addax_nasomaculatus/ASM1959352v1_HiC.hic</span></span>
<span><span class="co">##  4                                                                     &lt;NA&gt;</span></span>
<span><span class="co">##  5         https://dnazoo.s3.wasabisys.com/Aedes_aegypti__AaegL4/AaegL4.hic</span></span>
<span><span class="co">##  6     https://dnazoo.s3.wasabisys.com/Aedes_aegypti__AaegL5.0/AaegL5.0.hic</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>For example, we can directly fetch a Hi-C dataset generated from a tardigrade sample by specifying the right <code>species</code> argument.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hicfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DNAZooData/man/DNAZooData.html">DNAZooData</a></span><span class="op">(</span>species <span class="op">=</span> <span class="st">'Hypsibius_dujardini'</span><span class="op">)</span></span>
<span><span class="co">##  Fetching Hi-C data from DNAZoo</span></span>
<span><span class="co">##  Fetching Hi-C metadata from DNAZoo</span></span>
<span></span>
<span><span class="va">hicfile</span></span>
<span><span class="co">##  HicFile object</span></span>
<span><span class="co">##  .hic file: /root/.cache/R/DNAZooData/150e1fb4f43a_nHd_3.1_HiC.hic </span></span>
<span><span class="co">##  resolution: 5000 </span></span>
<span><span class="co">##  pairs file: </span></span>
<span><span class="co">##  metadata(6): organism draftAssembly ... credits assemblyURL</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here again, the resulting <code>HicFile</code> is populated with metadata parsed from the DNA Zoo data portal.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html">metadata</a></span><span class="op">(</span><span class="va">hicfile</span><span class="op">)</span><span class="op">$</span><span class="va">organism</span></span>
<span><span class="co">##  $vernacular</span></span>
<span><span class="co">##  [1] "Tardigrade"</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $binomial</span></span>
<span><span class="co">##  [1] "Hypsibius dujardini"</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $funFact</span></span>
<span><span class="co">##  [1] "&lt;i&gt;Hypsibius dujardini&lt;/i&gt; is a species of tardigrade, a tiny microscopic organism. They are also commonly called water bears. This species is found world-wide!"</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $extraInfo</span></span>
<span><span class="co">##  [1] "on BioKIDS website"</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $extraInfoLink</span></span>
<span><span class="co">##  [1] "http://www.biokids.umich.edu/critters/Hypsibius_dujardini/"</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $image</span></span>
<span><span class="co">##  [1] "https://static.wixstatic.com/media/2b9330_82db39c219f24b20a75cb38943aad1fb~mv2.jpg"</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $imageCredit</span></span>
<span><span class="co">##  [1] "By Willow Gabriel, Goldstein Lab - https://www.flickr.com/photos/waterbears/1614095719/ Template:Uploader Transferred from en.wikipedia to Commons., CC BY-SA 2.5, https://commons.wikimedia.org/w/index.php?curid=2261992"</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $isChromognomes</span></span>
<span><span class="co">##  [1] "FALSE"</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $taxonomy</span></span>
<span><span class="co">##  [1] "Species:202423-914154-914155-914158-155166-155362-710171-710179-710192-155390-155420"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><code>HiCFile</code> metadata also points to a URL to directly fetch the genome assembly corrected by the DNA Zoo consortium.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html">metadata</a></span><span class="op">(</span><span class="va">hicfile</span><span class="op">)</span><span class="op">$</span><span class="va">assemblyURL</span></span>
<span><span class="co">##  [1] "https://dnazoo.s3.wasabisys.com/Hypsibius_dujardini/nHd_3.1_HiC.fasta.gz"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="session-info" class="level1 unnumbered"><h1 class="unnumbered">Session info</h1>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to expand 👇
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">sessioninfo</span><span class="fu">::</span><span class="fu"><a href="https://r-lib.github.io/sessioninfo/reference/session_info.html">session_info</a></span><span class="op">(</span>include_base <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">##  ─ Session info ────────────────────────────────────────────────────────────</span></span>
<span><span class="co">##   setting  value</span></span>
<span><span class="co">##   version  R Under development (unstable) (2024-01-17 r85813)</span></span>
<span><span class="co">##   os       Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">##   system   x86_64, linux-gnu</span></span>
<span><span class="co">##   ui       X11</span></span>
<span><span class="co">##   language (EN)</span></span>
<span><span class="co">##   collate  C</span></span>
<span><span class="co">##   ctype    en_US.UTF-8</span></span>
<span><span class="co">##   tz       Etc/UTC</span></span>
<span><span class="co">##   date     2024-01-22</span></span>
<span><span class="co">##   pandoc   3.1.1 @ /usr/local/bin/ (via rmarkdown)</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  ─ Packages ────────────────────────────────────────────────────────────────</span></span>
<span><span class="co">##   package              * version     date (UTC) lib source</span></span>
<span><span class="co">##   abind                  1.4-5       2016-07-21 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   base                 * 4.4.0       2024-01-18 [3] local</span></span>
<span><span class="co">##   Biobase              * 2.63.0      2023-10-24 [2] Bioconductor</span></span>
<span><span class="co">##   BiocFileCache          2.11.1      2023-10-26 [2] Bioconductor</span></span>
<span><span class="co">##   BiocGenerics         * 0.49.1      2023-11-01 [2] Bioconductor</span></span>
<span><span class="co">##   BiocIO                 1.13.0      2023-10-24 [2] Bioconductor</span></span>
<span><span class="co">##   BiocParallel           1.37.0      2023-10-24 [2] Bioconductor</span></span>
<span><span class="co">##   Biostrings             2.71.1      2023-10-25 [2] Bioconductor</span></span>
<span><span class="co">##   bit                    4.0.5       2022-11-15 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   bit64                  4.0.5       2020-08-30 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   bitops                 1.0-7       2021-04-24 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   blob                   1.2.4       2023-03-17 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   cachem                 1.0.8       2023-05-01 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   cli                    3.6.2       2023-12-11 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   codetools              0.2-19      2023-02-01 [3] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   colorspace             2.1-0       2023-01-23 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   compiler               4.4.0       2024-01-18 [3] local</span></span>
<span><span class="co">##   crayon                 1.5.2       2022-09-29 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   curl                   5.2.0       2023-12-08 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   datasets             * 4.4.0       2024-01-18 [3] local</span></span>
<span><span class="co">##   DBI                    1.2.1       2024-01-12 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   dbplyr                 2.4.0       2023-10-26 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   DelayedArray           0.29.0      2023-10-24 [2] Bioconductor</span></span>
<span><span class="co">##   digest                 0.6.34      2024-01-11 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   DNAZooData           * 1.3.0       2023-10-31 [2] Bioconductor</span></span>
<span><span class="co">##   dplyr                  1.1.4       2023-11-17 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   evaluate               0.23        2023-11-01 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   fansi                  1.0.6       2023-12-08 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   fastmap                1.1.1       2023-02-24 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   filelock               1.0.3       2023-12-11 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   fourDNData           * 1.3.0       2023-10-31 [2] Bioconductor</span></span>
<span><span class="co">##   generics               0.1.3       2022-07-05 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   GenomeInfoDb         * 1.39.5      2024-01-01 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   GenomeInfoDbData       1.2.11      2024-01-22 [2] Bioconductor</span></span>
<span><span class="co">##   GenomicAlignments      1.39.2      2024-01-16 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   GenomicRanges        * 1.55.1      2023-10-29 [2] Bioconductor</span></span>
<span><span class="co">##   ggplot2              * 3.4.4       2023-10-12 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   glue                   1.7.0       2024-01-09 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   graphics             * 4.4.0       2024-01-18 [3] local</span></span>
<span><span class="co">##   grDevices            * 4.4.0       2024-01-18 [3] local</span></span>
<span><span class="co">##   grid                   4.4.0       2024-01-18 [3] local</span></span>
<span><span class="co">##   gtable                 0.3.4       2023-08-21 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   HiCExperiment        * 1.3.0       2023-10-24 [2] Bioconductor</span></span>
<span><span class="co">##   htmltools              0.5.7       2023-11-03 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   htmlwidgets            1.6.4       2023-12-06 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   httr                   1.4.7       2023-08-15 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   InteractionSet       * 1.31.0      2023-10-24 [2] Bioconductor</span></span>
<span><span class="co">##   IRanges              * 2.37.1      2024-01-19 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   jsonlite               1.8.8       2023-12-04 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   knitr                  1.45        2023-10-30 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   lattice                0.22-5      2023-10-24 [3] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   lifecycle              1.0.4       2023-11-07 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   magrittr               2.0.3       2022-03-30 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   Matrix                 1.6-5       2024-01-11 [3] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   MatrixGenerics       * 1.15.0      2023-10-24 [2] Bioconductor</span></span>
<span><span class="co">##   matrixStats          * 1.2.0       2023-12-11 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   memoise                2.0.1       2021-11-26 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   methods              * 4.4.0       2024-01-18 [3] local</span></span>
<span><span class="co">##   munsell                0.5.0       2018-06-12 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   parallel               4.4.0       2024-01-18 [3] local</span></span>
<span><span class="co">##   pillar                 1.9.0       2023-03-22 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   pkgconfig              2.0.3       2019-09-22 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   purrr                  1.0.2       2023-08-10 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   R6                     2.5.1       2021-08-19 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   Rcpp                   1.0.12      2024-01-09 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   RCurl                  1.98-1.14   2024-01-09 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   restfulr               0.0.15      2022-06-16 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   rhdf5                  2.47.2      2024-01-15 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   rhdf5filters           1.15.1      2023-11-06 [2] Bioconductor</span></span>
<span><span class="co">##   Rhdf5lib               1.25.1      2023-12-11 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   rjson                  0.2.21      2022-01-09 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   rlang                  1.1.3       2024-01-10 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   rmarkdown              2.25        2023-09-18 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   Rsamtools              2.19.3      2024-01-17 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   RSQLite                2.3.5       2024-01-21 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   rtracklayer          * 1.63.0      2024-01-22 [2] Github (lawremi/rtracklayer@86407bb)</span></span>
<span><span class="co">##   S4Arrays               1.3.2       2024-01-14 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   S4Vectors            * 0.41.3      2024-01-01 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   scales                 1.3.0       2023-11-28 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   sessioninfo            1.2.2       2021-12-06 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   SparseArray            1.3.3       2024-01-14 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   stats                * 4.4.0       2024-01-18 [3] local</span></span>
<span><span class="co">##   stats4               * 4.4.0       2024-01-18 [3] local</span></span>
<span><span class="co">##   strawr                 0.0.91      2023-03-29 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   SummarizedExperiment * 1.33.2      2024-01-07 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   tibble                 3.2.1       2023-03-20 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   tidyselect             1.2.0       2022-10-10 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   tools                  4.4.0       2024-01-18 [3] local</span></span>
<span><span class="co">##   tzdb                   0.4.0       2023-05-12 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   utf8                   1.2.4       2023-10-22 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   utils                * 4.4.0       2024-01-18 [3] local</span></span>
<span><span class="co">##   vctrs                  0.6.5       2023-12-01 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   vroom                  1.6.5       2023-12-05 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   withr                  3.0.0       2024-01-16 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   xfun                   0.41        2023-11-01 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   XML                    3.99-0.16.1 2024-01-22 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   XVector                0.43.1      2024-01-10 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   yaml                   2.3.8       2023-12-11 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   zlibbioc               1.49.0      2023-10-24 [2] Bioconductor</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##   [1] /tmp/Rtmpq5g2WV/Rinstb37571687</span></span>
<span><span class="co">##   [2] /usr/local/lib/R/site-library</span></span>
<span><span class="co">##   [3] /usr/local/lib/R/library</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  ───────────────────────────────────────────────────────────────────────────</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</section><section id="references" class="level1 unnumbered"><h1 class="unnumbered">References</h1>


<div id="refs" class="references csl-bib-body hanging-indent" data-line-spacing="2" role="list">
<div id="ref-Crane_2015" class="csl-entry" role="listitem">
Crane, E., Bian, Q., McCord, R. P., Lajoie, B. R., Wheeler, B. S., Ralston, E. J., Uzawa, S., Dekker, J., &amp; Meyer, B. J. (2015). Condensin-driven remodelling of x chromosome topology during dosage compensation. <em>Nature</em>, <em>523</em>(7559), 240–244. <a href="https://doi.org/10.1038/nature14450">https://doi.org/10.1038/nature14450</a>
</div>
</div>
</section><a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../pages/topological-features.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Finding topological features in Hi-C</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../pages/interoperability.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Interoperability: using HiCExperiment with other R packages</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left"><i>OHCA: Orchestrating Hi-C analysis with Bioconductor</i> was written by J. Serizay</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">This book was built with <a href="https://github.com/js2264/BiocBook/">BiocBook</a> with <span class="emoji" data-emoji="heart">❤️</span></div>
  </div>
</footer>


</body></html>