<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Orchestrating Hi-C analysis with Bioconductor - 9&nbsp; Interoperability: using HiCExperiment with other R packages</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../pages/workflow-yeast.html" rel="next">
<link href="../pages/disseminating.html" rel="prev">
<link href="../assets/cover.jpg" rel="icon" type="image/jpeg">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
</head>
<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/disseminating.html">Advanced Hi-C topics</a></li><li class="breadcrumb-item"><a href="../pages/interoperability.html"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Interoperability: using HiCExperiment with other R packages</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../">Orchestrating Hi-C analysis with Bioconductor</a> 
        <div class="sidebar-tools-main">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label=""><i class="bi bi-git"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
<li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://github.com/js2264/OHCA/">
            Source Code
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://js2264.github.io/OHCA/devel/">
            Browse version `devel`
            </a>
          </li>
      </ul>
</div>
    <a href="../Orchestrating-Hi-C-analysis-with-Bioconductor.pdf" rel="" title="Download PDF" class="quarto-navigation-tool px-1" aria-label="Download PDF"><i class="bi bi-file-pdf"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/preamble.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preamble</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Fundamentals concepts</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/principles.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Hi-C pre-processing steps</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/data-representation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Hi-C data structures in R</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/parsing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Manipulating Hi-C data in R</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/visualization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Hi-C data visualization</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">In-depth Hi-C analysis</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/matrix-centric.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Matrix-centric analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/interactions-centric.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Interactions-centric analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/topological-features.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Finding topological features in Hi-C</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Advanced Hi-C topics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/disseminating.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Data gateways: accessing public Hi-C data portals</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/interoperability.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Interoperability: using HiCExperiment with other R packages</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/workflow-yeast.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workflow 1: Distance-dependent interactions across yeast mutants</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/workflow-chicken.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workflow 2: Chromosome compartment cohesion upon mitosis entry</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/workflow-centros.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workflow 3: Inter-centromere interactions in yeast</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#diffhic" id="toc-diffhic" class="nav-link active" data-scroll-target="#diffhic"><span class="header-section-number">9.1</span> diffHic</a></li>
  <li><a href="#multihiccompare" id="toc-multihiccompare" class="nav-link" data-scroll-target="#multihiccompare"><span class="header-section-number">9.2</span> multiHiCcompare</a></li>
  <li><a href="#topdom" id="toc-topdom" class="nav-link" data-scroll-target="#topdom"><span class="header-section-number">9.3</span> TopDom</a></li>
  <li><a href="#gothic" id="toc-gothic" class="nav-link" data-scroll-target="#gothic"><span class="header-section-number">9.4</span> GOTHiC</a></li>
  <li><a href="#session-info" id="toc-session-info" class="nav-link" data-scroll-target="#session-info">Session info</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title">
<span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Interoperability: using HiCExperiment with other R packages</span>
</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Pre-loading packages and objects üì¶
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://purrr.tidyverse.org/">purrr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges">GenomicRanges</a></span><span class="op">)</span></span>
<span><span class="co">##  Loading required package: stats4</span></span>
<span><span class="co">##  Loading required package: BiocGenerics</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Attaching package: 'BiocGenerics'</span></span>
<span><span class="co">##  The following objects are masked from 'package:stats':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      IQR, mad, sd, var, xtabs</span></span>
<span><span class="co">##  The following objects are masked from 'package:base':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      Filter, Find, Map, Position, Reduce, anyDuplicated, aperm,</span></span>
<span><span class="co">##      append, as.data.frame, basename, cbind, colnames, dirname,</span></span>
<span><span class="co">##      do.call, duplicated, eval, evalq, get, grep, grepl, intersect,</span></span>
<span><span class="co">##      is.unsorted, lapply, mapply, match, mget, order, paste, pmax,</span></span>
<span><span class="co">##      pmax.int, pmin, pmin.int, rank, rbind, rownames, sapply,</span></span>
<span><span class="co">##      setdiff, table, tapply, union, unique, unsplit, which.max,</span></span>
<span><span class="co">##      which.min</span></span>
<span><span class="co">##  Loading required package: S4Vectors</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Attaching package: 'S4Vectors'</span></span>
<span><span class="co">##  The following object is masked from 'package:utils':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      findMatches</span></span>
<span><span class="co">##  The following objects are masked from 'package:base':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      I, expand.grid, unname</span></span>
<span><span class="co">##  Loading required package: IRanges</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Attaching package: 'IRanges'</span></span>
<span><span class="co">##  The following object is masked from 'package:purrr':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      reduce</span></span>
<span><span class="co">##  Loading required package: GenomeInfoDb</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">InteractionSet</span><span class="op">)</span></span>
<span><span class="co">##  Loading required package: SummarizedExperiment</span></span>
<span><span class="co">##  Loading required package: MatrixGenerics</span></span>
<span><span class="co">##  Loading required package: matrixStats</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Attaching package: 'MatrixGenerics'</span></span>
<span><span class="co">##  The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span></span>
<span><span class="co">##      colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span></span>
<span><span class="co">##      colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span></span>
<span><span class="co">##      colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span></span>
<span><span class="co">##      colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span></span>
<span><span class="co">##      colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span></span>
<span><span class="co">##      colWeightedMeans, colWeightedMedians, colWeightedSds,</span></span>
<span><span class="co">##      colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span></span>
<span><span class="co">##      rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span></span>
<span><span class="co">##      rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span></span>
<span><span class="co">##      rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span></span>
<span><span class="co">##      rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span></span>
<span><span class="co">##      rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs,</span></span>
<span><span class="co">##      rowVars, rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span></span>
<span><span class="co">##      rowWeightedSds, rowWeightedVars</span></span>
<span><span class="co">##  Loading required package: Biobase</span></span>
<span><span class="co">##  Welcome to Bioconductor</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      Vignettes contain introductory material; view with</span></span>
<span><span class="co">##      'browseVignettes()'. To cite Bioconductor, see</span></span>
<span><span class="co">##      'citation("Biobase")', and for packages 'citation("pkgname")'.</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Attaching package: 'Biobase'</span></span>
<span><span class="co">##  The following object is masked from 'package:MatrixGenerics':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      rowMedians</span></span>
<span><span class="co">##  The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      anyMissing, rowMedians</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/js2264/HiCExperiment">HiCExperiment</a></span><span class="op">)</span></span>
<span><span class="co">##  Consider using the `HiContacts` package to perform advanced genomic operations </span></span>
<span><span class="co">##  on `HiCExperiment` objects.</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Read "Orchestrating Hi-C analysis with Bioconductor" online book to learn more:</span></span>
<span><span class="co">##  https://js2264.github.io/OHCA/</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Attaching package: 'HiCExperiment'</span></span>
<span><span class="co">##  The following object is masked from 'package:SummarizedExperiment':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      metadata&lt;-</span></span>
<span><span class="co">##  The following object is masked from 'package:S4Vectors':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      metadata&lt;-</span></span>
<span><span class="co">##  The following object is masked from 'package:ggplot2':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      resolution</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/js2264/HiContactsData">HiContactsData</a></span><span class="op">)</span></span>
<span><span class="co">##  Loading required package: ExperimentHub</span></span>
<span><span class="co">##  Loading required package: AnnotationHub</span></span>
<span><span class="co">##  Loading required package: BiocFileCache</span></span>
<span><span class="co">##  Loading required package: dbplyr</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Attaching package: 'AnnotationHub'</span></span>
<span><span class="co">##  The following object is masked from 'package:Biobase':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      cache</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/js2264/plyinteractions">plyinteractions</a></span><span class="op">)</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Attaching package: 'plyinteractions'</span></span>
<span><span class="co">##  The following object is masked from 'package:matrixStats':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      count</span></span>
<span><span class="co">##  The following object is masked from 'package:IRanges':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      slice</span></span>
<span><span class="co">##  The following object is masked from 'package:S4Vectors':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      rename</span></span>
<span><span class="co">##  The following object is masked from 'package:ggplot2':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      annotate</span></span>
<span><span class="co">##  The following object is masked from 'package:stats':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      filter</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/dozmorovlab/multiHiCcompare">multiHiCcompare</a></span><span class="op">)</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Attaching package: 'multiHiCcompare'</span></span>
<span><span class="co">##  The following object is masked from 'package:HiCExperiment':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      resolution</span></span>
<span><span class="co">##  The following object is masked from 'package:ggplot2':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      resolution</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Attaching package: 'dplyr'</span></span>
<span><span class="co">##  The following objects are masked from 'package:dbplyr':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      ident, sql</span></span>
<span><span class="co">##  The following object is masked from 'package:Biobase':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      combine</span></span>
<span><span class="co">##  The following object is masked from 'package:matrixStats':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      count</span></span>
<span><span class="co">##  The following objects are masked from 'package:GenomicRanges':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      intersect, setdiff, union</span></span>
<span><span class="co">##  The following object is masked from 'package:GenomeInfoDb':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      intersect</span></span>
<span><span class="co">##  The following objects are masked from 'package:IRanges':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      collapse, desc, intersect, setdiff, slice, union</span></span>
<span><span class="co">##  The following objects are masked from 'package:S4Vectors':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      first, intersect, rename, setdiff, setequal, union</span></span>
<span><span class="co">##  The following objects are masked from 'package:BiocGenerics':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      combine, intersect, setdiff, union</span></span>
<span><span class="co">##  The following objects are masked from 'package:stats':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      filter, lag</span></span>
<span><span class="co">##  The following objects are masked from 'package:base':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      intersect, setdiff, setequal, union</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org">tidyr</a></span><span class="op">)</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Attaching package: 'tidyr'</span></span>
<span><span class="co">##  The following object is masked from 'package:S4Vectors':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      expand</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Aims
</div>
</div>
<div class="callout-body-container callout-body">
<p>This notebook illustrates how to use a range of popular Hi-C‚Äîrelated <code>R</code> packages with <code>HiCExperiment</code> objects. Conversion to the data structures supported by the following packages is illustrated here:</p>
<ul>
<li>
<code>diffHic</code> <!-- - `hicrep` -->
</li>
<li><code>multiHiCcompare</code></li>
<li><code>TopDom</code></li>
<li>
<code>GOTHiC</code> <!-- - `HiCDOC` -->
</li>
</ul>
</div>
</div>
<section id="diffhic" class="level2" data-number="9.1"><h2 data-number="9.1" class="anchored" data-anchor-id="diffhic">
<span class="header-section-number">9.1</span> diffHic</h2>
<p><code>diffHic</code> is the first R package dedicated to Hi-C processing and analysis (<span class="citation" data-cites="Lun_2015">Lun &amp; Smyth (<a href="#ref-Lun_2015" role="doc-biblioref">2015</a>)</span>). It is packed with useful functions to generate a contact matrix from read pairs and to perform downstream investigation, including normalization, 2D ‚Äúpeak‚Äù (i.e.&nbsp;loops) finding and aggregation, differential interaction between samples, etc. It works seamlessly with the <code>InteractionSet</code> class of object, which can be easily obtained from a <code>HiCExperiment</code> object.</p>
<p>To do so, we first need to extract <code>GInteractions</code> from one or several <code>HiCExperiment</code> objects and <strong>create</strong> a single <code>InteractionSet</code> object.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">InteractionSet</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges">GenomicRanges</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/js2264/HiCExperiment">HiCExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/js2264/HiContactsData">HiContactsData</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># ---- This downloads an example `.mcool` file and caches it locally </span></span>
<span><span class="va">coolf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/HiContactsData/man/HiContactsData.html">HiContactsData</a></span><span class="op">(</span><span class="st">'yeast_wt'</span>, <span class="st">'mcool'</span><span class="op">)</span></span>
<span><span class="co">##  see ?HiContactsData and browseVignettes('HiContactsData') for documentation</span></span>
<span><span class="co">##  loading from cache</span></span>
<span><span class="va">cool</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocIO/man/IO.html">import</a></span><span class="op">(</span><span class="va">coolf</span>, format <span class="op">=</span> <span class="st">'cool'</span><span class="op">)</span></span>
<span><span class="va">gi</span> <span class="op">&lt;-</span> <span class="va">cool</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">interactions</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/methods/as.html">as</a></span><span class="op">(</span><span class="st">"ReverseStrictGInteractions"</span><span class="op">)</span></span>
<span><span class="va">iset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/InteractionSet-class.html">InteractionSet</a></span><span class="op">(</span></span>
<span>    assays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>        counts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="va">gi</span><span class="op">$</span><span class="va">count</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>, </span>
<span>        balanced <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="va">gi</span><span class="op">$</span><span class="va">balanced</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span>    <span class="op">)</span>, </span>
<span>    interactions <span class="op">=</span> <span class="va">gi</span>, </span>
<span>    colData <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>lib <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"WT"</span><span class="op">)</span>, totals <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">gi</span><span class="op">$</span><span class="va">count</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>From there, we can <strong>filter</strong> interactions to only retain those with significant enrichment over background.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">diffHic</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">1234</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># --- Filter to find aggregated interactions</span></span>
<span><span class="va">enrichments</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/diffHic/man/enrichedPairs.html">enrichedPairs</a></span><span class="op">(</span><span class="va">iset</span><span class="op">)</span></span>
<span><span class="va">filter</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/diffHic/man/filterPeaks.html">filterPeaks</a></span><span class="op">(</span><span class="va">enrichments</span>, min.enrich <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="fl">1.2</span><span class="op">)</span>, min.diag <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">filtered_iset</span> <span class="op">&lt;-</span> <span class="va">iset</span><span class="op">[</span><span class="va">filter</span><span class="op">]</span></span>
<span><span class="va">filtered_iset</span></span>
<span><span class="co">##  class: InteractionSet </span></span>
<span><span class="co">##  dim: 41872 1 </span></span>
<span><span class="co">##  metadata(0):</span></span>
<span><span class="co">##  assays(2): counts balanced</span></span>
<span><span class="co">##  rownames: NULL</span></span>
<span><span class="co">##  rowData names(4): bin_id1 bin_id2 count balanced</span></span>
<span><span class="co">##  colnames: NULL</span></span>
<span><span class="co">##  colData names(2): lib totals</span></span>
<span><span class="co">##  type: ReverseStrictGInteractions</span></span>
<span><span class="co">##  regions: 12079</span></span>
<span></span>
<span><span class="co"># --- Visualize filtered interactions </span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/js2264/plyinteractions">plyinteractions</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/js2264/HiContacts">HiContacts</a></span><span class="op">)</span></span>
<span><span class="co">##  Registered S3 methods overwritten by 'readr':</span></span>
<span><span class="co">##    method                    from </span></span>
<span><span class="co">##    as.data.frame.spec_tbl_df vroom</span></span>
<span><span class="co">##    as_tibble.spec_tbl_df     vroom</span></span>
<span><span class="co">##    format.col_spec           vroom</span></span>
<span><span class="co">##    print.col_spec            vroom</span></span>
<span><span class="co">##    print.collector           vroom</span></span>
<span><span class="co">##    print.date_names          vroom</span></span>
<span><span class="co">##    print.locale              vroom</span></span>
<span><span class="co">##    str.col_spec              vroom</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">interactions</a></span><span class="op">(</span><span class="va">filtered_iset</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">seqnames2</span> <span class="op">==</span> <span class="st">'II'</span>, <span class="va">seqnames1</span> <span class="op">==</span> <span class="va">seqnames2</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/HiContacts/man/plotMatrix.html">plotMatrix</a></span><span class="op">(</span>use.scores <span class="op">=</span> <span class="st">'count'</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="interoperability_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Next, we can cluster filtered interactions that are next to each other.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># --- Cluster interactions to find loops</span></span>
<span><span class="va">clustered_iset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/diffHic/man/clusterPairs.html">clusterPairs</a></span><span class="op">(</span><span class="va">filtered_iset</span>, tol <span class="op">=</span> <span class="fl">5000</span><span class="op">)</span></span>
<span><span class="va">clustered_iset</span><span class="op">$</span><span class="va">interactions</span> </span>
<span><span class="co">##  ReverseStrictGInteractions object with 1644 interactions and 0 metadata columns:</span></span>
<span><span class="co">##           seqnames1       ranges1 strand1     seqnames2       ranges2 strand2</span></span>
<span><span class="co">##               &lt;Rle&gt;     &lt;IRanges&gt;   &lt;Rle&gt;         &lt;Rle&gt;     &lt;IRanges&gt;   &lt;Rle&gt;</span></span>
<span><span class="co">##       [1]         I  15001-149000       * ---         I      1-122000       *</span></span>
<span><span class="co">##       [2]         I 133001-148000       * ---         I 127001-139000       *</span></span>
<span><span class="co">##       [3]         I 154001-160000       * ---         I 128001-149000       *</span></span>
<span><span class="co">##       [4]         I 168001-173000       * ---         I 138001-148000       *</span></span>
<span><span class="co">##       [5]         I 184001-196000       * ---         I   15001-23000       *</span></span>
<span><span class="co">##       ...       ...           ...     ... ...       ...           ...     ...</span></span>
<span><span class="co">##    [1640]       XVI 897001-898000       * ---       XVI 831001-832000       *</span></span>
<span><span class="co">##    [1641]       XVI 907001-910000       * ---       XVI 840001-843000       *</span></span>
<span><span class="co">##    [1642]       XVI 926001-934000       * ---       XVI 872001-878000       *</span></span>
<span><span class="co">##    [1643]       XVI 933001-934000       * ---       XVI 858001-859000       *</span></span>
<span><span class="co">##    [1644]       XVI 933001-942000       * ---       XVI 928001-934000       *</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    regions: 2822 ranges and 0 metadata columns</span></span>
<span><span class="co">##    seqinfo: 16 sequences from an unspecified genome</span></span>
<span></span>
<span><span class="co"># --- Visualize clustered interactions </span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">interactions</a></span><span class="op">(</span><span class="va">filtered_iset</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>cluster <span class="op">=</span> <span class="va">clustered_iset</span><span class="op">$</span><span class="va">indices</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">seqnames2</span> <span class="op">==</span> <span class="st">'II'</span>, <span class="va">seqnames1</span> <span class="op">==</span> <span class="va">seqnames2</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/HiContacts/man/plotMatrix.html">plotMatrix</a></span><span class="op">(</span>use.scores <span class="op">=</span> <span class="st">'cluster'</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="interoperability_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Finally, we can visualize identified individual interaction clusters identified with <code>diffHic</code> using <code>HiContacts</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># --- Plot matrix at a clustered loops</span></span>
<span><span class="va">cgi</span> <span class="op">&lt;-</span> <span class="va">clustered_iset</span><span class="op">$</span><span class="va">interactions</span><span class="op">[</span><span class="fl">554</span><span class="op">]</span></span>
<span><span class="va">seqn</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html">seqnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">anchors</a></span><span class="op">(</span><span class="va">cgi</span>, type<span class="op">=</span><span class="st">"second"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">start</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/start.html">start</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">anchors</a></span><span class="op">(</span><span class="va">cgi</span>, type<span class="op">=</span><span class="st">"second"</span><span class="op">)</span><span class="op">)</span> <span class="op">-</span> <span class="fl">50000</span></span>
<span><span class="va">end</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/start.html">end</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">anchors</a></span><span class="op">(</span><span class="va">cgi</span>, type<span class="op">=</span><span class="st">"first"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fl">50000</span></span>
<span><span class="va">interactions_peak</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html">GRanges</a></span><span class="op">(</span><span class="va">seqn</span>, <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html">IRanges</a></span><span class="op">(</span><span class="va">start</span>, <span class="va">end</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/HiContacts/man/plotMatrix.html">plotMatrix</a></span><span class="op">(</span><span class="va">cool</span><span class="op">[</span><span class="va">interactions_peak</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="va">an</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">anchors</a></span><span class="op">(</span><span class="va">cgi</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_rect</a></span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>xmin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/start.html">start</a></span><span class="op">(</span><span class="va">an</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, xmax <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/start.html">end</a></span><span class="op">(</span><span class="va">an</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, ymin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/start.html">start</a></span><span class="op">(</span><span class="va">an</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, ymax <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/start.html">end</a></span><span class="op">(</span><span class="va">an</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>, </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>xmin <span class="op">=</span> <span class="va">xmin</span>, xmax <span class="op">=</span> <span class="va">xmax</span>, ymin <span class="op">=</span> <span class="va">ymin</span>, ymax <span class="op">=</span> <span class="va">ymax</span><span class="op">)</span>, </span>
<span>    inherit.aes <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>    fill <span class="op">=</span> <span class="cn">NA</span>, </span>
<span>    colour <span class="op">=</span> <span class="st">'cyan'</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="interoperability_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="multihiccompare" class="level2" data-number="9.2"><h2 data-number="9.2" class="anchored" data-anchor-id="multihiccompare">
<span class="header-section-number">9.2</span> multiHiCcompare</h2>
<p>The <code>multiHiCcompare</code> package provides functions for joint normalization and difference detection in multiple Hi-C datasets (<span class="citation" data-cites="Stansfield_2019">Stansfield et al. (<a href="#ref-Stansfield_2019" role="doc-biblioref">2019</a>)</span>). According to its excerpt, to perform differential interaction analysis, it requires a <code>list</code> of <strong>raw counts</strong> for different samples/replicates, stored in <strong>data frames with four columns</strong> (<code>chr</code>, <code>start1</code>, <code>start2</code>, <code>count</code>).<br>
Manipulate a <code>HiCExperiment</code> object to coerce it into such structure is straightforward.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org">tidyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://purrr.tidyverse.org/">purrr</a></span><span class="op">)</span></span>
<span><span class="va">coolf_wt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/HiContactsData/man/HiContactsData.html">HiContactsData</a></span><span class="op">(</span><span class="st">'yeast_wt'</span>, <span class="st">'mcool'</span><span class="op">)</span></span>
<span><span class="co">##  see ?HiContactsData and browseVignettes('HiContactsData') for documentation</span></span>
<span><span class="co">##  loading from cache</span></span>
<span><span class="va">coolf_eco1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/HiContactsData/man/HiContactsData.html">HiContactsData</a></span><span class="op">(</span><span class="st">'yeast_eco1'</span>, <span class="st">'mcool'</span><span class="op">)</span></span>
<span><span class="co">##  see ?HiContactsData and browseVignettes('HiContactsData') for documentation</span></span>
<span><span class="co">##  loading from cache</span></span>
<span><span class="va">hics</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"wt"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocIO/man/IO.html">import</a></span><span class="op">(</span><span class="va">coolf_wt</span>, format <span class="op">=</span> <span class="st">'cool'</span><span class="op">)</span>,</span>
<span>    <span class="st">"eco1"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocIO/man/IO.html">import</a></span><span class="op">(</span><span class="va">coolf_eco1</span>, format <span class="op">=</span> <span class="st">'cool'</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">hics_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span><span class="op">(</span><span class="va">hics</span>, <span class="op">~</span> <span class="va">.x</span><span class="op">[</span><span class="st">'XI'</span><span class="op">]</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>chr <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/relocate.html">relocate</a></span><span class="op">(</span><span class="va">chr</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">chr</span>, <span class="va">start1</span>, <span class="va">start2</span>, <span class="va">count</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">hics_list</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">##    chr start1 start2 count</span></span>
<span><span class="co">##  1   1      1      1     2</span></span>
<span><span class="co">##  2   1      1   1001     3</span></span>
<span><span class="co">##  3   1      1   2001     3</span></span>
<span><span class="co">##  4   1      1   3001    13</span></span>
<span><span class="co">##  5   1      1   4001     9</span></span>
<span><span class="co">##  6   1      1   5001    13</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Once this list is generated, the classical <code>multiHiCcompare</code> workflow can be applied: first run <code><a href="https://rdrr.io/pkg/multiHiCcompare/man/make_hicexp.html">make_hicexp()</a></code>, followed by <code><a href="https://rdrr.io/pkg/multiHiCcompare/man/cyclic_loess.html">cyclic_loess()</a></code>, then <code><a href="https://rdrr.io/pkg/multiHiCcompare/man/hic_exactTest.html">hic_exactTest()</a></code> and finally <code><a href="https://rdrr.io/pkg/multiHiCcompare/man/results.html">results()</a></code>:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">DI</span> <span class="op">&lt;-</span> <span class="va">hics_list</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/multiHiCcompare/man/make_hicexp.html">make_hicexp</a></span><span class="op">(</span></span>
<span>        data_list <span class="op">=</span> <span class="va">hics_list</span>, </span>
<span>        groups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/multiHiCcompare/man/cyclic_loess.html">cyclic_loess</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/multiHiCcompare/man/hic_exactTest.html">hic_exactTest</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/multiHiCcompare/man/results.html">results</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">DI</span></span>
<span><span class="co">##         chr region1 region2 D      logFC    logCPM    p.value     p.adj</span></span>
<span><span class="co">##      1:   1       1    1001 1  0.4279414  6.382927 0.78960192 1.0000000</span></span>
<span><span class="co">##      2:   1       1    3001 3  1.0325237  8.339327 0.06035705 0.9501367</span></span>
<span><span class="co">##      3:   1       1    4001 4  0.6862141  7.597689 0.34723639 1.0000000</span></span>
<span><span class="co">##      4:   1       1    5001 5  0.5124878  7.960339 0.43133791 1.0000000</span></span>
<span><span class="co">##      5:   1       1    6001 6 -0.3568672  8.563374 0.52289982 1.0000000</span></span>
<span><span class="co">##     ---                                                                </span></span>
<span><span class="co">##  22637:   1  663001  666001 3 -1.1680738  7.158551 0.17500113 1.0000000</span></span>
<span><span class="co">##  22638:   1  664001  664001 0  1.4530501  8.536212 0.16535151 1.0000000</span></span>
<span><span class="co">##  22639:   1  664001  665001 1 -0.1014769  8.166275 1.00000000 1.0000000</span></span>
<span><span class="co">##  22640:   1  665001  665001 0 -0.3110054 10.013750 0.60075706 1.0000000</span></span>
<span><span class="co">##  22641:   1  665001  666001 1 -0.4989794  7.750157 0.41481212 1.0000000</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="topdom" class="level2" data-number="9.3"><h2 data-number="9.3" class="anchored" data-anchor-id="topdom">
<span class="header-section-number">9.3</span> TopDom</h2>
<p>The <code>TopDom</code> method is widely used to annotate topological domains in genomes from Hi-C data (<span class="citation" data-cites="Shin_2015">Shin et al. (<a href="#ref-Shin_2015" role="doc-biblioref">2015</a>)</span>). The <code>TopDom</code> package was created to implement this method in <code>R</code> (<span class="citation" data-cites="Bengtsson_2020">Bengtsson et al. (<a href="#ref-Bengtsson_2020" role="doc-biblioref">2020</a>)</span>).</p>
<p>Unfortunately, the format of the input to <code>TopDom</code> is rather tricky (see <code><a href="https://rdrr.io/pkg/TopDom/man/readHiC.html">?TopDom::readHiC</a></code>). The following chunk of code shows how to coerce a <code>HiCExperiment</code> object into a <code>TopDom</code>-compatible object.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/HenrikBengtsson/TopDom">TopDom</a></span><span class="op">)</span></span>
<span><span class="va">hic</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocIO/man/IO.html">import</a></span><span class="op">(</span><span class="va">coolf_wt</span>, format <span class="op">=</span> <span class="st">'cool'</span><span class="op">)</span></span>
<span><span class="va">HiCExperiment2TopDom</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">hic</span>, <span class="va">chr</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="va">cm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html">as</a></span><span class="op">(</span><span class="va">hic</span><span class="op">[</span><span class="va">chr</span><span class="op">]</span>, <span class="st">'ContactMatrix'</span><span class="op">)</span></span>
<span>    <span class="va">data</span><span class="op">$</span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">cm</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">base</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="va">data</span><span class="op">$</span><span class="va">counts</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span>    <span class="va">data</span><span class="op">$</span><span class="va">bins</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">regions</a></span><span class="op">(</span><span class="va">cm</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>        <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>        <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">seqnames</span>, <span class="va">start</span>, <span class="va">end</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>        <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>seqnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">seqnames</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>        <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>id <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html">n</a></span><span class="op">(</span><span class="op">)</span>, start <span class="op">=</span> <span class="va">start</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>        <span class="fu"><a href="https://dplyr.tidyverse.org/reference/relocate.html">relocate</a></span><span class="op">(</span><span class="va">id</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>        <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html">rename</a></span><span class="op">(</span>chr <span class="op">=</span> <span class="va">seqnames</span>, from.coord <span class="op">=</span> <span class="va">start</span>, to.coord <span class="op">=</span> <span class="va">end</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">'TopDomData'</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">hic_topdom</span> <span class="op">&lt;-</span> <span class="fu">HiCExperiment2TopDom</span><span class="op">(</span><span class="va">hic</span>, <span class="st">"II"</span><span class="op">)</span></span>
<span><span class="va">hic_topdom</span></span>
<span><span class="co">##  TopDomData:</span></span>
<span><span class="co">##  bins:</span></span>
<span><span class="co">##  'data.frame':   813 obs. of  4 variables:</span></span>
<span><span class="co">##   $ id        : int  1 2 3 4 5 6 7 8 9 10 ...</span></span>
<span><span class="co">##   $ chr       : chr  "II" "II" "II" "II" ...</span></span>
<span><span class="co">##   $ from.coord: num  0 1000 2000 3000 4000 5000 6000 7000 8000 9000 ...</span></span>
<span><span class="co">##   $ to.coord  : int  1000 2000 3000 4000 5000 6000 7000 8000 9000 10000 ...</span></span>
<span><span class="co">##  counts:</span></span>
<span><span class="co">##   num [1:813, 1:813] 0 0 0 0 0 0 0 0 0 0 ...</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now that we have coerced a <code>HiCExperiment</code> object into a <code>TopDom</code>-compatible object, we can use the main <code>TopDom</code> function to annotate topological domains.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">domains</span> <span class="op">&lt;-</span> <span class="fu">TopDom</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/TopDom/man/TopDom.html">TopDom</a></span><span class="op">(</span><span class="va">hic_topdom</span>, window.size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">domains</span></span>
<span><span class="co">##  TopDom:</span></span>
<span><span class="co">##  Parameters:</span></span>
<span><span class="co">##  - window.size: 5</span></span>
<span><span class="co">##  - statFilter: TRUE</span></span>
<span><span class="co">##  binSignal:</span></span>
<span><span class="co">##  'data.frame':   813 obs. of  7 variables:</span></span>
<span><span class="co">##   $ id        : int  1 2 3 4 5 6 7 8 9 10 ...</span></span>
<span><span class="co">##   $ chr       : chr  "II" "II" "II" "II" ...</span></span>
<span><span class="co">##   $ from.coord: num  0 1000 2000 3000 4000 5000 6000 7000 8000 9000 ...</span></span>
<span><span class="co">##   $ to.coord  : int  1000 2000 3000 4000 5000 6000 7000 8000 9000 10000 ...</span></span>
<span><span class="co">##   $ local.ext : num  -0.5 -0.5 -0.5 -0.5 -0.5 -0.5 -0.5 -0.5 0 0 ...</span></span>
<span><span class="co">##   $ mean.cf   : num  0 0 0 0 0 ...</span></span>
<span><span class="co">##   $ pvalue    : num  1 1 1 1 1 ...</span></span>
<span><span class="co">##  domain:</span></span>
<span><span class="co">##  'data.frame':   61 obs. of  7 variables:</span></span>
<span><span class="co">##   $ chr       : chr  "II" "II" "II" "II" ...</span></span>
<span><span class="co">##   $ from.id   : int  1 9 31 36 47 61 76 82 91 102 ...</span></span>
<span><span class="co">##   $ from.coord: num  0 8000 30000 35000 46000 60000 75000 81000 90000 101000 ...</span></span>
<span><span class="co">##   $ to.id     : int  8 30 35 46 60 75 81 90 101 136 ...</span></span>
<span><span class="co">##   $ to.coord  : num  8000 30000 35000 46000 60000 75000 81000 90000 101000 136000 ...</span></span>
<span><span class="co">##   $ tag       : chr  "gap" "domain" "gap" "domain" ...</span></span>
<span><span class="co">##   $ size      : num  8000 22000 5000 11000 14000 15000 6000 9000 11000 35000 ...</span></span>
<span><span class="co">##  bed:</span></span>
<span><span class="co">##  'data.frame':   61 obs. of  4 variables:</span></span>
<span><span class="co">##   $ chrom     : chr  "II" "II" "II" "II" ...</span></span>
<span><span class="co">##   $ chromStart: num  0 8000 30000 35000 46000 60000 75000 81000 90000 101000 ...</span></span>
<span><span class="co">##   $ chromEnd  : num  8000 30000 35000 46000 60000 75000 81000 90000 101000 136000 ...</span></span>
<span><span class="co">##   $ name      : chr  "gap" "domain" "gap" "domain" ...</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The resulting <code>domains</code> object can be used to extract annotated domains, store them in <code>topologicalFeatures</code> of the original <code>HiCExperiment</code>, and optionally write a <code>bed</code> file to export them in text.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">topologicalFeatures</a></span><span class="op">(</span><span class="va">hic</span>, <span class="st">'domain'</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">domains</span><span class="op">$</span><span class="va">bed</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>chromStart <span class="op">=</span> <span class="va">chromStart</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">name</span> <span class="op">==</span> <span class="st">'domain'</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/makeGRangesFromDataFrame.html">makeGRangesFromDataFrame</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">topologicalFeatures</a></span><span class="op">(</span><span class="va">hic</span>, <span class="st">'domain'</span><span class="op">)</span></span>
<span><span class="co">##  GRanges object with 52 ranges and 0 metadata columns:</span></span>
<span><span class="co">##         seqnames        ranges strand</span></span>
<span><span class="co">##            &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##     [1]       II    8001-30000      *</span></span>
<span><span class="co">##     [2]       II   35001-46000      *</span></span>
<span><span class="co">##     [3]       II   46001-60000      *</span></span>
<span><span class="co">##     [4]       II   60001-75000      *</span></span>
<span><span class="co">##     [5]       II   75001-81000      *</span></span>
<span><span class="co">##     ...      ...           ...    ...</span></span>
<span><span class="co">##    [48]       II 664001-681000      *</span></span>
<span><span class="co">##    [49]       II 681001-707000      *</span></span>
<span><span class="co">##    [50]       II 707001-714000      *</span></span>
<span><span class="co">##    [51]       II 714001-761000      *</span></span>
<span><span class="co">##    [52]       II 761001-806000      *</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span>
<span></span>
<span><span class="fu">rtracklayer</span><span class="fu">::</span><span class="fu">export</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">topologicalFeatures</a></span><span class="op">(</span><span class="va">hic</span>, <span class="st">'domain'</span><span class="op">)</span>, <span class="st">'hic_domains.bed'</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="gothic" class="level2" data-number="9.4"><h2 data-number="9.4" class="anchored" data-anchor-id="gothic">
<span class="header-section-number">9.4</span> GOTHiC</h2>
<p><code>GOTHiC</code> relies on a cumulative binomial test to detect interactions between distal genomic loci that have significantly more reads than expected by chance in Hi-C experiments (<span class="citation" data-cites="Mifsud_2017">Mifsud et al. (<a href="#ref-Mifsud_2017" role="doc-biblioref">2017</a>)</span>).</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Using the <code>GOTHiC</code> function
</div>
</div>
<div class="callout-body-container callout-body">
<p>Unfortunately, the main <code>GOTHiC</code> function require two <code>.bam</code> files as input. These files are often deleted due to their larger size, while the filtered <code>pairs</code> file itself is retained.</p>
<p>Moreover, the internal nuts and bolts of the main <code>GOTHiC</code> function perform several operations that are not required in modern workflows:</p>
<ol type="1">
<li>
<a href="https://code.bioconductor.org/browse/GOTHiC/blob/RELEASE_3_17/R/GOTHiC.R#L814">Filtering pairs from same restriction fragment</a>; this step is now usually taken care of automatically, e.g.&nbsp;with <code>HiCool</code> Hi-C processing package.</li>
<li>
<a href="https://code.bioconductor.org/browse/GOTHiC/blob/RELEASE_3_17/R/GOTHiC.R#L826">Filtering short-range pairs</a>; the <code>GOTHiC</code> package hard-codes a 10kb lower threshold for minimum pair distance. More advanced optimized filtering approaches have been implemented since then, to circumvent the need for such hard-coded threshold.</li>
<li>
<a href="https://code.bioconductor.org/browse/GOTHiC/blob/RELEASE_3_17/R/GOTHiC.R#L834">Binning pairs</a>; this step is also already taken care of, when working with Hi-C matrices in modern formats, e.g.&nbsp;with <code>.(m)cool</code> files.</li>
</ol>
</div>
</div>
<p>Based on these facts, we can simplify the binomial test function provided by <code>GOTHiC</code> so that it can directly used binned interactions imported as a <code>HiCExperiment</code> object in <code>R</code>.</p>
<div class="cell" data-layout-align="center">
<details><summary>Show the code for <code>GOTHiC_binomial</code> function</summary><div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">GOTHiC_binomial</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">trans</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span> <span class="op">!=</span> <span class="fl">0</span><span class="op">)</span> <span class="kw"><a href="https://rdrr.io/r/base/stop.html">stop</a></span><span class="op">(</span><span class="st">"Only `cis` interactions can be used here."</span><span class="op">)</span></span>
<span>    <span class="va">ints</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">interactions</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>        <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">seqnames1</span>, <span class="va">start1</span>, <span class="va">seqnames2</span>, <span class="va">start2</span>, <span class="va">count</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>        <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html">rename</a></span><span class="op">(</span>chr1 <span class="op">=</span> <span class="va">seqnames1</span>, locus1 <span class="op">=</span> <span class="va">start1</span>, chr2 <span class="op">=</span> <span class="va">seqnames2</span>, locus2 <span class="op">=</span> <span class="va">start2</span>, frequencies <span class="op">=</span> <span class="va">count</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>        <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>locus1 <span class="op">=</span> <span class="va">locus1</span> <span class="op">-</span> <span class="fl">1</span>, locus2 <span class="op">=</span> <span class="va">locus2</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>        <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>int1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">chr1</span>, <span class="st">'_'</span>, <span class="va">locus1</span><span class="op">)</span>, int2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">chr2</span>, <span class="st">'_'</span>, <span class="va">locus2</span><span class="op">)</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="va">numberOfReadPairs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">ints</span><span class="op">$</span><span class="va">frequencies</span><span class="op">)</span></span>
<span>    <span class="va">all_bins</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html">unique</a></span><span class="op">(</span><span class="va">ints</span><span class="op">$</span><span class="va">int1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html">unique</a></span><span class="op">(</span><span class="va">ints</span><span class="op">$</span><span class="va">int2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">all_bins</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html">sort</a></span><span class="op">(</span><span class="va">all_bins</span><span class="op">)</span></span>
<span>    <span class="va">upperhalfBinNumber</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">all_bins</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">all_bins</span><span class="op">)</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span></span>
<span></span>
<span>    <span class="va">cov</span> <span class="op">&lt;-</span> <span class="va">ints</span> <span class="op">|&gt;</span> </span>
<span>        <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">int1</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>        <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">tally</a></span><span class="op">(</span><span class="va">frequencies</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>        <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">full_join</a></span><span class="op">(</span><span class="va">ints</span> <span class="op">|&gt;</span> </span>
<span>            <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">int2</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>            <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">tally</a></span><span class="op">(</span><span class="va">frequencies</span><span class="op">)</span>, </span>
<span>            by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'int1'</span> <span class="op">=</span> <span class="st">'int2'</span><span class="op">)</span></span>
<span>        <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>        <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rowwise.html">rowwise</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>        <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>coverage <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">n.x</span>, <span class="va">n.y</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>        <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">ungroup</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>        <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>relative_coverage <span class="op">=</span> <span class="va">coverage</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">coverage</span><span class="op">)</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span><span class="va">ints</span>,</span>
<span>        cov1 <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">ints</span>, <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">cov</span>, <span class="va">int1</span>, <span class="va">relative_coverage</span><span class="op">)</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'int1'</span> <span class="op">=</span> <span class="st">'int1'</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">relative_coverage</span>, </span>
<span>        cov2 <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">ints</span>, <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">cov</span>, <span class="va">int1</span>, <span class="va">relative_coverage</span><span class="op">)</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'int2'</span> <span class="op">=</span> <span class="st">'int1'</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">relative_coverage</span>,</span>
<span>        probability <span class="op">=</span> <span class="va">cov1</span> <span class="op">*</span> <span class="va">cov2</span> <span class="op">*</span> <span class="fl">2</span> <span class="op">*</span> <span class="fl">1</span><span class="op">/</span><span class="op">(</span><span class="fl">1</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">cov</span><span class="op">$</span><span class="va">relative_coverage</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span>,</span>
<span>        predicted <span class="op">=</span> <span class="va">probability</span> <span class="op">*</span> <span class="va">numberOfReadPairs</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rowwise.html">rowwise</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span></span>
<span>        pvalue <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/binom.test.html">binom.test</a></span><span class="op">(</span></span>
<span>            <span class="va">frequencies</span>, </span>
<span>            <span class="va">numberOfReadPairs</span>, </span>
<span>            <span class="va">probability</span>,</span>
<span>            alternative <span class="op">=</span> <span class="st">"greater"</span></span>
<span>        <span class="op">)</span><span class="op">$</span><span class="va">p.value</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">ungroup</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span></span>
<span>        logFoldChange <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="va">frequencies</span> <span class="op">/</span> <span class="va">predicted</span><span class="op">)</span>, </span>
<span>        qvalue <span class="op">=</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html">p.adjust</a></span><span class="op">(</span><span class="va">pvalue</span>, method <span class="op">=</span> <span class="st">"BH"</span>, n <span class="op">=</span> <span class="va">upperhalfBinNumber</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">scores</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"probability"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">results</span><span class="op">$</span><span class="va">probability</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">scores</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"predicted"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">results</span><span class="op">$</span><span class="va">predicted</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">scores</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"pvalue"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">results</span><span class="op">$</span><span class="va">pvalue</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">scores</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"qvalue"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">results</span><span class="op">$</span><span class="va">qvalue</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">scores</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"logFoldChange"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">results</span><span class="op">$</span><span class="va">logFoldChange</span></span>
<span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span></span>
<span><span class="op">}</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">GOTHiC_binomial</span><span class="op">(</span><span class="va">hic</span><span class="op">[</span><span class="st">"II"</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">res</span></span>
<span><span class="co">##  `HiCExperiment` object with 471,364 contacts over 802 regions </span></span>
<span><span class="co">##  -------</span></span>
<span><span class="co">##  fileName: "/root/.cache/R/ExperimentHub/f725e39d5_7752" </span></span>
<span><span class="co">##  focus: "II" </span></span>
<span><span class="co">##  resolutions(5): 1000 2000 4000 8000 16000</span></span>
<span><span class="co">##  active resolution: 1000 </span></span>
<span><span class="co">##  interactions: 74360 </span></span>
<span><span class="co">##  scores(7): count balanced probability predicted pvalue qvalue logFoldChange </span></span>
<span><span class="co">##  topologicalFeatures: compartments(0) borders(0) loops(0) viewpoints(0) domain(52) </span></span>
<span><span class="co">##  pairsFile: N/A </span></span>
<span><span class="co">##  metadata(0):</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/interaction-accessors.html">interactions</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span>
<span><span class="co">##  GInteractions object with 74360 interactions and 9 metadata columns:</span></span>
<span><span class="co">##            seqnames1       ranges1 strand1     seqnames2       ranges2</span></span>
<span><span class="co">##                &lt;Rle&gt;     &lt;IRanges&gt;   &lt;Rle&gt;         &lt;Rle&gt;     &lt;IRanges&gt;</span></span>
<span><span class="co">##        [1]        II        1-1000       * ---        II     1001-2000</span></span>
<span><span class="co">##        [2]        II        1-1000       * ---        II     5001-6000</span></span>
<span><span class="co">##        [3]        II        1-1000       * ---        II     6001-7000</span></span>
<span><span class="co">##        [4]        II        1-1000       * ---        II     8001-9000</span></span>
<span><span class="co">##        [5]        II        1-1000       * ---        II    9001-10000</span></span>
<span><span class="co">##        ...       ...           ...     ... ...       ...           ...</span></span>
<span><span class="co">##    [74356]        II 807001-808000       * ---        II 809001-810000</span></span>
<span><span class="co">##    [74357]        II 807001-808000       * ---        II 810001-811000</span></span>
<span><span class="co">##    [74358]        II 808001-809000       * ---        II 808001-809000</span></span>
<span><span class="co">##    [74359]        II 808001-809000       * ---        II 809001-810000</span></span>
<span><span class="co">##    [74360]        II 809001-810000       * ---        II 809001-810000</span></span>
<span><span class="co">##            strand2 |   bin_id1   bin_id2     count  balanced probability</span></span>
<span><span class="co">##              &lt;Rle&gt; | &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">##        [1]       * |       231       232         1       NaN 7.83580e-09</span></span>
<span><span class="co">##        [2]       * |       231       236         2       NaN 2.81318e-08</span></span>
<span><span class="co">##        [3]       * |       231       237         1       NaN 2.02960e-08</span></span>
<span><span class="co">##        [4]       * |       231       239         2       NaN 6.73108e-08</span></span>
<span><span class="co">##        [5]       * |       231       240         3       NaN 7.37336e-08</span></span>
<span><span class="co">##        ...     ... .       ...       ...       ...       ...         ...</span></span>
<span><span class="co">##    [74356]       * |      1038      1040         8 0.0472023 3.85638e-07</span></span>
<span><span class="co">##    [74357]       * |      1038      1041         1       NaN 5.03006e-08</span></span>
<span><span class="co">##    [74358]       * |      1039      1039         1       NaN 8.74604e-08</span></span>
<span><span class="co">##    [74359]       * |      1039      1040         7       NaN 1.02111e-07</span></span>
<span><span class="co">##    [74360]       * |      1040      1040         2 0.0411355 1.19216e-07</span></span>
<span><span class="co">##             predicted      pvalue      qvalue logFoldChange</span></span>
<span><span class="co">##             &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt;     &lt;numeric&gt;</span></span>
<span><span class="co">##        [1] 0.00369352 3.68670e-03 0.063385760       8.08079</span></span>
<span><span class="co">##        [2] 0.01326033 8.71446e-05 0.001926954       7.23674</span></span>
<span><span class="co">##        [3] 0.00956681 9.52120e-03 0.150288341       6.70775</span></span>
<span><span class="co">##        [4] 0.03172791 4.92808e-04 0.009806734       5.97810</span></span>
<span><span class="co">##        [5] 0.03475538 6.81713e-06 0.000173165       6.43158</span></span>
<span><span class="co">##        ...        ...         ...         ...           ...</span></span>
<span><span class="co">##    [74356]  0.1817758 2.51560e-11 1.07966e-09       5.45977</span></span>
<span><span class="co">##    [74357]  0.0237099 2.34310e-02 3.38098e-01       5.39837</span></span>
<span><span class="co">##    [74358]  0.0412257 4.03875e-02 5.49519e-01       4.60031</span></span>
<span><span class="co">##    [74359]  0.0481315 1.13834e-13 5.77259e-12       7.18423</span></span>
<span><span class="co">##    [74360]  0.0561941 1.52097e-03 2.79707e-02       5.15344</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    regions: 802 ranges and 4 metadata columns</span></span>
<span><span class="co">##    seqinfo: 16 sequences from an unspecified genome</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<!-- ## HiCDOC -->
</section><section id="session-info" class="level1 unnumbered"><h1 class="unnumbered">Session info</h1>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to expand üëá
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">sessioninfo</span><span class="fu">::</span><span class="fu"><a href="https://r-lib.github.io/sessioninfo/reference/session_info.html">session_info</a></span><span class="op">(</span>include_base <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">##  ‚îÄ Session info ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ</span></span>
<span><span class="co">##   setting  value</span></span>
<span><span class="co">##   version  R Under development (unstable) (2024-01-17 r85813)</span></span>
<span><span class="co">##   os       Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">##   system   x86_64, linux-gnu</span></span>
<span><span class="co">##   ui       X11</span></span>
<span><span class="co">##   language (EN)</span></span>
<span><span class="co">##   collate  C</span></span>
<span><span class="co">##   ctype    en_US.UTF-8</span></span>
<span><span class="co">##   tz       Etc/UTC</span></span>
<span><span class="co">##   date     2024-01-22</span></span>
<span><span class="co">##   pandoc   3.1.1 @ /usr/local/bin/ (via rmarkdown)</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  ‚îÄ Packages ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ</span></span>
<span><span class="co">##   package              * version     date (UTC) lib source</span></span>
<span><span class="co">##   abind                  1.4-5       2016-07-21 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   aggregation            1.0.1       2018-01-25 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   AnnotationDbi          1.65.2      2023-11-03 [2] Bioconductor</span></span>
<span><span class="co">##   AnnotationHub        * 3.11.1      2023-12-11 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   base                 * 4.4.0       2024-01-18 [3] local</span></span>
<span><span class="co">##   beeswarm               0.4.0       2021-06-01 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   Biobase              * 2.63.0      2023-10-24 [2] Bioconductor</span></span>
<span><span class="co">##   BiocFileCache        * 2.11.1      2023-10-26 [2] Bioconductor</span></span>
<span><span class="co">##   BiocGenerics         * 0.49.1      2023-11-01 [2] Bioconductor</span></span>
<span><span class="co">##   BiocIO                 1.13.0      2023-10-24 [2] Bioconductor</span></span>
<span><span class="co">##   BiocManager            1.30.22     2023-08-08 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   BiocParallel           1.37.0      2023-10-24 [2] Bioconductor</span></span>
<span><span class="co">##   BiocVersion            3.19.1      2023-10-26 [2] Bioconductor</span></span>
<span><span class="co">##   Biostrings             2.71.1      2023-10-25 [2] Bioconductor</span></span>
<span><span class="co">##   bit                    4.0.5       2022-11-15 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   bit64                  4.0.5       2020-08-30 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   bitops                 1.0-7       2021-04-24 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   blob                   1.2.4       2023-03-17 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   BSgenome               1.71.1      2023-11-01 [2] Bioconductor</span></span>
<span><span class="co">##   cachem                 1.0.8       2023-05-01 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   Cairo                  1.6-2       2023-11-28 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   calibrate              1.7.7       2020-06-19 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   cli                    3.6.2       2023-12-11 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   codetools              0.2-19      2023-02-01 [3] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   colorspace             2.1-0       2023-01-23 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   compiler               4.4.0       2024-01-18 [3] local</span></span>
<span><span class="co">##   crayon                 1.5.2       2022-09-29 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   csaw                   1.37.1      2023-11-30 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   curl                   5.2.0       2023-12-08 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   data.table             1.14.10     2023-12-08 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   datasets             * 4.4.0       2024-01-18 [3] local</span></span>
<span><span class="co">##   DBI                    1.2.1       2024-01-12 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   dbplyr               * 2.4.0       2023-10-26 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   DelayedArray           0.29.0      2023-10-24 [2] Bioconductor</span></span>
<span><span class="co">##   diffHic              * 1.35.0      2023-10-24 [2] Bioconductor</span></span>
<span><span class="co">##   digest                 0.6.34      2024-01-11 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   dplyr                * 1.1.4       2023-11-17 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   edgeR                  4.1.11      2024-01-18 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   evaluate               0.23        2023-11-01 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   ExperimentHub        * 2.11.1      2023-12-11 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   fansi                  1.0.6       2023-12-08 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   farver                 2.1.1       2022-07-06 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   fastmap                1.1.1       2023-02-24 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   filelock               1.0.3       2023-12-11 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   generics               0.1.3       2022-07-05 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   GenomeInfoDb         * 1.39.5      2024-01-01 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   GenomeInfoDbData       1.2.11      2024-01-22 [2] Bioconductor</span></span>
<span><span class="co">##   GenomicAlignments      1.39.2      2024-01-16 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   GenomicRanges        * 1.55.1      2023-10-29 [2] Bioconductor</span></span>
<span><span class="co">##   ggbeeswarm             0.7.2       2023-04-29 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   ggplot2              * 3.4.4       2023-10-12 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   ggrastr                1.0.2       2023-06-01 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   glue                   1.7.0       2024-01-09 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   graphics             * 4.4.0       2024-01-18 [3] local</span></span>
<span><span class="co">##   grDevices            * 4.4.0       2024-01-18 [3] local</span></span>
<span><span class="co">##   grid                   4.4.0       2024-01-18 [3] local</span></span>
<span><span class="co">##   gridExtra              2.3         2017-09-09 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   gtable                 0.3.4       2023-08-21 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   gtools                 3.9.5       2023-11-20 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   HiCcompare             1.25.0      2023-10-24 [2] Bioconductor</span></span>
<span><span class="co">##   HiCExperiment        * 1.3.0       2023-10-24 [2] Bioconductor</span></span>
<span><span class="co">##   HiContacts           * 1.5.0       2023-10-24 [2] Bioconductor</span></span>
<span><span class="co">##   HiContactsData       * 1.5.3       2024-01-22 [2] Github (js2264/HiContactsData@d5bebe7)</span></span>
<span><span class="co">##   hms                    1.1.3       2023-03-21 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   htmltools              0.5.7       2023-11-03 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   htmlwidgets            1.6.4       2023-12-06 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   httr                   1.4.7       2023-08-15 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   InteractionSet       * 1.31.0      2023-10-24 [2] Bioconductor</span></span>
<span><span class="co">##   IRanges              * 2.37.1      2024-01-19 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   jsonlite               1.8.8       2023-12-04 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   KEGGREST               1.43.0      2023-10-24 [2] Bioconductor</span></span>
<span><span class="co">##   KernSmooth             2.23-22     2023-07-10 [3] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   knitr                  1.45        2023-10-30 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   labeling               0.4.3       2023-08-29 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   lattice                0.22-5      2023-10-24 [3] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   lifecycle              1.0.4       2023-11-07 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   limma                  3.59.1      2023-10-30 [2] Bioconductor</span></span>
<span><span class="co">##   locfit                 1.5-9.8     2023-06-11 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   magrittr               2.0.3       2022-03-30 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   MASS                   7.3-60.2    2024-01-18 [3] local</span></span>
<span><span class="co">##   Matrix                 1.6-5       2024-01-11 [3] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   MatrixGenerics       * 1.15.0      2023-10-24 [2] Bioconductor</span></span>
<span><span class="co">##   matrixStats          * 1.2.0       2023-12-11 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   memoise                2.0.1       2021-11-26 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   metapod                1.11.1      2023-12-24 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   methods              * 4.4.0       2024-01-18 [3] local</span></span>
<span><span class="co">##   mgcv                   1.9-1       2023-12-21 [3] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   mime                   0.12        2021-09-28 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   multiHiCcompare      * 1.21.0      2023-10-24 [2] Bioconductor</span></span>
<span><span class="co">##   munsell                0.5.0       2018-06-12 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   nlme                   3.1-164     2023-11-27 [3] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   parallel               4.4.0       2024-01-18 [3] local</span></span>
<span><span class="co">##   pbapply                1.7-2       2023-06-27 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   pheatmap               1.0.12      2019-01-04 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   pillar                 1.9.0       2023-03-22 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   pkgconfig              2.0.3       2019-09-22 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   plyinteractions      * 1.1.0       2023-10-24 [2] Bioconductor</span></span>
<span><span class="co">##   plyr                   1.8.9       2023-10-02 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   plyranges              1.23.0      2023-10-24 [2] Bioconductor</span></span>
<span><span class="co">##   png                    0.1-8       2022-11-29 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   purrr                * 1.0.2       2023-08-10 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   qqman                  0.1.9       2023-08-23 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   R6                     2.5.1       2021-08-19 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   rappdirs               0.3.3       2021-01-31 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   RColorBrewer           1.1-3       2022-04-03 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   Rcpp                   1.0.12      2024-01-09 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   RCurl                  1.98-1.14   2024-01-09 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   readr                  2.1.5       2024-01-10 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   reshape2               1.4.4       2020-04-09 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   restfulr               0.0.15      2022-06-16 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   rhdf5                  2.47.2      2024-01-15 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   rhdf5filters           1.15.1      2023-11-06 [2] Bioconductor</span></span>
<span><span class="co">##   Rhdf5lib               1.25.1      2023-12-11 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   Rhtslib                2.99.2      2024-01-18 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   rjson                  0.2.21      2022-01-09 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   rlang                  1.1.3       2024-01-10 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   rmarkdown              2.25        2023-09-18 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   Rsamtools              2.19.3      2024-01-17 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   RSpectra               0.16-1      2022-04-24 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   RSQLite                2.3.5       2024-01-21 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   rtracklayer            1.63.0      2024-01-22 [2] Github (lawremi/rtracklayer@86407bb)</span></span>
<span><span class="co">##   S4Arrays               1.3.2       2024-01-14 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   S4Vectors            * 0.41.3      2024-01-01 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   scales                 1.3.0       2023-11-28 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   sessioninfo            1.2.2       2021-12-06 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   SparseArray            1.3.3       2024-01-14 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   splines                4.4.0       2024-01-18 [3] local</span></span>
<span><span class="co">##   statmod                1.5.0       2023-01-06 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   stats                * 4.4.0       2024-01-18 [3] local</span></span>
<span><span class="co">##   stats4               * 4.4.0       2024-01-18 [3] local</span></span>
<span><span class="co">##   strawr                 0.0.91      2023-03-29 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   stringi                1.8.3       2023-12-11 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   stringr                1.5.1       2023-11-14 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   SummarizedExperiment * 1.33.2      2024-01-07 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   tibble                 3.2.1       2023-03-20 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   tidyr                * 1.3.0       2023-01-24 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   tidyselect             1.2.0       2022-10-10 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   tools                  4.4.0       2024-01-18 [3] local</span></span>
<span><span class="co">##   TopDom               * 0.10.1      2021-05-06 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   tzdb                   0.4.0       2023-05-12 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   utf8                   1.2.4       2023-10-22 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   utils                * 4.4.0       2024-01-18 [3] local</span></span>
<span><span class="co">##   vctrs                  0.6.5       2023-12-01 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   vipor                  0.4.7       2023-12-18 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   vroom                  1.6.5       2023-12-05 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   withr                  3.0.0       2024-01-16 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   xfun                   0.41        2023-11-01 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   XML                    3.99-0.16.1 2024-01-22 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   XVector                0.43.1      2024-01-10 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   yaml                   2.3.8       2023-12-11 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   zlibbioc               1.49.0      2023-10-24 [2] Bioconductor</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##   [1] /tmp/Rtmp4kTMSb/Rinstb104d8a34</span></span>
<span><span class="co">##   [2] /usr/local/lib/R/site-library</span></span>
<span><span class="co">##   [3] /usr/local/lib/R/library</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</section><section id="references" class="level1 unnumbered"><h1 class="unnumbered">References</h1>


<div id="refs" class="references csl-bib-body hanging-indent" data-line-spacing="2" role="list">
<div id="ref-Bengtsson_2020" class="csl-entry" role="listitem">
Bengtsson, H., Shin, H., Lazaris, H., Hu, G., &amp; Zhou, X. (2020). <em>R package TopDom: An efficient and deterministic method for identifying topological domains in genomes</em>. <a href="https://github.com/HenrikBengtsson/TopDom">https://github.com/HenrikBengtsson/TopDom</a>
</div>
<div id="ref-Lun_2015" class="csl-entry" role="listitem">
Lun, A. T. L., &amp; Smyth, G. K. (2015). <span class="nocase">diffHic: a Bioconductor package to detect differential genomic interactions in Hi-C data</span>. <em>BMC Bioinf.</em>, <em>16</em>(1), 1‚Äì11. <a href="https://doi.org/10.1186/s12859-015-0683-0">https://doi.org/10.1186/s12859-015-0683-0</a>
</div>
<div id="ref-Mifsud_2017" class="csl-entry" role="listitem">
Mifsud, B., Martincorena, I., Darbo, E., Sugar, R., Schoenfelder, S., Fraser, P., &amp; Luscombe, N. M. (2017). <span>GOTHiC</span>, a probabilistic model to resolve complex biases and to identify real interactions in hi-c data. <em><span>PLOS</span> <span>ONE</span></em>, <em>12</em>(4), e0174744. <a href="https://doi.org/10.1371/journal.pone.0174744">https://doi.org/10.1371/journal.pone.0174744</a>
</div>
<div id="ref-Shin_2015" class="csl-entry" role="listitem">
Shin, H., Shi, Y., Dai, C., Tjong, H., Gong, K., Alber, F., &amp; Zhou, X. J. (2015). <span>TopDom</span>: An efficient and deterministic method for identifying topological domains in genomes. <em>Nucleic Acids Research</em>, <em>44</em>(7), e70‚Äìe70. <a href="https://doi.org/10.1093/nar/gkv1505">https://doi.org/10.1093/nar/gkv1505</a>
</div>
<div id="ref-Stansfield_2019" class="csl-entry" role="listitem">
Stansfield, J. C., Cresswell, K. G., &amp; Dozmorov, M. G. (2019). <span class="nocase">multiHiCcompare</span>: Joint normalization and comparative analysis of complex hi-c experiments. <em>Bioinformatics</em>, <em>35</em>(17), 2916‚Äì2923. <a href="https://doi.org/10.1093/bioinformatics/btz048">https://doi.org/10.1093/bioinformatics/btz048</a>
</div>
</div>
</section><a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "Óßã";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../pages/disseminating.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Data gateways: accessing public Hi-C data portals</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../pages/workflow-yeast.html" class="pagination-link">
        <span class="nav-page-text">Workflow 1: Distance-dependent interactions across yeast mutants</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left"><i>OHCA: Orchestrating Hi-C analysis with Bioconductor</i> was written by J. Serizay</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">This book was built with <a href="https://github.com/js2264/BiocBook/">BiocBook</a> with <span class="emoji" data-emoji="heart">‚ù§Ô∏è</span></div>
  </div>
</footer>


</body></html>