<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Orchestrating Hi-C analysis with Bioconductor - 5&nbsp; Matrix-centric analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../pages/interactions-centric.html" rel="next">
<link href="../pages/visualization.html" rel="prev">
<link href="../assets/cover.jpg" rel="icon" type="image/jpeg">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
</head>
<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/matrix-centric.html">In-depth Hi-C analysis</a></li><li class="breadcrumb-item"><a href="../pages/matrix-centric.html"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Matrix-centric analysis</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../">Orchestrating Hi-C analysis with Bioconductor</a> 
        <div class="sidebar-tools-main">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label=""><i class="bi bi-git"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
<li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://github.com/js2264/OHCA/">
            Source Code
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://js2264.github.io/OHCA/devel/">
            Browse version `devel`
            </a>
          </li>
      </ul>
</div>
    <a href="../Orchestrating-Hi-C-analysis-with-Bioconductor.pdf" rel="" title="Download PDF" class="quarto-navigation-tool px-1" aria-label="Download PDF"><i class="bi bi-file-pdf"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/preamble.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preamble</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Fundamentals concepts</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/principles.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Hi-C pre-processing steps</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/data-representation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Hi-C data structures in R</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/parsing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Manipulating Hi-C data in R</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/visualization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Hi-C data visualization</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">In-depth Hi-C analysis</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/matrix-centric.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Matrix-centric analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/interactions-centric.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Interactions-centric analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/topological-features.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Finding topological features in Hi-C</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Advanced Hi-C topics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/disseminating.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Data gateways: accessing public Hi-C data portals</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/interoperability.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Interoperability: using HiCExperiment with other R packages</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/workflow-yeast.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Workflow 1: Distance-dependent interactions across yeast mutants</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/workflow-chicken.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Workflow 2: Chromosome compartment cohesion upon mitosis entry</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/workflow-centros.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Workflow 3: Inter-centromere interactions in yeast</span></span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li>
<a href="#operations-in-an-individual-matrix" id="toc-operations-in-an-individual-matrix" class="nav-link active" data-scroll-target="#operations-in-an-individual-matrix"><span class="header-section-number">5.1</span> Operations in an individual matrix</a>
  <ul class="collapse">
<li><a href="#balancing-a-raw-interaction-count-map" id="toc-balancing-a-raw-interaction-count-map" class="nav-link" data-scroll-target="#balancing-a-raw-interaction-count-map"><span class="header-section-number">5.1.1</span> Balancing a raw interaction count map</a></li>
  <li><a href="#computing-observedexpected-oe-map" id="toc-computing-observedexpected-oe-map" class="nav-link" data-scroll-target="#computing-observedexpected-oe-map"><span class="header-section-number">5.1.2</span> Computing observed/expected (O/E) map</a></li>
  <li><a href="#computing-autocorrelated-map" id="toc-computing-autocorrelated-map" class="nav-link" data-scroll-target="#computing-autocorrelated-map"><span class="header-section-number">5.1.3</span> Computing autocorrelated map</a></li>
  <li><a href="#despeckling-smoothing-out-a-contact-map" id="toc-despeckling-smoothing-out-a-contact-map" class="nav-link" data-scroll-target="#despeckling-smoothing-out-a-contact-map"><span class="header-section-number">5.1.4</span> Despeckling (smoothing out) a contact map</a></li>
  </ul>
</li>
  <li>
<a href="#operations-between-multiple-matrices" id="toc-operations-between-multiple-matrices" class="nav-link" data-scroll-target="#operations-between-multiple-matrices"><span class="header-section-number">5.2</span> Operations between multiple matrices</a>
  <ul class="collapse">
<li><a href="#merging-maps" id="toc-merging-maps" class="nav-link" data-scroll-target="#merging-maps"><span class="header-section-number">5.2.1</span> Merging maps</a></li>
  <li><a href="#computing-ratio-between-two-maps" id="toc-computing-ratio-between-two-maps" class="nav-link" data-scroll-target="#computing-ratio-between-two-maps"><span class="header-section-number">5.2.2</span> Computing ratio between two maps</a></li>
  </ul>
</li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title">
<span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Matrix-centric analysis</span>
</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Aims
</div>
</div>
<div class="callout-body-container callout-body">
<p>This chapter focuses on the various analytical tools offered by <code>HiContacts</code> to compute matrix-related metrics from a <code>HiCExperiment</code> object.</p>
</div>
</div>
<p>In the first part of this book, we have seen how to query parts or all of the data contained in Hi-C contact matrices using the <code>HiCExperiment</code> object (<a href="../pages/data-representation.html#hicexperiment-class">Chapter 2</a>), how to manipulate <code>HiCExperiment</code> objects (<a href="../pages/parsing.html">Chapter 3</a>) and how to visualize Hi-C contact matrices as heatmaps (<a href="../pages/visualization.html">Chapter 4</a>).</p>
<p>The <code>HiContacts</code> package directly operates on <code>HiCExperiment</code> objects and extends its usability by providing a comprehensive toolkit to analyze Hi-C data, focusing on four main topics:</p>
<ul>
<li>Contact matrix-centric analyses (this chapter)</li>
<li>Interactions-centric analyses (<a href="../pages/interactions-centric.html">Chapter 6</a>)</li>
<li>Structural feature annotations (<a href="../pages/topological-features.html">Chapter 7</a>)</li>
<li>Hi-C visualization (see <a href="../pages/visualization.html">previous chapter</a>)</li>
</ul>
<p><strong>Matrix-centric</strong> analyses consider a <code>HiCExperiment</code> object from the “matrix” perspective to perform a range of matrix-based operations. This encompasses:</p>
<ul>
<li>Computing observed/expected (O/E) map</li>
<li>Computing auto-correlation map</li>
<li>Smoothing out a contact map</li>
<li>Merging multiple Hi-C maps together</li>
<li>Comparing two Hi-C maps to each other</li>
</ul>
<p><img src="images/20230421134800.jpg" class="img-fluid"></p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>All the functions described in this chapter are <strong>endomorphisms</strong>: they take <code>HiCExperiment</code> objects as input and return modified <code>HiCExperiment</code> objects.</li>
<li>Internally, most of the functions presented in this chapter make a call to <code>as.matrix(&lt;HiCExperiment&gt;)</code> to coerce it into a <code>matrix</code>.</li>
</ul>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Generating the example <code>hic</code> object 👇
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>To demonstrate <code>HiContacts</code> functionalities, we will create an <code>HiCExperiment</code> object from an example <code>.cool</code> file provided in the <code>HiContactsData</code> package.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/js2264/HiCExperiment">HiCExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/js2264/HiContactsData">HiContactsData</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># ---- This downloads an example `.mcool` file and caches it locally </span></span>
<span><span class="va">coolf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/HiContactsData/man/HiContactsData.html">HiContactsData</a></span><span class="op">(</span><span class="st">'yeast_wt'</span>, <span class="st">'mcool'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># ---- This creates a connection to the disk-stored `.mcool` file</span></span>
<span><span class="va">cf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/CoolFile-class.html">CoolFile</a></span><span class="op">(</span><span class="va">coolf</span><span class="op">)</span></span>
<span><span class="va">cf</span></span>
<span></span>
<span><span class="co"># ---- This imports contacts from the chromosome `II` at resolution `2000`</span></span>
<span><span class="va">hic</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocIO/man/IO.html">import</a></span><span class="op">(</span><span class="va">cf</span>, focus <span class="op">=</span> <span class="st">'II'</span>, resolution <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hic</span></span>
<span><span class="co">##  `HiCExperiment` object with 471,364 contacts over 407 regions </span></span>
<span><span class="co">##  -------</span></span>
<span><span class="co">##  fileName: "/root/.cache/R/ExperimentHub/17079abac61_7752" </span></span>
<span><span class="co">##  focus: "II" </span></span>
<span><span class="co">##  resolutions(5): 1000 2000 4000 8000 16000</span></span>
<span><span class="co">##  active resolution: 2000 </span></span>
<span><span class="co">##  interactions: 34063 </span></span>
<span><span class="co">##  scores(2): count balanced </span></span>
<span><span class="co">##  topologicalFeatures: compartments(0) borders(0) loops(0) viewpoints(0) centromeres(16) </span></span>
<span><span class="co">##  pairsFile: N/A </span></span>
<span><span class="co">##  metadata(0):</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<section id="operations-in-an-individual-matrix" class="level2 page-columns page-full" data-number="5.1"><h2 data-number="5.1" class="anchored" data-anchor-id="operations-in-an-individual-matrix">
<span class="header-section-number">5.1</span> Operations in an individual matrix</h2>
<section id="balancing-a-raw-interaction-count-map" class="level3 page-columns page-full" data-number="5.1.1"><h3 data-number="5.1.1" class="anchored" data-anchor-id="balancing-a-raw-interaction-count-map">
<span class="header-section-number">5.1.1</span> Balancing a raw interaction count map</h3>
<p>Hi-C sequencing coverage is systematically affected by multiple confounding factors, e.g.&nbsp; density of restriction sites, GC%, genome mappability, etc.. Overall, it generally ends up not homogenous throughout the entire genome and this leads to artifacts in un-normalized <code>count</code> matrices.</p>
<p>To correct for sequencing coverage heterogeneity of raw <code>count</code> maps, Hi-C data can be normalized using matrix balancing approaches (<span class="citation" data-cites="Cournac_2012">Cournac et al. (<a href="#ref-Cournac_2012" role="doc-biblioref">2012</a>)</span>, <span class="citation" data-cites="Imakaev_2012">Imakaev et al. (<a href="#ref-Imakaev_2012" role="doc-biblioref">2012</a>)</span>). This is generally done directly on the disk-stored matrices using out-of-memory strategies (e.g.&nbsp;with <code>cooler balance &lt;.cool&gt;</code>). However, if contact matrix files are imported into a <code>HiCExperiment</code> object but no <code>balanced</code> scores are available, in-memory balancing can be performed using the <code>normalize</code> function. This adds an extra <code>ICE</code> element in <code>scores</code> list (while the <code>interactions</code> themselves are unmodified).</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">normalized_hic</span> <span class="op">&lt;-</span> <span class="fu">normalize</span><span class="op">(</span><span class="va">hic</span><span class="op">)</span></span>
<span><span class="va">normalized_hic</span></span>
<span><span class="co">##  `HiCExperiment` object with 471,364 contacts over 407 regions </span></span>
<span><span class="co">##  -------</span></span>
<span><span class="co">##  fileName: "/root/.cache/R/ExperimentHub/17079abac61_7752" </span></span>
<span><span class="co">##  focus: "II" </span></span>
<span><span class="co">##  resolutions(5): 1000 2000 4000 8000 16000</span></span>
<span><span class="co">##  active resolution: 2000 </span></span>
<span><span class="co">##  interactions: 34063 </span></span>
<span><span class="co">##  scores(3): count balanced ICE </span></span>
<span><span class="co">##  topologicalFeatures: compartments(0) borders(0) loops(0) viewpoints(0) centromeres(16) </span></span>
<span><span class="co">##  pairsFile: N/A </span></span>
<span><span class="co">##  metadata(0):</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>It is possible to plot the different <code>scores</code> of the resulting object to visualize the newly computed <code>scores</code>. In this example, <code>ICE</code> scores should be nearly identical to <code>balanced</code> scores, which were originally imported from the disk-stored contact matrix.</p>
<div class="column-body-outset-right">
<div class="cell" data-layout-align="center" data-fig.asp="0.33">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html">wrap_plots</a></span><span class="op">(</span></span>
<span>    <span class="fu">plotMatrix</span><span class="op">(</span><span class="va">normalized_hic</span>, use.scores <span class="op">=</span> <span class="st">'count'</span>, caption <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>,</span>
<span>    <span class="fu">plotMatrix</span><span class="op">(</span><span class="va">normalized_hic</span>, use.scores <span class="op">=</span> <span class="st">'balanced'</span>, caption <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>,</span>
<span>    <span class="fu">plotMatrix</span><span class="op">(</span><span class="va">normalized_hic</span>, use.scores <span class="op">=</span> <span class="st">'ICE'</span>, caption <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>, </span>
<span>    nrow <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="matrix-centric_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</div>
</section><section id="computing-observedexpected-oe-map" class="level3 page-columns page-full" data-number="5.1.2"><h3 data-number="5.1.2" class="anchored" data-anchor-id="computing-observedexpected-oe-map">
<span class="header-section-number">5.1.2</span> Computing observed/expected (O/E) map</h3>
<p>The most prominent feature of a balanced Hi-C matrix is the strong main diagonal. This main diagonal is observed because interactions between immediate adjacent genomic loci are more prone to happen than interactions spanning longer genomic distances. This “expected” behavior is due to the polymer nature of the chromosomes being studied, and can be locally estimated using the distance-dependent interaction frequency (a.k.a. the “distance law”, or P(s)). It can be used to compute an <code>expected</code> matrix on interactions.</p>
<p>When it is desirable to “mask” this polymer behavior to emphasize topological structures formed by chromosomes, one can divide a given balanced matrix by its <code>expected</code> matrix, i.e.&nbsp;calculate the observed/expected (O/E) map. This is sometimes called “detrending”, as it effectively removes the average polymer behavior from the balanced matrix.</p>
<p>The <code>detrend</code> function performs this operation on a given <code>HiCExperiment</code> object. It adds two extra elements in <code>scores</code> list: <code>expected</code> and <code>detrended</code> metrics (while the <code>interactions</code> themselves are unmodified).</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">detrended_hic</span> <span class="op">&lt;-</span> <span class="fu">detrend</span><span class="op">(</span><span class="va">hic</span><span class="op">)</span></span>
<span><span class="va">detrended_hic</span></span>
<span><span class="co">##  `HiCExperiment` object with 471,364 contacts over 407 regions </span></span>
<span><span class="co">##  -------</span></span>
<span><span class="co">##  fileName: "/root/.cache/R/ExperimentHub/17079abac61_7752" </span></span>
<span><span class="co">##  focus: "II" </span></span>
<span><span class="co">##  resolutions(5): 1000 2000 4000 8000 16000</span></span>
<span><span class="co">##  active resolution: 2000 </span></span>
<span><span class="co">##  interactions: 34063 </span></span>
<span><span class="co">##  scores(4): count balanced expected detrended </span></span>
<span><span class="co">##  topologicalFeatures: compartments(0) borders(0) loops(0) viewpoints(0) centromeres(16) </span></span>
<span><span class="co">##  pairsFile: N/A </span></span>
<span><span class="co">##  metadata(0):</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Topological features will be visually more prominent in the O/E <code>detrended</code> Hi-C map.</p>
<div class="column-page-right">
<div class="cell" data-layout-align="center" data-fig.asp="0.33">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html">wrap_plots</a></span><span class="op">(</span></span>
<span>    <span class="fu">plotMatrix</span><span class="op">(</span><span class="va">detrended_hic</span>, use.scores <span class="op">=</span> <span class="st">'balanced'</span>, scale <span class="op">=</span> <span class="st">'log10'</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">3.5</span>, <span class="op">-</span><span class="fl">1.2</span><span class="op">)</span>, caption <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>,</span>
<span>    <span class="fu">plotMatrix</span><span class="op">(</span><span class="va">detrended_hic</span>, use.scores <span class="op">=</span> <span class="st">'expected'</span>, scale <span class="op">=</span> <span class="st">'log10'</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">3.5</span>, <span class="op">-</span><span class="fl">1.2</span><span class="op">)</span>, caption <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>,</span>
<span>    <span class="fu">plotMatrix</span><span class="op">(</span><span class="va">detrended_hic</span>, use.scores <span class="op">=</span> <span class="st">'detrended'</span>, scale <span class="op">=</span> <span class="st">'linear'</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>, cmap <span class="op">=</span> <span class="fu">bwrColors</span><span class="op">(</span><span class="op">)</span>, caption <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>, </span>
<span>    nrow <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="matrix-centric_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Scale for <code>detrended</code> scores
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>
<code>expected</code> scores are in <code>linear</code> scale and ± in the same amplitude than <code>balanced</code> scores;</li>
<li>
<code>detrended</code> scores are in <code>log2</code> scale, in general approximately centered around 0. When plotting <code>detrended</code> scores, <code>scale = linear</code> should be set to prevent the default <code>log10</code> scaling.</li>
</ul>
</div>
</div>
</section><section id="computing-autocorrelated-map" class="level3" data-number="5.1.3"><h3 data-number="5.1.3" class="anchored" data-anchor-id="computing-autocorrelated-map">
<span class="header-section-number">5.1.3</span> Computing autocorrelated map</h3>
<p>Correlation matrices are often calculated from balanced Hi-C matrices. For instance, in genomes composed of eu- and heterochromatin, a correlation matrix can be used to reveal a checkerboard pattern emphasizing the segregation of chromatin into two A/B compartments (<span class="citation" data-cites="Lieberman_Aiden_2009">Lieberman-Aiden et al. (<a href="#ref-Lieberman_Aiden_2009" role="doc-biblioref">2009</a>)</span>).</p>
<p>The <code>autocorrelate</code> function is used to compute a correlation matrix of a <code>HiCExperiment</code> object. For each pair of interacting loci, the <code>autocorrelated</code> score represents the correlation between their respective interaction profiles with the rest of the genome.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">autocorr_hic</span> <span class="op">&lt;-</span> <span class="fu">autocorrelate</span><span class="op">(</span><span class="va">hic</span><span class="op">)</span></span>
<span><span class="co">##  </span></span>
<span><span class="va">autocorr_hic</span></span>
<span><span class="co">##  `HiCExperiment` object with 471,364 contacts over 407 regions </span></span>
<span><span class="co">##  -------</span></span>
<span><span class="co">##  fileName: "/root/.cache/R/ExperimentHub/17079abac61_7752" </span></span>
<span><span class="co">##  focus: "II" </span></span>
<span><span class="co">##  resolutions(5): 1000 2000 4000 8000 16000</span></span>
<span><span class="co">##  active resolution: 2000 </span></span>
<span><span class="co">##  interactions: 34063 </span></span>
<span><span class="co">##  scores(5): count balanced expected detrended autocorrelated </span></span>
<span><span class="co">##  topologicalFeatures: compartments(0) borders(0) loops(0) viewpoints(0) centromeres(16) </span></span>
<span><span class="co">##  pairsFile: N/A </span></span>
<span><span class="co">##  metadata(0):</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Since these metrics represent correlation scores, they range between <code>-1</code> and <code>1</code>. Two loci with an <code>autocorrelated</code> score close to <code>-1</code> have anti-correlated interaction profiles, while two loci with a <code>autocorrelated</code> score close to <code>1</code> are likely to interact with shared targets.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">scores</a></span><span class="op">(</span><span class="va">autocorr_hic</span>, <span class="st">'autocorrelated'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##     Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's </span></span>
<span><span class="co">##  -0.4156  0.0025  0.0504  0.0645  0.1036  1.0000     564</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Correlated and anti-correlated loci will be visually represented in the <code>autocorrelated</code> Hi-C map in red and blue pixels, respectively.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Here we have illustrated how to compute an autocorrelation matrix from a <code>HiCExperiment</code> object using the example <strong>yeast</strong> Hi-C experiment. Bear in mind that this is unusual and not very useful, as yeast chromatin is not segregated in two compartments but rather follows a Rabl conformation (<span class="citation" data-cites="Duan_2010">Duan et al. (<a href="#ref-Duan_2010" role="doc-biblioref">2010</a>)</span>). An example of autocorrelation map from a vertebrate Hi-C experiment (for which chromatin is segregated in A/B compartments) is shown in <a href="../pages/workflow-chicken.html">Chapter 10</a>.</p>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">plotMatrix</span><span class="op">(</span></span>
<span>    <span class="va">autocorr_hic</span>, </span>
<span>    use.scores <span class="op">=</span> <span class="st">'autocorrelated'</span>, </span>
<span>    scale <span class="op">=</span> <span class="st">'linear'</span>, </span>
<span>    limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.4</span>, <span class="fl">0.4</span><span class="op">)</span>, </span>
<span>    cmap <span class="op">=</span> <span class="fu">bgrColors</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="matrix-centric_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Scale for <code>autocorrelated</code> scores
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>
<code>autocorrelated</code> scores are in <code>linear</code> scale, in general approximately centered around 0. When plotting <code>autocorrelated</code> scores, <code>scale = linear</code> should be set to prevent the default <code>log10</code> scaling.</li>
<li>
<code>limits</code> should be manually set to <code>c(-x, x)</code> (<code>0 &lt; x &lt;= 1</code>) to ensure that the color range is effectively centered on <code>0</code>.</li>
</ul>
</div>
</div>
</section><section id="despeckling-smoothing-out-a-contact-map" class="level3 page-columns page-full" data-number="5.1.4"><h3 data-number="5.1.4" class="anchored" data-anchor-id="despeckling-smoothing-out-a-contact-map">
<span class="header-section-number">5.1.4</span> Despeckling (smoothing out) a contact map</h3>
<p>Shallow-sequenced Hi-C libraries or matrices binned with an overly small bin size sometimes produce “grainy” Hi-C maps with noisy backgrounds. A grainy map may also be obtained when dividing two matrices, e.g.&nbsp;when computing the O/E ratio with <code>detrend</code>. This is particularly true for sparser long-range interactions. To overcome such limitations, <code>HiCExperiment</code> objects can be “<code>despeckle</code>d” to smooth out focal speckles.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hic2</span> <span class="op">&lt;-</span> <span class="fu">detrend</span><span class="op">(</span><span class="va">hic</span><span class="op">[</span><span class="st">'II:400000-700000'</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">hic2</span> <span class="op">&lt;-</span> <span class="fu">despeckle</span><span class="op">(</span><span class="va">hic2</span>, use.scores <span class="op">=</span> <span class="st">'detrended'</span>, focal.size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">hic2</span></span>
<span><span class="co">##  `HiCExperiment` object with 168,785 contacts over 150 regions </span></span>
<span><span class="co">##  -------</span></span>
<span><span class="co">##  fileName: "/root/.cache/R/ExperimentHub/17079abac61_7752" </span></span>
<span><span class="co">##  focus: "II:400,000-700,000" </span></span>
<span><span class="co">##  resolutions(5): 1000 2000 4000 8000 16000</span></span>
<span><span class="co">##  active resolution: 2000 </span></span>
<span><span class="co">##  interactions: 11325 </span></span>
<span><span class="co">##  scores(5): count balanced expected detrended detrended.despeckled </span></span>
<span><span class="co">##  topologicalFeatures: compartments(0) borders(0) loops(0) viewpoints(0) centromeres(16) </span></span>
<span><span class="co">##  pairsFile: N/A </span></span>
<span><span class="co">##  metadata(0):</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The added <code>&lt;use.scores&gt;.despeckled</code> scores correspond to scores averaged using a window, whose width is provided with the <code>focal.size</code> argument. This results in a smoother Hi-C heatmap, effectively removing the “speckles” observed at longer range.</p>
<div class="column-page-right">
<div class="cell" data-layout-align="center" data-fig.asp="0.33">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">InteractionSet</span><span class="op">)</span></span>
<span><span class="va">loops</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'S288C-loops.bedpe'</span>, package <span class="op">=</span> <span class="st">'HiCExperiment'</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocIO/man/IO.html">import</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/pairs.html">makeGInteractionsFromGRangesPairs</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">borders</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'S288C-borders.bed'</span>, package <span class="op">=</span> <span class="st">'HiCExperiment'</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocIO/man/IO.html">import</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html">wrap_plots</a></span><span class="op">(</span></span>
<span>    <span class="fu">plotMatrix</span><span class="op">(</span><span class="va">hic2</span>, caption <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>,</span>
<span>    <span class="fu">plotMatrix</span><span class="op">(</span><span class="va">hic2</span>, use.scores <span class="op">=</span> <span class="st">'detrended'</span>, scale <span class="op">=</span> <span class="st">'linear'</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>, caption <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>,</span>
<span>    <span class="fu">plotMatrix</span><span class="op">(</span></span>
<span>        <span class="va">hic2</span>, </span>
<span>        use.scores <span class="op">=</span> <span class="st">'detrended.despeckled'</span>, </span>
<span>        scale <span class="op">=</span> <span class="st">'linear'</span>, </span>
<span>        limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>, </span>
<span>        caption <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>        loops <span class="op">=</span> <span class="va">loops</span>, </span>
<span>        borders <span class="op">=</span> <span class="va">borders</span></span>
<span>    <span class="op">)</span>,</span>
<span>    nrow <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="matrix-centric_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Scale for <code>despeckled</code> scores
</div>
</div>
<div class="callout-body-container callout-body">
<p><code>despeckled</code> scores are in the same scale than the <code>scores</code> they were computed from.</p>
</div>
</div>
</section></section><section id="operations-between-multiple-matrices" class="level2" data-number="5.2"><h2 data-number="5.2" class="anchored" data-anchor-id="operations-between-multiple-matrices">
<span class="header-section-number">5.2</span> Operations between multiple matrices</h2>
<section id="merging-maps" class="level3" data-number="5.2.1"><h3 data-number="5.2.1" class="anchored" data-anchor-id="merging-maps">
<span class="header-section-number">5.2.1</span> Merging maps</h3>
<p>Hi-C libraries are often sequenced in multiple rounds, for example when high genome coverage is required. This results in multiple contact matrix files being generated. The <code>merge</code> function can be used to bind several <code>HiCExperiment</code> objects into a single one.</p>
<p>The different <code>HiCExperiment</code> objects do not need to all have identical <code>regions</code>, as shown in the following example.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hic_sub1</span> <span class="op">&lt;-</span> <span class="fu">subsetByOverlaps</span><span class="op">(</span><span class="va">hic</span>, <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html">GRanges</a></span><span class="op">(</span><span class="st">"II:100001-200000"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">hic_sub2</span> <span class="op">&lt;-</span> <span class="fu">subsetByOverlaps</span><span class="op">(</span><span class="va">hic</span>, <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html">GRanges</a></span><span class="op">(</span><span class="st">"II:300001-400000"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">bound_hic</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html">merge</a></span><span class="op">(</span><span class="va">hic_sub1</span>, <span class="va">hic_sub2</span><span class="op">)</span></span>
<span><span class="fu">plotMatrix</span><span class="op">(</span><span class="va">bound_hic</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="matrix-centric_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="computing-ratio-between-two-maps" class="level3" data-number="5.2.2"><h3 data-number="5.2.2" class="anchored" data-anchor-id="computing-ratio-between-two-maps">
<span class="header-section-number">5.2.2</span> Computing ratio between two maps</h3>
<p>Comparing two Hi-C maps can be useful to infer which genomic loci are differentially interacting between experimental conditions. Comparing two <code>HiCExperiment</code> objects can be done in <code>R</code> using the <code>divide</code> function.</p>
<p>For example, we can divide the <em>eco1</em> mutant Hi-C data by wild-type Hi-C dataset using the <code>divide</code> function.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hic_eco1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocIO/man/IO.html">import</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/CoolFile-class.html">CoolFile</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/HiContactsData/man/HiContactsData.html">HiContactsData</a></span><span class="op">(</span><span class="st">'yeast_eco1'</span>, <span class="st">'mcool'</span><span class="op">)</span><span class="op">)</span>, </span>
<span>    focus <span class="op">=</span> <span class="st">'II'</span>, </span>
<span>    resolution <span class="op">=</span> <span class="fl">2000</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">div_contacts</span> <span class="op">&lt;-</span> <span class="fu">divide</span><span class="op">(</span><span class="va">hic_eco1</span>, by <span class="op">=</span> <span class="va">hic</span><span class="op">)</span> </span>
<span><span class="va">div_contacts</span></span>
<span><span class="co">##  `HiCExperiment` object with 996,154 contacts over 407 regions </span></span>
<span><span class="co">##  -------</span></span>
<span><span class="co">##  fileName: N/A </span></span>
<span><span class="co">##  focus: "II" </span></span>
<span><span class="co">##  resolutions(1): 2000</span></span>
<span><span class="co">##  active resolution: 2000 </span></span>
<span><span class="co">##  interactions: 60894 </span></span>
<span><span class="co">##  scores(6): count.x balanced.x count.by balanced.by balanced.fc balanced.l2fc </span></span>
<span><span class="co">##  topologicalFeatures: () </span></span>
<span><span class="co">##  pairsFile: N/A </span></span>
<span><span class="co">##  metadata(2): hce_list operation</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can visually compare wild-type and <em>eco1</em> maps side by side (left) and their ratio map (right). This highlights the depletion of short-range and increase of long-range interactions in the <em>eco1</em> dataset.</p>
<div class="cell" data-layout-align="center" data-fig.asp="0.5">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span></span>
<span>    <span class="fu">plotMatrix</span><span class="op">(</span><span class="va">hic_eco1</span>, compare.to <span class="op">=</span> <span class="va">hic</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">4</span>, <span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span>, </span>
<span>    <span class="fu">plotMatrix</span><span class="op">(</span></span>
<span>        <span class="va">div_contacts</span>, </span>
<span>        use.scores <span class="op">=</span> <span class="st">'balanced.fc'</span>, </span>
<span>        scale <span class="op">=</span> <span class="st">'log2'</span>, </span>
<span>        limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>        cmap <span class="op">=</span> <span class="fu">bwrColors</span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="matrix-centric_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid figure-img" style="width:100.0%"></p>
</figure>
</div>
</div>
</div>
</section></section><section id="references" class="level1 unnumbered"><h1 class="unnumbered">References</h1>


<div id="refs" class="references csl-bib-body hanging-indent" data-line-spacing="2" role="list">
<div id="ref-Cournac_2012" class="csl-entry" role="listitem">
Cournac, A., Marie-Nelly, H., Marbouty, M., Koszul, R., &amp; Mozziconacci, J. (2012). Normalization of a chromosomal contact map. <em><span>BMC</span> Genomics</em>, <em>13</em>(1). <a href="https://doi.org/10.1186/1471-2164-13-436">https://doi.org/10.1186/1471-2164-13-436</a>
</div>
<div id="ref-Duan_2010" class="csl-entry" role="listitem">
Duan, Z., Andronescu, M., Schutz, K., McIlwain, S., Kim, Y. J., Lee, C., Shendure, J., Fields, S., Blau, C. A., &amp; Noble, W. S. (2010). A three-dimensional model of the yeast genome. <em>Nature</em>, <em>465</em>(7296), 363–367. <a href="https://doi.org/10.1038/nature08973">https://doi.org/10.1038/nature08973</a>
</div>
<div id="ref-Imakaev_2012" class="csl-entry" role="listitem">
Imakaev, M., Fudenberg, G., McCord, R. P., Naumova, N., Goloborodko, A., Lajoie, B. R., Dekker, J., &amp; Mirny, L. A. (2012). Iterative correction of hi-c data reveals hallmarks of chromosome organization. <em>Nature Methods</em>, <em>9</em>(10), 999–1003. <a href="https://doi.org/10.1038/nmeth.2148">https://doi.org/10.1038/nmeth.2148</a>
</div>
<div id="ref-Lieberman_Aiden_2009" class="csl-entry" role="listitem">
Lieberman-Aiden, E., Berkum, N. L. van, Williams, L., Imakaev, M., Ragoczy, T., Telling, A., Amit, I., Lajoie, B. R., Sabo, P. J., Dorschner, M. O., Sandstrom, R., Bernstein, B., Bender, M. A., Groudine, M., Gnirke, A., Stamatoyannopoulos, J., Mirny, L. A., Lander, E. S., &amp; Dekker, J. (2009). Comprehensive mapping of long-range interactions reveals folding principles of the human genome. <em>Science</em>, <em>326</em>(5950), 289–293. <a href="https://doi.org/10.1126/science.1181369">https://doi.org/10.1126/science.1181369</a>
</div>
</div>
</section><a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../pages/visualization.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Hi-C data visualization</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../pages/interactions-centric.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Interactions-centric analysis</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left"><i>OHCA: Orchestrating Hi-C analysis with Bioconductor</i> was written by J. Serizay</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">This book was built with <a href="https://github.com/js2264/BiocBook/">BiocBook</a> with <span class="emoji" data-emoji="heart">❤️</span></div>
  </div>
</footer>


</body></html>