<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Orchestrating Hi-C analysis with Bioconductor - 3&nbsp; Manipulating Hi-C data in R</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../pages/visualization.html" rel="next">
<link href="../pages/data-representation.html" rel="prev">
<link href="../assets/cover.jpg" rel="icon" type="image/jpeg">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
</head>
<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/principles.html">Fundamentals concepts</a></li><li class="breadcrumb-item"><a href="../pages/parsing.html"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Manipulating Hi-C data in R</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../">Orchestrating Hi-C analysis with Bioconductor</a> 
        <div class="sidebar-tools-main">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label=""><i class="bi bi-git"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
<li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://github.com/js2264/OHCA/">
            Source Code
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://js2264.github.io/OHCA/devel/">
            Browse version `devel`
            </a>
          </li>
      </ul>
</div>
    <a href="../Orchestrating-Hi-C-analysis-with-Bioconductor.pdf" rel="" title="Download PDF" class="quarto-navigation-tool px-1" aria-label="Download PDF"><i class="bi bi-file-pdf"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/preamble.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preamble</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Fundamentals concepts</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/principles.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Hi-C pre-processing steps</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/data-representation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Hi-C data structures in R</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/parsing.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Manipulating Hi-C data in R</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/visualization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Hi-C data visualization</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">In-depth Hi-C analysis</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/matrix-centric.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Matrix-centric analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/interactions-centric.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Interactions-centric analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/topological-features.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Finding topological features in Hi-C</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Advanced Hi-C topics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/disseminating.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Data gateways: accessing public Hi-C data portals</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/interoperability.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Interoperability: using HiCExperiment with other R packages</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/workflow-yeast.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workflow 1: Distance-dependent interactions across yeast mutants</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/workflow-chicken.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workflow 2: Chromosome compartment cohesion upon mitosis entry</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/workflow-centros.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workflow 3: Inter-centromere interactions in yeast</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li>
<a href="#subsetting-a-contact-matrix" id="toc-subsetting-a-contact-matrix" class="nav-link active" data-scroll-target="#subsetting-a-contact-matrix"><span class="header-section-number">3.1</span> Subsetting a contact matrix</a>
  <ul class="collapse">
<li><a href="#subsetting-before-import-with-focus" id="toc-subsetting-before-import-with-focus" class="nav-link" data-scroll-target="#subsetting-before-import-with-focus"><span class="header-section-number">3.1.1</span> Subsetting <strong>before</strong> import: with <code>focus</code></a></li>
  <li><a href="#subsetting-after-import" id="toc-subsetting-after-import" class="nav-link" data-scroll-target="#subsetting-after-import"><span class="header-section-number">3.1.2</span> Subsetting <strong>after</strong> import</a></li>
  <li><a href="#zooming-on-a-hicexperiment" id="toc-zooming-on-a-hicexperiment" class="nav-link" data-scroll-target="#zooming-on-a-hicexperiment"><span class="header-section-number">3.1.3</span> Zooming on a <code>HiCExperiment</code></a></li>
  </ul>
</li>
  <li>
<a href="#updating-an-hicexperiment-object" id="toc-updating-an-hicexperiment-object" class="nav-link" data-scroll-target="#updating-an-hicexperiment-object"><span class="header-section-number">3.2</span> Updating an <code>HiCExperiment</code> object</a>
  <ul class="collapse">
<li><a href="#immutable-slots" id="toc-immutable-slots" class="nav-link" data-scroll-target="#immutable-slots"><span class="header-section-number">3.2.1</span> Immutable slots</a></li>
  <li><a href="#mutable-slots" id="toc-mutable-slots" class="nav-link" data-scroll-target="#mutable-slots"><span class="header-section-number">3.2.2</span> Mutable slots</a></li>
  </ul>
</li>
  <li><a href="#coercing-hicexperiment-objects" id="toc-coercing-hicexperiment-objects" class="nav-link" data-scroll-target="#coercing-hicexperiment-objects"><span class="header-section-number">3.3</span> Coercing <code>HiCExperiment</code> objects</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title">
<span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Manipulating Hi-C data in R</span>
</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Aims
</div>
</div>
<div class="callout-body-container callout-body">
<p>This chapter focuses on:</p>
<ul>
<li>Modifying information associated with an existing <code>HiCExperiment</code> object</li>
<li>Subsetting a <code>HiCExperiment</code> object</li>
<li>Coercing a <code>HiCExperiment</code> object in a base data structure</li>
</ul>
</div>
</div>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important reminder
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>An <code>HiCExperiment</code> object allows random access parsing of a disk-stored contact matrix.</li>
<li>An <code>HiCExperiment</code> object operates by wrapping together (1) a <code>ContactFile</code> (i.e.&nbsp;a connection to a disk-stored data file) and (2) a <code>GInteractions</code> generated by parsing the data file.</li>
</ul>
</div>
</div>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Recap on <code>HiCExperiment</code> objects 👇
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ul>
<li>Creating a connection to a disk-stored contact matrix:</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># coolf &lt;- "&lt;path-to-disk-stored-contact-matrix.cool&gt;"</span></span>
<span><span class="va">coolf</span> <span class="op">&lt;-</span> <span class="fu">HiContactsData</span><span class="op">(</span><span class="st">'yeast_wt'</span>, <span class="st">'mcool'</span><span class="op">)</span></span>
<span><span class="va">cf</span> <span class="op">&lt;-</span> <span class="fu">CoolFile</span><span class="op">(</span><span class="va">coolf</span><span class="op">)</span></span>
<span><span class="fu">availableResolutions</span><span class="op">(</span><span class="va">cf</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">availableChromosomes</span><span class="op">(</span><span class="va">cf</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Importing a contact matrix over a specific genomic location, at a given resolution:</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hic</span> <span class="op">&lt;-</span> <span class="fu">import</span><span class="op">(</span><span class="va">cf</span>, focus <span class="op">=</span> <span class="st">'II:10000-50000'</span>, resolution <span class="op">=</span> <span class="fl">4000</span><span class="op">)</span></span>
<span><span class="va">hic</span></span>
<span><span class="co">##  `HiCExperiment` object with 10,801 contacts over 11 regions </span></span>
<span><span class="co">##  -------</span></span>
<span><span class="co">##  fileName: "/root/.cache/R/ExperimentHub/f56b079dc5_7752" </span></span>
<span><span class="co">##  focus: "II:10,000-50,000" </span></span>
<span><span class="co">##  resolutions(5): 1000 2000 4000 8000 16000</span></span>
<span><span class="co">##  active resolution: 4000 </span></span>
<span><span class="co">##  interactions: 45 </span></span>
<span><span class="co">##  scores(2): count balanced </span></span>
<span><span class="co">##  topologicalFeatures: compartments(0) borders(0) loops(0) viewpoints(0) </span></span>
<span><span class="co">##  pairsFile: N/A </span></span>
<span><span class="co">##  metadata(0):</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Recovering genomic interactions stored in a <code>HiCExperiment</code>:</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">interactions</span><span class="op">(</span><span class="va">hic</span><span class="op">)</span></span>
<span><span class="co">##  GInteractions object with 45 interactions and 4 metadata columns:</span></span>
<span><span class="co">##         seqnames1     ranges1     seqnames2     ranges2 |   bin_id1   bin_id2</span></span>
<span><span class="co">##             &lt;Rle&gt;   &lt;IRanges&gt;         &lt;Rle&gt;   &lt;IRanges&gt; | &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##     [1]        II 12001-16000 ---        II 12001-16000 |        61        61</span></span>
<span><span class="co">##     [2]        II 12001-16000 ---        II 16001-20000 |        61        62</span></span>
<span><span class="co">##     [3]        II 12001-16000 ---        II 20001-24000 |        61        63</span></span>
<span><span class="co">##     [4]        II 12001-16000 ---        II 24001-28000 |        61        64</span></span>
<span><span class="co">##     [5]        II 12001-16000 ---        II 28001-32000 |        61        65</span></span>
<span><span class="co">##     ...       ...         ... ...       ...         ... .       ...       ...</span></span>
<span><span class="co">##    [41]        II 36001-40000 ---        II 40001-44000 |        67        68</span></span>
<span><span class="co">##    [42]        II 36001-40000 ---        II 44001-48000 |        67        69</span></span>
<span><span class="co">##    [43]        II 40001-44000 ---        II 40001-44000 |        68        68</span></span>
<span><span class="co">##    [44]        II 40001-44000 ---        II 44001-48000 |        68        69</span></span>
<span><span class="co">##    [45]        II 44001-48000 ---        II 44001-48000 |        69        69</span></span>
<span><span class="co">##             count  balanced</span></span>
<span><span class="co">##         &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##     [1]       213  0.249303</span></span>
<span><span class="co">##     [2]       673  0.449271</span></span>
<span><span class="co">##     [3]       325  0.210001</span></span>
<span><span class="co">##     [4]       137  0.125732</span></span>
<span><span class="co">##     [5]        77  0.106917</span></span>
<span><span class="co">##     ...       ...       ...</span></span>
<span><span class="co">##    [41]       941  0.358860</span></span>
<span><span class="co">##    [42]       275  0.114972</span></span>
<span><span class="co">##    [43]       675  0.253868</span></span>
<span><span class="co">##    [44]       497  0.204920</span></span>
<span><span class="co">##    [45]       295  0.133344</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    regions: 11 ranges and 4 metadata columns</span></span>
<span><span class="co">##    seqinfo: 16 sequences from an unspecified genome</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Generating the example <code>hic</code> object 👇
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>To demonstrate how to manipulate a <code>HiCExperiment</code> object, we will create an <code>HiCExperiment</code> object from an example <code>.cool</code> file provided in the <code>HiContactsData</code> package.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/js2264/HiCExperiment">HiCExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/js2264/HiContactsData">HiContactsData</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># ---- This downloads an example `.mcool` file and caches it locally </span></span>
<span><span class="va">coolf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/HiContactsData/man/HiContactsData.html">HiContactsData</a></span><span class="op">(</span><span class="st">'yeast_wt'</span>, <span class="st">'mcool'</span><span class="op">)</span></span>
<span><span class="co">##  see ?HiContactsData and browseVignettes('HiContactsData') for documentation</span></span>
<span><span class="co">##  loading from cache</span></span>
<span></span>
<span><span class="co"># ---- This creates a connection to the disk-stored `.mcool` file</span></span>
<span><span class="va">cf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/CoolFile-class.html">CoolFile</a></span><span class="op">(</span><span class="va">coolf</span><span class="op">)</span></span>
<span><span class="va">cf</span></span>
<span><span class="co">##  CoolFile object</span></span>
<span><span class="co">##  .mcool file: /root/.cache/R/ExperimentHub/f56b079dc5_7752 </span></span>
<span><span class="co">##  resolution: 1000 </span></span>
<span><span class="co">##  pairs file: </span></span>
<span><span class="co">##  metadata(0):</span></span>
<span></span>
<span><span class="co"># ---- This imports contacts from the long arm of chromosome `II`, at resolution `2000`</span></span>
<span><span class="va">hic</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocIO/man/IO.html">import</a></span><span class="op">(</span><span class="va">cf</span>, focus <span class="op">=</span> <span class="st">'II:300001-813184'</span>, resolution <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span>
<span><span class="va">hic</span></span>
<span><span class="co">##  `HiCExperiment` object with 306,212 contacts over 257 regions </span></span>
<span><span class="co">##  -------</span></span>
<span><span class="co">##  fileName: "/root/.cache/R/ExperimentHub/f56b079dc5_7752" </span></span>
<span><span class="co">##  focus: "II:300,001-813,184" </span></span>
<span><span class="co">##  resolutions(5): 1000 2000 4000 8000 16000</span></span>
<span><span class="co">##  active resolution: 2000 </span></span>
<span><span class="co">##  interactions: 18513 </span></span>
<span><span class="co">##  scores(2): count balanced </span></span>
<span><span class="co">##  topologicalFeatures: compartments(0) borders(0) loops(0) viewpoints(0) </span></span>
<span><span class="co">##  pairsFile: N/A </span></span>
<span><span class="co">##  metadata(0):</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<section id="subsetting-a-contact-matrix" class="level2" data-number="3.1"><h2 data-number="3.1" class="anchored" data-anchor-id="subsetting-a-contact-matrix">
<span class="header-section-number">3.1</span> Subsetting a contact matrix</h2>
<p>Two entirely different approaches are possible to subset of a Hi-C contact matrix:</p>
<ul>
<li><p><strong>Subsetting before importing</strong>: leveraging random access to a <strong>disk-stored</strong> contact matrix to <strong>only import interactions overlapping with a genomic locus of interest</strong>.</p></li>
<li><p><strong>Subsetting after importing</strong>: parsing the <strong>entire contact matrix</strong> in memory, and <strong>subsequently</strong> subset interactions overlapping with a genomic locus of interest.</p></li>
</ul>
<p><img src="images/20230525134200.png" class="img-fluid"></p>
<section id="subsetting-before-import-with-focus" class="level3" data-number="3.1.1"><h3 data-number="3.1.1" class="anchored" data-anchor-id="subsetting-before-import-with-focus">
<span class="header-section-number">3.1.1</span> Subsetting <strong>before</strong> import: with <code>focus</code>
</h3>
<p>Specifying a <code>focus</code> <strong>when</strong> importing a dataset in R (i.e.&nbsp;<code>"Subset first, then parse"</code>) is generally the recommended approach to import Hi-C data in R.</p>
<p>The <code>focus</code> argument can be set when <code>import</code>ing a <code>ContactFile</code> in R, as follows:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocIO/man/IO.html">import</a></span><span class="op">(</span><span class="va">cf</span>, focus <span class="op">=</span> <span class="st">"..."</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This ensures that only the needed data is parsed in R, reducing memory load and accelerating the import. Thus, this should be the preferred way of parsing <code>HiCExperiment</code> data, as disk-stored contact matrices allow <strong>efficient random access to indexed data</strong>.</p>
<p><code>focus</code> can be any of the following string types:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#   "II"                                  --&gt; import contacts over an entire chromosome</span></span>
<span><span class="co">#   "II:300001-800000"                    --&gt; import on-diagonal contacts within a chromosome</span></span>
<span><span class="co">#   "II:300001-400000|II:600001-700000"   --&gt; import off-diagonal contacts within a chromosome</span></span>
<span><span class="co">#   "II|III"                              --&gt; import contacts between two chromosomes</span></span>
<span><span class="co">#   "II:300001-800000|V:1-500000"         --&gt; import contacts between segments of two chromosomes</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-5-contents" aria-controls="callout-5" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
More examples for import with <code>focus</code> argument 👇
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-5" class="callout-5-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ul>
<li>Subsetting to a specific <strong>on-diagonal</strong> genomic location using standard UCSC coordinates query:</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocIO/man/IO.html">import</a></span><span class="op">(</span><span class="va">cf</span>, focus <span class="op">=</span> <span class="st">'II:300001-800000'</span>, resolution <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span>
<span><span class="co">##  `HiCExperiment` object with 301,018 contacts over 250 regions </span></span>
<span><span class="co">##  -------</span></span>
<span><span class="co">##  fileName: "/root/.cache/R/ExperimentHub/f56b079dc5_7752" </span></span>
<span><span class="co">##  focus: "II:300,001-800,000" </span></span>
<span><span class="co">##  resolutions(5): 1000 2000 4000 8000 16000</span></span>
<span><span class="co">##  active resolution: 2000 </span></span>
<span><span class="co">##  interactions: 17974 </span></span>
<span><span class="co">##  scores(2): count balanced </span></span>
<span><span class="co">##  topologicalFeatures: compartments(0) borders(0) loops(0) viewpoints(0) </span></span>
<span><span class="co">##  pairsFile: N/A </span></span>
<span><span class="co">##  metadata(0):</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Subsetting to a specific <strong>off-diagonal</strong> genomic location using pairs of coordinates query:</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocIO/man/IO.html">import</a></span><span class="op">(</span><span class="va">cf</span>, focus <span class="op">=</span> <span class="st">'II:300001-400000|II:600001-700000'</span>, resolution <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span>
<span><span class="co">##  `HiCExperiment` object with 402 contacts over 100 regions </span></span>
<span><span class="co">##  -------</span></span>
<span><span class="co">##  fileName: "/root/.cache/R/ExperimentHub/f56b079dc5_7752" </span></span>
<span><span class="co">##  focus: "II:300001-400000|II:600001-700000" </span></span>
<span><span class="co">##  resolutions(5): 1000 2000 4000 8000 16000</span></span>
<span><span class="co">##  active resolution: 2000 </span></span>
<span><span class="co">##  interactions: 357 </span></span>
<span><span class="co">##  scores(2): count balanced </span></span>
<span><span class="co">##  topologicalFeatures: compartments(0) borders(0) loops(0) viewpoints(0) </span></span>
<span><span class="co">##  pairsFile: N/A </span></span>
<span><span class="co">##  metadata(0):</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Subsetting interactions to retain those constrained within a single chromosome:</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocIO/man/IO.html">import</a></span><span class="op">(</span><span class="va">cf</span>, focus <span class="op">=</span> <span class="st">'II'</span>, resolution <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span>
<span><span class="co">##  `HiCExperiment` object with 471,364 contacts over 407 regions </span></span>
<span><span class="co">##  -------</span></span>
<span><span class="co">##  fileName: "/root/.cache/R/ExperimentHub/f56b079dc5_7752" </span></span>
<span><span class="co">##  focus: "II" </span></span>
<span><span class="co">##  resolutions(5): 1000 2000 4000 8000 16000</span></span>
<span><span class="co">##  active resolution: 2000 </span></span>
<span><span class="co">##  interactions: 34063 </span></span>
<span><span class="co">##  scores(2): count balanced </span></span>
<span><span class="co">##  topologicalFeatures: compartments(0) borders(0) loops(0) viewpoints(0) </span></span>
<span><span class="co">##  pairsFile: N/A </span></span>
<span><span class="co">##  metadata(0):</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Subsetting interactions to retain those between two chromosomes:</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocIO/man/IO.html">import</a></span><span class="op">(</span><span class="va">cf</span>, focus <span class="op">=</span> <span class="st">'II|III'</span>, resolution <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span>
<span><span class="co">##  `HiCExperiment` object with 9,092 contacts over 566 regions </span></span>
<span><span class="co">##  -------</span></span>
<span><span class="co">##  fileName: "/root/.cache/R/ExperimentHub/f56b079dc5_7752" </span></span>
<span><span class="co">##  focus: "II|III" </span></span>
<span><span class="co">##  resolutions(5): 1000 2000 4000 8000 16000</span></span>
<span><span class="co">##  active resolution: 2000 </span></span>
<span><span class="co">##  interactions: 7438 </span></span>
<span><span class="co">##  scores(2): count balanced </span></span>
<span><span class="co">##  topologicalFeatures: compartments(0) borders(0) loops(0) viewpoints(0) </span></span>
<span><span class="co">##  pairsFile: N/A </span></span>
<span><span class="co">##  metadata(0):</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Subsetting interactions to retain those between parts of two chromosomes:</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocIO/man/IO.html">import</a></span><span class="op">(</span><span class="va">cf</span>, focus <span class="op">=</span> <span class="st">'II:300001-800000|V:1-500000'</span>, resolution <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span>
<span><span class="co">##  `HiCExperiment` object with 7,147 contacts over 500 regions </span></span>
<span><span class="co">##  -------</span></span>
<span><span class="co">##  fileName: "/root/.cache/R/ExperimentHub/f56b079dc5_7752" </span></span>
<span><span class="co">##  focus: "II:300001-800000|V:1-500000" </span></span>
<span><span class="co">##  resolutions(5): 1000 2000 4000 8000 16000</span></span>
<span><span class="co">##  active resolution: 2000 </span></span>
<span><span class="co">##  interactions: 6523 </span></span>
<span><span class="co">##  scores(2): count balanced </span></span>
<span><span class="co">##  topologicalFeatures: compartments(0) borders(0) loops(0) viewpoints(0) </span></span>
<span><span class="co">##  pairsFile: N/A </span></span>
<span><span class="co">##  metadata(0):</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<!-- 
### `refocus(<HiCExperiment>, focus = "...")`

Bear in mind that every call to `import` creates a `HiCExperiment` from 
scratch. However, in some cases, one may want to **change** the `focus` of a 
pre-existing `HiCExperiment` object, rather than creating a new one. This 
allows to preserve `metadata`, `topologicalFeatures` and `pairsFile` slots of 
the existing `HiCExperiment` object. This can be achieved with the `refocus()` 
function. `refocus` takes a `HiCExperiment` object rather than a `ContactFile`, 
as well as a `focus`. 


::: {.cell layout-align="center"}

```{.r .cell-code}
# --- Note existing `metadata`, `topologicalFeatures` and `pairsFile` fields 
hic 
##  `HiCExperiment` object with 306,212 contacts over 257 regions 
##  -------
##  fileName: "/root/.cache/R/ExperimentHub/f56b079dc5_7752" 
##  focus: "II:300,001-813,184" 
##  resolutions(5): 1000 2000 4000 8000 16000
##  active resolution: 2000 
##  interactions: 18513 
##  scores(2): count balanced 
##  topologicalFeatures: compartments(0) borders(0) loops(0) viewpoints(0) 
##  pairsFile: N/A 
##  metadata(0):

# --- Note how these fields are erased upon parsing of the original 
#     `cf` contact matrix with a new focus
import(cf, focus = 'III', resolution = 2000)
##  `HiCExperiment` object with 151,990 contacts over 159 regions 
##  -------
##  fileName: "/root/.cache/R/ExperimentHub/f56b079dc5_7752" 
##  focus: "III" 
##  resolutions(5): 1000 2000 4000 8000 16000
##  active resolution: 2000 
##  interactions: 8090 
##  scores(2): count balanced 
##  topologicalFeatures: compartments(0) borders(0) loops(0) viewpoints(0) 
##  pairsFile: N/A 
##  metadata(0):

# --- However, these fields can be preserved using the `refocus` 
#     function on the original `HiCExperiment` object
refocus(hic, 'III')
##  `HiCExperiment` object with 151,990 contacts over 159 regions 
##  -------
##  fileName: "/root/.cache/R/ExperimentHub/f56b079dc5_7752" 
##  focus: "III" 
##  resolutions(5): 1000 2000 4000 8000 16000
##  active resolution: 2000 
##  interactions: 8090 
##  scores(2): count balanced 
##  topologicalFeatures: compartments(0) borders(0) loops(0) viewpoints(0) 
##  pairsFile: N/A 
##  metadata(0):
```
:::

-->
</section><section id="subsetting-after-import" class="level3" data-number="3.1.2"><h3 data-number="3.1.2" class="anchored" data-anchor-id="subsetting-after-import">
<span class="header-section-number">3.1.2</span> Subsetting <strong>after</strong> import</h3>
<p>It may sometimes be desirable to import a full dataset from disk first, and <strong>only then</strong> perform in-memory subsetting of the <code>HiCExperiment</code> object (i.e.&nbsp;<code>"Parse first, then subset"</code>). <strong>This is for example necessary when the end user aims to investigate subsets of interactions across a large number of different areas of a contact matrix.</strong></p>
<p>Several strategies are possible to allow subsetting of imported data, either with <code>subsetByOverlaps</code> or <code>[</code>.</p>
<section id="subsetbyoverlapshicexperiment-granges" class="level4" data-number="3.1.2.1"><h4 data-number="3.1.2.1" class="anchored" data-anchor-id="subsetbyoverlapshicexperiment-granges">
<span class="header-section-number">3.1.2.1</span> <code>subsetByOverlaps(&lt;HiCExperiment&gt;, &lt;GRanges&gt;)</code>
</h4>
<p><code>subsetByOverlaps</code> can take a <code>HiCExperiment</code> as a <code>query</code> and a <code>GRanges</code> as a query. In this case, the <code>GRanges</code> is used to extract a subset of a <code>HiCExperiment</code> <strong>constrained</strong> within a specific genomic location.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">telomere</span> <span class="op">&lt;-</span> <span class="fu">GRanges</span><span class="op">(</span><span class="st">"II:700001-813184"</span><span class="op">)</span></span>
<span><span class="fu">subsetByOverlaps</span><span class="op">(</span><span class="va">hic</span>, <span class="va">telomere</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">interactions</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">##  GInteractions object with 1540 interactions and 4 metadata columns:</span></span>
<span><span class="co">##           seqnames1       ranges1     seqnames2       ranges2 |   bin_id1</span></span>
<span><span class="co">##               &lt;Rle&gt;     &lt;IRanges&gt;         &lt;Rle&gt;     &lt;IRanges&gt; | &lt;numeric&gt;</span></span>
<span><span class="co">##       [1]        II 700001-702000 ---        II 700001-702000 |       466</span></span>
<span><span class="co">##       [2]        II 700001-702000 ---        II 702001-704000 |       466</span></span>
<span><span class="co">##       [3]        II 700001-702000 ---        II 704001-706000 |       466</span></span>
<span><span class="co">##       [4]        II 700001-702000 ---        II 706001-708000 |       466</span></span>
<span><span class="co">##       [5]        II 700001-702000 ---        II 708001-710000 |       466</span></span>
<span><span class="co">##       ...       ...           ... ...       ...           ... .       ...</span></span>
<span><span class="co">##    [1536]        II 804001-806000 ---        II 810001-812000 |       518</span></span>
<span><span class="co">##    [1537]        II 806001-808000 ---        II 806001-808000 |       519</span></span>
<span><span class="co">##    [1538]        II 806001-808000 ---        II 808001-810000 |       519</span></span>
<span><span class="co">##    [1539]        II 806001-808000 ---        II 810001-812000 |       519</span></span>
<span><span class="co">##    [1540]        II 808001-810000 ---        II 808001-810000 |       520</span></span>
<span><span class="co">##             bin_id2     count  balanced</span></span>
<span><span class="co">##           &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##       [1]       466        30 0.0283618</span></span>
<span><span class="co">##       [2]       467       145 0.0709380</span></span>
<span><span class="co">##       [3]       468       124 0.0704979</span></span>
<span><span class="co">##       [4]       469        59 0.0510221</span></span>
<span><span class="co">##       [5]       470        59 0.0384004</span></span>
<span><span class="co">##       ...       ...       ...       ...</span></span>
<span><span class="co">##    [1536]       521         1       NaN</span></span>
<span><span class="co">##    [1537]       519        15 0.0560633</span></span>
<span><span class="co">##    [1538]       520        25       NaN</span></span>
<span><span class="co">##    [1539]       521         1       NaN</span></span>
<span><span class="co">##    [1540]       520        10       NaN</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    regions: 57 ranges and 4 metadata columns</span></span>
<span><span class="co">##    seqinfo: 16 sequences from an unspecified genome</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>By default, <code>subsetByOverlaps(hic, telomere)</code> will only recover interactions <strong>constrained</strong> within <code>telomere</code>, i.e.&nbsp;interactions for which both ends are in <code>telomere</code>.</p>
<p>Alternatively, <code>type = "any"</code> can be specified to get all interactions with at least one of their anchors within <code>telomere</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">subsetByOverlaps</span><span class="op">(</span><span class="va">hic</span>, <span class="va">telomere</span>, type <span class="op">=</span> <span class="st">"any"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">interactions</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">##  GInteractions object with 6041 interactions and 4 metadata columns:</span></span>
<span><span class="co">##           seqnames1       ranges1     seqnames2       ranges2 |   bin_id1</span></span>
<span><span class="co">##               &lt;Rle&gt;     &lt;IRanges&gt;         &lt;Rle&gt;     &lt;IRanges&gt; | &lt;numeric&gt;</span></span>
<span><span class="co">##       [1]        II 300001-302000 ---        II 702001-704000 |       266</span></span>
<span><span class="co">##       [2]        II 300001-302000 ---        II 704001-706000 |       266</span></span>
<span><span class="co">##       [3]        II 300001-302000 ---        II 768001-770000 |       266</span></span>
<span><span class="co">##       [4]        II 300001-302000 ---        II 784001-786000 |       266</span></span>
<span><span class="co">##       [5]        II 302001-304000 ---        II 740001-742000 |       267</span></span>
<span><span class="co">##       ...       ...           ... ...       ...           ... .       ...</span></span>
<span><span class="co">##    [6037]        II 804001-806000 ---        II 810001-812000 |       518</span></span>
<span><span class="co">##    [6038]        II 806001-808000 ---        II 806001-808000 |       519</span></span>
<span><span class="co">##    [6039]        II 806001-808000 ---        II 808001-810000 |       519</span></span>
<span><span class="co">##    [6040]        II 806001-808000 ---        II 810001-812000 |       519</span></span>
<span><span class="co">##    [6041]        II 808001-810000 ---        II 808001-810000 |       520</span></span>
<span><span class="co">##             bin_id2     count    balanced</span></span>
<span><span class="co">##           &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">##       [1]       467         1 0.000590999</span></span>
<span><span class="co">##       [2]       468         1 0.000686799</span></span>
<span><span class="co">##       [3]       500         1 0.000728215</span></span>
<span><span class="co">##       [4]       508         1 0.000923092</span></span>
<span><span class="co">##       [5]       486         1 0.000382222</span></span>
<span><span class="co">##       ...       ...       ...         ...</span></span>
<span><span class="co">##    [6037]       521         1         NaN</span></span>
<span><span class="co">##    [6038]       519        15   0.0560633</span></span>
<span><span class="co">##    [6039]       520        25         NaN</span></span>
<span><span class="co">##    [6040]       521         1         NaN</span></span>
<span><span class="co">##    [6041]       520        10         NaN</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    regions: 257 ranges and 4 metadata columns</span></span>
<span><span class="co">##    seqinfo: 16 sequences from an unspecified genome</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="hicexperiment..." class="level4" data-number="3.1.2.2"><h4 data-number="3.1.2.2" class="anchored" data-anchor-id="hicexperiment...">
<span class="header-section-number">3.1.2.2</span> <code>&lt;HiCExperiment&gt;["..."]</code>
</h4>
<p>The square bracket operator <code>[</code> allows for more advanced textual queries, similarly to <code>focus</code> arguments that can be used when importing contact matrices in memory.</p>
<p><strong>This ensures that only the needed data is parsed in R, reducing memory load and accelerating the import.</strong> Thus, this should be the preferred way of parsing <code>HiCExperiment</code> data, as disk-stored contact matrices allow efficient random access to indexed data.</p>
<p>The following string types can be used to subset a <code>HiCExperiment</code> object with the <code>[</code> notation:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#   "II"                                  --&gt; import contacts over an entire chromosome</span></span>
<span><span class="co">#   "II:300001-800000"                    --&gt; import on-diagonal contacts within a chromosome</span></span>
<span><span class="co">#   "II:300001-400000|II:600001-700000"   --&gt; import off-diagonal contacts within a chromosome</span></span>
<span><span class="co">#   "II|III"                              --&gt; import contacts between two chromosomes</span></span>
<span><span class="co">#   "II:300001-800000|V:1-500000"         --&gt; import contacts between segments of two chromosomes</span></span>
<span><span class="co">#   c("II", "III", "IV")                  --&gt; import contacts within and between several chromosomes</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-6-contents" aria-controls="callout-6" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
More examples for subsetting with <code>[</code> 👇
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-6" class="callout-6-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ul>
<li>Subsetting to a specific <strong>on-diagonal</strong> genomic location using standard UCSC coordinates query:</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hic</span><span class="op">[</span><span class="st">"II:800001-813184"</span><span class="op">]</span></span>
<span><span class="co">##  `HiCExperiment` object with 1,040 contacts over 6 regions </span></span>
<span><span class="co">##  -------</span></span>
<span><span class="co">##  fileName: "/root/.cache/R/ExperimentHub/f56b079dc5_7752" </span></span>
<span><span class="co">##  focus: "II:800,001-813,184" </span></span>
<span><span class="co">##  resolutions(5): 1000 2000 4000 8000 16000</span></span>
<span><span class="co">##  active resolution: 2000 </span></span>
<span><span class="co">##  interactions: 19 </span></span>
<span><span class="co">##  scores(2): count balanced </span></span>
<span><span class="co">##  topologicalFeatures: compartments(0) borders(0) loops(0) viewpoints(0) </span></span>
<span><span class="co">##  pairsFile: N/A </span></span>
<span><span class="co">##  metadata(0):</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Subsetting to a specific <strong>off-diagonal</strong> genomic location using pairs of coordinates query:</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hic</span><span class="op">[</span><span class="st">"II:300001-320000|II:800001-813184"</span><span class="op">]</span></span>
<span><span class="co">##  `HiCExperiment` object with 3 contacts over 6 regions </span></span>
<span><span class="co">##  -------</span></span>
<span><span class="co">##  fileName: "/root/.cache/R/ExperimentHub/f56b079dc5_7752" </span></span>
<span><span class="co">##  focus: "II:300001-320000|II:800001-813184" </span></span>
<span><span class="co">##  resolutions(5): 1000 2000 4000 8000 16000</span></span>
<span><span class="co">##  active resolution: 2000 </span></span>
<span><span class="co">##  interactions: 3 </span></span>
<span><span class="co">##  scores(2): count balanced </span></span>
<span><span class="co">##  topologicalFeatures: compartments(0) borders(0) loops(0) viewpoints(0) </span></span>
<span><span class="co">##  pairsFile: N/A </span></span>
<span><span class="co">##  metadata(0):</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Subsetting interactions to retain those constrained within a single chromosome:</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hic</span><span class="op">[</span><span class="st">"II"</span><span class="op">]</span></span>
<span><span class="co">##  `HiCExperiment` object with 306,212 contacts over 257 regions </span></span>
<span><span class="co">##  -------</span></span>
<span><span class="co">##  fileName: "/root/.cache/R/ExperimentHub/f56b079dc5_7752" </span></span>
<span><span class="co">##  focus: "II" </span></span>
<span><span class="co">##  resolutions(5): 1000 2000 4000 8000 16000</span></span>
<span><span class="co">##  active resolution: 2000 </span></span>
<span><span class="co">##  interactions: 18513 </span></span>
<span><span class="co">##  scores(2): count balanced </span></span>
<span><span class="co">##  topologicalFeatures: compartments(0) borders(0) loops(0) viewpoints(0) </span></span>
<span><span class="co">##  pairsFile: N/A </span></span>
<span><span class="co">##  metadata(0):</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Subsetting interactions to retain those between two chromosomes:</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hic</span><span class="op">[</span><span class="st">"II|IV"</span><span class="op">]</span></span>
<span><span class="co">##  `HiCExperiment` object with 0 contacts over 0 regions </span></span>
<span><span class="co">##  -------</span></span>
<span><span class="co">##  fileName: "/root/.cache/R/ExperimentHub/f56b079dc5_7752" </span></span>
<span><span class="co">##  focus: "II:1-813184|IV:1-1531933" </span></span>
<span><span class="co">##  resolutions(5): 1000 2000 4000 8000 16000</span></span>
<span><span class="co">##  active resolution: 2000 </span></span>
<span><span class="co">##  interactions: 0 </span></span>
<span><span class="co">##  scores(2): count balanced </span></span>
<span><span class="co">##  topologicalFeatures: compartments(0) borders(0) loops(0) viewpoints(0) </span></span>
<span><span class="co">##  pairsFile: N/A </span></span>
<span><span class="co">##  metadata(0):</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Subsetting interactions to retain those between segments of two chromosomes:</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hic</span><span class="op">[</span><span class="st">"II:300001-320000|IV:1-100000"</span><span class="op">]</span></span>
<span><span class="co">##  `HiCExperiment` object with 0 contacts over 0 regions </span></span>
<span><span class="co">##  -------</span></span>
<span><span class="co">##  fileName: "/root/.cache/R/ExperimentHub/f56b079dc5_7752" </span></span>
<span><span class="co">##  focus: "II:300001-320000|IV:1-100000" </span></span>
<span><span class="co">##  resolutions(5): 1000 2000 4000 8000 16000</span></span>
<span><span class="co">##  active resolution: 2000 </span></span>
<span><span class="co">##  interactions: 0 </span></span>
<span><span class="co">##  scores(2): count balanced </span></span>
<span><span class="co">##  topologicalFeatures: compartments(0) borders(0) loops(0) viewpoints(0) </span></span>
<span><span class="co">##  pairsFile: N/A </span></span>
<span><span class="co">##  metadata(0):</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Subsetting interactions to retain those constrained within several chromosomes:</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hic</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'II'</span>, <span class="st">'III'</span>, <span class="st">'IV'</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">##  `HiCExperiment` object with 306,212 contacts over 257 regions </span></span>
<span><span class="co">##  -------</span></span>
<span><span class="co">##  fileName: "/root/.cache/R/ExperimentHub/f56b079dc5_7752" </span></span>
<span><span class="co">##  focus: "II, III, IV" </span></span>
<span><span class="co">##  resolutions(5): 1000 2000 4000 8000 16000</span></span>
<span><span class="co">##  active resolution: 2000 </span></span>
<span><span class="co">##  interactions: 18513 </span></span>
<span><span class="co">##  scores(2): count balanced </span></span>
<span><span class="co">##  topologicalFeatures: compartments(0) borders(0) loops(0) viewpoints(0) </span></span>
<span><span class="co">##  pairsFile: N/A </span></span>
<span><span class="co">##  metadata(0):</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Some notes:</p>
<ul>
<li>This last example (subsetting for a vector of several chromosomes) is the only scenario for which <code>[</code>-based in-memory subsetting of pre-imported data is the only way to go, as such subsetting is not possible with <code>focus</code> from disk-stored data.</li>
<li>All the other <code>[</code> subsetting scenarii illustrated above can be achieved more efficiently using the <code>focus</code> argument when <code>import</code>ing data into a <code>HiCExperiment</code> object.</li>
<li>However, keep in mind that subsetting preserves extra data, e.g.&nbsp;added <code>scores</code>, <code>topologicalFeatures</code>, <code>metadata</code> or <code>pairsFile</code>, whereas this information is lost using <code>focus</code> with <code>import</code>.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="zooming-on-a-hicexperiment" class="level3" data-number="3.1.3"><h3 data-number="3.1.3" class="anchored" data-anchor-id="zooming-on-a-hicexperiment">
<span class="header-section-number">3.1.3</span> Zooming on a <code>HiCExperiment</code>
</h3>
<p>“Zooming” refers to dynamically changing the <strong>resolution</strong> of a <code>HiCExperiment</code>. By <code>zoom</code>ing a <code>HiCExperiment</code>, one can refine or coarsen the contact matrix. This operation takes a<code>ContactFile</code> and <code>focus</code> from an existing <code>HiCExperiment</code> input and re-generates a new <code>HiCExperiment</code> with updated <code>resolution</code>, <code>interactions</code> and <code>scores</code>. Note that <code>zoom</code> will preserve existing <code>metadata</code>, <code>topologicalFeatures</code> and <code>pairsFile</code> information.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hic</span></span>
<span><span class="co">##  `HiCExperiment` object with 306,212 contacts over 257 regions </span></span>
<span><span class="co">##  -------</span></span>
<span><span class="co">##  fileName: "/root/.cache/R/ExperimentHub/f56b079dc5_7752" </span></span>
<span><span class="co">##  focus: "II:300,001-813,184" </span></span>
<span><span class="co">##  resolutions(5): 1000 2000 4000 8000 16000</span></span>
<span><span class="co">##  active resolution: 2000 </span></span>
<span><span class="co">##  interactions: 18513 </span></span>
<span><span class="co">##  scores(2): count balanced </span></span>
<span><span class="co">##  topologicalFeatures: compartments(0) borders(0) loops(0) viewpoints(0) </span></span>
<span><span class="co">##  pairsFile: N/A </span></span>
<span><span class="co">##  metadata(0):</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">zoom</a></span><span class="op">(</span><span class="va">hic</span>, <span class="fl">4000</span><span class="op">)</span></span>
<span><span class="co">##  `HiCExperiment` object with 306,212 contacts over 129 regions </span></span>
<span><span class="co">##  -------</span></span>
<span><span class="co">##  fileName: "/root/.cache/R/ExperimentHub/f56b079dc5_7752" </span></span>
<span><span class="co">##  focus: "II:300,001-813,184" </span></span>
<span><span class="co">##  resolutions(5): 1000 2000 4000 8000 16000</span></span>
<span><span class="co">##  active resolution: 4000 </span></span>
<span><span class="co">##  interactions: 6800 </span></span>
<span><span class="co">##  scores(2): count balanced </span></span>
<span><span class="co">##  topologicalFeatures: compartments(0) borders(0) loops(0) viewpoints(0) </span></span>
<span><span class="co">##  pairsFile: N/A </span></span>
<span><span class="co">##  metadata(0):</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">zoom</a></span><span class="op">(</span><span class="va">hic</span>, <span class="fl">1000</span><span class="op">)</span></span>
<span><span class="co">##  `HiCExperiment` object with 306,212 contacts over 514 regions </span></span>
<span><span class="co">##  -------</span></span>
<span><span class="co">##  fileName: "/root/.cache/R/ExperimentHub/f56b079dc5_7752" </span></span>
<span><span class="co">##  focus: "II:300,001-813,184" </span></span>
<span><span class="co">##  resolutions(5): 1000 2000 4000 8000 16000</span></span>
<span><span class="co">##  active resolution: 1000 </span></span>
<span><span class="co">##  interactions: 44363 </span></span>
<span><span class="co">##  scores(2): count balanced </span></span>
<span><span class="co">##  topologicalFeatures: compartments(0) borders(0) loops(0) viewpoints(0) </span></span>
<span><span class="co">##  pairsFile: N/A </span></span>
<span><span class="co">##  metadata(0):</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>The sum of raw counts do not change after <code>zoom</code>ing, however the number of individual <code>interactions</code> and <code>regions</code> changes.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">hic</span><span class="op">)</span></span>
<span><span class="co">##  [1] 18513</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">zoom</a></span><span class="op">(</span><span class="va">hic</span>, <span class="fl">1000</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  [1] 44363</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">zoom</a></span><span class="op">(</span><span class="va">hic</span>, <span class="fl">4000</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  [1] 6800</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">scores</a></span><span class="op">(</span><span class="va">hic</span>, <span class="st">"count"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  [1] 306212</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">scores</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">zoom</a></span><span class="op">(</span><span class="va">hic</span>, <span class="fl">1000</span><span class="op">)</span>, <span class="st">"count"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  [1] 306212</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">scores</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">zoom</a></span><span class="op">(</span><span class="va">hic</span>, <span class="fl">4000</span><span class="op">)</span>, <span class="st">"count"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  [1] 306212</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>
<code>zoom</code> does not change the <code>focus</code>! It only affects the <code>resolution</code> (and consequently, the <code>interactions</code>).</li>
<li>
<code>zoom</code> will only work for multi-resolution contact matrices, e.g.&nbsp;<code>.mcool</code> or <code>.hic</code>.</li>
</ul>
</div>
</div>
</section></section><section id="updating-an-hicexperiment-object" class="level2" data-number="3.2"><h2 data-number="3.2" class="anchored" data-anchor-id="updating-an-hicexperiment-object">
<span class="header-section-number">3.2</span> Updating an <code>HiCExperiment</code> object</h2>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
TL;DR: Which <code>HiCExperiment</code> slots are mutable (✅) / immutable (⛔️)?
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>
<code>fileName(hic)</code>: ⛔️ (obtained from disk-stored file)</li>
<li>
<code>focus(hic)</code>: 🤔 (see <a href="#subsetting-methods">subsetting section</a>)</li>
<li>
<code>resolutions(hic)</code>: ⛔️ (obtained from disk-stored file)</li>
<li>
<code>resolution(hic)</code>: 🤔 (see <a href="#zooming-on-a-hicexperiment">zooming section</a>)</li>
<li>
<code>interactions(hic)</code>: ⛔️ (obtained from disk-stored file)</li>
<li>
<code>scores(hic)</code>: ✅</li>
<li>
<code>topologicalFeatures(hic)</code>: ✅</li>
<li>
<code>pairsFile(hic)</code>: ✅</li>
<li>
<code>metadata(hic)</code>: ✅</li>
</ul>
</div>
</div>
<section id="immutable-slots" class="level3" data-number="3.2.1"><h3 data-number="3.2.1" class="anchored" data-anchor-id="immutable-slots">
<span class="header-section-number">3.2.1</span> Immutable slots</h3>
<p>An <code>HiCExperiment</code> object acts as an interface exposing disk-stored data. This implies that the <code>fileName</code> slot itself is immutable (i.e.&nbsp;<strong>cannot</strong> be changed). This should be obvious, as a <code>HiCExperiment</code> <em>has to</em> be associated with a disk-stored contact matrix to properly function (except in some advanced cases developed in next chapters).</p>
<p>For this reason, methods to manually modify <code>interactions</code> and <code>resolutions</code> slots are also <strong>not</strong> exposed in the <code>HiCExperiment</code> package.</p>
<p>A corollary of this is that the associated <code>regions</code> and <code>anchors</code> of an <code>HiCExperiment</code> should <strong>not</strong> be modified by hand either, since they are directly linked to <code>interactions</code>.</p>
</section><section id="mutable-slots" class="level3" data-number="3.2.2"><h3 data-number="3.2.2" class="anchored" data-anchor-id="mutable-slots">
<span class="header-section-number">3.2.2</span> Mutable slots</h3>
<p>That being said, <code>HiCExperiment</code> objects are flexible and can be partially modified in memory without having to change/overwrite the original, disk-stored contact matrix.</p>
<p>Several <code>slots</code> can be modified in memory: <code>slots</code>, <code>topologicalFeatures</code>, <code>pairsFile</code> and <code>metadata</code>.</p>
<section id="scores" class="level4" data-number="3.2.2.1"><h4 data-number="3.2.2.1" class="anchored" data-anchor-id="scores">
<span class="header-section-number">3.2.2.1</span> <code>scores</code>
</h4>
<p>We have seen in the previous chapter that scores are stored in a <code>list</code> and are available using the <code>scores</code> function.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">scores</a></span><span class="op">(</span><span class="va">hic</span><span class="op">)</span></span>
<span><span class="co">##  List of length 2</span></span>
<span><span class="co">##  names(2): count balanced</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">scores</a></span><span class="op">(</span><span class="va">hic</span>, <span class="st">"count"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  [1]  7 92 75 61 38 43</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">scores</a></span><span class="op">(</span><span class="va">hic</span>, <span class="st">"balanced"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  [1] 0.009657438 0.076622340 0.054101992 0.042940512 0.040905212 0.029293930</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Extra scores can be added to this list, e.g.&nbsp;to describe the “expected” interaction frequency for each interaction stored in the <code>HiCExperiment</code> object). This can be achieved using the <code>scores()&lt;-</code> function.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">scores</a></span><span class="op">(</span><span class="va">hic</span>, <span class="st">"random"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">hic</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">scores</a></span><span class="op">(</span><span class="va">hic</span><span class="op">)</span></span>
<span><span class="co">##  List of length 3</span></span>
<span><span class="co">##  names(3): count balanced random</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">scores</a></span><span class="op">(</span><span class="va">hic</span>, <span class="st">"random"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  [1] 0.4614872 0.6568060 0.7589135 0.3992278 0.3551184 0.3534359</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="topologicalfeatures" class="level4" data-number="3.2.2.2"><h4 data-number="3.2.2.2" class="anchored" data-anchor-id="topologicalfeatures">
<span class="header-section-number">3.2.2.2</span> <code>topologicalFeatures</code>
</h4>
<p>The end-user can create additional <code>topologicalFeatures</code> or modify the existing ones using the <code>topologicalFeatures()&lt;-</code> function.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">topologicalFeatures</a></span><span class="op">(</span><span class="va">hic</span>, <span class="st">'CTCF'</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">GRanges</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"II:340-352"</span>, </span>
<span>    <span class="st">"II:3520-3532"</span>, </span>
<span>    <span class="st">"II:7980-7992"</span>, </span>
<span>    <span class="st">"II:9240-9252"</span> </span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">topologicalFeatures</a></span><span class="op">(</span><span class="va">hic</span>, <span class="st">'CTCF'</span><span class="op">)</span></span>
<span><span class="co">##  GRanges object with 4 ranges and 0 metadata columns:</span></span>
<span><span class="co">##        seqnames    ranges strand</span></span>
<span><span class="co">##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##    [1]       II   340-352      *</span></span>
<span><span class="co">##    [2]       II 3520-3532      *</span></span>
<span><span class="co">##    [3]       II 7980-7992      *</span></span>
<span><span class="co">##    [4]       II 9240-9252      *</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">topologicalFeatures</a></span><span class="op">(</span><span class="va">hic</span>, <span class="st">'loops'</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">GInteractions</span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">topologicalFeatures</a></span><span class="op">(</span><span class="va">hic</span>, <span class="st">'CTCF'</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, each <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">]</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">topologicalFeatures</a></span><span class="op">(</span><span class="va">hic</span>, <span class="st">'CTCF'</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fl">3</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">topologicalFeatures</a></span><span class="op">(</span><span class="va">hic</span>, <span class="st">'loops'</span><span class="op">)</span></span>
<span><span class="co">##  GInteractions object with 9 interactions and 0 metadata columns:</span></span>
<span><span class="co">##        seqnames1   ranges1     seqnames2   ranges2</span></span>
<span><span class="co">##            &lt;Rle&gt; &lt;IRanges&gt;         &lt;Rle&gt; &lt;IRanges&gt;</span></span>
<span><span class="co">##    [1]        II   340-352 ---        II   340-352</span></span>
<span><span class="co">##    [2]        II   340-352 ---        II 3520-3532</span></span>
<span><span class="co">##    [3]        II   340-352 ---        II 7980-7992</span></span>
<span><span class="co">##    [4]        II 3520-3532 ---        II   340-352</span></span>
<span><span class="co">##    [5]        II 3520-3532 ---        II 3520-3532</span></span>
<span><span class="co">##    [6]        II 3520-3532 ---        II 7980-7992</span></span>
<span><span class="co">##    [7]        II 7980-7992 ---        II   340-352</span></span>
<span><span class="co">##    [8]        II 7980-7992 ---        II 3520-3532</span></span>
<span><span class="co">##    [9]        II 7980-7992 ---        II 7980-7992</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    regions: 3 ranges and 0 metadata columns</span></span>
<span><span class="co">##    seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span>
<span></span>
<span><span class="va">hic</span></span>
<span><span class="co">##  `HiCExperiment` object with 306,212 contacts over 257 regions </span></span>
<span><span class="co">##  -------</span></span>
<span><span class="co">##  fileName: "/root/.cache/R/ExperimentHub/f56b079dc5_7752" </span></span>
<span><span class="co">##  focus: "II:300,001-813,184" </span></span>
<span><span class="co">##  resolutions(5): 1000 2000 4000 8000 16000</span></span>
<span><span class="co">##  active resolution: 2000 </span></span>
<span><span class="co">##  interactions: 18513 </span></span>
<span><span class="co">##  scores(3): count balanced random </span></span>
<span><span class="co">##  topologicalFeatures: compartments(0) borders(0) loops(9) viewpoints(0) CTCF(4) </span></span>
<span><span class="co">##  pairsFile: N/A </span></span>
<span><span class="co">##  metadata(0):</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>All these objects can be used in <code>*Overlap</code> methods, as they all extend the <code>GRanges</code> class of objects.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># ---- This counts the number of times `CTCF` anchors are being used in the </span></span>
<span><span class="co">#      `loops` `GInteractions` object</span></span>
<span><span class="fu">countOverlaps</span><span class="op">(</span></span>
<span>    query <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">topologicalFeatures</a></span><span class="op">(</span><span class="va">hic</span>, <span class="st">'CTCF'</span><span class="op">)</span>, </span>
<span>    subject <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">topologicalFeatures</a></span><span class="op">(</span><span class="va">hic</span>, <span class="st">'loops'</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">##  [1] 5 5 5 0</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="pairsfile" class="level4" data-number="3.2.2.3"><h4 data-number="3.2.2.3" class="anchored" data-anchor-id="pairsfile">
<span class="header-section-number">3.2.2.3</span> <code>pairsFile</code>
</h4>
<p>If <code>pairsFile</code> is not specified when importing the <code>ContactFile</code> into a <code>HiCExperiment</code> object, one can add it later.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pairsf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/HiContactsData/man/HiContactsData.html">HiContactsData</a></span><span class="op">(</span><span class="st">'yeast_wt'</span>, <span class="st">'pairs.gz'</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">pairsFile</a></span><span class="op">(</span><span class="va">hic</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">pairsf</span></span>
<span><span class="va">hic</span></span>
<span><span class="co">##  `HiCExperiment` object with 306,212 contacts over 257 regions </span></span>
<span><span class="co">##  -------</span></span>
<span><span class="co">##  fileName: "/root/.cache/R/ExperimentHub/f56b079dc5_7752" </span></span>
<span><span class="co">##  focus: "II:300,001-813,184" </span></span>
<span><span class="co">##  resolutions(5): 1000 2000 4000 8000 16000</span></span>
<span><span class="co">##  active resolution: 2000 </span></span>
<span><span class="co">##  interactions: 18513 </span></span>
<span><span class="co">##  scores(3): count balanced random </span></span>
<span><span class="co">##  topologicalFeatures: compartments(0) borders(0) loops(9) viewpoints(0) CTCF(4) </span></span>
<span><span class="co">##  pairsFile: /root/.cache/R/ExperimentHub/f53d30fcec_7753 </span></span>
<span><span class="co">##  metadata(0):</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="metadata" class="level4" data-number="3.2.2.4"><h4 data-number="3.2.2.4" class="anchored" data-anchor-id="metadata">
<span class="header-section-number">3.2.2.4</span> <code>metadata</code>
</h4>
<p>Metadata associated with a <code>HiCExperiment</code> can be updated at any point.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html">metadata</a></span><span class="op">(</span><span class="va">hic</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    info <span class="op">=</span> <span class="st">"HiCExperiment created from an example .mcool file from `HiContactsData`"</span>, </span>
<span>    date <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/date.html">date</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html">metadata</a></span><span class="op">(</span><span class="va">hic</span><span class="op">)</span></span>
<span><span class="co">##  $info</span></span>
<span><span class="co">##  [1] "HiCExperiment created from an example .mcool file from `HiContactsData`"</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  $date</span></span>
<span><span class="co">##  [1] "Tue Nov 28 08:07:24 2023"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section></section><section id="coercing-hicexperiment-objects" class="level2" data-number="3.3"><h2 data-number="3.3" class="anchored" data-anchor-id="coercing-hicexperiment-objects">
<span class="header-section-number">3.3</span> Coercing <code>HiCExperiment</code> objects</h2>
<p>Convenient coercing functions exist to transform data stored as a <code>HiCExperiment</code> into another class.</p>
<ul>
<li>
<code><a href="https://rdrr.io/r/base/matrix.html">as.matrix()</a></code>: allows to coerce the <code>HiCExperiment</code> into a <code>sparse</code> or <code>dense</code> matrix (using the <code>sparse</code> logical argument, <code>TRUE</code> by default) and choosing specific <code>scores</code> of interest (using the <code>use.scores</code> argument, <code>"balanced"</code> by default).</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># ----- `as.matrix` coerces a `HiCExperiment` into a `sparseMatrix` by default </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">hic</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">##  [1] "dgTMatrix"</span></span>
<span><span class="co">##  attr(,"package")</span></span>
<span><span class="co">##  [1] "Matrix"</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">hic</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">##  [1] 257 257</span></span>
<span></span>
<span><span class="co"># ----- One can specify which scores should be used when coercing into a matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">hic</span>, use.scores <span class="op">=</span> <span class="st">"balanced"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">##  5 x 5 sparse Matrix of class "dgTMatrix"</span></span>
<span><span class="co">##                                                              </span></span>
<span><span class="co">##  [1,] 0.009657438 0.07662234 0.05410199 0.04294051 0.04090521</span></span>
<span><span class="co">##  [2,] 0.076622340 0.05128277 0.09841564 0.06926737 0.05263611</span></span>
<span><span class="co">##  [3,] 0.054101992 0.09841564 0.05657589 0.08723160 0.07316890</span></span>
<span><span class="co">##  [4,] 0.042940512 0.06926737 0.08723160 0.03699543 0.08403496</span></span>
<span><span class="co">##  [5,] 0.040905212 0.05263611 0.07316890 0.08403496 0.04787415</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">hic</span>, use.scores <span class="op">=</span> <span class="st">"count"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">##  5 x 5 sparse Matrix of class "dgTMatrix"</span></span>
<span><span class="co">##                         </span></span>
<span><span class="co">##  [1,]  7  92  75  61  38</span></span>
<span><span class="co">##  [2,] 92 102 226 163  81</span></span>
<span><span class="co">##  [3,] 75 226 150 237 130</span></span>
<span><span class="co">##  [4,] 61 163 237 103 153</span></span>
<span><span class="co">##  [5,] 38  81 130 153  57</span></span>
<span></span>
<span><span class="co"># ----- If **expressly required**, one can coerce a HiCExperiment into a dense matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">hic</span>, use.scores <span class="op">=</span> <span class="st">"count"</span>, sparse <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">##       [,1] [,2] [,3] [,4] [,5]</span></span>
<span><span class="co">##  [1,]    7   92   75   61   38</span></span>
<span><span class="co">##  [2,]   92  102  226  163   81</span></span>
<span><span class="co">##  [3,]   75  226  150  237  130</span></span>
<span><span class="co">##  [4,]   61  163  237  103  153</span></span>
<span><span class="co">##  [5,]   38   81  130  153   57</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>
<code><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame()</a></code>: simply coercing <code>interactions</code> into a rectangular data frame</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">hic</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">##    seqnames1 start1   end1 width1 strand1 bin_id1    weight1 center1</span></span>
<span><span class="co">##  1        II 300001 302000   2000       *     266 0.03714342  301000</span></span>
<span><span class="co">##  2        II 300001 302000   2000       *     266 0.03714342  301000</span></span>
<span><span class="co">##  3        II 300001 302000   2000       *     266 0.03714342  301000</span></span>
<span><span class="co">##  4        II 300001 302000   2000       *     266 0.03714342  301000</span></span>
<span><span class="co">##  5        II 300001 302000   2000       *     266 0.03714342  301000</span></span>
<span><span class="co">##  6        II 300001 302000   2000       *     266 0.03714342  301000</span></span>
<span><span class="co">##    seqnames2 start2   end2 width2 strand2 bin_id2    weight2 center2 count</span></span>
<span><span class="co">##  1        II 300001 302000   2000       *     266 0.03714342  301000     7</span></span>
<span><span class="co">##  2        II 302001 304000   2000       *     267 0.02242258  303000    92</span></span>
<span><span class="co">##  3        II 304001 306000   2000       *     268 0.01942093  305000    75</span></span>
<span><span class="co">##  4        II 306001 308000   2000       *     269 0.01895202  307000    61</span></span>
<span><span class="co">##  5        II 308001 310000   2000       *     270 0.02898098  309000    38</span></span>
<span><span class="co">##  6        II 310001 312000   2000       *     271 0.01834118  311000    43</span></span>
<span><span class="co">##       balanced    random</span></span>
<span><span class="co">##  1 0.009657438 0.4614872</span></span>
<span><span class="co">##  2 0.076622340 0.6568060</span></span>
<span><span class="co">##  3 0.054101992 0.7589135</span></span>
<span><span class="co">##  4 0.042940512 0.3992278</span></span>
<span><span class="co">##  5 0.040905212 0.3551184</span></span>
<span><span class="co">##  6 0.029293930 0.3534359</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Warning
</div>
</div>
<div class="callout-body-container callout-body">
<p>These coercing methods only operate on <code>interactions</code> and <code>scores</code>, and discard all other information, e.g.&nbsp;regarding genomic <code>regions</code>, available <code>resolutions</code>, associated <code>metadata</code>, <code>pairsFile</code> or <code>topologicalFeatures</code>.</p>
</div>
</div>
</section><section id="references" class="level1 unnumbered"><h1 class="unnumbered">References</h1>


</section><a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../pages/data-representation.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Hi-C data structures in R</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../pages/visualization.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Hi-C data visualization</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left"><i>OHCA: Orchestrating Hi-C analysis with Bioconductor</i> was written by J. Serizay</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">This book was built with <a href="https://github.com/js2264/BiocBook/">BiocBook</a> with <span class="emoji" data-emoji="heart">❤️</span></div>
  </div>
</footer>


</body></html>