<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Orchestrating Hi-C analysis with Bioconductor - Preamble</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../pages/principles.html" rel="next">
<link href="../index.html" rel="prev">
<link href="../assets/cover.jpg" rel="icon" type="image/jpeg">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/preamble.html">Preamble</a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../">Orchestrating Hi-C analysis with Bioconductor</a> 
        <div class="sidebar-tools-main">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label=""><i class="bi bi-git"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://github.com/js2264/OHCA/">
            Source Code
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://js2264.github.io/OHCA/devel/">
            Browse version `devel`
            </a>
          </li>
      </ul>
    </div>
    <a href="../Orchestrating-Hi-C-analysis-with-Bioconductor.pdf" rel="" title="Download PDF" class="quarto-navigation-tool px-1" aria-label="Download PDF"><i class="bi bi-file-pdf"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/preamble.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Preamble</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Fundamentals concepts</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/principles.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Hi-C pre-processing steps</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/data-representation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Hi-C data structures in R</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/parsing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Manipulating Hi-C data in R</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/visualization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Hi-C data visualization</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">In-depth Hi-C analysis</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/matrix-centric.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Matrix-centric analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/interactions-centric.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Interactions-centric analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/topological-features.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Finding topological features in Hi-C</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Advanced Hi-C topics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/disseminating.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Data gateways: accessing public Hi-C data portals</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/interoperability.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Interoperability: using HiCExperiment with other R packages</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/workflow-yeast.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workflow 1: Distance-dependent interactions across yeast mutants</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/workflow-chicken.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workflow 2: Chromosome compartment cohesion upon mitosis entry</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/workflow-centros.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workflow 3: Inter-centromere interactions in yeast</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#package-status" id="toc-package-status" class="nav-link active" data-scroll-target="#package-status">Package status</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Preamble</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<p>Hi-C is an experimental method to quantify spatial interactions between any pair of genomic loci. While a number of command-line interfaces (CLI) exist to process and manipulate Hi-C data (e.g.&nbsp;<code>cooler</code> (<span class="citation" data-cites="Abdennur_2019">Abdennur &amp; Mirny (<a href="#ref-Abdennur_2019" role="doc-biblioref">2019</a>)</span>), <code>juicer</code> (<span class="citation" data-cites="Durand_2016">Durand et al. (<a href="#ref-Durand_2016" role="doc-biblioref">2016</a>)</span>) and HiC-Pro (<span class="citation" data-cites="Servant_2015">Servant et al. (<a href="#ref-Servant_2015" role="doc-biblioref">2015</a>)</span>)), they generally suffer from several limitations often found in emerging genomics techniques:</p>
<ul>
<li><strong>No genomic representation of Hi-C processed data</strong>: the existing CLIs can efficiently parse Hi-C data as a numerical matrix and perform a few standard quantitative operations (e.g.&nbsp;contact matrix binning and normalization, dimensionality reduction, etc). However, they systematically fail to <strong>represent a Hi-C contact matrix as a genomic object</strong>. Qualitative analyses (e.g.&nbsp;intersecting chromatin loops with genomic features, finding genes overlapping with domains, etc) therefore remain extremely tedious.</li>
<li><strong>No format-agnostic analysis libraries</strong>. Three competing file format standards (<code>.(m)cool</code>, <code>.hic</code> and <code>HiC-Pro</code> files) currently exist to store Hi-C processed data and dedicated CLIs propose sets of tools specifically working with their corresponding Hi-C processed data file format. This has curbed the development of generic Hi-C data analysis libraries by favoring the emergence of several redundant tools.</li>
<li><strong>Lack of integration within a biology-oriented community</strong>. While rapid development of Hi-C analysis methodology is ongoing, it is primarily driven by small-scale teams rather than by a community as a whole. This oriented development is less likely to fulfill the needs met by other investigators.</li>
</ul>
<p>In this book, we provide an <strong>overview of a set of tools that enable processing, visualization and in-depth investigation of Hi-C data in R, ensuring intuitive integration of Hi-C data in the existing Bioconductor ecosystem</strong>. We introduce a high-level <code>HiCExperiment</code> data structure to represent Hi-C data, directly extending robust, pre-existing core genomic classes offered by <code>Bioconductor</code>. This guarantees a stable and intuitive Hi-C data representation in R as a genomic entity, which is highly interoperable and can be used by all existing analysis packages in R.</p>
<p><img src="images/20230309114202.png" class="img-fluid"></p>
<p>On top of the <code>HiCExperiment</code> data structure, the <code>HiContacts</code> package offers extended functionalities to perform matrix-centric and interaction-centric analysis directly on <code>HiCExperiment</code> objects and provides powerful visualization tools specifically designed for Hi-C data to facilitate exploratory data analysis. In addition, the <code>HiCool</code> package implements a processing workflow based on a lightweight library to process raw Hi-C data into binned Hi-C contact matrices ready to be imported as <code>HiCExperiment</code> objects. Finally, the <code>fourDNData</code> and <code>DNAZooData</code> packages offer a gateway to major public data repositories directly in R.</p>
<p><img src="images/20230421134800.jpg" class="img-fluid"></p>
<section id="package-status" class="level1 unnumbered">
<h1 class="unnumbered">Package status</h1>
<table class="table">
<thead>
<tr class="header">
<th>Github repo üíæ</th>
<th>Doc üìò</th>
<th>Github checks ‚úÖ</th>
<th>Bioc builds üèó</th>
<th>Lifecycle üå±</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><a href="https://github.com/js2264/HiCExperiment">HiCExperiment</a></td>
<td><a href="https://js2264.github.io/HiCExperiment">Doc</a></td>
<td><a href="https://github.com/js2264/HiCExperiment/actions/workflows/rworkflows.yml"><img src="https://github.com/js2264/HiCExperiment/actions/workflows/rworkflows.yml/badge.svg" class="img-fluid" alt="rworkflows"></a></td>
<td>Bioc release:<br><a href="https://bioconductor.org/checkResults/release/bioc-LATEST/HiCExperiment/"><img src="https://bioconductor.org/shields/build/release/bioc/HiCExperiment.svg" class="img-fluid" alt="Release"></a> <br>Bioc devel:<br><a href="https://bioconductor.org/checkResults/devel/bioc-LATEST/HiCExperiment/"><img src="https://bioconductor.org/shields/build/devel/bioc/HiCExperiment.svg" class="img-fluid" alt="Devel"></a></td>
<td><a href="https://www.tidyverse.org/lifecycle/#maturing"><img src="https://img.shields.io/badge/lifecycle-maturing-blue.svg" class="img-fluid"></a></td>
</tr>
<tr class="even">
<td><a href="https://github.com/js2264/HiContacts">HiContacts</a></td>
<td><a href="https://js2264.github.io/HiContacts">Doc</a></td>
<td><a href="https://github.com/js2264/HiContacts/actions/workflows/rworkflows.yml"><img src="https://github.com/js2264/HiContacts/actions/workflows/rworkflows.yml/badge.svg" class="img-fluid" alt="rworkflows"></a></td>
<td>Bioc release:<br><a href="https://bioconductor.org/checkResults/release/bioc-LATEST/HiContacts/"><img src="https://bioconductor.org/shields/build/release/bioc/HiContacts.svg" class="img-fluid" alt="Release"></a> <br>Bioc devel:<br><a href="https://bioconductor.org/checkResults/devel/bioc-LATEST/HiContacts/"><img src="https://bioconductor.org/shields/build/devel/bioc/HiContacts.svg" class="img-fluid" alt="Devel"></a></td>
<td><a href="https://www.tidyverse.org/lifecycle/#maturing"><img src="https://img.shields.io/badge/lifecycle-maturing-blue.svg" class="img-fluid"></a></td>
</tr>
<tr class="odd">
<td><a href="https://github.com/js2264/HiCool">HiCool</a></td>
<td><a href="https://js2264.github.io/HiCool">Doc</a></td>
<td><a href="https://github.com/js2264/HiCool/actions/workflows/rworkflows.yml"><img src="https://github.com/js2264/HiCool/actions/workflows/rworkflows.yml/badge.svg" class="img-fluid" alt="rworkflows"></a></td>
<td>Bioc release:<br><a href="https://bioconductor.org/checkResults/release/bioc-LATEST/HiCool/"><img src="https://bioconductor.org/shields/build/release/bioc/HiCool.svg" class="img-fluid" alt="Release"></a> <br>Bioc devel:<br><a href="https://bioconductor.org/checkResults/devel/bioc-LATEST/HiCool/"><img src="https://bioconductor.org/shields/build/devel/bioc/HiCool.svg" class="img-fluid" alt="Devel"></a></td>
<td><a href="https://www.tidyverse.org/lifecycle/#maturing"><img src="https://img.shields.io/badge/lifecycle-maturing-blue.svg" class="img-fluid"></a></td>
</tr>
<tr class="even">
<td><a href="https://github.com/js2264/HiContactsData">HiContactsData</a></td>
<td><a href="https://js2264.github.io/HiContactsData">Doc</a></td>
<td><a href="https://github.com/js2264/HiContactsData/actions/workflows/check-bioc.yml"><img src="https://github.com/js2264/HiContactsData/actions/workflows/check-bioc.yml/badge.svg" class="img-fluid" alt="R-CMD-check-bioc"></a></td>
<td>Bioc release:<br><a href="https://bioconductor.org/checkResults/release/data-experiment-LATEST/HiContactsData/"><img src="https://bioconductor.org/shields/build/release/data-experiment/HiContactsData.svg" class="img-fluid" alt="Release"></a> <br>Bioc devel:<br><a href="https://bioconductor.org/checkResults/devel/data-experiment-LATEST/HiContactsData/"><img src="https://bioconductor.org/shields/build/release/data-experiment/HiContactsData.svg" class="img-fluid" alt="Devel"></a></td>
<td><a href="https://lifecycle.r-lib.org/articles/stages.html#stable"><img src="https://img.shields.io/badge/lifecycle-stable-brightgreen.svg" class="img-fluid"></a></td>
</tr>
<tr class="odd">
<td><a href="https://github.com/js2264/DNAZooData">DNAZooData</a></td>
<td><a href="https://js2264.github.io/DNAZooData">Doc</a></td>
<td><a href="https://github.com/js2264/DNAZooData/actions/workflows/rworkflows.yml"><img src="https://github.com/js2264/DNAZooData/actions/workflows/rworkflows.yml/badge.svg" class="img-fluid" alt="rworkflows"></a></td>
<td>Bioc release:<br><a href="https://bioconductor.org/checkResults/release/data-experiment-LATEST/DNAZooData/"><img src="https://bioconductor.org/shields/build/release/data-experiment/DNAZooData.svg" class="img-fluid" alt="Release"></a> <br>Bioc devel:<br><a href="https://bioconductor.org/checkResults/devel/data-experiment-LATEST/DNAZooData/"><img src="https://bioconductor.org/shields/build/release/data-experiment/DNAZooData.svg" class="img-fluid" alt="Devel"></a></td>
<td><a href="https://lifecycle.r-lib.org/articles/stages.html#stable"><img src="https://img.shields.io/badge/lifecycle-stable-brightgreen.svg" class="img-fluid"></a></td>
</tr>
<tr class="even">
<td><a href="https://github.com/js2264/fourDNData">fourDNData</a></td>
<td><a href="https://js2264.github.io/fourDNData">Doc</a></td>
<td><a href="https://github.com/js2264/fourDNData/actions/workflows/rworkflows.yml"><img src="https://github.com/js2264/fourDNData/actions/workflows/rworkflows.yml/badge.svg" class="img-fluid" alt="rworkflows"></a></td>
<td>Bioc release:<br><a href="https://bioconductor.org/checkResults/release/data-experiment-LATEST/fourDNData/"><img src="https://bioconductor.org/shields/build/release/data-experiment/fourDNData.svg" class="img-fluid" alt="Release"></a> <br>Bioc devel:<br><a href="https://bioconductor.org/checkResults/devel/data-experiment-LATEST/fourDNData/"><img src="https://bioconductor.org/shields/build/release/data-experiment/fourDNData.svg" class="img-fluid" alt="Devel"></a></td>
<td><a href="https://lifecycle.r-lib.org/articles/stages.html#stable"><img src="https://img.shields.io/badge/lifecycle-stable-brightgreen.svg" class="img-fluid"></a></td>
</tr>
</tbody>
</table>
</section>
<section id="references" class="level1 unnumbered">
<h1 class="unnumbered">References</h1>


<div id="refs" class="references csl-bib-body hanging-indent" data-line-spacing="2" role="list">
<div id="ref-Abdennur_2019" class="csl-entry" role="listitem">
Abdennur, N., &amp; Mirny, L. A. (2019). Cooler: Scalable storage for hi-c data and other genomically labeled arrays. <em>Bioinformatics</em>, <em>36</em>(1), 311‚Äì316. <a href="https://doi.org/10.1093/bioinformatics/btz540">https://doi.org/10.1093/bioinformatics/btz540</a>
</div>
<div id="ref-Durand_2016" class="csl-entry" role="listitem">
Durand, N. C., Shamim, M. S., Machol, I., Rao, S. S. P., Huntley, M. H., Lander, E. S., &amp; Aiden, E. L. (2016). Juicer provides a one-click system for analyzing loop-resolution hi-c experiments. <em>Cell Systems</em>, <em>3</em>(1), 95‚Äì98. <a href="https://doi.org/10.1016/j.cels.2016.07.002">https://doi.org/10.1016/j.cels.2016.07.002</a>
</div>
<div id="ref-Servant_2015" class="csl-entry" role="listitem">
Servant, N., Varoquaux, N., Lajoie, B. R., Viara, E., Chen, C.-J., Vert, J.-P., Heard, E., Dekker, J., &amp; Barillot, E. (2015). <span>HiC</span>-pro: An optimized and flexible pipeline for hi-c data processing. <em>Genome Biology</em>, <em>16</em>(1). <a href="https://doi.org/10.1186/s13059-015-0831-x">https://doi.org/10.1186/s13059-015-0831-x</a>
</div>
</div>
</section>

<a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "Óßã";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../index.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Welcome</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../pages/principles.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Hi-C pre-processing steps</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left"><i>OHCA: Orchestrating Hi-C analysis with Bioconductor</i> was written by J. Serizay</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">This book was built with <a href="https://github.com/js2264/BiocBook/">BiocBook</a> with <span class="emoji" data-emoji="heart">‚ù§Ô∏è</span></div>
  </div>
</footer>



</body></html>