<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Orchestrating Hi-C analysis with Bioconductor - 4&nbsp; Hi-C data visualization</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../pages/matrix-centric.html" rel="next">
<link href="../pages/parsing.html" rel="prev">
<link href="../assets/cover.jpg" rel="icon" type="image/jpeg">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
</head>
<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/principles.html">Fundamentals concepts</a></li><li class="breadcrumb-item"><a href="../pages/visualization.html"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Hi-C data visualization</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../">Orchestrating Hi-C analysis with Bioconductor</a> 
        <div class="sidebar-tools-main">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label=""><i class="bi bi-git"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
<li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://github.com/js2264/OHCA/">
            Source Code
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://js2264.github.io/OHCA/devel/">
            Browse version `devel`
            </a>
          </li>
      </ul>
</div>
    <a href="../Orchestrating-Hi-C-analysis-with-Bioconductor.pdf" rel="" title="Download PDF" class="quarto-navigation-tool px-1" aria-label="Download PDF"><i class="bi bi-file-pdf"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/preamble.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preamble</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Fundamentals concepts</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/principles.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Hi-C pre-processing steps</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/data-representation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Hi-C data structures in R</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/parsing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Manipulating Hi-C data in R</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/visualization.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Hi-C data visualization</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">In-depth Hi-C analysis</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/matrix-centric.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Matrix-centric analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/interactions-centric.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Interactions-centric analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/topological-features.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Finding topological features in Hi-C</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Advanced Hi-C topics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/disseminating.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Data gateways: accessing public Hi-C data portals</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/interoperability.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Interoperability: using HiCExperiment with other R packages</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/workflow-yeast.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workflow 1: Distance-dependent interactions across yeast mutants</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/workflow-chicken.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workflow 2: Chromosome compartment cohesion upon mitosis entry</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/workflow-centros.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workflow 3: Inter-centromere interactions in yeast</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li>
<a href="#visualizing-hi-c-contact-maps" id="toc-visualizing-hi-c-contact-maps" class="nav-link active" data-scroll-target="#visualizing-hi-c-contact-maps"><span class="header-section-number">4.1</span> Visualizing Hi-C contact maps</a>
  <ul class="collapse">
<li><a href="#single-map" id="toc-single-map" class="nav-link" data-scroll-target="#single-map"><span class="header-section-number">4.1.1</span> Single map</a></li>
  <li><a href="#horizontal-map" id="toc-horizontal-map" class="nav-link" data-scroll-target="#horizontal-map"><span class="header-section-number">4.1.2</span> Horizontal map</a></li>
  <li><a href="#side-by-side-maps" id="toc-side-by-side-maps" class="nav-link" data-scroll-target="#side-by-side-maps"><span class="header-section-number">4.1.3</span> Side-by-side maps</a></li>
  <li><a href="#plotting-multiple-chromosomes" id="toc-plotting-multiple-chromosomes" class="nav-link" data-scroll-target="#plotting-multiple-chromosomes"><span class="header-section-number">4.1.4</span> Plotting multiple chromosomes</a></li>
  </ul>
</li>
  <li>
<a href="#hi-c-maps-customization-options" id="toc-hi-c-maps-customization-options" class="nav-link" data-scroll-target="#hi-c-maps-customization-options"><span class="header-section-number">4.2</span> Hi-C maps customization options</a>
  <ul class="collapse">
<li><a href="#choosing-scores" id="toc-choosing-scores" class="nav-link" data-scroll-target="#choosing-scores"><span class="header-section-number">4.2.1</span> Choosing scores</a></li>
  <li><a href="#choosing-scale" id="toc-choosing-scale" class="nav-link" data-scroll-target="#choosing-scale"><span class="header-section-number">4.2.2</span> Choosing scale</a></li>
  <li><a href="#choosing-color-map" id="toc-choosing-color-map" class="nav-link" data-scroll-target="#choosing-color-map"><span class="header-section-number">4.2.3</span> Choosing color map</a></li>
  </ul>
</li>
  <li>
<a href="#advanced-visualization" id="toc-advanced-visualization" class="nav-link" data-scroll-target="#advanced-visualization"><span class="header-section-number">4.3</span> Advanced visualization</a>
  <ul class="collapse">
<li><a href="#overlaying-topological-features" id="toc-overlaying-topological-features" class="nav-link" data-scroll-target="#overlaying-topological-features"><span class="header-section-number">4.3.1</span> Overlaying topological features</a></li>
  <li><a href="#aggregated-hi-c-maps" id="toc-aggregated-hi-c-maps" class="nav-link" data-scroll-target="#aggregated-hi-c-maps"><span class="header-section-number">4.3.2</span> Aggregated Hi-C maps</a></li>
  </ul>
</li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title">
<span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Hi-C data visualization</span>
</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Aims
</div>
</div>
<div class="callout-body-container callout-body">
<p>This chapter focuses on the various visualization tools offered by <code>HiContacts</code> to plot <code>HiCExperiment</code> contact matrices in R.</p>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Generating the example <code>hic</code> object üëá
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>To demonstrate how to visualize a <code>HiCExperiment</code> contact matrix, we will create an <code>HiCExperiment</code> object from an example <code>.cool</code> file provided in the <code>HiContactsData</code> package.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/js2264/HiCExperiment">HiCExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/js2264/HiContactsData">HiContactsData</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># ---- This downloads an example `.mcool` file and caches it locally </span></span>
<span><span class="va">coolf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/HiContactsData/man/HiContactsData.html">HiContactsData</a></span><span class="op">(</span><span class="st">'yeast_wt'</span>, <span class="st">'mcool'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># ---- This creates a connection to the disk-stored `.mcool` file</span></span>
<span><span class="va">cf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/CoolFile-class.html">CoolFile</a></span><span class="op">(</span><span class="va">coolf</span><span class="op">)</span></span>
<span><span class="va">cf</span></span>
<span></span>
<span><span class="co"># ---- This imports contacts from the chromosome `V` at resolution `2000`</span></span>
<span><span class="va">hic</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocIO/man/IO.html">import</a></span><span class="op">(</span><span class="va">cf</span>, focus <span class="op">=</span> <span class="st">'V'</span>, resolution <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hic</span></span>
<span><span class="co">##  `HiCExperiment` object with 303,545 contacts over 289 regions </span></span>
<span><span class="co">##  -------</span></span>
<span><span class="co">##  fileName: "/root/.cache/R/ExperimentHub/f56b079dc5_7752" </span></span>
<span><span class="co">##  focus: "V" </span></span>
<span><span class="co">##  resolutions(5): 1000 2000 4000 8000 16000</span></span>
<span><span class="co">##  active resolution: 2000 </span></span>
<span><span class="co">##  interactions: 20177 </span></span>
<span><span class="co">##  scores(2): count balanced </span></span>
<span><span class="co">##  topologicalFeatures: compartments(0) borders(0) loops(0) viewpoints(0) centromeres(16) </span></span>
<span><span class="co">##  pairsFile: N/A </span></span>
<span><span class="co">##  metadata(0):</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<section id="visualizing-hi-c-contact-maps" class="level2" data-number="4.1"><h2 data-number="4.1" class="anchored" data-anchor-id="visualizing-hi-c-contact-maps">
<span class="header-section-number">4.1</span> Visualizing Hi-C contact maps</h2>
<p>Visualizing Hi-C contact maps is often a necessary step in exploratory data analysis. A Hi-C contact map is usually displayed as a heatmap, in which:</p>
<ul>
<li>Each axis represents a section of the genome of interest (either a segment of a chromosome, or several chromosomes, ‚Ä¶).</li>
<li>The color code aims to represent ‚Äúinteraction frequency‚Äù, which can be expressed in ‚Äúraw‚Äù counts or normalized (balanced).</li>
<li>Other metrics can also be displayed in Hi-C heatmaps, e.g.&nbsp;ratios of interaction frequency between two Hi-C experiments, p-values of differential interaction analysis, ‚Ä¶</li>
<li>Axes are often identical, representing interactions constrained within a single genomic window, a.k.a. <strong>on-diagonal</strong> matrices.</li>
<li>However, axes <em>can</em> be different: this is the case when <strong>off-diagonal</strong> matrices are displayed.</li>
</ul>
<section id="single-map" class="level3" data-number="4.1.1"><h3 data-number="4.1.1" class="anchored" data-anchor-id="single-map">
<span class="header-section-number">4.1.1</span> Single map</h3>
<p>Simple visualization of disk-stored Hi-C contact matrices can be done by:</p>
<ol type="1">
<li>Importing the interactions over the genomic location of interest into a <code>HiCExperiment</code> object;</li>
<li>Using <code>plotMatrix</code> function (provided by <code>HiContacts</code>) to generate a plot.</li>
</ol>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/js2264/HiContacts">HiContacts</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/HiContacts/man/plotMatrix.html">plotMatrix</a></span><span class="op">(</span><span class="va">hic</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="visualization_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>A caption summarizing the plotting parameters is added below the heatmap. This can be removed with <code>caption = FALSE</code>.</p>
</div>
</div>
</section><section id="horizontal-map" class="level3" data-number="4.1.2"><h3 data-number="4.1.2" class="anchored" data-anchor-id="horizontal-map">
<span class="header-section-number">4.1.2</span> Horizontal map</h3>
<p>Hi-C maps are sometimes visualized in a ‚Äúhorizontal‚Äù style, where a square <strong>on-diagonal</strong> heatmap is tilted by 45Àö and truncated to only show interactions up to a certain distance from the main diagonal.</p>
<p>When a <code>maxDistance</code> argument is provided to <code>plotMatrix</code>, it automatically generates a horizontal-style heatmap.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/HiContacts/man/plotMatrix.html">plotMatrix</a></span><span class="op">(</span><span class="va">hic</span>, maxDistance <span class="op">=</span> <span class="fl">200000</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="visualization_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="side-by-side-maps" class="level3" data-number="4.1.3"><h3 data-number="4.1.3" class="anchored" data-anchor-id="side-by-side-maps">
<span class="header-section-number">4.1.3</span> Side-by-side maps</h3>
<p>Sometimes, one may want to visually plot 2 Hi-C samples side by side to compare the interaction landscapes over the same genomic locus. This can be done by adding a second <code>HiCExperiment</code> (imported with the same <code>focus</code>) with the <code>compare.to</code> argument.</p>
<p>Here, we are importing a second <code>.mcool</code> file corresponding to a Hi-C experiment performed in a <em>eco1</em> yeast mutant:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hic2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocIO/man/IO.html">import</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/CoolFile-class.html">CoolFile</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/HiContactsData/man/HiContactsData.html">HiContactsData</a></span><span class="op">(</span><span class="st">'yeast_eco1'</span>, <span class="st">'mcool'</span><span class="op">)</span><span class="op">)</span>, </span>
<span>    focus <span class="op">=</span> <span class="st">'V'</span>, </span>
<span>    resolution <span class="op">=</span> <span class="fl">2000</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We then plot the 2 matrices side by side. The first will be displayed in the top right corner and the second (provided with <code>compare.to</code>) will be in the bottom left corner.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/HiContacts/man/plotMatrix.html">plotMatrix</a></span><span class="op">(</span><span class="va">hic</span>, compare.to <span class="op">=</span> <span class="va">hic2</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="visualization_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="plotting-multiple-chromosomes" class="level3" data-number="4.1.4"><h3 data-number="4.1.4" class="anchored" data-anchor-id="plotting-multiple-chromosomes">
<span class="header-section-number">4.1.4</span> Plotting multiple chromosomes</h3>
<p>Interactions from multiple chromosomes can be visualized in a Hi-C heatmap. One needs to (1) first parse the entire contact matrix in <code>R</code>, (2) then subset interactions over chromosomes of interest with <code>[</code> and (3) use <code>plotMatrix</code> to generate the multi-chromosome plot.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">full_hic</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocIO/man/IO.html">import</a></span><span class="op">(</span><span class="va">cf</span>, resolution <span class="op">=</span> <span class="fl">4000</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/HiContacts/man/plotMatrix.html">plotMatrix</a></span><span class="op">(</span><span class="va">full_hic</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="visualization_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hic_subset</span> <span class="op">&lt;-</span> <span class="va">full_hic</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"II"</span>, <span class="st">"III"</span>, <span class="st">"IV"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/HiContacts/man/plotMatrix.html">plotMatrix</a></span><span class="op">(</span><span class="va">hic_subset</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="visualization_files/figure-html/unnamed-chunk-8-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section></section><section id="hi-c-maps-customization-options" class="level2" data-number="4.2"><h2 data-number="4.2" class="anchored" data-anchor-id="hi-c-maps-customization-options">
<span class="header-section-number">4.2</span> Hi-C maps customization options</h2>
<p>A number of customization options are available for the <code>plotMatrix</code> function. The next subsections focus on how to:</p>
<ul>
<li>Pick the <code>scores</code> of interest to represent in a Hi-C heatmap;</li>
<li>Change the numeric scale and boundaries;</li>
<li>Change the color map;</li>
<li>Extra customization options</li>
</ul>
<section id="choosing-scores" class="level3" data-number="4.2.1"><h3 data-number="4.2.1" class="anchored" data-anchor-id="choosing-scores">
<span class="header-section-number">4.2.1</span> Choosing scores</h3>
<p>By default, <code>plotMatrix</code> will attempt to plot <code>balanced</code> (coverage normalized) Hi-C matrices. However, extra scores may be associated with interactions in a <code>HiCExperiment</code> object (more on this in the <a href="../pages/matrix-centric.html">next chapter</a>)</p>
<p>For instance, we can plot the <code>count</code> scores, which are un-normalized raw contact counts directly obtained when binning a <code>.pairs</code> file:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/HiContacts/man/plotMatrix.html">plotMatrix</a></span><span class="op">(</span><span class="va">hic</span>, use.scores <span class="op">=</span> <span class="st">'count'</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="visualization_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="choosing-scale" class="level3" data-number="4.2.2"><h3 data-number="4.2.2" class="anchored" data-anchor-id="choosing-scale">
<span class="header-section-number">4.2.2</span> Choosing scale</h3>
<p>The color scale is automatically adjusted to range from the minimum to the maximum <code>scores</code> of the <code>HiCExperiment</code> being plotted. This can be adjusted using the <code>limits</code> argument.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/HiContacts/man/plotMatrix.html">plotMatrix</a></span><span class="op">(</span><span class="va">hic</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">3.5</span>, <span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="visualization_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="choosing-color-map" class="level3" data-number="4.2.3"><h3 data-number="4.2.3" class="anchored" data-anchor-id="choosing-color-map">
<span class="header-section-number">4.2.3</span> Choosing color map</h3>
<p><code><a href="https://rdrr.io/pkg/HiContacts/man/palettes.html">?HiContacts::palettes</a></code> returns a list of available color maps to use with <code>plotMatrix</code>. Any custom color map can also be used by manually specifying a vector of colors.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># ----- `afmhotr` color map is shipped in the `HiContacts` package</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/HiContacts/man/palettes.html">afmhotrColors</a></span><span class="op">(</span><span class="op">)</span> </span>
<span><span class="co">##  [1] "#ffffff" "#f8f5c3" "#f4ee8d" "#f6be35" "#ee7d32" "#c44228" "#821d19"</span></span>
<span><span class="co">##  [8] "#381211" "#050606"</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/HiContacts/man/plotMatrix.html">plotMatrix</a></span><span class="op">(</span></span>
<span>    <span class="va">hic</span>, </span>
<span>    use.scores <span class="op">=</span> <span class="st">'balanced'</span>,</span>
<span>    limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">4</span>, <span class="op">-</span><span class="fl">1</span><span class="op">)</span>,</span>
<span>    cmap <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/HiContacts/man/palettes.html">afmhotrColors</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="visualization_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section></section><section id="advanced-visualization" class="level2" data-number="4.3"><h2 data-number="4.3" class="anchored" data-anchor-id="advanced-visualization">
<span class="header-section-number">4.3</span> Advanced visualization</h2>
<section id="overlaying-topological-features" class="level3" data-number="4.3.1"><h3 data-number="4.3.1" class="anchored" data-anchor-id="overlaying-topological-features">
<span class="header-section-number">4.3.1</span> Overlaying topological features</h3>
<p>Topological features (e.g.&nbsp;chromatin loops, domain borders, A/B compartments, e.g.&nbsp;‚Ä¶) are often displayed over a Hi-C heatmap.</p>
<p>To illustrate how to do this, let‚Äôs import pre-computed chromatin loops in <code>R</code>. These loops have been identified using <code>chromosight</code> (<span class="citation" data-cites="Matthey_Doret_2020">Matthey-Doret et al. (<a href="#ref-Matthey_Doret_2020" role="doc-biblioref">2020</a>)</span>) on the contact matrix which we imported interactions from.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">InteractionSet</span><span class="op">)</span></span>
<span><span class="va">loops</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'S288C-loops.bedpe'</span>, package <span class="op">=</span> <span class="st">'HiCExperiment'</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocIO/man/IO.html">import</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/pairs.html">makeGInteractionsFromGRangesPairs</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">loops</span></span>
<span><span class="co">##  GInteractions object with 162 interactions and 0 metadata columns:</span></span>
<span><span class="co">##          seqnames1       ranges1     seqnames2       ranges2</span></span>
<span><span class="co">##              &lt;Rle&gt;     &lt;IRanges&gt;         &lt;Rle&gt;     &lt;IRanges&gt;</span></span>
<span><span class="co">##      [1]         I     3001-4000 ---         I   29001-30000</span></span>
<span><span class="co">##      [2]         I   29001-30000 ---         I   50001-51000</span></span>
<span><span class="co">##      [3]         I   95001-96000 ---         I 128001-129000</span></span>
<span><span class="co">##      [4]         I 133001-134000 ---         I 157001-158000</span></span>
<span><span class="co">##      [5]        II     8001-9000 ---        II   46001-47000</span></span>
<span><span class="co">##      ...       ...           ... ...       ...           ...</span></span>
<span><span class="co">##    [158]       XVI 773001-774000 ---       XVI 803001-804000</span></span>
<span><span class="co">##    [159]       XVI 834001-835000 ---       XVI 859001-860000</span></span>
<span><span class="co">##    [160]       XVI 860001-861000 ---       XVI 884001-885000</span></span>
<span><span class="co">##    [161]       XVI 901001-902000 ---       XVI 940001-941000</span></span>
<span><span class="co">##    [162]       XVI 917001-918000 ---       XVI 939001-940000</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    regions: 316 ranges and 0 metadata columns</span></span>
<span><span class="co">##    seqinfo: 16 sequences from an unspecified genome; no seqlengths</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Similarly, borders have also been mapped with <code>chromosight</code>. We can also import them in <code>R</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">borders</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'S288C-borders.bed'</span>, package <span class="op">=</span> <span class="st">'HiCExperiment'</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocIO/man/IO.html">import</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">borders</span></span>
<span><span class="co">##  GRanges object with 814 ranges and 0 metadata columns:</span></span>
<span><span class="co">##          seqnames        ranges strand</span></span>
<span><span class="co">##             &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##      [1]        I   73001-74000      *</span></span>
<span><span class="co">##      [2]        I 108001-109000      *</span></span>
<span><span class="co">##      [3]        I 181001-182000      *</span></span>
<span><span class="co">##      [4]       II   90001-91000      *</span></span>
<span><span class="co">##      [5]       II 119001-120000      *</span></span>
<span><span class="co">##      ...      ...           ...    ...</span></span>
<span><span class="co">##    [810]      XVI 777001-778000      *</span></span>
<span><span class="co">##    [811]      XVI 796001-797000      *</span></span>
<span><span class="co">##    [812]      XVI 811001-812000      *</span></span>
<span><span class="co">##    [813]      XVI 890001-891000      *</span></span>
<span><span class="co">##    [814]      XVI 933001-934000      *</span></span>
<span><span class="co">##    -------</span></span>
<span><span class="co">##    seqinfo: 16 sequences from an unspecified genome; no seqlengths</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Chromatin loops are stored in <code>GInteractions</code> while borders are <code>GRanges</code>. The former will be displayed as <strong>off-diagonal circles</strong> and the later as <strong>on-diagonal diamonds</strong> on the Hi-C heatmap.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/HiContacts/man/plotMatrix.html">plotMatrix</a></span><span class="op">(</span><span class="va">hic</span>, loops <span class="op">=</span> <span class="va">loops</span>, borders <span class="op">=</span> <span class="va">borders</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="visualization_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="aggregated-hi-c-maps" class="level3" data-number="4.3.2"><h3 data-number="4.3.2" class="anchored" data-anchor-id="aggregated-hi-c-maps">
<span class="header-section-number">4.3.2</span> Aggregated Hi-C maps</h3>
<p>Finally, Hi-C map ‚Äúsnippets‚Äù (i.e.&nbsp;extracts) are often aggregated together to show an average signal. This analysis is sometimes referred to as APA (Aggregated Plot Analysis).</p>
<p>Aggregated Hi-C maps can be computed over a collection of <code>targets</code> using the <code>aggregate</code> function. These targets can be <code>GRanges</code> (to extract on-diagonal snippets) or <code>GInteractions</code> (to extract off-diagonal snippets). The <code>flankingBins</code> specifies how many matrix bins should be extracted on each side of the <code>targets</code> of interest.</p>
<p>Here, we compute the aggregated Hi-C snippets of ¬± 15kb around each chromatin loop listed in <code>loops</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hic</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">zoom</a></span><span class="op">(</span><span class="va">hic</span>, <span class="fl">1000</span><span class="op">)</span></span>
<span><span class="va">aggr_loops</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/aggregate.html">aggregate</a></span><span class="op">(</span><span class="va">hic</span>, targets <span class="op">=</span> <span class="va">loops</span>, flankingBins <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></span>
<span><span class="co">##  Going through preflight checklist...</span></span>
<span><span class="co">##  Parsing the entire contact matrice as a sparse matrix...</span></span>
<span><span class="co">##  Modeling distance decay...</span></span>
<span><span class="co">##  Filtering for contacts within provided targets...</span></span>
<span><span class="va">aggr_loops</span></span>
<span><span class="co">##  `AggrHiCExperiment` object over 148 targets </span></span>
<span><span class="co">##  -------</span></span>
<span><span class="co">##  fileName: "/root/.cache/R/ExperimentHub/f56b079dc5_7752" </span></span>
<span><span class="co">##  focus: 148 targets </span></span>
<span><span class="co">##  resolutions(5): 1000 2000 4000 8000 16000</span></span>
<span><span class="co">##  active resolution: 1000 </span></span>
<span><span class="co">##  interactions: 961 </span></span>
<span><span class="co">##  scores(4): count balanced expected detrended </span></span>
<span><span class="co">##  slices(4): count balanced expected detrended </span></span>
<span><span class="co">##  topologicalFeatures: targets(148) compartments(0) borders(0) loops(0) viewpoints(0) centromeres(16) </span></span>
<span><span class="co">##  pairsFile: N/A </span></span>
<span><span class="co">##  metadata(0):</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><code>aggregate</code> generates a <code>AggrHiCExperiment</code> object, a flavor of <code>HiCExperiment</code> class of objects.</p>
<ul>
<li>
<code>AggrHiCExperiment</code> objects have an extra <code>slices</code> slot. This stores a list of <code>array</code>s, one per <code>scores</code>. Each <code>array</code> is of 3 dimensions, <code>x</code> and <code>y</code> representing the heatmap axes, and <code>z</code> representing the index of the <code>target</code>.</li>
<li>
<code>AggrHiCExperiment</code> objects also have a mandatory <code>topologicalFeatures</code> element named <code>targets</code>, storing the genomic loci provided in <code>aggregate</code>.</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">slices</a></span><span class="op">(</span><span class="va">aggr_loops</span><span class="op">)</span></span>
<span><span class="co">##  List of length 4</span></span>
<span><span class="co">##  names(4): count balanced expected detrended</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">slices</a></span><span class="op">(</span><span class="va">aggr_loops</span>, <span class="st">'count'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  [1]  31  31 148</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">topologicalFeatures</a></span><span class="op">(</span><span class="va">aggr_loops</span>, <span class="st">'targets'</span><span class="op">)</span></span>
<span><span class="co">##  Pairs object with 148 pairs and 0 metadata columns:</span></span>
<span><span class="co">##                      first            second</span></span>
<span><span class="co">##                  &lt;GRanges&gt;         &lt;GRanges&gt;</span></span>
<span><span class="co">##      [1]     I:14501-44500     I:35501-65500</span></span>
<span><span class="co">##      [2]    I:80501-110500   I:113501-143500</span></span>
<span><span class="co">##      [3]   I:118501-148500   I:142501-172500</span></span>
<span><span class="co">##      [4]    II:33501-63500    II:63501-93500</span></span>
<span><span class="co">##      [5]  II:134501-164500  II:159501-189500</span></span>
<span><span class="co">##      ...               ...               ...</span></span>
<span><span class="co">##    [144] XVI:586501-616500 XVI:606501-636500</span></span>
<span><span class="co">##    [145] XVI:733501-763500 XVI:754501-784500</span></span>
<span><span class="co">##    [146] XVI:758501-788500 XVI:788501-818500</span></span>
<span><span class="co">##    [147] XVI:819501-849500 XVI:844501-874500</span></span>
<span><span class="co">##    [148] XVI:845501-875500 XVI:869501-899500</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The resulting <code>AggrHiCExperiment</code> can be plotted using the same <code>plotMatrix</code> function with the arguments described above.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/HiContacts/man/plotMatrix.html">plotMatrix</a></span><span class="op">(</span></span>
<span>    <span class="va">aggr_loops</span>, </span>
<span>    use.scores <span class="op">=</span> <span class="st">'detrended'</span>, </span>
<span>    scale <span class="op">=</span> <span class="st">'linear'</span>, </span>
<span>    limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>, </span>
<span>    cmap <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/HiContacts/man/palettes.html">bgrColors</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="visualization_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section></section><section id="references" class="level1 unnumbered"><h1 class="unnumbered">References</h1>


<div id="refs" class="references csl-bib-body hanging-indent" data-line-spacing="2" role="list">
<div id="ref-Matthey_Doret_2020" class="csl-entry" role="listitem">
Matthey-Doret, C., Baudry, L., Breuer, A., Montagne, R., Guiglielmoni, N., Scolari, V., Jean, E., Campeas, A., Chanut, P. H., Oriol, E., M√©ot, A., Politis, L., Vigouroux, A., Moreau, P., Koszul, R., &amp; Cournac, A. (2020). Computer vision for pattern detection in chromosome contact maps. <em>Nature Communications</em>, <em>11</em>(1). <a href="https://doi.org/10.1038/s41467-020-19562-7">https://doi.org/10.1038/s41467-020-19562-7</a>
</div>
</div>
</section><a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "Óßã";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../pages/parsing.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Manipulating Hi-C data in R</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../pages/matrix-centric.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Matrix-centric analysis</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left"><i>OHCA: Orchestrating Hi-C analysis with Bioconductor</i> was written by J. Serizay</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">This book was built with <a href="https://github.com/js2264/BiocBook/">BiocBook</a> with <span class="emoji" data-emoji="heart">‚ù§Ô∏è</span></div>
  </div>
</footer>


</body></html>