<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Orchestrating Hi-C analysis with Bioconductor - Workflow 1: Distance-dependent interactions across yeast mutants</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../pages/workflow-chicken.html" rel="next">
<link href="../pages/interoperability.html" rel="prev">
<link href="../assets/cover.jpg" rel="icon" type="image/jpeg">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
</head>
<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/disseminating.html">Advanced Hi-C topics</a></li><li class="breadcrumb-item"><a href="../pages/workflow-yeast.html">Workflow 1: Distance-dependent interactions across yeast mutants</a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../">Orchestrating Hi-C analysis with Bioconductor</a> 
        <div class="sidebar-tools-main">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label=""><i class="bi bi-git"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
<li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://github.com/js2264/OHCA/">
            Source Code
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://js2264.github.io/OHCA/devel/">
            Browse version `devel`
            </a>
          </li>
      </ul>
</div>
    <a href="../Orchestrating-Hi-C-analysis-with-Bioconductor.pdf" rel="" title="Download PDF" class="quarto-navigation-tool px-1" aria-label="Download PDF"><i class="bi bi-file-pdf"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/preamble.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preamble</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Fundamentals concepts</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/principles.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Hi-C pre-processing steps</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/data-representation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Hi-C data structures in R</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/parsing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Manipulating Hi-C data in R</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/visualization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Hi-C data visualization</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">In-depth Hi-C analysis</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/matrix-centric.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Matrix-centric analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/interactions-centric.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Interactions-centric analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/topological-features.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Finding topological features in Hi-C</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Advanced Hi-C topics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/disseminating.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Data gateways: accessing public Hi-C data portals</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/interoperability.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Interoperability: using HiCExperiment with other R packages</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/workflow-yeast.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Workflow 1: Distance-dependent interactions across yeast mutants</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/workflow-chicken.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workflow 2: Chromosome compartment cohesion upon mitosis entry</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/workflow-centros.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workflow 3: Inter-centromere interactions in yeast</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#recovering-data-from-sra" id="toc-recovering-data-from-sra" class="nav-link active" data-scroll-target="#recovering-data-from-sra">Recovering data from SRA</a></li>
  <li><a href="#processing-reads-with-hicool" id="toc-processing-reads-with-hicool" class="nav-link" data-scroll-target="#processing-reads-with-hicool">Processing reads with HiCool</a></li>
  <li><a href="#plotting-chromosome-wide-matrices-of-merged-replicates" id="toc-plotting-chromosome-wide-matrices-of-merged-replicates" class="nav-link" data-scroll-target="#plotting-chromosome-wide-matrices-of-merged-replicates">Plotting chromosome-wide matrices of merged replicates</a></li>
  <li><a href="#compute-ps-per-replicate-and-plot-it" id="toc-compute-ps-per-replicate-and-plot-it" class="nav-link" data-scroll-target="#compute-ps-per-replicate-and-plot-it">Compute P(s) per replicate and plot it</a></li>
  <li><a href="#correlation-between-replicates-with-hicrep" id="toc-correlation-between-replicates-with-hicrep" class="nav-link" data-scroll-target="#correlation-between-replicates-with-hicrep">Correlation between replicates with <code>hicrep</code></a></li>
  <li><a href="#differential-interaction-di-analysis-with-multihiccompare" id="toc-differential-interaction-di-analysis-with-multihiccompare" class="nav-link" data-scroll-target="#differential-interaction-di-analysis-with-multihiccompare">Differential interaction (DI) analysis with <code>multiHiCcompare</code></a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title">Workflow 1: Distance-dependent interactions across yeast mutants</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Aims
</div>
</div>
<div class="callout-body-container callout-body">
<p>This chapter illustrates how to:</p>
<ul>
<li>Compute P(s) of several samples and compare them</li>
<li>Compute distance-adjusted correlation between Hi-C datasets with <code>HiCRep</code>
</li>
<li>Perform differential interaction analysis between Hi-C datasets with <code>multiHiCcompare</code>
</li>
</ul>
</div>
</div>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Datasets
</div>
</div>
<div class="callout-body-container callout-body">
<p>We leverage seven yeast datasets in this notebook. They are all available from SRA:</p>
<ul>
<li>
<code>SRR8769554</code>: WT yeast strain, G1 phase (rep1)</li>
<li>
<code>SRR10687276</code>: WT yeast strain, G1 phase (rep12)</li>
<li>
<code>SRR8769549</code>: WT yeast strain, G2/M phase (rep1)</li>
<li>
<code>SRR10687281</code>: WT yeast strain, G2/M phase (rep12)</li>
<li>
<code>SRR8769551</code>: <em>wpl1</em> mutant yeast strain, G2/M phase (rep1)</li>
<li>
<code>SRR10687278</code>: <em>wpl1</em> mutant yeast strain, G2/M phase (rep2)</li>
<li>
<code>SRR8769555</code>: <em>wpl1/eco1</em> mutant yeast strain, G2/M phase</li>
</ul>
</div>
</div>
<section id="recovering-data-from-sra" class="level2"><h2 class="anchored" data-anchor-id="recovering-data-from-sra">Recovering data from SRA</h2>
<p>The easiest for this is to directly fetch files from SRA from their FTP server. We can do so using the base <code>download.file</code> function.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>The next two code chunks illustrate how to do download and process Hi-C reads from SRA, but they are not actually executed when rendering this website as it would take a significant amount of time.</p>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># !! This code is not actually executed !!</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span><span class="op">(</span><span class="st">'data'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="st">"ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR876/004/SRR8769554/SRR8769554_1.fastq.gz"</span>, <span class="st">"data/WT_G1_WT_rep1_R1.fastq.gz"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="st">"ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR876/004/SRR8769554/SRR8769554_2.fastq.gz"</span>, <span class="st">"data/WT_G1_WT_rep1_R2.fastq.gz"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="st">"ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR106/076/SRR10687276/SRR10687276_1.fastq.gz"</span>, <span class="st">"data/WT_G1_WT_rep2_R1.fastq.gz"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="st">"ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR106/076/SRR10687276/SRR10687276_2.fastq.gz"</span>, <span class="st">"data/WT_G1_WT_rep2_R2.fastq.gz"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="st">"ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR876/009/SRR8769549/SRR8769549_1.fastq.gz"</span>, <span class="st">"data/WT_G2M_WT_rep1_R1.fastq.gz"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="st">"ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR876/009/SRR8769549/SRR8769549_2.fastq.gz"</span>, <span class="st">"data/WT_G2M_WT_rep1_R2.fastq.gz"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="st">"ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR106/081/SRR10687281/SRR10687281_1.fastq.gz"</span>, <span class="st">"data/WT_G2M_WT_rep2_R1.fastq.gz"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="st">"ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR106/081/SRR10687281/SRR10687281_2.fastq.gz"</span>, <span class="st">"data/WT_G2M_WT_rep2_R2.fastq.gz"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="st">"ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR876/001/SRR8769551/SRR8769551_1.fastq.gz"</span>, <span class="st">"data/wpl1_G2M_rep1_R1.fastq.gz"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="st">"ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR876/001/SRR8769551/SRR8769551_2.fastq.gz"</span>, <span class="st">"data/wpl1_G2M_rep1_R2.fastq.gz"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="st">"ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR106/078/SRR10687278/SRR10687278_1.fastq.gz"</span>, <span class="st">"data/wpl1_G2M_rep2_R1.fastq.gz"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="st">"ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR106/078/SRR10687278/SRR10687278_2.fastq.gz"</span>, <span class="st">"data/wpl1_G2M_rep2_R2.fastq.gz"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="st">"ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR876/005/SRR8769555/SRR8769555_1.fastq.gz"</span>, <span class="st">"data/wpl1eco1_G2M_R1.fastq.gz"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="st">"ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR876/005/SRR8769555/SRR8769555_2.fastq.gz"</span>, <span class="st">"data/wpl1eco1_G2M_R2.fastq.gz"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="processing-reads-with-hicool" class="level2"><h2 class="anchored" data-anchor-id="processing-reads-with-hicool">Processing reads with HiCool</h2>
<p>We will map each pair of fastqs on the yeast genome reference (<code>R64-1-1</code>) using <code>HiCool</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># !! This code is not actually executed !!</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/js2264/HiCool">HiCool</a></span><span class="op">)</span></span>
<span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>    <span class="st">'WT_G1_rep1'</span>, </span>
<span>    <span class="st">'WT_G1_rep2'</span>, </span>
<span>    <span class="st">'WT_G2M_rep1'</span>, </span>
<span>    <span class="st">'WT_G2M_rep2'</span>, </span>
<span>    <span class="st">'wpl1_G2M_rep1'</span>, </span>
<span>    <span class="st">'wpl1_G2M_rep2'</span>, </span>
<span>    <span class="st">'wpl1eco1_G2M'</span> </span>
<span><span class="op">)</span></span>
<span><span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span><span class="op">(</span><span class="va">samples</span>, <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/pkg/HiCool/man/HiCool.html">HiCool</a></span><span class="op">(</span></span>
<span>    r1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'data/'</span>, <span class="va">.x</span>, <span class="st">'_R1.fastq.gz'</span><span class="op">)</span>, </span>
<span>    r2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'data/'</span>, <span class="va">.x</span>, <span class="st">'_R2.fastq.gz'</span><span class="op">)</span>, </span>
<span>    genome <span class="op">=</span> <span class="st">'R64-1-1'</span>, </span>
<span>    restriction <span class="op">=</span> <span class="st">'DpnII'</span>, </span>
<span>    iterative <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>    threads <span class="op">=</span> <span class="fl">15</span>, </span>
<span>    output <span class="op">=</span> <span class="st">'data/HiCool/'</span>, </span>
<span>    scratch <span class="op">=</span> <span class="st">'/data/scratch/'</span></span>
<span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Processed samples are put in <code>data/HiCool</code> directory. <code>CoolFile</code> objects are pointers to individual contact matrices. We can create such objects by using the <code>importHiCoolFolder</code> utility function.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cfs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    WT_G1_rep1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/HiCool/man/HiCool.html">importHiCoolFolder</a></span><span class="op">(</span><span class="st">'data/HiCool'</span>, <span class="st">'GK8ISZ'</span><span class="op">)</span>, </span>
<span>    WT_G1_rep2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/HiCool/man/HiCool.html">importHiCoolFolder</a></span><span class="op">(</span><span class="st">'data/HiCool'</span>, <span class="st">'SWZTO0'</span><span class="op">)</span>, </span>
<span>    WT_G2M_rep1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/HiCool/man/HiCool.html">importHiCoolFolder</a></span><span class="op">(</span><span class="st">'data/HiCool'</span>, <span class="st">'3KHHUE'</span><span class="op">)</span>, </span>
<span>    WT_G2M_rep2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/HiCool/man/HiCool.html">importHiCoolFolder</a></span><span class="op">(</span><span class="st">'data/HiCool'</span>, <span class="st">'UVNG7M'</span><span class="op">)</span>, </span>
<span>    wpl1_G2M_rep1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/HiCool/man/HiCool.html">importHiCoolFolder</a></span><span class="op">(</span><span class="st">'data/HiCool'</span>, <span class="st">'Q4KX6Z'</span><span class="op">)</span>, </span>
<span>    wpl1_G2M_rep2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/HiCool/man/HiCool.html">importHiCoolFolder</a></span><span class="op">(</span><span class="st">'data/HiCool'</span>, <span class="st">'3N0L25'</span><span class="op">)</span>, </span>
<span>    wpl1eco1_G2M <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/HiCool/man/HiCool.html">importHiCoolFolder</a></span><span class="op">(</span><span class="st">'data/HiCool'</span>, <span class="st">'LHMXWE'</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">cfs</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now that these pointers have been defined, Hi-C contact matrices can be seamlessly imported in R with <code>import</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://purrr.tidyverse.org/">purrr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/js2264/HiCExperiment">HiCExperiment</a></span><span class="op">)</span></span>
<span><span class="va">hics</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span><span class="op">(</span><span class="va">cfs</span>, <span class="va">import</span><span class="op">)</span></span>
<span><span class="va">hics</span></span>
<span><span class="co">## $WT_G1_rep1</span></span>
<span><span class="co">## `HiCExperiment` object with 5,454,145 contacts over 12,079 regions</span></span>
<span><span class="co">## -------</span></span>
<span><span class="co">## fileName: "../OHCA-data/HiCool/matrices/W303_G1_WT_rep1^mapped-S288c^GK8ISZ.mcool"</span></span>
<span><span class="co">## focus: "whole genome"</span></span>
<span><span class="co">## resolutions(5): 1000 2000 4000 8000 16000</span></span>
<span><span class="co">## active resolution: 1000</span></span>
<span><span class="co">## interactions: 3347524</span></span>
<span><span class="co">## scores(2): count balanced</span></span>
<span><span class="co">## topologicalFeatures: compartments(0) borders(0) loops(0) viewpoints(0)</span></span>
<span><span class="co">## pairsFile: ../OHCA-data/HiCool/pairs/W303_G1_WT_rep1^mapped-S288c^GK8ISZ.pairs</span></span>
<span><span class="co">## metadata(3): log args stats</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $WT_G1_rep2</span></span>
<span><span class="co">## `HiCExperiment` object with 12,068,214 contacts over 12,079 regions</span></span>
<span><span class="co">## -------</span></span>
<span><span class="co">## fileName: "../OHCA-data/HiCool/matrices/W303_G1_WT_rep2^mapped-S288c^SWZTO0.mcool"</span></span>
<span><span class="co">## focus: "whole genome"</span></span>
<span><span class="co">## resolutions(5): 1000 2000 4000 8000 16000</span></span>
<span><span class="co">## active resolution: 1000</span></span>
<span><span class="co">## interactions: 6756099</span></span>
<span><span class="co">## scores(2): count balanced</span></span>
<span><span class="co">## topologicalFeatures: compartments(0) borders(0) loops(0) viewpoints(0)</span></span>
<span><span class="co">## pairsFile: ../OHCA-data/HiCool/pairs/W303_G1_WT_rep2^mapped-S288c^SWZTO0.pairs</span></span>
<span><span class="co">## metadata(3): log args stats</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## ... </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="plotting-chromosome-wide-matrices-of-merged-replicates" class="level2"><h2 class="anchored" data-anchor-id="plotting-chromosome-wide-matrices-of-merged-replicates">Plotting chromosome-wide matrices of merged replicates</h2>
<p>We can merge replicates with the <code>merge</code> function, and <code>map</code> the <code>plotMatrix</code> function over the resulting list of <code>HiCExperiment</code>s.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/js2264/HiContacts">HiContacts</a></span><span class="op">)</span></span>
<span><span class="va">chr</span> <span class="op">&lt;-</span> <span class="st">'X'</span></span>
<span><span class="va">merged_replicates</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    WT_G1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html">merge</a></span><span class="op">(</span><span class="va">hics</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">chr</span><span class="op">]</span>, <span class="va">hics</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">chr</span><span class="op">]</span><span class="op">)</span>, </span>
<span>    WT_G2M <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html">merge</a></span><span class="op">(</span><span class="va">hics</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">chr</span><span class="op">]</span>, <span class="va">hics</span><span class="op">[[</span><span class="fl">4</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">chr</span><span class="op">]</span><span class="op">)</span>, </span>
<span>    wpl1_G2M <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html">merge</a></span><span class="op">(</span><span class="va">hics</span><span class="op">[[</span><span class="fl">5</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">chr</span><span class="op">]</span>, <span class="va">hics</span><span class="op">[[</span><span class="fl">6</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">chr</span><span class="op">]</span><span class="op">)</span>, </span>
<span>    wpl1eco1_G2M <span class="op">=</span> <span class="va">hics</span><span class="op">[[</span><span class="fl">7</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">chr</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="va">maps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/imap.html">imap</a></span><span class="op">(</span><span class="va">merged_replicates</span>, <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/pkg/HiContacts/man/plotMatrix.html">plotMatrix</a></span><span class="op">(</span></span>
<span>    <span class="va">.x</span>, use.scores <span class="op">=</span> <span class="st">'balanced'</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">3.5</span>, <span class="op">-</span><span class="fl">1.5</span><span class="op">)</span>, caption <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="va">.y</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">maps</span>, nrow <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><img src="images/20230321230247.png" class="img-fluid"></p>
<p>We can already note that long-range contacts seem to increase in frequency, in G2/M vs G1, in <em>wpl1</em> vs WT and in <em>wpl1/eco1</em> vs <em>wpl1</em>.</p>
</section><section id="compute-ps-per-replicate-and-plot-it" class="level2"><h2 class="anchored" data-anchor-id="compute-ps-per-replicate-and-plot-it">Compute P(s) per replicate and plot it</h2>
<p>Still using the <code>map</code> function, we can compute average P(s) for each replicate.<br>
Computation of the P(s) will take some time, as millions of pairs have to be imported in memory, but it will be accurate at the base resolution, rather than bin resolution from matrices.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Since matrices were imported after <code>HiCool</code> processing with the <code>importHiCoolFolder</code>, the associated <code>.pairs</code> file has been automatically added to each <code>HiCExperiment</code> object!</p>
</div>
</div>
<p>The computed P(s) is stored for each sample as a <code>tibble</code>.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">pairsFile</a></span><span class="op">(</span><span class="va">hics</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">ps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/imap.html">imap</a></span><span class="op">(</span><span class="va">hics</span>, <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/pkg/HiContacts/man/Ps.html">distanceLaw</a></span><span class="op">(</span><span class="va">.x</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>sample <span class="op">=</span> <span class="va">.y</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## Importing pairs file ../OHCA-data/HiCool/pairs/W303_G1_WT_rep1^mapped-S288c^GK8ISZ.pairs in memory. This may take a while...</span></span>
<span><span class="co">## |===============================================================| 100% 318 MB</span></span>
<span><span class="co">## Importing pairs file ../OHCA-data/HiCool/pairs/W303_G1_WT_rep2^mapped-S288c^SWZTO0.pairs in memory. This may take a while...</span></span>
<span><span class="co">## |===============================================================| 100% 674 MB</span></span>
<span><span class="co">## Importing pairs file ../OHCA-data/HiCool/pairs/W303_G2M_WT_rep1^mapped-S288c^3KHHUE.pairs in memory. This may take a while...</span></span>
<span><span class="co">## |===============================================================| 100% 709 MB</span></span>
<span><span class="co">## Importing pairs file ../OHCA-data/HiCool/pairs/W303_G2M_WT_rep2^mapped-S288c^UVNG7M.pairs in memory. This may take a while...</span></span>
<span><span class="co">## |==============================================================| 100% 1683 MB</span></span>
<span><span class="co">## Importing pairs file ../OHCA-data/HiCool/pairs/W303_G2M_wpl1_rep1^mapped-S288c^Q4KX6Z.pairs in memory. This may take a while...</span></span>
<span><span class="co">## |==============================================================| 100% 1269 MB</span></span>
<span><span class="co">## Importing pairs file ../OHCA-data/HiCool/pairs/W303_G2M_wpl1_rep2^mapped-S288c^3N0L25.pairs in memory. This may take a while...</span></span>
<span><span class="co">## |==============================================================| 100% 1529 MB</span></span>
<span><span class="co">## Importing pairs file ../OHCA-data/HiCool/pairs/W303_G2M_wpl1-eco1^mapped-S288c^LHMXWE.pairs in memory. This may take a while...</span></span>
<span><span class="co">## |==============================================================| 100% 1036 MB</span></span>
<span><span class="va">ps</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">## # A tibble: 133 x 6</span></span>
<span><span class="co">##   binned_distance          p     norm_p norm_p_unity slope sample</span></span>
<span><span class="co">##             &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;        &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;</span></span>
<span><span class="co">## 1               1 0.000154   0.000154         249.   0     WT_G1_rep1</span></span>
<span><span class="co">## 2               2 0.0000563  0.0000563         91.2  0.702 WT_G1_rep1</span></span>
<span><span class="co">## 3               3 0.0000417  0.0000417         67.5  0.699 WT_G1_rep1</span></span>
<span><span class="co">## 4               4 0.00000835 0.00000835        13.5  0.696 WT_G1_rep1</span></span>
<span><span class="co">## 5               5 0.00000501 0.00000501         8.10 0.693 WT_G1_rep1</span></span>
<span><span class="co">## 6               6 0.00000250 0.00000250         4.05 0.690 WT_G1_rep1</span></span>
<span><span class="co">## # ... with 127 more rows</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can bind all <code>tibbles</code> together and plot P(s) and their slope for each sample.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/list_c.html">list_rbind</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/HiContacts/man/plotPs.html">plotPs</a></span><span class="op">(</span></span>
<span>    <span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">binned_distance</span>, y <span class="op">=</span> <span class="va">norm_p</span>, </span>
<span>    group <span class="op">=</span> <span class="va">sample</span>, color <span class="op">=</span> <span class="va">sample</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/HiContacts/man/plotPs.html">plotPsSlope</a></span><span class="op">(</span></span>
<span>    <span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">binned_distance</span>, y <span class="op">=</span> <span class="va">slope</span>, </span>
<span>    group <span class="op">=</span> <span class="va">sample</span>, color <span class="op">=</span> <span class="va">sample</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><img src="images/20230321233026.png" class="img-fluid"></p>
</section><section id="correlation-between-replicates-with-hicrep" class="level2"><h2 class="anchored" data-anchor-id="correlation-between-replicates-with-hicrep">Correlation between replicates with <code>hicrep</code>
</h2>
<p><code>hicrep</code> is a popular package to compute stratum-adjusted correlations between Hi-C datasets. “Stratum” refers to the distance from the main diagonal: with increase distance from the main diagonal, interactions of the DNA polymer are bound to decrease. <code>hicrep</code> computes a “per-stratum” correlation score and computes a weighted average correlation for entire chromosomes.</p>
<p>We can check the documentation for <code>hicrep</code> main function, <code>get.scc</code>. This tells us that <code>mat1</code> and <code>mat2</code> <code>n*n</code> intrachromosomal contact maps of raw <code>counts</code> should be provided. Fortunately, <code>HiCExperiment</code> objects can easily be coerced into actual dense matrices using <code><a href="https://rdrr.io/r/base/matrix.html">as.matrix()</a></code> function.</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p>Make sure to use the <code>count</code> scores, which are required by <code>hicrep</code>.</p>
</div>
</div>
<p>We can calculate the overall stratum-corrected correlation score over the chromosome <code>IV</code> between the two <code>G2M</code> WT replicates.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">hicrep</span><span class="op">)</span></span>
<span><span class="va">scc</span> <span class="op">&lt;-</span> <span class="fu">get.scc</span><span class="op">(</span></span>
<span>    <span class="va">hics</span><span class="op">[[</span><span class="st">'WT_G2M_rep1'</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="st">"IV"</span><span class="op">]</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span>use.scores <span class="op">=</span> <span class="st">'count'</span><span class="op">)</span>, </span>
<span>    <span class="va">hics</span><span class="op">[[</span><span class="st">'WT_G2M_rep2'</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="st">"IV"</span><span class="op">]</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span>use.scores <span class="op">=</span> <span class="st">'count'</span><span class="op">)</span>, </span>
<span>    resol <span class="op">=</span> <span class="fl">1000</span>, h <span class="op">=</span> <span class="fl">2</span>, lbr <span class="op">=</span> <span class="fl">5000</span>, ubr <span class="op">=</span> <span class="fl">50000</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">scc</span><span class="op">)</span></span>
<span><span class="co">## [1] "corr" "wei"  "scc"  "std"</span></span>
<span><span class="va">scc</span><span class="op">$</span><span class="va">scc</span></span>
<span><span class="co">##           [,1]</span></span>
<span><span class="co">## [1,] 0.9785691</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This can be generalized to all pairwise combinations of Hi-C datasets.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://purrr.tidyverse.org/">purrr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="va">mats</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span><span class="op">(</span><span class="va">hics</span>, <span class="op">~</span> <span class="va">.x</span><span class="op">[</span><span class="st">"IV"</span><span class="op">]</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span>use.scores <span class="op">=</span> <span class="st">'count'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">7</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">7</span>, <span class="kw">function</span><span class="op">(</span><span class="va">j</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span></span>
<span>            i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">hics</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, </span>
<span>            j <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">hics</span><span class="op">)</span><span class="op">[</span><span class="va">j</span><span class="op">]</span>, </span>
<span>            scc <span class="op">=</span> <span class="fu">hicrep</span><span class="fu">::</span><span class="fu">get.scc</span><span class="op">(</span><span class="va">mats</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, <span class="va">mats</span><span class="op">[[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span>, resol <span class="op">=</span> <span class="fl">1000</span>, h <span class="op">=</span> <span class="fl">2</span>, lbr <span class="op">=</span> <span class="fl">5000</span>, ubr <span class="op">=</span> <span class="fl">200000</span><span class="op">)</span><span class="op">$</span><span class="va">scc</span></span>
<span>        <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>            <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">i</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">cfs</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>            <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>j <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">j</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">cfs</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">}</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/list_c.html">list_rbind</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/list_c.html">list_rbind</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">i</span>, y <span class="op">=</span> <span class="va">j</span>, fill <span class="op">=</span> <span class="va">scc</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_discrete.html">scale_x_discrete</a></span><span class="op">(</span>guide <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_axis.html">guide_axis</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html">coord_fixed</a></span><span class="op">(</span>ratio <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html">scale_fill_gradientn</a></span><span class="op">(</span>colours <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/HiContacts/man/palettes.html">bgrColors</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><img src="images/20230322002953.png" class="img-fluid" style="width:50.0%"></p>
<p>We can even iterate over an extra level, to compute stratum-corrected correlation for all chromosomes. Here, we will only compute correlation scores between any sample and <code>WT_G2M_rep1</code> sample.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Parallelizing over chromosomes
</div>
</div>
<div class="callout-body-container callout-body">
<p><code><a href="https://rdrr.io/pkg/BiocParallel/man/bplapply.html">BiocParallel::bplapply()</a></code> replaces <code><a href="https://purrr.tidyverse.org/reference/map.html">purrr::map()</a></code> here, as it allows parallelization of independent correlation computation runs over multiple CPUs.</p>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Some chromosomes will be ignored as they are too small for this analysis </span></span>
<span><span class="va">chrs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'II'</span>, <span class="st">'IV'</span>, <span class="st">'V'</span>, <span class="st">'VII'</span>, <span class="st">'VIII'</span>, <span class="st">'IX'</span>, <span class="st">'X'</span>, <span class="st">'XI'</span>, <span class="st">'XIII'</span>, <span class="st">'XIV'</span>, <span class="st">'XVI'</span><span class="op">)</span></span>
<span><span class="va">bpparam</span> <span class="op">&lt;-</span> <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html">MulticoreParam</a></span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">6</span>, progressbar <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/bplapply.html">bplapply</a></span><span class="op">(</span><span class="va">chrs</span>, <span class="kw">function</span><span class="op">(</span><span class="va">CHR</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">mats</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span><span class="op">(</span><span class="va">hics</span>, <span class="op">~</span> <span class="va">.x</span><span class="op">[</span><span class="va">CHR</span><span class="op">]</span> <span class="op">|&gt;</span> <span class="fu">interactions</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/as.html">gi2cm</a></span><span class="op">(</span><span class="st">'count'</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/as.html">cm2matrix</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>        <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">4</span>, <span class="fl">5</span>, <span class="fl">6</span>, <span class="fl">7</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">j</span><span class="op">)</span> <span class="op">{</span></span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span></span>
<span>                chr <span class="op">=</span> <span class="va">CHR</span>,</span>
<span>                i <span class="op">=</span> <span class="st">"WT_G2M_rep1"</span>, </span>
<span>                j <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">mats</span><span class="op">)</span><span class="op">[</span><span class="va">j</span><span class="op">]</span>, </span>
<span>                dist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">5000</span>, <span class="fl">200000</span>, <span class="fl">1000</span><span class="op">)</span>,</span>
<span>                scc <span class="op">=</span> <span class="fu">hicrep</span><span class="fu">::</span><span class="fu">get.scc</span><span class="op">(</span><span class="va">mats</span><span class="op">[[</span><span class="st">"WT_G2M_rep1"</span><span class="op">]</span><span class="op">]</span>, <span class="va">mats</span><span class="op">[[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span>, resol <span class="op">=</span> <span class="fl">1000</span>, h <span class="op">=</span> <span class="fl">2</span>, lbr <span class="op">=</span> <span class="fl">5000</span>, ubr <span class="op">=</span> <span class="fl">200000</span><span class="op">)</span> </span>
<span>            <span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>j <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">j</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">mats</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>        <span class="op">}</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/list_c.html">list_rbind</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="op">}</span>, BPPARAM <span class="op">=</span> <span class="va">bpparam</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/list_c.html">list_rbind</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>A tiny bit of data wrangling will allow us to plot the mean +/- confidence interval (90%) of stratum-adjusted correlations across the different chromosomes.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">df</span>, <span class="va">j</span>, <span class="va">dist</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize</a></span><span class="op">(</span></span>
<span>        mean <span class="op">=</span> <span class="fu">Rmisc</span><span class="fu">::</span><span class="fu">CI</span><span class="op">(</span><span class="va">scc.corr</span>, ci <span class="op">=</span> <span class="fl">0.90</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, </span>
<span>        CI_up <span class="op">=</span> <span class="fu">Rmisc</span><span class="fu">::</span><span class="fu">CI</span><span class="op">(</span><span class="va">scc.corr</span>, ci <span class="op">=</span> <span class="fl">0.90</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, </span>
<span>        CI_down <span class="op">=</span> <span class="fu">Rmisc</span><span class="fu">::</span><span class="fu">CI</span><span class="op">(</span><span class="va">scc.corr</span>, ci <span class="op">=</span> <span class="fl">0.90</span><span class="op">)</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span></span>
<span>    <span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">results</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">dist</span>, y <span class="op">=</span> <span class="va">mean</span>, ymax <span class="op">=</span> <span class="va">CI_up</span>, ymin <span class="op">=</span> <span class="va">CI_down</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>col <span class="op">=</span> <span class="va">j</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_ribbon.html">geom_ribbon</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="va">j</span><span class="op">)</span>, alpha <span class="op">=</span> <span class="fl">0.2</span>, col <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Stratum (genomic distance)"</span>, y <span class="op">=</span> <span class="st">'Stratum-corrected correlation'</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><img src="images/20230322003342.png" class="img-fluid"></p>
</section><section id="differential-interaction-di-analysis-with-multihiccompare" class="level2"><h2 class="anchored" data-anchor-id="differential-interaction-di-analysis-with-multihiccompare">Differential interaction (DI) analysis with <code>multiHiCcompare</code>
</h2>
<p>We will now focus on the chromosome <code>XI</code> and identify differentially interacting (DI) loci between WT and <em>wpl1</em> mutant in G2/M.</p>
<p>To do this, we can use the <code>multiHiCcompare</code> package. The required input for the main <code><a href="https://rdrr.io/pkg/multiHiCcompare/man/make_hicexp.html">make_hicexp()</a></code> function is a <code>list</code> of raw counts for different samples/replicates, stored in data frames with four columns (<code>chr</code>, <code>start1</code>, <code>start2</code>, <code>count</code>).<br>
Although this data structure does not correspond to a standard HiC format, it is easy to manipulate a <code>HiCExperiment</code> object to coerce it into such structure.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/dozmorovlab/multiHiCcompare">multiHiCcompare</a></span><span class="op">)</span></span>
<span><span class="va">hics_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span><span class="op">(</span><span class="va">hics</span>, <span class="op">~</span> <span class="va">.x</span><span class="op">[</span><span class="st">'XI'</span><span class="op">]</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">zoom</a></span><span class="op">(</span><span class="fl">2000</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">start1</span>, <span class="va">start2</span>, <span class="va">count</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>chr <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/relocate.html">relocate</a></span><span class="op">(</span><span class="va">chr</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">mhicc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/multiHiCcompare/man/make_hicexp.html">make_hicexp</a></span><span class="op">(</span></span>
<span>    data_list <span class="op">=</span> <span class="va">hics_list</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">4</span>, <span class="fl">5</span>, <span class="fl">6</span><span class="op">)</span><span class="op">]</span>, </span>
<span>    groups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span></span>
<span><span class="op">)</span>, A.min <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code>mhicc</code> object contains data over the chromosome <code>XI</code> binned at 2kb for two pairs of replicates (WT or <code>wpl1</code> G2/M HiC, each in duplicates):</p>
<ul>
<li>Group1 contains WT data</li>
<li>Group2 contains <code>wpl1</code> data</li>
</ul>
<p>To identify differential interactions, the actual statistical comparison is performed with the <code><a href="https://rdrr.io/pkg/multiHiCcompare/man/hic_exactTest.html">hic_exactTest()</a></code> function.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/multiHiCcompare/man/cyclic_loess.html">cyclic_loess</a></span><span class="op">(</span><span class="va">mhicc</span>, span <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/multiHiCcompare/man/hic_exactTest.html">hic_exactTest</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">## |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed=00s</span></span>
<span><span class="co">## |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed=05s</span></span>
<span><span class="va">results</span> </span>
<span><span class="co">## Hi-C Experiment Object</span></span>
<span><span class="co">## 2 experimental groups</span></span>
<span><span class="co">## Group 1 has 2 samples</span></span>
<span><span class="co">## Group 2 has 2 samples</span></span>
<span><span class="co">## Data has been normalized</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code><a href="https://rdrr.io/pkg/multiHiCcompare/man/results.html">results()</a></code> output is not very informative as it is. It requires a little bit of reformating to be able to extract valuable insights from it.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">results</span><span class="op">@</span><span class="va">hic_table</span>, <span class="fu"><a href="https://rdrr.io/pkg/multiHiCcompare/man/results.html">results</a></span><span class="op">(</span><span class="va">results</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>dist <span class="op">=</span> <span class="va">region2</span> <span class="op">-</span> <span class="va">region1</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>group <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/case_when.html">case_when</a></span><span class="op">(</span></span>
<span>        <span class="va">region1</span> <span class="op">&lt;</span> <span class="fl">430000</span> <span class="op">&amp;</span> <span class="va">region2</span> <span class="op">&gt;</span> <span class="fl">450000</span> <span class="op">~</span> <span class="st">'inter_arms'</span>,</span>
<span>        <span class="va">region1</span> <span class="op">&gt;=</span> <span class="fl">430000</span> <span class="op">&amp;</span> <span class="va">region2</span> <span class="op">&lt;=</span> <span class="fl">450000</span> <span class="op">~</span> <span class="st">'at_centro'</span>,</span>
<span>        <span class="cn">TRUE</span> <span class="op">~</span> <span class="st">'arms'</span></span>
<span>    <span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">group</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'arms'</span>, <span class="st">'inter_arms'</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>sign <span class="op">=</span> <span class="va">p.value</span> <span class="op">&lt;=</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">logFC</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">df</span></span>
<span><span class="co">## chr region1 region2 D   IF1   IF2   IF3   IF4   logFC  logCPM  p.value    p.adj dist group  sign</span></span>
<span><span class="co">##   1       1       1 0  6.16  2.09  7.96  5.43  0.5401 4.81329 5.38e-01 7.94e-01    0  arms FALSE</span></span>
<span><span class="co">##   1       1    2001 1 16.38 10.25 12.96 12.16 -0.2257 5.82484 7.00e-01 8.81e-01 2000  arms FALSE</span></span>
<span><span class="co">##   1       1    4001 2 41.41 40.72 84.41 45.14  0.5064 7.69885 5.94e-02 2.16e-01 4000  arms FALSE</span></span>
<span><span class="co">##   1       1    6001 3 22.26 30.51 73.83 48.48  1.2726 8.10243 6.48e-07 5.83e-05 6000  arms  TRUE</span></span>
<span><span class="co">##   1       1    8001 4 26.63 31.20 33.39 25.92  0.0998 7.55207 8.02e-01 9.34e-01 8000  arms FALSE</span></span>
<span><span class="co">## ...</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">logFC</span>, y <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">p.value</span><span class="op">)</span>, col <span class="op">=</span> <span class="va">sign</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">group</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html">ylim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">6</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">'none'</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'grey'</span>, <span class="st">'black'</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><img src="images/20230322102803.png" class="img-fluid"></p>
<p>In this volcano plot, we can visually appreciate the fold-change of interaction frequency in WT or <code>wpl1</code>, for interactions constrained within the chromosome <code>XI</code> arms (left) or <em>spanning</em> the chr. <code>XI</code> centromere (right). This clearly highlights that interactions within arms are increased in <code>wpl1</code> mutant while those spanning the centromere strongly decreased.</p>
<!-- 
Another analysis one can do is to investigate the relationship between 
genomic interaction distance and interaction fold-change. 


::: {.cell layout-align="center"}

```{.r .cell-code}
ggplot(filter(df, sign), aes(x = dist, y = logFC)) + 
    geom_point(size = 0.2) + 
    geom_smooth() + 
    theme_bw() + 
    facet_wrap(~group) + 
    theme(legend.position = 'none')
```
:::

-->
<p>One of the strengths of <code>HiContacts</code> is that it can be leveraged to visualize any quantification related to genomic interactions as a HiC heatmap, since <code>plotMatrix</code> can take a <code>GInteractions</code> object with any score saved in <code>mcols</code> as input.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gis</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html">rename</a></span><span class="op">(</span><span class="va">df</span>, seqnames1 <span class="op">=</span> <span class="va">chr</span>, start1 <span class="op">=</span> <span class="va">region1</span>, start2 <span class="op">=</span> <span class="va">region2</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span></span>
<span>        seqnames2 <span class="op">=</span> <span class="va">seqnames1</span>, </span>
<span>        end1 <span class="op">=</span> <span class="va">start1</span> <span class="op">+</span> <span class="fl">1999</span>, </span>
<span>        end2 <span class="op">=</span> <span class="va">start2</span> <span class="op">+</span> <span class="fl">1999</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">logFC</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/as.html">df2gi</a></span><span class="op">(</span><span class="op">)</span> </span>
<span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/HiContacts/man/plotMatrix.html">plotMatrix</a></span><span class="op">(</span><span class="va">merged_replicates</span><span class="op">[[</span><span class="st">'WT_G2M'</span><span class="op">]</span><span class="op">]</span>, use.scores <span class="op">=</span> <span class="st">'balanced'</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">3.5</span>, <span class="op">-</span><span class="fl">1</span><span class="op">)</span>, caption <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/HiContacts/man/plotMatrix.html">plotMatrix</a></span><span class="op">(</span><span class="va">merged_replicates</span><span class="op">[[</span><span class="st">'wpl1_G2M'</span><span class="op">]</span><span class="op">]</span>, use.scores <span class="op">=</span> <span class="st">'balanced'</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">3.5</span>, <span class="op">-</span><span class="fl">1</span><span class="op">)</span>, caption <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/HiContacts/man/plotMatrix.html">plotMatrix</a></span><span class="op">(</span><span class="va">gis</span>, use.scores <span class="op">=</span> <span class="st">'logFC'</span>, scale <span class="op">=</span> <span class="st">'linear'</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span>, cmap <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/HiContacts/man/palettes.html">bgrColors</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>, </span>
<span>    align <span class="op">=</span> <span class="st">"hv"</span>, axis <span class="op">=</span> <span class="st">'tblr'</span>, nrow <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><img src="images/20230322102919.png" class="img-fluid"></p>
</section><section id="references" class="level1 unnumbered"><h1 class="unnumbered">References</h1>


</section><a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../pages/interoperability.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Interoperability: using HiCExperiment with other R packages</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../pages/workflow-chicken.html" class="pagination-link">
        <span class="nav-page-text">Workflow 2: Chromosome compartment cohesion upon mitosis entry</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left"><i>OHCA: Orchestrating Hi-C analysis with Bioconductor</i> was written by J. Serizay</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">This book was built with <a href="https://github.com/js2264/BiocBook/">BiocBook</a> with <span class="emoji" data-emoji="heart">❤️</span></div>
  </div>
</footer>


</body></html>