<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Orchestrating Hi-C analysis with Bioconductor - Workflow 2: Chromosome compartment cohesion upon mitosis entry</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../pages/workflow-centros.html" rel="next">
<link href="../pages/workflow-yeast.html" rel="prev">
<link href="../assets/cover.jpg" rel="icon" type="image/jpeg">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
</head>
<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../pages/disseminating.html">Advanced Hi-C topics</a></li><li class="breadcrumb-item"><a href="../pages/workflow-chicken.html">Workflow 2: Chromosome compartment cohesion upon mitosis entry</a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../">Orchestrating Hi-C analysis with Bioconductor</a> 
        <div class="sidebar-tools-main">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label=""><i class="bi bi-git"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
<li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://github.com/js2264/OHCA/">
            Source Code
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="https://js2264.github.io/OHCA/devel/">
            Browse version `devel`
            </a>
          </li>
      </ul>
</div>
    <a href="../Orchestrating-Hi-C-analysis-with-Bioconductor.pdf" rel="" title="Download PDF" class="quarto-navigation-tool px-1" aria-label="Download PDF"><i class="bi bi-file-pdf"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/preamble.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preamble</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Fundamentals concepts</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/principles.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Hi-C pre-processing steps</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/data-representation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Hi-C data structures in R</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/parsing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Manipulating Hi-C data in R</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/visualization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Hi-C data visualization</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">In-depth Hi-C analysis</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/matrix-centric.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Matrix-centric analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/interactions-centric.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Interactions-centric analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/topological-features.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Finding topological features in Hi-C</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Advanced Hi-C topics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/disseminating.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Data gateways: accessing public Hi-C data portals</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/interoperability.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Interoperability: using HiCExperiment with other R packages</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/workflow-yeast.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workflow 1: Distance-dependent interactions across yeast mutants</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/workflow-chicken.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Workflow 2: Chromosome compartment cohesion upon mitosis entry</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../pages/workflow-centros.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Workflow 3: Inter-centromere interactions in yeast</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#importing-data" id="toc-importing-data" class="nav-link active" data-scroll-target="#importing-data">Importing data</a></li>
  <li><a href="#plotting-whole-chromosome-matrices" id="toc-plotting-whole-chromosome-matrices" class="nav-link" data-scroll-target="#plotting-whole-chromosome-matrices">Plotting whole chromosome matrices</a></li>
  <li><a href="#zooming-on-a-chromosome-section" id="toc-zooming-on-a-chromosome-section" class="nav-link" data-scroll-target="#zooming-on-a-chromosome-section">Zooming on a chromosome section</a></li>
  <li><a href="#generating-saddle-plots" id="toc-generating-saddle-plots" class="nav-link" data-scroll-target="#generating-saddle-plots">Generating saddle plots</a></li>
  <li><a href="#quantifying-interactions-within-and-between-compartments" id="toc-quantifying-interactions-within-and-between-compartments" class="nav-link" data-scroll-target="#quantifying-interactions-within-and-between-compartments">Quantifying interactions within and between compartments</a></li>
  <li><a href="#session-info" id="toc-session-info" class="nav-link" data-scroll-target="#session-info">Session info</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title">Workflow 2: Chromosome compartment cohesion upon mitosis entry</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Pre-loading packages and objects ðŸ“¦
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/">cowplot</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://purrr.tidyverse.org/">purrr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/js2264/HiCExperiment">HiCExperiment</a></span><span class="op">)</span></span>
<span><span class="co">##  Consider using the `HiContacts` package to perform advanced genomic operations </span></span>
<span><span class="co">##  on `HiCExperiment` objects.</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Read "Orchestrating Hi-C analysis with Bioconductor" online book to learn more:</span></span>
<span><span class="co">##  https://js2264.github.io/OHCA/</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Attaching package: 'HiCExperiment'</span></span>
<span><span class="co">##  The following object is masked from 'package:ggplot2':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      resolution</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/js2264/HiContactsData">HiContactsData</a></span><span class="op">)</span></span>
<span><span class="co">##  Loading required package: ExperimentHub</span></span>
<span><span class="co">##  Loading required package: BiocGenerics</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Attaching package: 'BiocGenerics'</span></span>
<span><span class="co">##  The following object is masked from 'package:HiCExperiment':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      as.data.frame</span></span>
<span><span class="co">##  The following objects are masked from 'package:stats':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      IQR, mad, sd, var, xtabs</span></span>
<span><span class="co">##  The following objects are masked from 'package:base':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      Filter, Find, Map, Position, Reduce, anyDuplicated, aperm,</span></span>
<span><span class="co">##      append, as.data.frame, basename, cbind, colnames, dirname,</span></span>
<span><span class="co">##      do.call, duplicated, eval, evalq, get, grep, grepl, intersect,</span></span>
<span><span class="co">##      is.unsorted, lapply, mapply, match, mget, order, paste, pmax,</span></span>
<span><span class="co">##      pmax.int, pmin, pmin.int, rank, rbind, rownames, sapply,</span></span>
<span><span class="co">##      setdiff, table, tapply, union, unique, unsplit, which.max,</span></span>
<span><span class="co">##      which.min</span></span>
<span><span class="co">##  Loading required package: AnnotationHub</span></span>
<span><span class="co">##  Loading required package: BiocFileCache</span></span>
<span><span class="co">##  Loading required package: dbplyr</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/js2264/fourDNData">fourDNData</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Aims
</div>
</div>
<div class="callout-body-container callout-body">
<p>This chapter illustrates how to:</p>
<ul>
<li>Annotate compartments for a list of HiC experiments</li>
<li>Generate saddle plots for a list of HiC experiments</li>
<li>Quantify changes in interactions between compartments between different timepoints</li>
</ul>
</div>
</div>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Datasets
</div>
</div>
<div class="callout-body-container callout-body">
<p>We leverage five chicken datasets in this notebook, published in <span class="citation" data-cites="Gibcus_2018">Gibcus et al. (<a href="#ref-Gibcus_2018" role="doc-biblioref">2018</a>)</span>. They are all available from the 4DN data portal using the <code>fourDNData</code> package.</p>
<ul>
<li>
<code>4DNES9LEZXN7</code>: chicken cell culture blocked in G2</li>
<li>
<code>4DNESNWWIFZU</code>: chicken cell culture released from G2 block (5min)</li>
<li>
<code>4DNESGDXKM2I</code>: chicken cell culture released from G2 block (10min)</li>
<li>
<code>4DNESIR416OW</code>: chicken cell culture released from G2 block (15min)</li>
<li>
<code>4DNESS8PTK6F</code>: chicken cell culture released from G2 block (30min)</li>
</ul>
</div>
</div>
<section id="importing-data" class="level2"><h2 class="anchored" data-anchor-id="importing-data">Importing data</h2>
<p>The 4DN consortium provides access to the datasets published in <span class="citation" data-cites="Gibcus_2018">Gibcus et al. (<a href="#ref-Gibcus_2018" role="doc-biblioref">2018</a>)</span>. in <code>R</code>, they can be obtained thanks to the <code>fourDNData</code> gateway package.</p>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Beware
</div>
</div>
<div class="callout-body-container callout-body">
<p>The first time the following chunk of code is executed, it will cache a large amount of data (mostly consisting of contact matrices stored in <code>.mcool</code> files).</p>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/js2264/HiCExperiment">HiCExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/js2264/fourDNData">fourDNData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel">BiocParallel</a></span><span class="op">)</span></span>
<span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    <span class="st">'4DNES9LEZXN7'</span> <span class="op">=</span> <span class="st">'G2 block'</span>, </span>
<span>    <span class="st">'4DNESNWWIFZU'</span> <span class="op">=</span> <span class="st">'prophase (5m)'</span>, </span>
<span>    <span class="st">'4DNESGDXKM2I'</span> <span class="op">=</span> <span class="st">'prophase (10m)'</span>, </span>
<span>    <span class="st">'4DNESIR416OW'</span> <span class="op">=</span> <span class="st">'prometaphase (15m)'</span>, </span>
<span>    <span class="st">'4DNESS8PTK6F'</span> <span class="op">=</span> <span class="st">'prometaphase (30m)'</span> </span>
<span><span class="op">)</span></span>
<span><span class="va">bpparam</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html">MulticoreParam</a></span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">5</span>, progressbar <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">hics</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/bplapply.html">bplapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">samples</span><span class="op">)</span>, <span class="va">fourDNHiCExperiment</span>, BPPARAM <span class="op">=</span> <span class="va">bpparam</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">hics</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">samples</span></span>
<span></span>
<span><span class="va">hics</span><span class="op">[[</span><span class="st">"G2 block"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">##  `HiCExperiment` object with 150,494,008 contacts over 4,109 regions </span></span>
<span><span class="co">##  -------</span></span>
<span><span class="co">##  fileName: "/root/.cache/R/fourDNData/15f21f56ea78_4DNFIT479GDR.mcool" </span></span>
<span><span class="co">##  focus: "whole genome" </span></span>
<span><span class="co">##  resolutions(13): 1000 2000 ... 5000000 10000000</span></span>
<span><span class="co">##  active resolution: 250000 </span></span>
<span><span class="co">##  interactions: 7262748 </span></span>
<span><span class="co">##  scores(2): count balanced </span></span>
<span><span class="co">##  topologicalFeatures: compartments(891) borders(3465) </span></span>
<span><span class="co">##  pairsFile: N/A </span></span>
<span><span class="co">##  metadata(3): 4DN_info eigens diamond_insulation</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="plotting-whole-chromosome-matrices" class="level2"><h2 class="anchored" data-anchor-id="plotting-whole-chromosome-matrices">Plotting whole chromosome matrices</h2>
<p>We can visualize the five different Hi-C maps on the entire chromosome <code>3</code> with <code>HiContacts</code> by iterating over each of the <code>HiCExperiment</code> objects.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://purrr.tidyverse.org/">purrr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/js2264/HiContacts">HiContacts</a></span><span class="op">)</span></span>
<span><span class="co">##  Registered S3 methods overwritten by 'readr':</span></span>
<span><span class="co">##    method                    from </span></span>
<span><span class="co">##    as.data.frame.spec_tbl_df vroom</span></span>
<span><span class="co">##    as_tibble.spec_tbl_df     vroom</span></span>
<span><span class="co">##    format.col_spec           vroom</span></span>
<span><span class="co">##    print.col_spec            vroom</span></span>
<span><span class="co">##    print.collector           vroom</span></span>
<span><span class="co">##    print.date_names          vroom</span></span>
<span><span class="co">##    print.locale              vroom</span></span>
<span><span class="co">##    str.col_spec              vroom</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="va">pl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/imap.html">imap</a></span><span class="op">(</span><span class="va">hics</span>, <span class="op">~</span> <span class="va">.x</span><span class="op">[</span><span class="st">'chr3'</span><span class="op">]</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">zoom</a></span><span class="op">(</span><span class="fl">100000</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/HiContacts/man/plotMatrix.html">plotMatrix</a></span><span class="op">(</span>use.scores <span class="op">=</span> <span class="st">'balanced'</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">4</span>, <span class="op">-</span><span class="fl">1</span><span class="op">)</span>, caption <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="va">.y</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/">cowplot</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">pl</span>, nrow <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="workflow-chicken_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>This highlights the progressive remodeling of chromatin into condensed chromosomes, starting as soon as 5â€™ after release from G2 phase.</p>
</section><section id="zooming-on-a-chromosome-section" class="level2"><h2 class="anchored" data-anchor-id="zooming-on-a-chromosome-section">Zooming on a chromosome section</h2>
<p>Zooming on a chromosome section, we can plot the Hi-C autocorrelation matrix for each timepoint. These matrices are generally used to highlight the overall correlation of interaction profiles between different segments of a chromosome section (see <a href="../pages/matrix-centric.html#computing-autocorrelated-map">Chapter 5</a> for more details).</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## --- Format compartment positions of chr. 4 segment</span></span>
<span><span class="va">.chr</span> <span class="op">&lt;-</span> <span class="st">'chr4'</span></span>
<span><span class="va">.start</span> <span class="op">&lt;-</span> <span class="fl">59000000L</span></span>
<span><span class="va">.stop</span> <span class="op">&lt;-</span> <span class="fl">75000000L</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges">GenomicRanges</a></span><span class="op">)</span></span>
<span><span class="co">##  Loading required package: stats4</span></span>
<span><span class="co">##  Loading required package: S4Vectors</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Attaching package: 'S4Vectors'</span></span>
<span><span class="co">##  The following object is masked from 'package:HiCExperiment':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      metadata&lt;-</span></span>
<span><span class="co">##  The following object is masked from 'package:utils':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      findMatches</span></span>
<span><span class="co">##  The following objects are masked from 'package:base':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      I, expand.grid, unname</span></span>
<span><span class="co">##  Loading required package: IRanges</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Attaching package: 'IRanges'</span></span>
<span><span class="co">##  The following object is masked from 'package:purrr':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      reduce</span></span>
<span><span class="co">##  Loading required package: GenomeInfoDb</span></span>
<span><span class="va">coords</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html">GRanges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">.chr</span>, <span class="st">':'</span>, <span class="va">.start</span>, <span class="st">'-'</span>, <span class="va">.stop</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">compts_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">topologicalFeatures</a></span><span class="op">(</span><span class="va">hics</span><span class="op">[[</span><span class="st">"G2 block"</span><span class="op">]</span><span class="op">]</span>, <span class="st">"compartments"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html">subsetByOverlaps</a></span><span class="op">(</span><span class="va">coords</span>, type <span class="op">=</span> <span class="st">'within'</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">compts_gg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_rect</a></span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="va">compts_df</span>, </span>
<span>    mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>xmin <span class="op">=</span> <span class="va">start</span>, xmax <span class="op">=</span> <span class="va">end</span>, ymin <span class="op">=</span> <span class="op">-</span><span class="fl">500000</span>, ymax <span class="op">=</span> <span class="fl">0</span>, alpha <span class="op">=</span> <span class="va">compartment</span><span class="op">)</span>, </span>
<span>    col <span class="op">=</span> <span class="st">'black'</span>, inherit.aes <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">## --- Subset contact matrices to chr. 4 segment and computing autocorrelation scores</span></span>
<span><span class="va">g2</span> <span class="op">&lt;-</span> <span class="va">hics</span><span class="op">[[</span><span class="st">"G2 block"</span><span class="op">]</span><span class="op">]</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">zoom</a></span><span class="op">(</span><span class="fl">100000</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html">subsetByOverlaps</a></span><span class="op">(</span><span class="va">coords</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/HiContacts/man/arithmetics.html">autocorrelate</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">##  </span></span>
<span><span class="va">pro5</span> <span class="op">&lt;-</span> <span class="va">hics</span><span class="op">[[</span><span class="st">"prophase (5m)"</span><span class="op">]</span><span class="op">]</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">zoom</a></span><span class="op">(</span><span class="fl">100000</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html">subsetByOverlaps</a></span><span class="op">(</span><span class="va">coords</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/HiContacts/man/arithmetics.html">autocorrelate</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">pro30</span> <span class="op">&lt;-</span> <span class="va">hics</span><span class="op">[[</span><span class="st">"prometaphase (30m)"</span><span class="op">]</span><span class="op">]</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">zoom</a></span><span class="op">(</span><span class="fl">100000</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html">subsetByOverlaps</a></span><span class="op">(</span><span class="va">coords</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/HiContacts/man/arithmetics.html">autocorrelate</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## --- Plot autocorrelation matrices</span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/HiContacts/man/plotMatrix.html">plotMatrix</a></span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html">subsetByOverlaps</a></span><span class="op">(</span><span class="va">g2</span>, <span class="va">coords</span><span class="op">)</span>,</span>
<span>        use.scores <span class="op">=</span> <span class="st">'autocorrelated'</span>, </span>
<span>        scale <span class="op">=</span> <span class="st">'linear'</span>, </span>
<span>        limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>, </span>
<span>        cmap <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/HiContacts/man/palettes.html">bwrColors</a></span><span class="op">(</span><span class="op">)</span>, </span>
<span>        maxDistance <span class="op">=</span> <span class="fl">10000000</span>, </span>
<span>        caption <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>    <span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">'G2'</span><span class="op">)</span> <span class="op">+</span> <span class="va">compts_gg</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/HiContacts/man/plotMatrix.html">plotMatrix</a></span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html">subsetByOverlaps</a></span><span class="op">(</span><span class="va">pro5</span>, <span class="va">coords</span><span class="op">)</span>,</span>
<span>        use.scores <span class="op">=</span> <span class="st">'autocorrelated'</span>, </span>
<span>        scale <span class="op">=</span> <span class="st">'linear'</span>, </span>
<span>        limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>, </span>
<span>        cmap <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/HiContacts/man/palettes.html">bwrColors</a></span><span class="op">(</span><span class="op">)</span>, </span>
<span>        maxDistance <span class="op">=</span> <span class="fl">10000000</span>, </span>
<span>        caption <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>    <span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">'Prophase 5min'</span><span class="op">)</span> <span class="op">+</span> <span class="va">compts_gg</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/HiContacts/man/plotMatrix.html">plotMatrix</a></span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html">subsetByOverlaps</a></span><span class="op">(</span><span class="va">pro30</span>, <span class="va">coords</span><span class="op">)</span>,</span>
<span>        use.scores <span class="op">=</span> <span class="st">'autocorrelated'</span>, </span>
<span>        scale <span class="op">=</span> <span class="st">'linear'</span>, </span>
<span>        limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>, </span>
<span>        cmap <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/HiContacts/man/palettes.html">bwrColors</a></span><span class="op">(</span><span class="op">)</span>, </span>
<span>        maxDistance <span class="op">=</span> <span class="fl">10000000</span>, </span>
<span>        caption <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>    <span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">'Prometaphase 30min'</span><span class="op">)</span> <span class="op">+</span> <span class="va">compts_gg</span>,</span>
<span>    nrow <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span>
<span><span class="co">##  Warning: Using alpha for a discrete variable is not advised.</span></span>
<span><span class="co">##  Using alpha for a discrete variable is not advised.</span></span>
<span><span class="co">##  Using alpha for a discrete variable is not advised.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="workflow-chicken_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>These correlation matrices suggest that there are two different regimes of chromatin compartment remodeling in this chromosome section:</p>
<ol type="1">
<li>Correlation scores between genomic bins within the compartment A remain positive 5â€™ after G2 release (albeit reduced compared to G2 block) and eventually become null 30â€™ after G2 release.</li>
<li>Correlation scores between genomic bins within the compartment B are overall null as soon as 5â€™ after G2 release.</li>
</ol></section><section id="generating-saddle-plots" class="level2"><h2 class="anchored" data-anchor-id="generating-saddle-plots">Generating saddle plots</h2>
<p>Saddle plots are typically used to measure the <code>observed</code> vs.&nbsp;<code>expected</code> interaction scores within or between genomic loci belonging to A and B compartments. Here, they can be used to check whether the two regimes of chromatin compartment remodeling are observed genome-wide.</p>
<p>Non-overlapping genomic windows are grouped by <code>nbins</code> quantiles (typically between 10 and 50 bins) according to their A/B compartment eigenvector value, from lowest eigenvector values (i.e.&nbsp;strongest B compartments) to highest eigenvector values (i.e.&nbsp;strongest A compartments). The average <code>observed</code> vs.&nbsp;<code>expected</code> interaction scores are computed for pairwise eigenvector quantiles and plotted in a 2D heatmap.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">pl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/imap.html">imap</a></span><span class="op">(</span><span class="va">hics</span>, <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/pkg/HiContacts/man/plotSaddle.html">plotSaddle</a></span><span class="op">(</span><span class="va">.x</span>, nbins <span class="op">=</span> <span class="fl">38</span>, BPPARAM <span class="op">=</span> <span class="va">bpparam</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="va">.y</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span>plotlist <span class="op">=</span> <span class="va">pl</span>, nrow <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="workflow-chicken_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>These plots confirm the previous observation made on chr. <code>4</code> and reveal that intra-B compartment interactions are generally lost 5â€™ after G2 release, while intra-A interactions take up to 15â€™ after G2 release to disappear.</p>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Beware
</div>
</div>
<div class="callout-body-container callout-body">
<p>The <code><a href="https://rdrr.io/pkg/HiContacts/man/plotSaddle.html">plotSaddle()</a></code> function requires an eigenvector corresponding to A/B compartments. In this example, this eigenvector is recovered from the 4DN data portal. If not already available, this eigenvector can be computed from the contact matrix using the <code><a href="https://rdrr.io/pkg/HiContacts/man/getCompartments.html">getCompartments()</a></code> function.</p>
</div>
</div>
</section><section id="quantifying-interactions-within-and-between-compartments" class="level2"><h2 class="anchored" data-anchor-id="quantifying-interactions-within-and-between-compartments">Quantifying interactions within and between compartments</h2>
<p>We can leverage the replicate-merged contact matrices to quantify the interaction frequencies within A or B compartments or between A and B compartments, at different timepoints.</p>
<p>We can use the A/B compartment annotations obtained at the <code>G2 block</code> timepoint and extract <code>O/E</code> (observed vs expected) scores for interactions within A or B compartments or between A and B compartments, at different timepoints.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## --- Extract the A/B compartments identified in G2 block</span></span>
<span><span class="va">compts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">topologicalFeatures</a></span><span class="op">(</span><span class="va">hics</span><span class="op">[[</span><span class="st">"G2 block"</span><span class="op">]</span><span class="op">]</span>, <span class="st">"compartments"</span><span class="op">)</span></span>
<span><span class="va">compts</span><span class="op">$</span><span class="va">ID</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">compts</span><span class="op">$</span><span class="va">compartment</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_along</a></span><span class="op">(</span><span class="va">compts</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## --- Iterate over timepoints to extract `detrended` (O/E) scores and </span></span>
<span><span class="co">##     compartment annotations</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/">tibble</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">plyranges</span><span class="op">)</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  Attaching package: 'plyranges'</span></span>
<span><span class="co">##  The following object is masked from 'package:IRanges':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      slice</span></span>
<span><span class="co">##  The following object is masked from 'package:stats':</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##      filter</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/imap.html">imap</a></span><span class="op">(</span><span class="va">hics</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">5</span><span class="op">)</span><span class="op">]</span>, <span class="op">~</span> <span class="op">{</span></span>
<span>    <span class="va">ints</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/HiCExperiment/man/AllGenerics.html">cis</a></span><span class="op">(</span><span class="va">.x</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="co">## Filter out trans interactions</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/pkg/HiContacts/man/arithmetics.html">detrend</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="co">## Compute O/E scores</span></span>
<span>        <span class="fu">interactions</span><span class="op">(</span><span class="op">)</span> <span class="co">## Recover interactions </span></span>
<span>    <span class="va">ints</span><span class="op">$</span><span class="va">comp_first</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/overlap-joins.html">join_overlap_left</a></span><span class="op">(</span><span class="fu">anchors</span><span class="op">(</span><span class="va">ints</span>, <span class="st">"first"</span><span class="op">)</span>, <span class="va">compts</span><span class="op">)</span><span class="op">$</span><span class="va">ID</span></span>
<span>    <span class="va">ints</span><span class="op">$</span><span class="va">comp_second</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/plyranges/man/overlap-joins.html">join_overlap_left</a></span><span class="op">(</span><span class="fu">anchors</span><span class="op">(</span><span class="va">ints</span>, <span class="st">"second"</span><span class="op">)</span>, <span class="va">compts</span><span class="op">)</span><span class="op">$</span><span class="va">ID</span></span>
<span>    <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span></span>
<span>        sample <span class="op">=</span> <span class="va">.y</span>, </span>
<span>        bin1 <span class="op">=</span> <span class="va">ints</span><span class="op">$</span><span class="va">comp_first</span>, </span>
<span>        bin2 <span class="op">=</span> <span class="va">ints</span><span class="op">$</span><span class="va">comp_second</span>, </span>
<span>        dist <span class="op">=</span> <span class="fu">InteractionSet</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/distances.html">pairdist</a></span><span class="op">(</span><span class="va">ints</span><span class="op">)</span>, </span>
<span>        OE <span class="op">=</span> <span class="va">ints</span><span class="op">$</span><span class="va">detrended</span> </span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>        <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">dist</span> <span class="op">&gt;</span> <span class="fl">5e6</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>        <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>type <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/case_when.html">case_when</a></span><span class="op">(</span></span>
<span>            <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html">grepl</a></span><span class="op">(</span><span class="st">'A'</span>, <span class="va">bin1</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html">grepl</a></span><span class="op">(</span><span class="st">'A'</span>, <span class="va">bin2</span><span class="op">)</span> <span class="op">~</span> <span class="st">'AA'</span>,</span>
<span>            <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html">grepl</a></span><span class="op">(</span><span class="st">'B'</span>, <span class="va">bin1</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html">grepl</a></span><span class="op">(</span><span class="st">'B'</span>, <span class="va">bin2</span><span class="op">)</span> <span class="op">~</span> <span class="st">'BB'</span>,</span>
<span>            <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html">grepl</a></span><span class="op">(</span><span class="st">'A'</span>, <span class="va">bin1</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html">grepl</a></span><span class="op">(</span><span class="st">'B'</span>, <span class="va">bin2</span><span class="op">)</span> <span class="op">~</span> <span class="st">'AB'</span>,</span>
<span>            <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html">grepl</a></span><span class="op">(</span><span class="st">'B'</span>, <span class="va">bin1</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html">grepl</a></span><span class="op">(</span><span class="st">'A'</span>, <span class="va">bin2</span><span class="op">)</span> <span class="op">~</span> <span class="st">'BA'</span></span>
<span>        <span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>        <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">bin1</span> <span class="op">!=</span> <span class="va">bin2</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/list_c.html">list_rbind</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span></span>
<span>    sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">sample</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">hics</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">5</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can now plot the changes in O/E scores for intra-A, intra-B, A-B or B-A interactions, splitting boxplots by timepoint.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">type</span>, y <span class="op">=</span> <span class="va">OE</span>, group <span class="op">=</span> <span class="va">type</span>, fill <span class="op">=</span> <span class="va">type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span>outlier.shape <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html">facet_grid</a></span><span class="op">(</span><span class="op">~</span><span class="va">sample</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html">ylim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##  Warning: Removed 66307 rows containing non-finite values (`stat_boxplot()`).</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="workflow-chicken_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>This visualization suggests that interactions between genomic loci belonging to the B compartment are lost more rapidly than those between genomic loci belonging to the A compartment, when cells are released from G2 to enter mitosis.</p>
</section><section id="session-info" class="level1 unnumbered"><h1 class="unnumbered">Session info</h1>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-6-contents" aria-controls="callout-6" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Click to expand ðŸ‘‡
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-6" class="callout-6-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">sessioninfo</span><span class="fu">::</span><span class="fu"><a href="https://r-lib.github.io/sessioninfo/reference/session_info.html">session_info</a></span><span class="op">(</span>include_base <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">##  â”€ Session info â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€</span></span>
<span><span class="co">##   setting  value</span></span>
<span><span class="co">##   version  R Under development (unstable) (2024-01-17 r85813)</span></span>
<span><span class="co">##   os       Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">##   system   x86_64, linux-gnu</span></span>
<span><span class="co">##   ui       X11</span></span>
<span><span class="co">##   language (EN)</span></span>
<span><span class="co">##   collate  C</span></span>
<span><span class="co">##   ctype    en_US.UTF-8</span></span>
<span><span class="co">##   tz       Etc/UTC</span></span>
<span><span class="co">##   date     2024-01-22</span></span>
<span><span class="co">##   pandoc   3.1.1 @ /usr/local/bin/ (via rmarkdown)</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  â”€ Packages â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€</span></span>
<span><span class="co">##   package              * version     date (UTC) lib source</span></span>
<span><span class="co">##   abind                  1.4-5       2016-07-21 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   AnnotationDbi          1.65.2      2023-11-03 [2] Bioconductor</span></span>
<span><span class="co">##   AnnotationHub        * 3.11.1      2023-12-11 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   backports              1.4.1       2021-12-13 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   base                 * 4.4.0       2024-01-18 [3] local</span></span>
<span><span class="co">##   base64enc              0.1-3       2015-07-28 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   beeswarm               0.4.0       2021-06-01 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   Biobase                2.63.0      2023-10-24 [2] Bioconductor</span></span>
<span><span class="co">##   BiocFileCache        * 2.11.1      2023-10-26 [2] Bioconductor</span></span>
<span><span class="co">##   BiocGenerics         * 0.49.1      2023-11-01 [2] Bioconductor</span></span>
<span><span class="co">##   BiocIO                 1.13.0      2023-10-24 [2] Bioconductor</span></span>
<span><span class="co">##   BiocManager            1.30.22     2023-08-08 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   BiocParallel         * 1.37.0      2023-10-24 [2] Bioconductor</span></span>
<span><span class="co">##   BiocVersion            3.19.1      2023-10-26 [2] Bioconductor</span></span>
<span><span class="co">##   Biostrings             2.71.1      2023-10-25 [2] Bioconductor</span></span>
<span><span class="co">##   bit                    4.0.5       2022-11-15 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   bit64                  4.0.5       2020-08-30 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   bitops                 1.0-7       2021-04-24 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   blob                   1.2.4       2023-03-17 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   cachem                 1.0.8       2023-05-01 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   Cairo                  1.6-2       2023-11-28 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   checkmate              2.3.1       2023-12-04 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   cli                    3.6.2       2023-12-11 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   cluster                2.1.6       2023-12-01 [3] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   codetools              0.2-19      2023-02-01 [3] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   colorspace             2.1-0       2023-01-23 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   compiler               4.4.0       2024-01-18 [3] local</span></span>
<span><span class="co">##   cowplot              * 1.1.2       2023-12-15 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   crayon                 1.5.2       2022-09-29 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   curl                   5.2.0       2023-12-08 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   data.table             1.14.10     2023-12-08 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   datasets             * 4.4.0       2024-01-18 [3] local</span></span>
<span><span class="co">##   DBI                    1.2.1       2024-01-12 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   dbplyr               * 2.4.0       2023-10-26 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   DelayedArray           0.29.0      2023-10-24 [2] Bioconductor</span></span>
<span><span class="co">##   digest                 0.6.34      2024-01-11 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   doParallel             1.0.17      2022-02-07 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   dplyr                  1.1.4       2023-11-17 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   dynamicTreeCut         1.63-1      2016-03-11 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   evaluate               0.23        2023-11-01 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   ExperimentHub        * 2.11.1      2023-12-11 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   fansi                  1.0.6       2023-12-08 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   farver                 2.1.1       2022-07-06 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   fastcluster            1.2.6       2024-01-12 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   fastmap                1.1.1       2023-02-24 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   filelock               1.0.3       2023-12-11 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   foreach                1.5.2       2022-02-02 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   foreign                0.8-86      2023-11-28 [3] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   Formula                1.2-5       2023-02-24 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   fourDNData           * 1.3.0       2023-10-31 [2] Bioconductor</span></span>
<span><span class="co">##   generics               0.1.3       2022-07-05 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   GenomeInfoDb         * 1.39.5      2024-01-01 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   GenomeInfoDbData       1.2.11      2024-01-22 [2] Bioconductor</span></span>
<span><span class="co">##   GenomicAlignments      1.39.2      2024-01-16 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   GenomicRanges        * 1.55.1      2023-10-29 [2] Bioconductor</span></span>
<span><span class="co">##   ggbeeswarm             0.7.2       2023-04-29 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   ggplot2              * 3.4.4       2023-10-12 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   ggrastr                1.0.2       2023-06-01 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   glue                   1.7.0       2024-01-09 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   GO.db                  3.18.0      2024-01-22 [2] Bioconductor</span></span>
<span><span class="co">##   graphics             * 4.4.0       2024-01-18 [3] local</span></span>
<span><span class="co">##   grDevices            * 4.4.0       2024-01-18 [3] local</span></span>
<span><span class="co">##   grid                   4.4.0       2024-01-18 [3] local</span></span>
<span><span class="co">##   gridExtra              2.3         2017-09-09 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   gtable                 0.3.4       2023-08-21 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   HiCExperiment        * 1.3.0       2023-10-24 [2] Bioconductor</span></span>
<span><span class="co">##   HiContacts           * 1.5.0       2023-10-24 [2] Bioconductor</span></span>
<span><span class="co">##   HiContactsData       * 1.5.3       2024-01-22 [2] Github (js2264/HiContactsData@d5bebe7)</span></span>
<span><span class="co">##   Hmisc                  5.1-1       2023-09-12 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   hms                    1.1.3       2023-03-21 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   htmlTable              2.4.2       2023-10-29 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   htmltools              0.5.7       2023-11-03 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   htmlwidgets            1.6.4       2023-12-06 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   httr                   1.4.7       2023-08-15 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   impute                 1.77.0      2023-10-24 [2] Bioconductor</span></span>
<span><span class="co">##   InteractionSet         1.31.0      2023-10-24 [2] Bioconductor</span></span>
<span><span class="co">##   IRanges              * 2.37.1      2024-01-19 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   iterators              1.0.14      2022-02-05 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   jsonlite               1.8.8       2023-12-04 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   KEGGREST               1.43.0      2023-10-24 [2] Bioconductor</span></span>
<span><span class="co">##   knitr                  1.45        2023-10-30 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   labeling               0.4.3       2023-08-29 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   lattice                0.22-5      2023-10-24 [3] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   lifecycle              1.0.4       2023-11-07 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   magrittr               2.0.3       2022-03-30 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   Matrix                 1.6-5       2024-01-11 [3] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   MatrixGenerics         1.15.0      2023-10-24 [2] Bioconductor</span></span>
<span><span class="co">##   matrixStats            1.2.0       2023-12-11 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   memoise                2.0.1       2021-11-26 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   methods              * 4.4.0       2024-01-18 [3] local</span></span>
<span><span class="co">##   munsell                0.5.0       2018-06-12 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   nnet                   7.3-19      2023-05-03 [3] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   parallel               4.4.0       2024-01-18 [3] local</span></span>
<span><span class="co">##   pillar                 1.9.0       2023-03-22 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   pkgconfig              2.0.3       2019-09-22 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   plyranges            * 1.23.0      2023-10-24 [2] Bioconductor</span></span>
<span><span class="co">##   png                    0.1-8       2022-11-29 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   preprocessCore         1.65.0      2023-10-24 [2] Bioconductor</span></span>
<span><span class="co">##   purrr                * 1.0.2       2023-08-10 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   R6                     2.5.1       2021-08-19 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   rappdirs               0.3.3       2021-01-31 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   Rcpp                   1.0.12      2024-01-09 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   RCurl                  1.98-1.14   2024-01-09 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   readr                  2.1.5       2024-01-10 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   restfulr               0.0.15      2022-06-16 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   rhdf5                  2.47.2      2024-01-15 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   rhdf5filters           1.15.1      2023-11-06 [2] Bioconductor</span></span>
<span><span class="co">##   Rhdf5lib               1.25.1      2023-12-11 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   rjson                  0.2.21      2022-01-09 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   rlang                  1.1.3       2024-01-10 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   rmarkdown              2.25        2023-09-18 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   rpart                  4.1.23      2023-12-05 [3] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   Rsamtools              2.19.3      2024-01-17 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   RSpectra               0.16-1      2022-04-24 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   RSQLite                2.3.5       2024-01-21 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   rstudioapi             0.15.0      2023-07-07 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   rtracklayer            1.63.0      2024-01-22 [2] Github (lawremi/rtracklayer@86407bb)</span></span>
<span><span class="co">##   S4Arrays               1.3.2       2024-01-14 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   S4Vectors            * 0.41.3      2024-01-01 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   scales                 1.3.0       2023-11-28 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   sessioninfo            1.2.2       2021-12-06 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   SparseArray            1.3.3       2024-01-14 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   splines                4.4.0       2024-01-18 [3] local</span></span>
<span><span class="co">##   stats                * 4.4.0       2024-01-18 [3] local</span></span>
<span><span class="co">##   stats4               * 4.4.0       2024-01-18 [3] local</span></span>
<span><span class="co">##   strawr                 0.0.91      2023-03-29 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   stringi                1.8.3       2023-12-11 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   stringr                1.5.1       2023-11-14 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   SummarizedExperiment   1.33.2      2024-01-07 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   survival               3.5-7       2023-08-14 [3] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   tibble               * 3.2.1       2023-03-20 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   tidyr                  1.3.0       2023-01-24 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   tidyselect             1.2.0       2022-10-10 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   tools                  4.4.0       2024-01-18 [3] local</span></span>
<span><span class="co">##   tzdb                   0.4.0       2023-05-12 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   utf8                   1.2.4       2023-10-22 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   utils                * 4.4.0       2024-01-18 [3] local</span></span>
<span><span class="co">##   vctrs                  0.6.5       2023-12-01 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   vipor                  0.4.7       2023-12-18 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   vroom                  1.6.5       2023-12-05 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   WGCNA                  1.72-5      2023-12-07 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   withr                  3.0.0       2024-01-16 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   xfun                   0.41        2023-11-01 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   XML                    3.99-0.16.1 2024-01-22 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   XVector                0.43.1      2024-01-10 [2] Bioconductor 3.19 (R 4.4.0)</span></span>
<span><span class="co">##   yaml                   2.3.8       2023-12-11 [2] CRAN (R 4.4.0)</span></span>
<span><span class="co">##   zlibbioc               1.49.0      2023-10-24 [2] Bioconductor</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##   [1] /tmp/Rtmpq5g2WV/Rinstb37571687</span></span>
<span><span class="co">##   [2] /usr/local/lib/R/site-library</span></span>
<span><span class="co">##   [3] /usr/local/lib/R/library</span></span>
<span><span class="co">##  </span></span>
<span><span class="co">##  â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</section><section id="references" class="level1 unnumbered"><h1 class="unnumbered">References</h1>


<div id="refs" class="references csl-bib-body hanging-indent" data-line-spacing="2" role="list">
<div id="ref-Gibcus_2018" class="csl-entry" role="listitem">
Gibcus, J. H., Samejima, K., Goloborodko, A., Samejima, I., Naumova, N., Nuebler, J., Kanemaki, M. T., Xie, L., Paulson, J. R., Earnshaw, W. C., Mirny, L. A., &amp; Dekker, J. (2018). A pathway for mitotic chromosome formation. <em>Science</em>, <em>359</em>(6376). <a href="https://doi.org/10.1126/science.aao6135">https://doi.org/10.1126/science.aao6135</a>
</div>
</div>
</section><a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "î§‹";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../pages/workflow-yeast.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Workflow 1: Distance-dependent interactions across yeast mutants</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../pages/workflow-centros.html" class="pagination-link">
        <span class="nav-page-text">Workflow 3: Inter-centromere interactions in yeast</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left"><i>OHCA: Orchestrating Hi-C analysis with Bioconductor</i> was written by J. Serizay</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">This book was built with <a href="https://github.com/js2264/BiocBook/">BiocBook</a> with <span class="emoji" data-emoji="heart">â¤ï¸</span></div>
  </div>
</footer>


</body></html>